{
 "cells": [
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:40.824903Z",
     "start_time": "2024-11-14T20:35:36.604074Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from operator import index\n",
    "\n",
    "from IPython.core.inputtransformer import cellmagic\n",
    "from oaklib import get_adapter\n",
    "from oaklib.datamodels.vocabulary import IS_A, PART_OF\n",
    "import pandas as pd\n",
    "from linkml_runtime import SchemaView\n",
    "from linkml_runtime.dumpers import yaml_dumper\n",
    "import requests\n",
    "import pprint\n",
    "\n",
    "from typing import Dict, Any\n",
    "import csv\n",
    "\n",
    "from sklearn.feature_extraction.text import CountVectorizer  # from scikit-learn\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import classification_report\n",
    "from scipy.sparse import hstack\n",
    "\n",
    "import re\n",
    "\n",
    "import os\n",
    "\n",
    "import duckdb\n",
    "import sqlite3\n",
    "\n",
    "import gzip\n",
    "import shutil\n",
    "\n",
    "from urllib.parse import urlparse"
   ],
   "id": "630126a7790cc3a7",
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:40.834913Z",
     "start_time": "2024-11-14T20:35:40.830947Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Approved prefixes (case-insensitive)\n",
    "approved_prefixes = ['ENVO']"
   ],
   "id": "914fdae0d02dad58",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:40.923189Z",
     "start_time": "2024-11-14T20:35:40.920293Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# make a biomes curie -> label dict?\n",
    "BIOME_CURIE = 'ENVO:00000428'"
   ],
   "id": "95f7efc06f20987a",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:40.966102Z",
     "start_time": "2024-11-14T20:35:40.963053Z"
    }
   },
   "cell_type": "code",
   "source": "ENV_BROAD_SCALE_ENUM = \"EnvBroadScaleSoilEnum\"",
   "id": "20750df32ebb2033",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.015648Z",
     "start_time": "2024-11-14T20:35:41.009989Z"
    }
   },
   "cell_type": "code",
   "source": [
    "ncbi_query = \"\"\"\n",
    "SELECT content, COUNT(1) AS sample_count \n",
    "FROM attributes \n",
    "WHERE harmonized_name = 'env_broad_scale' AND package_content = 'MIMS.me.soil.6.0' \n",
    "GROUP BY content \n",
    "ORDER BY COUNT(1) DESC\n",
    "\"\"\""
   ],
   "id": "d33879ec4c6ca1df",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.062242Z",
     "start_time": "2024-11-14T20:35:41.059110Z"
    }
   },
   "cell_type": "code",
   "source": "envo_adapter_string = \"sqlite:obo:envo\"",
   "id": "8240c0de4cbcec24",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.109126Z",
     "start_time": "2024-11-14T20:35:41.106028Z"
    }
   },
   "cell_type": "code",
   "source": "# goldterms_adapter_string = \"sqlite:obo:envo\"",
   "id": "eb27b996bb24645a",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.156379Z",
     "start_time": "2024-11-14T20:35:41.152829Z"
    }
   },
   "cell_type": "code",
   "source": [
    "goldterms_semsql_url = \"https://s3.amazonaws.com/bbop-sqlite/goldterms.db.gz\"\n",
    "\n",
    "# https://s3.amazonaws.com/bbop-sqlite/\n",
    "# <Contents>\n",
    "# <Key>goldterms.db.gz</Key>\n",
    "# <LastModified>2024-11-03T17:24:56.000Z</LastModified>\n",
    "# <ETag>\"fe8e35b215786cb9fc347b7fadbe055f\"</ETag>\n",
    "# <Size>2935781</Size>\n",
    "# <StorageClass>STANDARD</StorageClass>\n",
    "# </Contents>\n"
   ],
   "id": "2ae64976b48e359",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.203584Z",
     "start_time": "2024-11-14T20:35:41.200108Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo could this have been done with a OAK query, eliminating the need to explicitly download the file?\n",
    "\n",
    "goldterms_envo_query = \"\"\"\n",
    "SELECT\n",
    "\t*\n",
    "FROM\n",
    "\tstatements s\n",
    "WHERE\n",
    "\tpredicate = 'mixs:env_broad'\"\"\""
   ],
   "id": "ae751e002f65be71",
   "outputs": [],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.249770Z",
     "start_time": "2024-11-14T20:35:41.246858Z"
    }
   },
   "cell_type": "code",
   "source": "GOLDTERMS_SOIL = 'GOLDTERMS:4212'",
   "id": "eda591e14a91cd01",
   "outputs": [],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.297185Z",
     "start_time": "2024-11-14T20:35:41.293956Z"
    }
   },
   "cell_type": "code",
   "source": [
    "goldterms_soil_subclass_query = f\"\"\"\n",
    "select\n",
    "\tsubject\n",
    "from\n",
    "\tentailed_edge ee\n",
    "where\n",
    "\tpredicate = 'rdfs:subClassOf'\n",
    "\tand object = '{GOLDTERMS_SOIL}'\n",
    "\"\"\""
   ],
   "id": "6b672d93e1f5c53",
   "outputs": [],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.344414Z",
     "start_time": "2024-11-14T20:35:41.341092Z"
    }
   },
   "cell_type": "code",
   "source": "previous_submission_schema_url = \"https://raw.githubusercontent.com/microbiomedata/submission-schema/v10.7.0/src/nmdc_submission_schema/schema/nmdc_submission_schema.yaml\"",
   "id": "a1e9e99c5ac41ef4",
   "outputs": [],
   "execution_count": 12
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.391453Z",
     "start_time": "2024-11-14T20:35:41.388181Z"
    }
   },
   "cell_type": "code",
   "source": [
    "NMDC_RUNTIME_BASE_URL = 'https://api.microbiomedata.org/nmdcschema/'\n",
    "STUDY_SET_COLLECTION = 'study_set'\n",
    "BIOSAMPLE_SET_COLLECTION = 'biosample_set'"
   ],
   "id": "cde9a57dca22a7c1",
   "outputs": [],
   "execution_count": 13
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.439003Z",
     "start_time": "2024-11-14T20:35:41.435707Z"
    }
   },
   "cell_type": "code",
   "source": [
    "env_package_override_file = '../../mam-env-package-overrides.tsv'\n",
    "override_column = 'mam_inferred_env_package'"
   ],
   "id": "375bac6b84ab6198",
   "outputs": [],
   "execution_count": 14
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.487115Z",
     "start_time": "2024-11-14T20:35:41.483201Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# ncbi_duckdb_file = '../../ncbi_biosamples.duckdb'\n",
    "\n",
    "ncbi_duckdb_url = 'https://portal.nersc.gov/project/m3408/biosamples_duckdb/ncbi_biosamples_2024-09-23.duckdb.gz'"
   ],
   "id": "8a8be3a9e6123df5",
   "outputs": [],
   "execution_count": 15
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.534412Z",
     "start_time": "2024-11-14T20:35:41.531010Z"
    }
   },
   "cell_type": "code",
   "source": [
    "gold_data_url = \"https://gold.jgi.doe.gov/download?mode=site_excel\"\n",
    "BIOSAMPLES_SHEET = \"Biosample\""
   ],
   "id": "fb9c800afcc86146",
   "outputs": [],
   "execution_count": 16
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.582163Z",
     "start_time": "2024-11-14T20:35:41.578065Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Initialize cache dictionaries for predict_from_normalized_env_packages\n",
    "ancestor_cache = {}\n",
    "descendant_cache = {}"
   ],
   "id": "f952b5888d2c5f0d",
   "outputs": [],
   "execution_count": 17
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.631891Z",
     "start_time": "2024-11-14T20:35:41.626525Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo is filling memory with things like this a good idea? for understandability? or performance?\n",
    "# todo they should be aggregated somewhere, as specified by the config.yaml\n",
    "# todo or should we going straight to data frames? in which case a dlist of dicts might be preferable\n",
    "def get_curie_descendants_label_dict(curie, predicates, adapter):\n",
    "    curie_label_dict = {}\n",
    "    for descendant in adapter.descendants(curie, predicates=predicates):\n",
    "        curie_label_dict[descendant] = adapter.label(descendant)\n",
    "    return curie_label_dict"
   ],
   "id": "342c71e8b7a3366f",
   "outputs": [],
   "execution_count": 18
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.680322Z",
     "start_time": "2024-11-14T20:35:41.676410Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def curie_descendants_label_dict_to_lod(curie_label_dict):\n",
    "    return [{'curie': k, 'label': v} for k, v in curie_label_dict.items()]"
   ],
   "id": "1f4304edb0904f25",
   "outputs": [],
   "execution_count": 19
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.834509Z",
     "start_time": "2024-11-14T20:35:41.828935Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def curie_descendants_label_lod_to_df(curie_label_lod):\n",
    "    return pd.DataFrame(curie_label_lod)"
   ],
   "id": "c43fc53d70ff43e8",
   "outputs": [],
   "execution_count": 20
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.902425Z",
     "start_time": "2024-11-14T20:35:41.899324Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def get_schemaview_from_source(source):\n",
    "    return SchemaView(source)"
   ],
   "id": "28e160b747e7480e",
   "outputs": [],
   "execution_count": 21
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:41.951534Z",
     "start_time": "2024-11-14T20:35:41.948172Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# def get_schema_from_schemaview(schemaview):\n",
    "#     return schemaview.schema"
   ],
   "id": "993d01723365ec07",
   "outputs": [],
   "execution_count": 22
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.001236Z",
     "start_time": "2024-11-14T20:35:41.996339Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def parse_hierarchically_underscored_strings(hierarchically_underscored_string_list):\n",
    "    result = []\n",
    "    for item in hierarchically_underscored_string_list:\n",
    "        # Remove leading underscores for label, split on '[' to separate curie\n",
    "        label, curie = item.lstrip('_').split(' [')\n",
    "        # Remove the trailing ']' from curie\n",
    "        curie = curie.rstrip(']')\n",
    "        # Append dictionary with label and curie\n",
    "        result.append({'label': label.strip(), 'curie': curie.strip()})\n",
    "    return result"
   ],
   "id": "8ce34ecfbae82a23",
   "outputs": [],
   "execution_count": 23
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.049723Z",
     "start_time": "2024-11-14T20:35:42.045219Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def dedupe_underscoreless_pvs(underscoreless_pvs):\n",
    "    # Dictionary to store CURIE as key and list of unique labels as values\n",
    "    curie_to_labels = {}\n",
    "\n",
    "    for item in underscoreless_pvs:\n",
    "        curie = item['curie']\n",
    "        label = item['label']\n",
    "\n",
    "        # Initialize the list if curie is not yet a key\n",
    "        if curie not in curie_to_labels:\n",
    "            curie_to_labels[curie] = []\n",
    "\n",
    "        # Add label if it is not already in the list for this curie\n",
    "        if label not in curie_to_labels[curie]:\n",
    "            curie_to_labels[curie].append(label)\n",
    "    return curie_to_labels\n"
   ],
   "id": "ce2b4b74d1fdc0d4",
   "outputs": [],
   "execution_count": 24
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.099794Z",
     "start_time": "2024-11-14T20:35:42.093704Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def validate_curie_label_list_dict(curie_label_dict, adapter, print_flag=False):\n",
    "    problem_curies = []\n",
    "    valid_curies = []\n",
    "    for curie, labels in curie_label_dict.items():\n",
    "        true_label = adapter.label(curie)\n",
    "        if true_label not in labels:\n",
    "            problem_curies.append(curie)\n",
    "            if print_flag:\n",
    "                print(f\"Error: {curie} has true label {true_label} which doesn't appear in {labels}\")\n",
    "        else:\n",
    "            valid_curies.append({\"curie\": curie, \"label\": true_label})\n",
    "    return {\"problems\": problem_curies, \"valids\": valid_curies}"
   ],
   "id": "cbd3ed3aec38ee16",
   "outputs": [],
   "execution_count": 25
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.152391Z",
     "start_time": "2024-11-14T20:35:42.144111Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo could pre-determine the collection sizes\n",
    "# todo could report elapsed time\n",
    "\n",
    "def get_docs_from_nmdc_collection(base_url, collection_name, max_page_size=1000, stop_after=None):\n",
    "    \"\"\"\n",
    "    Fetch all documents from a paginated API. Defaults to fetching a large number of documents per page.\n",
    "    Optionally stop after a specified number of documents.\n",
    "\n",
    "    Parameters:\n",
    "    - base_url: The base URL of the API endpoint (e.g., 'https://api.microbiomedata.org/nmdcschema/').\n",
    "    - collection_name: The name of the collection to fetch (e.g., 'biosample_set').\n",
    "    - max_page_size: The maximum number of items to retrieve per page (default 1000).\n",
    "    - stop_after: Optional parameter to stop fetching after a certain number of documents (default None).\n",
    "\n",
    "    Returns:\n",
    "    - A list of documents fetched from the API.\n",
    "    \"\"\"\n",
    "    documents = []\n",
    "    page_token = None\n",
    "    total_documents = 0\n",
    "    page_count = 0\n",
    "\n",
    "    # Construct the full URL with the collection name\n",
    "    url = f\"{base_url}{collection_name}\"\n",
    "\n",
    "    while True:\n",
    "        page_count += 1\n",
    "        # Prepare the query parameters\n",
    "        params = {\n",
    "            'collection_name': collection_name,\n",
    "            'max_page_size': max_page_size,  # Set large max_page_size to reduce pagination\n",
    "        }\n",
    "\n",
    "        if page_token:\n",
    "            params['page_token'] = page_token  # Add the page token for pagination\n",
    "\n",
    "        # Send the request to the API\n",
    "        response = requests.get(url, params=params)\n",
    "\n",
    "        if response.status_code != 200:\n",
    "            print(f\"Error fetching data: {response.status_code}\")\n",
    "            break\n",
    "\n",
    "        data = response.json()\n",
    "\n",
    "        # Add the current page of documents to the list\n",
    "        num_documents_on_page = len(data['resources'])\n",
    "        documents.extend(data['resources'])\n",
    "        total_documents += num_documents_on_page\n",
    "\n",
    "        # Status reporting\n",
    "        print(f\"Fetched page {page_count} with {num_documents_on_page} documents. Total fetched: {total_documents}\")\n",
    "\n",
    "        # If stop_after is provided, stop fetching after reaching the specified number of documents\n",
    "        if stop_after and total_documents >= stop_after:\n",
    "            documents = documents[:stop_after]  # Trim to the required number\n",
    "            print(f\"Reached stop_after limit of {stop_after} documents.\")\n",
    "            break\n",
    "\n",
    "        # Check if there is a next page\n",
    "        page_token = data.get('next_page_token')\n",
    "        if not page_token:\n",
    "            print(\"All documents fetched.\")\n",
    "            break  # Exit the loop if no more pages are available\n",
    "\n",
    "    return documents"
   ],
   "id": "69ea6069b4c4abbb",
   "outputs": [],
   "execution_count": 26
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.201125Z",
     "start_time": "2024-11-14T20:35:42.196627Z"
    }
   },
   "cell_type": "code",
   "source": [
    "\n",
    "def get_name_or_rawval(env_scale: Dict[str, Any]) -> str:\n",
    "    \"\"\"Safely extract label from environmental scale data.\"\"\"\n",
    "    if env_scale:\n",
    "        term = env_scale.get('term')\n",
    "        if term:\n",
    "            return term.get('name', term.get('has_raw_value', ''))\n",
    "    return ''"
   ],
   "id": "590502abe99de710",
   "outputs": [],
   "execution_count": 27
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.252557Z",
     "start_time": "2024-11-14T20:35:42.247822Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def tsv_to_dict_of_dicts(tsv_file, outer_key_column):\n",
    "    \"\"\"\n",
    "    Reads a TSV file into a dictionary of dictionaries.\n",
    "\n",
    "    :param tsv_file: Path to the TSV file.\n",
    "    :param outer_key_column: The column name or index to be used as the key for the outer dictionary.\n",
    "    :return: A dictionary of dictionaries, with outer keys being the values from the specified column.\n",
    "    \"\"\"\n",
    "    with open(tsv_file, newline='', encoding='utf-8') as f:\n",
    "        reader = csv.DictReader(f, delimiter='\\t')\n",
    "\n",
    "        result = {}\n",
    "\n",
    "        for row in reader:\n",
    "            outer_key = row[outer_key_column]\n",
    "            result[outer_key] = {key: value for key, value in row.items() if key != outer_key_column}\n",
    "\n",
    "    return result"
   ],
   "id": "6668e06afbd16f4",
   "outputs": [],
   "execution_count": 28
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.305221Z",
     "start_time": "2024-11-14T20:35:42.296565Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo only gets authoritative labels from the passed adapter, which is presumably EnvO only\n",
    "# todo would benefit from caching of labels\n",
    "\n",
    "def biosamples_lod_context_extractor(biosamples_lod, adapter, env_pacakge_overrides=None):\n",
    "    new_lod = []\n",
    "    for biosample in biosamples_lod:\n",
    "        insdc_identifiers = biosample.get('insdc_biosample_identifiers', [])\n",
    "\n",
    "        env_broad_scale_label = get_name_or_rawval(biosample.get('env_broad_scale'))\n",
    "        env_local_scale_label = get_name_or_rawval(biosample.get('env_local_scale'))\n",
    "        env_medium_label = get_name_or_rawval(biosample.get('env_medium'))\n",
    "\n",
    "        # Extracting optional scalar env_package.has_raw_value\n",
    "        env_package_has_raw_value = biosample.get('env_package', {}).get('has_raw_value', '')\n",
    "\n",
    "        # Extracting required multivalued part_of\n",
    "        associated_studies = '|'.join(biosample.get('associated_studies', []))  # Assuming part_of is a list of strings\n",
    "\n",
    "        row: Dict[str, str] = {\n",
    "            'id': biosample['id'],\n",
    "            'insdc_biosample_identifiers': '|'.join(insdc_identifiers) if insdc_identifiers else '',\n",
    "\n",
    "            'env_broad_scale_id': biosample['env_broad_scale']['term']['id'],\n",
    "            'env_broad_scale_mongo_label': env_broad_scale_label,\n",
    "            'env_broad_scale_auth_label': adapter.label(biosample['env_broad_scale']['term']['id']),\n",
    "\n",
    "            'env_local_scale_id': biosample['env_local_scale']['term']['id'],\n",
    "            'env_local_scale_mongo_label': env_local_scale_label,\n",
    "            'env_local_scale_auth_label': adapter.label(biosample['env_local_scale']['term']['id']),\n",
    "\n",
    "            'env_medium_id': biosample['env_medium']['term']['id'],\n",
    "            'env_medium_mongo_label': env_medium_label,\n",
    "            'env_medium_auth_label': adapter.label(biosample['env_medium']['term']['id']),\n",
    "\n",
    "            'env_package_has_raw_value': env_package_has_raw_value,\n",
    "            'normalized_env_package': 'soil' if env_package_has_raw_value == 'ENVO:00001998' else env_package_has_raw_value.lower(),\n",
    "            # todo abstract this though label search, or at least providing a lookup structure\n",
    "\n",
    "            'associated_studies': associated_studies\n",
    "        }\n",
    "\n",
    "        if env_pacakge_overrides and biosample['id'] in env_pacakge_overrides:\n",
    "            print(\n",
    "                f\"Overriding env_package for biosample {biosample['id']} from {row['normalized_env_package']} to {env_pacakge_overrides[biosample['id']]['mam_inferred_env_package']}\")\n",
    "            row['normalized_env_package'] = env_pacakge_overrides[biosample['id']]['mam_inferred_env_package']\n",
    "\n",
    "        new_lod.append(row)\n",
    "    return new_lod\n"
   ],
   "id": "53b9c7e7d15845f6",
   "outputs": [],
   "execution_count": 29
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.356145Z",
     "start_time": "2024-11-14T20:35:42.349276Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def get_hierarchy_terms(curie: str, adapter) -> dict:\n",
    "    \"\"\"\n",
    "    Extract ancestor and descendant terms from the ontology for a given CURIE,\n",
    "    using caching to improve performance and filtering by 'is_a' relationships.\n",
    "\n",
    "    Args:\n",
    "        curie (str): CURIE identifier for the ontology term.\n",
    "        adapter: Ontology adapter.\n",
    "\n",
    "    Returns:\n",
    "        dict: Dictionary containing lists of ancestor and descendant terms.\n",
    "    \"\"\"\n",
    "    if curie in ancestor_cache:\n",
    "        ancestors = ancestor_cache[curie]\n",
    "    else:\n",
    "        try:\n",
    "            ancestors = list(adapter.ancestors(curie, predicates=[IS_A]))\n",
    "            ancestor_cache[curie] = [adapter.label(ancestor) for ancestor in ancestors if ancestor]\n",
    "        except Exception as e:\n",
    "            print(f\"Error retrieving ancestors for {curie}: {e}\")\n",
    "            ancestor_cache[curie] = []\n",
    "\n",
    "    if curie in descendant_cache:\n",
    "        descendants = descendant_cache[curie]\n",
    "    else:\n",
    "        try:\n",
    "            descendants = list(adapter.descendants(curie, predicates=[IS_A]))\n",
    "            descendant_cache[curie] = [adapter.label(descendant) for descendant in descendants if descendant]\n",
    "        except Exception as e:\n",
    "            print(f\"Error retrieving descendants for {curie}: {e}\")\n",
    "            descendant_cache[curie] = []\n",
    "\n",
    "    return {\n",
    "        'ancestors': ancestor_cache[curie],\n",
    "        'descendants': descendant_cache[curie],\n",
    "    }"
   ],
   "id": "3efcfc9fcc2fe391",
   "outputs": [],
   "execution_count": 30
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.406087Z",
     "start_time": "2024-11-14T20:35:42.400464Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def vectorize_terms(df, column):\n",
    "    \"\"\"\n",
    "    Vectorize the ancestor or descendant terms for a given column.\n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame): The input dataframe.\n",
    "        column (str): The column name to vectorize.\n",
    "\n",
    "    Returns:\n",
    "        sparse matrix: The vectorized term matrix.\n",
    "    \"\"\"\n",
    "    vectorizer = CountVectorizer()\n",
    "    return vectorizer.fit_transform(\n",
    "        df[column].apply(lambda x: ' '.join([str(term) for term in x if term is not None]) if x is not None else '')\n",
    "    )"
   ],
   "id": "e9be0a2e69ee7409",
   "outputs": [],
   "execution_count": 31
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.485702Z",
     "start_time": "2024-11-14T20:35:42.476605Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def predict_from_normalized_env_packages(df_raw, adapter):\n",
    "    # Apply the function to the relevant columns\n",
    "\n",
    "    df = df_raw.copy()\n",
    "    for column in ['env_broad_scale_id', 'env_local_scale_id', 'env_medium_id']:\n",
    "        df[f'{column}_ancestors'] = df[column].apply(lambda x: get_hierarchy_terms(x, adapter)['ancestors'])\n",
    "        df[f'{column}_descendants'] = df[column].apply(lambda x: get_hierarchy_terms(x, adapter)['descendants'])\n",
    "\n",
    "    # Vectorize each set of terms separately\n",
    "    broad_scale_ancestors = vectorize_terms(df, 'env_broad_scale_id_ancestors')\n",
    "    broad_scale_descendants = vectorize_terms(df, 'env_broad_scale_id_descendants')\n",
    "\n",
    "    local_scale_ancestors = vectorize_terms(df, 'env_local_scale_id_ancestors')\n",
    "    local_scale_descendants = vectorize_terms(df, 'env_local_scale_id_descendants')\n",
    "\n",
    "    medium_ancestors = vectorize_terms(df, 'env_medium_id_ancestors')\n",
    "    medium_descendants = vectorize_terms(df, 'env_medium_id_descendants')\n",
    "\n",
    "    # Combine all feature matrices\n",
    "    X = hstack([\n",
    "        broad_scale_ancestors,\n",
    "        broad_scale_descendants,\n",
    "        local_scale_ancestors,\n",
    "        local_scale_descendants,\n",
    "        medium_ancestors,\n",
    "        medium_descendants\n",
    "    ])\n",
    "\n",
    "    # Filter the DataFrame to only include non-null rows for the target column\n",
    "    df_filtered = df[df['normalized_env_package'].notnull() & (df['normalized_env_package'] != \"\")]\n",
    "\n",
    "    # Extract the target variable\n",
    "    y = df_filtered['normalized_env_package']\n",
    "\n",
    "    # Ensure X corresponds to the filtered rows\n",
    "    X_filtered = X[df_filtered.index]\n",
    "\n",
    "    # Split the data into training and testing sets\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X_filtered, y, test_size=0.3, random_state=42)\n",
    "\n",
    "    # Train a Random Forest Classifier\n",
    "    clf = RandomForestClassifier(n_estimators=100, random_state=42)\n",
    "    clf.fit(X_train, y_train)\n",
    "\n",
    "    # Make predictions on the test set\n",
    "    y_pred = clf.predict(X_test)\n",
    "\n",
    "    # Evaluate the model\n",
    "    print(classification_report(y_test, y_pred))\n",
    "\n",
    "    # # Predict the normalized_env_package for all rows\n",
    "    # df['predicted_normalized_env_package'] = clf.predict(X)\n",
    "\n",
    "    # # If you want to add confidence scores for each class\n",
    "    # class_probabilities = clf.predict_proba(X)\n",
    "    # \n",
    "    # # Get the class labels from the model\n",
    "    # class_labels = clf.classes_\n",
    "    # \n",
    "    # # Add a column for each class with the corresponding confidence score\n",
    "    # for i, class_label in enumerate(class_labels):\n",
    "    #     df[f'confidence_{class_label}'] = class_probabilities[:, i]\n",
    "    # \n",
    "    # return df\n",
    "\n",
    "    return clf.predict(X)"
   ],
   "id": "f03bb583dfde6d48",
   "outputs": [],
   "execution_count": 32
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.536787Z",
     "start_time": "2024-11-14T20:35:42.531464Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def parse_curie_label(text, approved_prefixes=['ENVO']):\n",
    "    # Case-insensitive pattern for matching approved prefixes followed by an ID\n",
    "    pattern = r'\\b(?:' + '|'.join(approved_prefixes) + r')\\s*[:_]\\s*(\\d+)\\b'\n",
    "    curie_match = re.search(pattern, text, re.IGNORECASE)\n",
    "    \n",
    "    if curie_match:\n",
    "        curie = f\"{approved_prefixes[0].upper()}:{curie_match.group(1)}\"  # standardize prefix to 'ENVO:ID'\n",
    "        label = re.sub(pattern, \"\", text).strip(\"[]() \")\n",
    "        # replace any colons in the label with a whitespace\n",
    "        return pd.Series([label, curie])\n",
    "    else:\n",
    "        label = re.sub(r':', ' ', text)\n",
    "        return pd.Series([label, None])  # No CURIE found, return original label and None for CURIE\n"
   ],
   "id": "9edf0d5477083503",
   "outputs": [],
   "execution_count": 33
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.585278Z",
     "start_time": "2024-11-14T20:35:42.581190Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def get_longest_annotation_curie(text, adapter):\n",
    "    annotations = adapter.annotate_text(text)\n",
    "    if not annotations:  # Check if annotations list is empty\n",
    "        return None\n",
    "    try:\n",
    "        longest_annotation = max(annotations, key=lambda x: x.subject_end - x.subject_start)\n",
    "        return longest_annotation.object_id\n",
    "    except ValueError:\n",
    "        return None  # Return None if there's an unexpected issue with finding the max\n"
   ],
   "id": "8db01eb510b460c2",
   "outputs": [],
   "execution_count": 34
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:35:42.633131Z",
     "start_time": "2024-11-14T20:35:42.628938Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Determine the filenames and target directory\n",
    "ncbi_compressed_filename = urlparse(ncbi_duckdb_url).path.split('/')[-1]\n",
    "ncbi_filename = os.path.splitext(ncbi_compressed_filename)[0]\n",
    "target_dir = os.path.join(\"..\", \"..\")  # Two levels up"
   ],
   "id": "6b934c793818f5bc",
   "outputs": [],
   "execution_count": 35
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:37:19.015440Z",
     "start_time": "2024-11-14T20:35:42.677655Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Fetch the contents from the URL and save compressed file in target directory\n",
    "ncbi_response = requests.get(ncbi_duckdb_url)\n",
    "ncbi_compressed_file_path = os.path.join(target_dir, ncbi_compressed_filename)\n",
    "with open(ncbi_compressed_file_path, \"wb\") as f:\n",
    "    f.write(ncbi_response.content)\n",
    "    \n",
    "# ~ 2 minutes @ 250 Mbps"
   ],
   "id": "8c6147ebd9dcd69d",
   "outputs": [],
   "execution_count": 36
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:38:52.606804Z",
     "start_time": "2024-11-14T20:37:19.028969Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Unzip the compressed file and save the extracted file in target directory\n",
    "ncbi_uncompressed_file_path = os.path.join(target_dir, ncbi_filename)\n",
    "with gzip.open(ncbi_compressed_file_path, \"rb\") as f_in:\n",
    "    with open(ncbi_uncompressed_file_path, \"wb\") as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)\n",
    "        \n",
    "# ~ 1 minute"
   ],
   "id": "a65971b89486ffe8",
   "outputs": [],
   "execution_count": 37
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:38:52.704858Z",
     "start_time": "2024-11-14T20:38:52.656711Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_conn = duckdb.connect(database=ncbi_uncompressed_file_path, read_only=True)",
   "id": "a3c0ec1db5a4f847",
   "outputs": [],
   "execution_count": 38
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:38:52.771276Z",
     "start_time": "2024-11-14T20:38:52.722908Z"
    }
   },
   "cell_type": "code",
   "source": "envo_adapter = get_adapter(envo_adapter_string)",
   "id": "2f5ec46c2396082c",
   "outputs": [],
   "execution_count": 39
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:38:52.937405Z",
     "start_time": "2024-11-14T20:38:52.787300Z"
    }
   },
   "cell_type": "code",
   "source": "biome_descendants = get_curie_descendants_label_dict(BIOME_CURIE, [IS_A], envo_adapter)",
   "id": "67808c8134cd1f9a",
   "outputs": [],
   "execution_count": 40
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:38:52.954814Z",
     "start_time": "2024-11-14T20:38:52.951105Z"
    }
   },
   "cell_type": "code",
   "source": "biome_descendants_lod = curie_descendants_label_dict_to_lod(biome_descendants)",
   "id": "97f53975986137a2",
   "outputs": [],
   "execution_count": 41
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:38:53.001111Z",
     "start_time": "2024-11-14T20:38:52.995536Z"
    }
   },
   "cell_type": "code",
   "source": "biome_descendants_frame = curie_descendants_label_lod_to_df(biome_descendants_lod)",
   "id": "6ff81c86ce7c4064",
   "outputs": [],
   "execution_count": 42
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:38:53.069137Z",
     "start_time": "2024-11-14T20:38:53.046115Z"
    }
   },
   "cell_type": "code",
   "source": "biome_descendants_frame",
   "id": "7d9f5e51cb21844",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "             curie                          label\n",
       "0    ENVO:01001505            alpine tundra biome\n",
       "1    ENVO:01000024           marine benthic biome\n",
       "2    ENVO:01000252          freshwater lake biome\n",
       "3    ENVO:01000180                   tundra biome\n",
       "4    ENVO:01000123       marine sponge reef biome\n",
       "..             ...                            ...\n",
       "123  ENVO:01000858         marine upwelling biome\n",
       "124  ENVO:01000188         tropical savanna biome\n",
       "125  ENVO:01000042  neritic epipelagic zone biome\n",
       "126  ENVO:01000045              epeiric sea biome\n",
       "127  ENVO:01000049        marine coral reef biome\n",
       "\n",
       "[128 rows x 2 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:01001505</td>\n",
       "      <td>alpine tundra biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>marine benthic biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:01000252</td>\n",
       "      <td>freshwater lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENVO:01000180</td>\n",
       "      <td>tundra biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:01000123</td>\n",
       "      <td>marine sponge reef biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>ENVO:01000858</td>\n",
       "      <td>marine upwelling biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>ENVO:01000188</td>\n",
       "      <td>tropical savanna biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>ENVO:01000042</td>\n",
       "      <td>neritic epipelagic zone biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>ENVO:01000045</td>\n",
       "      <td>epeiric sea biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>ENVO:01000049</td>\n",
       "      <td>marine coral reef biome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128 rows Ã— 2 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 43
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "**Use `biome_descendants_frame` as an approximation of `local/biome-ids.tsv`**",
   "id": "4c418e6f7c75da52"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:38:57.796942Z",
     "start_time": "2024-11-14T20:38:53.169081Z"
    }
   },
   "cell_type": "code",
   "source": "sv = get_schemaview_from_source(previous_submission_schema_url)",
   "id": "3f73cf88902b5ca6",
   "outputs": [],
   "execution_count": 44
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:38:57.865443Z",
     "start_time": "2024-11-14T20:38:57.857971Z"
    }
   },
   "cell_type": "code",
   "source": [
    "soil_env_broad_scale_enum = sv.get_enum(ENV_BROAD_SCALE_ENUM)\n",
    "soil_env_broad_scale_pvs_keys = list(soil_env_broad_scale_enum.permissible_values.keys())"
   ],
   "id": "c6355aee16d6a340",
   "outputs": [],
   "execution_count": 45
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:38:57.903944Z",
     "start_time": "2024-11-14T20:38:57.900178Z"
    }
   },
   "cell_type": "code",
   "source": "initially_parsed_soil_env_broad_scale_pvs = parse_hierarchically_underscored_strings(soil_env_broad_scale_pvs_keys)",
   "id": "f0631ea631f6c807",
   "outputs": [],
   "execution_count": 46
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:39:04.137244Z",
     "start_time": "2024-11-14T20:39:04.133416Z"
    }
   },
   "cell_type": "code",
   "source": "deduped_soil_env_broad_scale_pvs = dedupe_underscoreless_pvs(initially_parsed_soil_env_broad_scale_pvs)",
   "id": "4f9efbd6e1e2c834",
   "outputs": [],
   "execution_count": 47
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:39:10.647865Z",
     "start_time": "2024-11-14T20:39:10.626443Z"
    }
   },
   "cell_type": "code",
   "source": "pv_validation_results = validate_curie_label_list_dict(deduped_soil_env_broad_scale_pvs, envo_adapter, print_flag=True)",
   "id": "eb81442383577618",
   "outputs": [],
   "execution_count": 48
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:39:17.883912Z",
     "start_time": "2024-11-14T20:39:17.873448Z"
    }
   },
   "cell_type": "code",
   "source": "pv_validation_results",
   "id": "aa0d1bdc45bb3f3b",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'problems': [],\n",
       " 'valids': [{'curie': 'ENVO:01001838', 'label': 'arid biome'},\n",
       "  {'curie': 'ENVO:01001837', 'label': 'subalpine biome'},\n",
       "  {'curie': 'ENVO:01001836', 'label': 'montane biome'},\n",
       "  {'curie': 'ENVO:01000223', 'label': 'montane savanna biome'},\n",
       "  {'curie': 'ENVO:01000216', 'label': 'montane shrubland biome'},\n",
       "  {'curie': 'ENVO:01001835', 'label': 'alpine biome'},\n",
       "  {'curie': 'ENVO:01001505', 'label': 'alpine tundra biome'},\n",
       "  {'curie': 'ENVO:01001834', 'label': 'subpolar biome'},\n",
       "  {'curie': 'ENVO:01001832', 'label': 'subtropical biome'},\n",
       "  {'curie': 'ENVO:01001833', 'label': 'mediterranean biome'},\n",
       "  {'curie': 'ENVO:01000229', 'label': 'mediterranean savanna biome'},\n",
       "  {'curie': 'ENVO:01000217', 'label': 'mediterranean shrubland biome'},\n",
       "  {'curie': 'ENVO:01000208', 'label': 'mediterranean woodland biome'},\n",
       "  {'curie': 'ENVO:01000222', 'label': 'subtropical woodland biome'},\n",
       "  {'curie': 'ENVO:01000213', 'label': 'subtropical shrubland biome'},\n",
       "  {'curie': 'ENVO:01000187', 'label': 'subtropical savanna biome'},\n",
       "  {'curie': 'ENVO:01001831', 'label': 'temperate biome'},\n",
       "  {'curie': 'ENVO:01000221', 'label': 'temperate woodland biome'},\n",
       "  {'curie': 'ENVO:01000215', 'label': 'temperate shrubland biome'},\n",
       "  {'curie': 'ENVO:01000189', 'label': 'temperate savanna biome'},\n",
       "  {'curie': 'ENVO:01001830', 'label': 'tropical biome'},\n",
       "  {'curie': 'ENVO:01000220', 'label': 'tropical woodland biome'},\n",
       "  {'curie': 'ENVO:01000214', 'label': 'tropical shrubland biome'},\n",
       "  {'curie': 'ENVO:01000188', 'label': 'tropical savanna biome'},\n",
       "  {'curie': 'ENVO:01000339', 'label': 'polar biome'},\n",
       "  {'curie': 'ENVO:00000446', 'label': 'terrestrial biome'},\n",
       "  {'curie': 'ENVO:01000219', 'label': 'anthropogenic terrestrial biome'},\n",
       "  {'curie': 'ENVO:01000248', 'label': 'dense settlement biome'},\n",
       "  {'curie': 'ENVO:01000249', 'label': 'urban biome'},\n",
       "  {'curie': 'ENVO:01000247', 'label': 'rangeland biome'},\n",
       "  {'curie': 'ENVO:01000246', 'label': 'village biome'},\n",
       "  {'curie': 'ENVO:01000181', 'label': 'mangrove biome'},\n",
       "  {'curie': 'ENVO:01000180', 'label': 'tundra biome'},\n",
       "  {'curie': 'ENVO:01000176', 'label': 'shrubland biome'},\n",
       "  {'curie': 'ENVO:01001369', 'label': 'tidal mangrove shrubland'},\n",
       "  {'curie': 'ENVO:01000218', 'label': 'xeric shrubland biome'},\n",
       "  {'curie': 'ENVO:01000175', 'label': 'woodland biome'},\n",
       "  {'curie': 'ENVO:01000178', 'label': 'savanna biome'},\n",
       "  {'curie': 'ENVO:01000190', 'label': 'flooded savanna biome'}]}"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 49
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "**Use `pv_validation_results['valids']` as an approximation of `local/EnvBroadScaleSoilEnum-pvs-keys-parsed-unique.csv`**",
   "id": "cd05c55ab2a84b0d"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:40:13.978610Z",
     "start_time": "2024-11-14T20:39:24.418116Z"
    }
   },
   "cell_type": "code",
   "source": [
    "all_biosamples = get_docs_from_nmdc_collection(NMDC_RUNTIME_BASE_URL,\n",
    "                                               BIOSAMPLE_SET_COLLECTION)  # Example with stop_after"
   ],
   "id": "53d8c3995e597527",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched page 1 with 1000 documents. Total fetched: 1000\n",
      "Fetched page 2 with 1000 documents. Total fetched: 2000\n",
      "Fetched page 3 with 1000 documents. Total fetched: 3000\n",
      "Fetched page 4 with 1000 documents. Total fetched: 4000\n",
      "Fetched page 5 with 1000 documents. Total fetched: 5000\n",
      "Fetched page 6 with 1000 documents. Total fetched: 6000\n",
      "Fetched page 7 with 1000 documents. Total fetched: 7000\n",
      "Fetched page 8 with 1000 documents. Total fetched: 8000\n",
      "Fetched page 9 with 320 documents. Total fetched: 8320\n",
      "All documents fetched.\n"
     ]
    }
   ],
   "execution_count": 50
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:40:20.511223Z",
     "start_time": "2024-11-14T20:40:20.011430Z"
    }
   },
   "cell_type": "code",
   "source": "all_studies = get_docs_from_nmdc_collection(NMDC_RUNTIME_BASE_URL, STUDY_SET_COLLECTION)  # Example with stop_after",
   "id": "95a21f388c30f596",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched page 1 with 29 documents. Total fetched: 29\n",
      "All documents fetched.\n"
     ]
    }
   ],
   "execution_count": 51
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:40:27.089708Z",
     "start_time": "2024-11-14T20:40:27.084537Z"
    }
   },
   "cell_type": "code",
   "source": "env_pacakge_overrides = tsv_to_dict_of_dicts(env_package_override_file, 'id')",
   "id": "c8f6e276e7e29c15",
   "outputs": [],
   "execution_count": 52
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:40:34.704927Z",
     "start_time": "2024-11-14T20:40:34.701116Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# env_pacakge_overrides\n",
    "# todo or show as frame\n",
    "# todo include some other columns for context?"
   ],
   "id": "d3bd77c4549f2db3",
   "outputs": [],
   "execution_count": 53
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:40:50.464426Z",
     "start_time": "2024-11-14T20:40:40.510994Z"
    }
   },
   "cell_type": "code",
   "source": [
    "biosample_contexts_lod = biosamples_lod_context_extractor(all_biosamples, envo_adapter,\n",
    "                                                          env_pacakge_overrides=env_pacakge_overrides)"
   ],
   "id": "cdd0699ed7bfa22d",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overriding env_package for biosample nmdc:bsm-11-0k8nkx16 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-19v98823 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-1yvac190 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-28kgw077 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-2hswww54 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-34przm31 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-35m0rm03 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-3636w778 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3nffqc45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3nhng665 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3r4g4610 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-3tr7f795 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3xc5ft48 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3xthwx28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-42c9kj17 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-43rwvm93 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-4838sk73 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5ak8ja54 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5g2kjf44 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5j742c82 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5qq80d09 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5rgay328 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5w4ba396 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-61xfks69 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-62fxv990 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-674m9194 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-678ndj04 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-69g1rw32 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6k8cmy53 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6m5dma62 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6qgwc522 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6xa4vj59 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6y26bz06 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6ya0cs71 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6z11g163 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-73zqmh40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-766dwa91 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7b566r47 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-7cfwd964 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7pyrek52 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7yg2ne22 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7z1qvf63 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-82qw5v38 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-899bm454 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8aktkq28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8bewbp90 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8cgpqn03 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8e43d898 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8g7xbe66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-91bdea61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-94c2xj88 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-9c0hk014 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-9f2ahd94 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-a6p0kd54 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-acehgb61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-acfvxq24 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-aesqd647 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-af8kqq09 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-ahfq0n74 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-b2w6cr26 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-b7pv2a25 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-b9cjnj29 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bcns6e03 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bgar5122 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bja22k98 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bm9haf23 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bqp0m764 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-byzpm179 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-c12hyq86 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-c52npq37 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-c756vg97 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-can3rz84 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cjp6kd42 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ck98jg55 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cqbpqf62 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cv5vxb38 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cvqsze66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cxprmv88 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cz68sy30 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-d8cgz835 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-djqze645 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dppkv536 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dqnfvd06 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dtyb1w45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dva76848 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dz264p50 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-e0rz6p82 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-e747ch69 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-epzqzg45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-evfbfp31 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-f8tzvk82 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fbm8fa50 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-feynq214 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fgjpsy56 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fm4p4558 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fqchmc77 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-fy3j8w19 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fyey4107 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fzvzb345 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g35dnf26 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g3r2dm95 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g4my3419 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g4rxzh95 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-gch9pp73 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-gd31q856 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-gnbsqd30 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-gvp5zp51 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-gwx12029 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-h2k7nv93 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-h3e54f51 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-h96d6881 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ht2c1k71 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ht2v0s47 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-hv1hv533 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j1bay453 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j1hqcz15 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j2pfz414 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j5y27f16 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j6h8pw46 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j6j1p639 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jjchgw36 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jp678560 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jq2e3v76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jx6ar740 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-jxazjs92 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jy6ece20 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-k0vz1j87 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-k1twxb65 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-k9192006 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kdja3v28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kesw6757 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kmpwgz16 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-kqbbyb62 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-krcjqt76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kv966g71 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-m0h9cb45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-m4e83773 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-m4p1ck60 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-m5prkp97 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-m69k6p77 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-m9k8tx16 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ma7ry826 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-mb9s8c17 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-mbcv4b48 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n2a7p974 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n6vwc542 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n8t7zs15 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n9sx1a80 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nher1m25 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-nm187g36 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nr7vje40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nrkha596 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nwh9ez71 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nx7zwq61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-p4wss361 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-p9j7rw40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pebqj065 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pgmhxj86 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pgxc4f08 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ppp4fk41 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pq105b36 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pvr0ez88 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pygdem51 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-q3qsx507 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-q3sa0t28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-q94v7385 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qam3hk28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qe13q836 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qg6kxr61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qhq2kh40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qms4zn61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qqc8e352 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qvm0vt03 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qy2fef59 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qzncy838 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r3b0ak56 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r3tm7493 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r4kwpw90 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r7k97q10 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-rk62cv25 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-rpj4zm44 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-rzz0z710 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-s6r1q327 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-s70cv630 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sa4z4f71 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-sdjt4464 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sdjyhj66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-shjn2p67 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sjtd0y05 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-spav5m37 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sv771k17 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-syb42s66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-t7xb8h63 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ta8dt754 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tcxb3w32 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-te5fxe60 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-th3sr643 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tkwn9650 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tn8q1269 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-tnbabs47 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tpatza67 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-trbf6m66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-trpbtt49 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ttg2t449 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-tz96dg48 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-v0p16d89 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-v9k6q078 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-veffrc25 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-vme3w877 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-vp7n5531 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-vpe6j733 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-vrsw1c02 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-w1t8bv92 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wh35sj59 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wnrp0p76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wqy7mc31 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wr7h1t97 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-x0sc5x81 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-xy5ern23 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-xy8t6808 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-y31mxv90 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-y5w1j489 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-y80cxc26 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-yd4dqe75 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-yfc2at70 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-yvzvbr38 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-z4hspp43 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-z66zfk92 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-z75j2v52 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-zf5qxz83 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-zsdrmw76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-zvh2hc82 from  to plant-associated\n"
     ]
    }
   ],
   "execution_count": 54
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:40:59.292224Z",
     "start_time": "2024-11-14T20:40:59.266391Z"
    }
   },
   "cell_type": "code",
   "source": "nmdc_biosample_contexts_frame = pd.DataFrame(biosample_contexts_lod)",
   "id": "4b93dbbb4912a825",
   "outputs": [],
   "execution_count": 55
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:41:08.192911Z",
     "start_time": "2024-11-14T20:41:08.186803Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# print a value count for the normalized_env_package column\n",
    "print(\"Value counts for normalized_env_package column:\")\n",
    "print(nmdc_biosample_contexts_frame['normalized_env_package'].value_counts(dropna=False))"
   ],
   "id": "f7cc947dcc620637",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Value counts for normalized_env_package column:\n",
      "normalized_env_package\n",
      "                                                   5838\n",
      "soil                                               1665\n",
      "plant-associated                                    401\n",
      "water                                               192\n",
      "miscellaneous natural or artificial environment     140\n",
      "host-associated                                      61\n",
      "hydrocarbon resources-fluids_swabs                   23\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "execution_count": 56
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:41:23.460323Z",
     "start_time": "2024-11-14T20:41:15.166080Z"
    }
   },
   "cell_type": "code",
   "source": "package_predictions = predict_from_normalized_env_packages(nmdc_biosample_contexts_frame, envo_adapter)",
   "id": "c09274a258eaa131",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                 precision    recall  f1-score   support\n",
      "\n",
      "                                host-associated       1.00      1.00      1.00        21\n",
      "             hydrocarbon resources-fluids_swabs       1.00      0.83      0.91         6\n",
      "miscellaneous natural or artificial environment       1.00      1.00      1.00        44\n",
      "                               plant-associated       1.00      1.00      1.00       132\n",
      "                                           soil       1.00      1.00      1.00       489\n",
      "                                          water       0.98      1.00      0.99        53\n",
      "\n",
      "                                       accuracy                           1.00       745\n",
      "                                      macro avg       1.00      0.97      0.98       745\n",
      "                                   weighted avg       1.00      1.00      1.00       745\n",
      "\n"
     ]
    }
   ],
   "execution_count": 57
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:41:30.064347Z",
     "start_time": "2024-11-14T20:41:30.060292Z"
    }
   },
   "cell_type": "code",
   "source": "nmdc_biosample_contexts_frame['predicted_env_package'] = package_predictions",
   "id": "8db3619186f07681",
   "outputs": [],
   "execution_count": 58
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:41:36.933370Z",
     "start_time": "2024-11-14T20:41:36.927169Z"
    }
   },
   "cell_type": "code",
   "source": "nmdc_biosample_contexts_frame.shape",
   "id": "764a152a8c9fb9ff",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8320, 15)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 59
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:41:44.409615Z",
     "start_time": "2024-11-14T20:41:44.403009Z"
    }
   },
   "cell_type": "code",
   "source": "nmdc_soil_biosample_contexts_frame = nmdc_biosample_contexts_frame[nmdc_biosample_contexts_frame['predicted_env_package'] == 'soil']",
   "id": "f8e447afbe4734f",
   "outputs": [],
   "execution_count": 60
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:41:53.997409Z",
     "start_time": "2024-11-14T20:41:53.993371Z"
    }
   },
   "cell_type": "code",
   "source": "nmdc_soil_biosample_contexts_frame.shape",
   "id": "8c10b65e05f9df26",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5860, 15)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 61
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "**filter and count, then use `nmdc_soil_biosample_contexts_frame` as an approximation of `local/nmdc-production-biosamples-soil-env_broad_scale.tsv`**",
   "id": "38d09a3bb6bfb3c3"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:42:08.919761Z",
     "start_time": "2024-11-14T20:42:06.504354Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame = ncbi_conn.execute(ncbi_query).fetchdf()",
   "id": "5789ecb82c53fd3a",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))"
      ],
      "application/vnd.jupyter.widget-view+json": {
       "version_major": 2,
       "version_minor": 0,
       "model_id": "24c5a089dd1c486fbe92d2aebc80ded4"
      }
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "execution_count": 63
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:42:17.241040Z",
     "start_time": "2024-11-14T20:42:17.236460Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame.insert(0, 'serial_number', range(1, len(ncbi_frame) + 1))",
   "id": "dd689a5d175c9fd9",
   "outputs": [],
   "execution_count": 64
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:42:24.869191Z",
     "start_time": "2024-11-14T20:42:24.865803Z"
    }
   },
   "cell_type": "code",
   "source": "# includes env broad scale values with counts of one... useful for discovering drag-down submissions?",
   "id": "c33fd385877863f3",
   "outputs": [],
   "execution_count": 65
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:42:33.988679Z",
     "start_time": "2024-11-14T20:42:33.981191Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['content_list'] = ncbi_frame['content'].str.split('|')",
   "id": "a2492565d3934896",
   "outputs": [],
   "execution_count": 66
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:42:43.025441Z",
     "start_time": "2024-11-14T20:42:43.018675Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['content_count'] = ncbi_frame['content_list'].apply(len)",
   "id": "20aa3fa342102bef",
   "outputs": [],
   "execution_count": 67
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:42:49.003752Z",
     "start_time": "2024-11-14T20:42:48.999394Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame.shape",
   "id": "34682934e6d4b38d",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4427, 5)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 68
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:42:55.597244Z",
     "start_time": "2024-11-14T20:42:55.587044Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame = ncbi_frame.explode('content_list').reset_index(drop=True)",
   "id": "d708e65b1ef28344",
   "outputs": [],
   "execution_count": 69
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:43:02.113239Z",
     "start_time": "2024-11-14T20:43:02.107797Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame.shape",
   "id": "3feabb64849b8402",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4492, 5)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 70
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:43:10.463156Z",
     "start_time": "2024-11-14T20:43:10.460592Z"
    }
   },
   "cell_type": "code",
   "source": "# how many content_list strings contain envo multiple times now?",
   "id": "a76934c8555f4923",
   "outputs": [],
   "execution_count": 71
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:43:16.866934Z",
     "start_time": "2024-11-14T20:43:16.859133Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['envo_count'] = ncbi_frame['content_list'].str.lower().str.count(\"envo\")",
   "id": "ec3f672a6982e836",
   "outputs": [],
   "execution_count": 72
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:43:22.616129Z",
     "start_time": "2024-11-14T20:43:22.610677Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['envo_count'].value_counts()",
   "id": "927b2041141d82ad",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "envo_count\n",
       "0    3582\n",
       "1     909\n",
       "2       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 73
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "doesn't account for multiple label strings delimited with something other than '|'",
   "id": "36d4d1c696c1a5d5"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:43:29.276760Z",
     "start_time": "2024-11-14T20:43:28.597621Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame[['extracted_label', 'extracted_curie']] = ncbi_frame['content_list'].apply(parse_curie_label)",
   "id": "393d2f52a7b7155c",
   "outputs": [],
   "execution_count": 74
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:43:29.338076Z",
     "start_time": "2024-11-14T20:43:29.331670Z"
    }
   },
   "cell_type": "code",
   "source": "parse_failures = ncbi_frame[(ncbi_frame['envo_count'] > 0) & (ncbi_frame['extracted_curie'].isna() | (ncbi_frame['extracted_curie'] == ''))]\n",
   "id": "2b310df2655ee870",
   "outputs": [],
   "execution_count": 75
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:43:29.369759Z",
     "start_time": "2024-11-14T20:43:29.355686Z"
    }
   },
   "cell_type": "code",
   "source": "parse_failures",
   "id": "f0ec687bf5e3843c",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      serial_number                                            content  \\\n",
       "415             411  ENVO:Temperate grasslands, savannas, and shrub...   \n",
       "418             414  ENVO:Tropical and subtropical moist broadleaf ...   \n",
       "428             423           ENVO:subtropical coniferous forest biome   \n",
       "432             427                                         ENVO:taiga   \n",
       "433             428                        ENVO:Agricultural ecosystem   \n",
       "664             652              rhizosphere environment[ENVO01000999]   \n",
       "786             770                                      ENVOï¼š00002255   \n",
       "1024           1006  tropical biome [ENVO:01001830]|histosol [ENVO:...   \n",
       "\n",
       "      sample_count                                       content_list  \\\n",
       "415             37  ENVO:Temperate grasslands, savannas, and shrub...   \n",
       "418             36  ENVO:Tropical and subtropical moist broadleaf ...   \n",
       "428             36           ENVO:subtropical coniferous forest biome   \n",
       "432             36                                         ENVO:taiga   \n",
       "433             36                        ENVO:Agricultural ecosystem   \n",
       "664             15              rhizosphere environment[ENVO01000999]   \n",
       "786              9                                      ENVOï¼š00002255   \n",
       "1024             4                   wetland ecosystem [ENVO01001209]   \n",
       "\n",
       "      content_count  envo_count  \\\n",
       "415               1           1   \n",
       "418               1           1   \n",
       "428               1           1   \n",
       "432               1           1   \n",
       "433               1           1   \n",
       "664               1           1   \n",
       "786               1           1   \n",
       "1024              3           1   \n",
       "\n",
       "                                        extracted_label extracted_curie  \n",
       "415   ENVO Temperate grasslands, savannas, and shrub...            None  \n",
       "418   ENVO Tropical and subtropical moist broadleaf ...            None  \n",
       "428            ENVO subtropical coniferous forest biome            None  \n",
       "432                                          ENVO taiga            None  \n",
       "433                         ENVO Agricultural ecosystem            None  \n",
       "664               rhizosphere environment[ENVO01000999]            None  \n",
       "786                                       ENVOï¼š00002255            None  \n",
       "1024                   wetland ecosystem [ENVO01001209]            None  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>serial_number</th>\n",
       "      <th>content</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>content_list</th>\n",
       "      <th>content_count</th>\n",
       "      <th>envo_count</th>\n",
       "      <th>extracted_label</th>\n",
       "      <th>extracted_curie</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>411</td>\n",
       "      <td>ENVO:Temperate grasslands, savannas, and shrub...</td>\n",
       "      <td>37</td>\n",
       "      <td>ENVO:Temperate grasslands, savannas, and shrub...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO Temperate grasslands, savannas, and shrub...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>414</td>\n",
       "      <td>ENVO:Tropical and subtropical moist broadleaf ...</td>\n",
       "      <td>36</td>\n",
       "      <td>ENVO:Tropical and subtropical moist broadleaf ...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO Tropical and subtropical moist broadleaf ...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>423</td>\n",
       "      <td>ENVO:subtropical coniferous forest biome</td>\n",
       "      <td>36</td>\n",
       "      <td>ENVO:subtropical coniferous forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO subtropical coniferous forest biome</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>427</td>\n",
       "      <td>ENVO:taiga</td>\n",
       "      <td>36</td>\n",
       "      <td>ENVO:taiga</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO taiga</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>428</td>\n",
       "      <td>ENVO:Agricultural ecosystem</td>\n",
       "      <td>36</td>\n",
       "      <td>ENVO:Agricultural ecosystem</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO Agricultural ecosystem</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>664</th>\n",
       "      <td>652</td>\n",
       "      <td>rhizosphere environment[ENVO01000999]</td>\n",
       "      <td>15</td>\n",
       "      <td>rhizosphere environment[ENVO01000999]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere environment[ENVO01000999]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>786</th>\n",
       "      <td>770</td>\n",
       "      <td>ENVOï¼š00002255</td>\n",
       "      <td>9</td>\n",
       "      <td>ENVOï¼š00002255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVOï¼š00002255</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1024</th>\n",
       "      <td>1006</td>\n",
       "      <td>tropical biome [ENVO:01001830]|histosol [ENVO:...</td>\n",
       "      <td>4</td>\n",
       "      <td>wetland ecosystem [ENVO01001209]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>wetland ecosystem [ENVO01001209]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 76
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:43:39.915239Z",
     "start_time": "2024-11-14T20:43:37.678841Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['real_label'] = ncbi_frame['extracted_curie'].apply(envo_adapter.label)",
   "id": "6be66d60a992144b",
   "outputs": [],
   "execution_count": 77
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:44:08.154662Z",
     "start_time": "2024-11-14T20:43:39.966119Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Apply the function to each row in the 'label' column\n",
    "ncbi_frame['longest_annotation_curie'] = ncbi_frame['extracted_label'].apply(lambda x: get_longest_annotation_curie(x, envo_adapter))\n"
   ],
   "id": "8bc7d8d00aeab56",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Skipping statements(subject=ENVO:00000112,predicate=oio:hasDbXref,object=<http://www.eionet.europa.eu/gemet/concept/8704>,value=None,datatype=None,language=None,); ValueError: <http://www.eionet.europa.eu/gemet/concept/8704> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:00001996,predicate=oio:hasDbXref,object=<https://en.wikipedia.org/wiki/Acid_mine_drainage>,value=None,datatype=None,language=None,); ValueError: <https://en.wikipedia.org/wiki/Acid_mine_drainage> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01000225,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01000227,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001495,predicate=oio:hasDbXref,object=<https://www.goosocean.org/components/com_oe/oe.php?task=download&id=35904&version=2.0&lang=1&format=1>,value=None,datatype=None,language=None,); ValueError: <https://www.goosocean.org/components/com_oe/oe.php?task=download&id=35904&version=2.0&lang=1&format=1> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001621,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001626,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/deserts-and-xeric-shrublands>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/deserts-and-xeric-shrublands> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001627,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/ecoregions/im1304>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/ecoregions/im1304> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001628,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/ecoregions/im1303>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/ecoregions/im1303> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001644,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C021/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C021/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001644,predicate=oio:hasDbXref,object=None,value=Carbonate which is formed as the result of some biological process.,datatype=None,language=None,); ValueError: Carbonate which is formed as the result of some biological process. is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001645,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C013/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C013/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001646,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C014/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C014/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001647,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C004/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C004/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001649,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C011/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C011/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001650,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C012/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C012/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001651,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C003/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C003/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001652,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C005/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C005/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001685,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001692,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001743,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001799,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001801,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001823,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001878,predicate=oio:hasDbXref,object=<https://www.goosocean.org/index.php?option=com_oe&task=viewDocumentRecord&docID=17476>,value=None,datatype=None,language=None,); ValueError: <https://www.goosocean.org/index.php?option=com_oe&task=viewDocumentRecord&docID=17476> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001879,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001880,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001881,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:02000093,predicate=oio:hasDbXref,object=<https://en.wikipedia.org/wiki/Uranium_mining>,value=None,datatype=None,language=None,); ValueError: <https://en.wikipedia.org/wiki/Uranium_mining> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:02000094,predicate=oio:hasDbXref,object=<https://en.wikipedia.org/wiki/Ore>,value=None,datatype=None,language=None,); ValueError: <https://en.wikipedia.org/wiki/Ore> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:03501363,predicate=oio:hasDbXref,object=<https://www.conveyorsystemsltd.co.uk/news/168/126/Roller-Conveyor-vs-Belt-Conveyor.htm>,value=None,datatype=None,language=None,); ValueError: <https://www.conveyorsystemsltd.co.uk/news/168/126/Roller-Conveyor-vs-Belt-Conveyor.htm> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:06105112,predicate=oio:hasDbXref,object=<http://eurovoc.europa.eu/1837>,value=None,datatype=None,language=None,); ValueError: <http://eurovoc.europa.eu/1837> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:06105112,predicate=oio:hasDbXref,object=<http://aims.fao.org/aos/agrovoc/c_5998>,value=None,datatype=None,language=None,); ValueError: <http://aims.fao.org/aos/agrovoc/c_5998> is not a valid URI or CURIE\n",
      "WARNING:root:Skipping <http://geneontology.org/foo/applies-pattern> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <http://schema.org/image> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <http://swrl.stanford.edu/ontologies/3.3/swrla.owl#isRuleEnabled> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q2306597> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q2> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q525> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q715269> as it is not a valid CURIE\n"
     ]
    }
   ],
   "execution_count": 78
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:44:10.837829Z",
     "start_time": "2024-11-14T20:44:08.202618Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['longest_annotation_label'] = ncbi_frame['longest_annotation_curie'].apply(envo_adapter.label)",
   "id": "38e6a45ffdafc058",
   "outputs": [],
   "execution_count": 79
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:44:19.579153Z",
     "start_time": "2024-11-14T20:44:19.565596Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame",
   "id": "d393c31729bdfb34",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      serial_number                content  sample_count  \\\n",
       "0                 1                missing          9203   \n",
       "1                 2         not applicable          3869   \n",
       "2                 3                   soil          3472   \n",
       "3                 4          not collected          3464   \n",
       "4                 5           forest biome          2674   \n",
       "...             ...                    ...           ...   \n",
       "4487           4423   alpine grassland 170             1   \n",
       "4488           4424                Saltpan             1   \n",
       "4489           4425  sludge[ENVO:00002058]             1   \n",
       "4490           4426  sludge[ENVO:00002060]             1   \n",
       "4491           4427  sludge[ENVO:00002063]             1   \n",
       "\n",
       "               content_list  content_count  envo_count       extracted_label  \\\n",
       "0                   missing              1           0               missing   \n",
       "1            not applicable              1           0        not applicable   \n",
       "2                      soil              1           0                  soil   \n",
       "3             not collected              1           0         not collected   \n",
       "4              forest biome              1           0          forest biome   \n",
       "...                     ...            ...         ...                   ...   \n",
       "4487   alpine grassland 170              1           0  alpine grassland 170   \n",
       "4488                Saltpan              1           0               Saltpan   \n",
       "4489  sludge[ENVO:00002058]              1           1                sludge   \n",
       "4490  sludge[ENVO:00002060]              1           1                sludge   \n",
       "4491  sludge[ENVO:00002063]              1           1                sludge   \n",
       "\n",
       "     extracted_curie        real_label longest_annotation_curie  \\\n",
       "0               None              None                     None   \n",
       "1               None              None              CHEBI:25555   \n",
       "2               None              None            ENVO:00001998   \n",
       "3               None              None            ENVO:00000084   \n",
       "4               None              None            ENVO:01000174   \n",
       "...              ...               ...                      ...   \n",
       "4487            None              None            ENVO:00000114   \n",
       "4488            None              None              CHEBI:24866   \n",
       "4489   ENVO:00002058  secondary sludge            ENVO:00002044   \n",
       "4490   ENVO:00002060        oil sludge            ENVO:00002044   \n",
       "4491   ENVO:00002063          oil seep            ENVO:00002044   \n",
       "\n",
       "     longest_annotation_label  \n",
       "0                        None  \n",
       "1               nitrogen atom  \n",
       "2                        soil  \n",
       "3               mountain pass  \n",
       "4                forest biome  \n",
       "...                       ...  \n",
       "4487       agricultural field  \n",
       "4488                     salt  \n",
       "4489                   sludge  \n",
       "4490                   sludge  \n",
       "4491                   sludge  \n",
       "\n",
       "[4492 rows x 11 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>serial_number</th>\n",
       "      <th>content</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>content_list</th>\n",
       "      <th>content_count</th>\n",
       "      <th>envo_count</th>\n",
       "      <th>extracted_label</th>\n",
       "      <th>extracted_curie</th>\n",
       "      <th>real_label</th>\n",
       "      <th>longest_annotation_curie</th>\n",
       "      <th>longest_annotation_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>missing</td>\n",
       "      <td>9203</td>\n",
       "      <td>missing</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>missing</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>3869</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CHEBI:25555</td>\n",
       "      <td>nitrogen atom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>soil</td>\n",
       "      <td>3472</td>\n",
       "      <td>soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>soil</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>not collected</td>\n",
       "      <td>3464</td>\n",
       "      <td>not collected</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>not collected</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000084</td>\n",
       "      <td>mountain pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>forest biome</td>\n",
       "      <td>2674</td>\n",
       "      <td>forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>forest biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:01000174</td>\n",
       "      <td>forest biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4487</th>\n",
       "      <td>4423</td>\n",
       "      <td>alpine grassland 170</td>\n",
       "      <td>1</td>\n",
       "      <td>alpine grassland 170</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>alpine grassland 170</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000114</td>\n",
       "      <td>agricultural field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4488</th>\n",
       "      <td>4424</td>\n",
       "      <td>Saltpan</td>\n",
       "      <td>1</td>\n",
       "      <td>Saltpan</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Saltpan</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>CHEBI:24866</td>\n",
       "      <td>salt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4489</th>\n",
       "      <td>4425</td>\n",
       "      <td>sludge[ENVO:00002058]</td>\n",
       "      <td>1</td>\n",
       "      <td>sludge[ENVO:00002058]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>sludge</td>\n",
       "      <td>ENVO:00002058</td>\n",
       "      <td>secondary sludge</td>\n",
       "      <td>ENVO:00002044</td>\n",
       "      <td>sludge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4490</th>\n",
       "      <td>4426</td>\n",
       "      <td>sludge[ENVO:00002060]</td>\n",
       "      <td>1</td>\n",
       "      <td>sludge[ENVO:00002060]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>sludge</td>\n",
       "      <td>ENVO:00002060</td>\n",
       "      <td>oil sludge</td>\n",
       "      <td>ENVO:00002044</td>\n",
       "      <td>sludge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4491</th>\n",
       "      <td>4427</td>\n",
       "      <td>sludge[ENVO:00002063]</td>\n",
       "      <td>1</td>\n",
       "      <td>sludge[ENVO:00002063]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>sludge</td>\n",
       "      <td>ENVO:00002063</td>\n",
       "      <td>oil seep</td>\n",
       "      <td>ENVO:00002044</td>\n",
       "      <td>sludge</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4492 rows Ã— 11 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 80
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "**Use `ncbi_frame` as an approximation of `local/ncbi-mims-soil-biosamples-env_broad_scale-annotated.tsv`**",
   "id": "a34f812b53b3b945"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:46:49.733276Z",
     "start_time": "2024-11-14T20:44:26.341317Z"
    }
   },
   "cell_type": "code",
   "source": [
    "gold_biosamples_frame = pd.read_excel(gold_data_url, sheet_name=BIOSAMPLES_SHEET)\n",
    "# 2 minutes"
   ],
   "id": "a66b83e1abdbe5c",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mark/.cache/pypoetry/virtualenvs/nmdc-submission-schema-DC6HKp4p-py3.10/lib/python3.10/site-packages/openpyxl/styles/stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default\n",
      "  warn(\"Workbook contains no default style, apply openpyxl's default\")\n"
     ]
    }
   ],
   "execution_count": 81
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:46:56.431861Z",
     "start_time": "2024-11-14T20:46:56.425638Z"
    }
   },
   "cell_type": "code",
   "source": [
    "gold_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'] = gold_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'].fillna(\n",
    "    0).astype(int)\n"
   ],
   "id": "e56753c932a76fc9",
   "outputs": [],
   "execution_count": 82
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:47:04.736766Z",
     "start_time": "2024-11-14T20:47:04.734133Z"
    }
   },
   "cell_type": "code",
   "source": "# gold_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'] = gold_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'].astype(int)",
   "id": "42229a18c7aa03f3",
   "outputs": [],
   "execution_count": 83
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:47:28.374380Z",
     "start_time": "2024-11-14T20:47:28.355517Z"
    }
   },
   "cell_type": "code",
   "source": "gold_biosamples_frame",
   "id": "6a4bc761aa25612e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "       BIOSAMPLE GOLD ID                                     BIOSAMPLE NAME  \\\n",
       "0              Gb0011929  GEBA_MDM Biosample from Great Boiling Spring, ...   \n",
       "1              Gb0035601  Compost enrichment cellulose adapted microbial...   \n",
       "2              Gb0035602  Compost enrichment cellulose adapted microbial...   \n",
       "3              Gb0035635  Compost enrichment cellulose adapted microbial...   \n",
       "4              Gb0035638  Compost enrichment cellulose adapted microbial...   \n",
       "...                  ...                                                ...   \n",
       "210862         Gb0405291  Freshwater biofilm microbial communities from ...   \n",
       "210863         Gb0405292  Freshwater biofilm microbial communities from ...   \n",
       "210864         Gb0405293  Freshwater microbial communities from Lake Sug...   \n",
       "210865         Gb0405294  Freshwater microbial communities from Prairie ...   \n",
       "210866         Gb0405295  Freshwater biofilm microbial communities from ...   \n",
       "\n",
       "        BIOSAMPLE NCBI TAX ID BIOSAMPLE NCBI TAX NAME  \\\n",
       "0                    749907.0     sediment metagenome   \n",
       "1                    702656.0      compost metagenome   \n",
       "2                    702656.0      compost metagenome   \n",
       "3                    702656.0      compost metagenome   \n",
       "4                    702656.0      compost metagenome   \n",
       "...                       ...                     ...   \n",
       "210862               718308.0      biofilm metagenome   \n",
       "210863               718308.0      biofilm metagenome   \n",
       "210864               449393.0   freshwater metagenome   \n",
       "210865               449393.0   freshwater metagenome   \n",
       "210866               718308.0      biofilm metagenome   \n",
       "\n",
       "       BIOSAMPLE SAMPLE COLLECTION SITE BIOSAMPLE SAMPLE COLLECTION DATE  \\\n",
       "0                              Sediment                              NaN   \n",
       "1                           single cell                       2012-10-31   \n",
       "2                           single cell                       2012-10-31   \n",
       "3                           single cell                       2012-10-31   \n",
       "4                           single cell                       2012-10-31   \n",
       "...                                 ...                              ...   \n",
       "210862                    creek biofilm                       2023-07-26   \n",
       "210863                    creek biofilm                       2023-07-26   \n",
       "210864                       lake water                       2023-08-01   \n",
       "210865                       lake water                       2023-08-08   \n",
       "210866                    creek biofilm                       2023-07-24   \n",
       "\n",
       "                            BIOSAMPLE GEOGRAPHIC LOCATION  BIOSAMPLE LATITUDE  \\\n",
       "0                      Great Boiling Spring (GBS), Nevada           40.661433   \n",
       "1       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "2       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "3       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "4       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "...                                                   ...                 ...   \n",
       "210862   USA: Lewis Run NEON Field Site, Briggs, Virginia           39.095630   \n",
       "210863   USA: Lewis Run NEON Field Site, Briggs, Virginia           39.095630   \n",
       "210864  USA: Lake Suggs NEON Field Site, Melrose, Florida           29.688140   \n",
       "210865  USA: Prairie Lake NEON Field Site, Vashti, Nor...           47.159710   \n",
       "210866  USA: Kings Creek NEON Field Site, Manhattan, K...           39.105060   \n",
       "\n",
       "        BIOSAMPLE LONGITUDE  BIOSAMPLE ECOSYSTEM PATH ID BIOSAMPLE ECOSYSTEM  \\\n",
       "0               -119.366250                         3992       Environmental   \n",
       "1               -122.289600                         6039          Engineered   \n",
       "2               -122.289600                         6039          Engineered   \n",
       "3               -122.289600                         6039          Engineered   \n",
       "4               -122.289600                         6039          Engineered   \n",
       "...                     ...                          ...                 ...   \n",
       "210862           -77.983216                         8389       Environmental   \n",
       "210863           -77.983216                         8389       Environmental   \n",
       "210864           -82.017318                         4179       Environmental   \n",
       "210865           -99.118723                         4179       Environmental   \n",
       "210866           -96.603829                         8389       Environmental   \n",
       "\n",
       "       BIOSAMPLE ECOSYSTEM CATEGORY BIOSAMPLE ECOSYSTEM TYPE  \\\n",
       "0                           Aquatic          Thermal springs   \n",
       "1                       Solid waste              Green waste   \n",
       "2                       Solid waste              Green waste   \n",
       "3                       Solid waste              Green waste   \n",
       "4                       Solid waste              Green waste   \n",
       "...                             ...                      ...   \n",
       "210862                      Aquatic               Freshwater   \n",
       "210863                      Aquatic               Freshwater   \n",
       "210864                      Aquatic               Freshwater   \n",
       "210865                      Aquatic               Freshwater   \n",
       "210866                      Aquatic               Freshwater   \n",
       "\n",
       "       BIOSAMPLE ECOSYSTEM SUBTYPE BIOSAMPLE SPECIFIC ECOSYSTEM  \n",
       "0                     Hot (42-90C)                 Unclassified  \n",
       "1                       Composting                 Unclassified  \n",
       "2                       Composting                 Unclassified  \n",
       "3                       Composting                 Unclassified  \n",
       "4                       Composting                 Unclassified  \n",
       "...                            ...                          ...  \n",
       "210862                       Creek                      Biofilm  \n",
       "210863                       Creek                      Biofilm  \n",
       "210864                        Lake                 Unclassified  \n",
       "210865                        Lake                 Unclassified  \n",
       "210866                       Creek                      Biofilm  \n",
       "\n",
       "[210867 rows x 15 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOSAMPLE GOLD ID</th>\n",
       "      <th>BIOSAMPLE NAME</th>\n",
       "      <th>BIOSAMPLE NCBI TAX ID</th>\n",
       "      <th>BIOSAMPLE NCBI TAX NAME</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION SITE</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION DATE</th>\n",
       "      <th>BIOSAMPLE GEOGRAPHIC LOCATION</th>\n",
       "      <th>BIOSAMPLE LATITUDE</th>\n",
       "      <th>BIOSAMPLE LONGITUDE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM PATH ID</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM CATEGORY</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM TYPE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM SUBTYPE</th>\n",
       "      <th>BIOSAMPLE SPECIFIC ECOSYSTEM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gb0011929</td>\n",
       "      <td>GEBA_MDM Biosample from Great Boiling Spring, ...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Great Boiling Spring (GBS), Nevada</td>\n",
       "      <td>40.661433</td>\n",
       "      <td>-119.366250</td>\n",
       "      <td>3992</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal springs</td>\n",
       "      <td>Hot (42-90C)</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gb0035601</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gb0035602</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gb0035635</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gb0035638</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210862</th>\n",
       "      <td>Gb0405291</td>\n",
       "      <td>Freshwater biofilm microbial communities from ...</td>\n",
       "      <td>718308.0</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "      <td>creek biofilm</td>\n",
       "      <td>2023-07-26</td>\n",
       "      <td>USA: Lewis Run NEON Field Site, Briggs, Virginia</td>\n",
       "      <td>39.095630</td>\n",
       "      <td>-77.983216</td>\n",
       "      <td>8389</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Creek</td>\n",
       "      <td>Biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210863</th>\n",
       "      <td>Gb0405292</td>\n",
       "      <td>Freshwater biofilm microbial communities from ...</td>\n",
       "      <td>718308.0</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "      <td>creek biofilm</td>\n",
       "      <td>2023-07-26</td>\n",
       "      <td>USA: Lewis Run NEON Field Site, Briggs, Virginia</td>\n",
       "      <td>39.095630</td>\n",
       "      <td>-77.983216</td>\n",
       "      <td>8389</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Creek</td>\n",
       "      <td>Biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210864</th>\n",
       "      <td>Gb0405293</td>\n",
       "      <td>Freshwater microbial communities from Lake Sug...</td>\n",
       "      <td>449393.0</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>lake water</td>\n",
       "      <td>2023-08-01</td>\n",
       "      <td>USA: Lake Suggs NEON Field Site, Melrose, Florida</td>\n",
       "      <td>29.688140</td>\n",
       "      <td>-82.017318</td>\n",
       "      <td>4179</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Lake</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210865</th>\n",
       "      <td>Gb0405294</td>\n",
       "      <td>Freshwater microbial communities from Prairie ...</td>\n",
       "      <td>449393.0</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>lake water</td>\n",
       "      <td>2023-08-08</td>\n",
       "      <td>USA: Prairie Lake NEON Field Site, Vashti, Nor...</td>\n",
       "      <td>47.159710</td>\n",
       "      <td>-99.118723</td>\n",
       "      <td>4179</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Lake</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210866</th>\n",
       "      <td>Gb0405295</td>\n",
       "      <td>Freshwater biofilm microbial communities from ...</td>\n",
       "      <td>718308.0</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "      <td>creek biofilm</td>\n",
       "      <td>2023-07-24</td>\n",
       "      <td>USA: Kings Creek NEON Field Site, Manhattan, K...</td>\n",
       "      <td>39.105060</td>\n",
       "      <td>-96.603829</td>\n",
       "      <td>8389</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Creek</td>\n",
       "      <td>Biofilm</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210867 rows Ã— 15 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 86
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:47:34.867667Z",
     "start_time": "2024-11-14T20:47:34.861917Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Determine the filenames and target directory\n",
    "goldterms_compressed_filename = urlparse(goldterms_semsql_url).path.split('/')[-1]\n",
    "goldterms_filename = os.path.splitext(goldterms_compressed_filename)[0]\n",
    "target_dir = os.path.join(\"..\", \"..\")  # Two levels up\n",
    "\n",
    "# Print to confirm the filenames\n",
    "print(goldterms_filename)"
   ],
   "id": "900a1ccb2e55da85",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "goldterms.db\n"
     ]
    }
   ],
   "execution_count": 87
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:47:35.363033Z",
     "start_time": "2024-11-14T20:47:34.969046Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Fetch the contents from the URL and save compressed file in target directory\n",
    "goldterms_response = requests.get(goldterms_semsql_url)\n",
    "goldterms_compressed_file_path = os.path.join(target_dir, goldterms_compressed_filename)\n",
    "with open(goldterms_compressed_file_path, \"wb\") as f:\n",
    "    f.write(goldterms_response.content)"
   ],
   "id": "ba5fb12553bf4825",
   "outputs": [],
   "execution_count": 88
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:47:35.535739Z",
     "start_time": "2024-11-14T20:47:35.445950Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Unzip the compressed file and save the extracted file in target directory\n",
    "goldterms_uncompressed_file_path = os.path.join(target_dir, goldterms_filename)\n",
    "with gzip.open(goldterms_compressed_file_path, \"rb\") as f_in:\n",
    "    with open(goldterms_uncompressed_file_path, \"wb\") as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)"
   ],
   "id": "f60c064a587dd407",
   "outputs": [],
   "execution_count": 89
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:47:35.561172Z",
     "start_time": "2024-11-14T20:47:35.553098Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_conn = sqlite3.connect(goldterms_uncompressed_file_path)",
   "id": "1f663ead96b88e75",
   "outputs": [],
   "execution_count": 90
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:47:45.107993Z",
     "start_time": "2024-11-14T20:47:45.098017Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_soil_subjects = pd.read_sql_query(goldterms_soil_subclass_query, goldterms_conn)",
   "id": "bcedfe9ddb7254fd",
   "outputs": [],
   "execution_count": 91
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:02.207471Z",
     "start_time": "2024-11-14T20:48:02.202220Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_soil_subjects['path_id'] = goldterms_soil_subjects['subject'].str.extract(r'GOLDTERMS:(\\d+)')",
   "id": "82a240e419eb635d",
   "outputs": [],
   "execution_count": 92
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:02.235918Z",
     "start_time": "2024-11-14T20:48:02.214645Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_soil_subjects",
   "id": "8e90c8dfb6b820b4",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                              subject path_id\n",
       "0                                      GOLDTERMS:5820    5820\n",
       "1                                      GOLDTERMS:5421    5421\n",
       "2                                      GOLDTERMS:5617    5617\n",
       "3   GOLDTERMS:Environmental-Terrestrial-Soil-Natur...     NaN\n",
       "4    GOLDTERMS:Environmental-Terrestrial-Soil-Pasture     NaN\n",
       "..                                                ...     ...\n",
       "81                                     GOLDTERMS:4203    4203\n",
       "82  GOLDTERMS:Environmental-Terrestrial-Soil-Uncla...     NaN\n",
       "83                                     GOLDTERMS:5804    5804\n",
       "84                                     GOLDTERMS:4241    4241\n",
       "85                                     GOLDTERMS:4209    4209\n",
       "\n",
       "[86 rows x 2 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>path_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOLDTERMS:5820</td>\n",
       "      <td>5820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOLDTERMS:5421</td>\n",
       "      <td>5421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOLDTERMS:5617</td>\n",
       "      <td>5617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOLDTERMS:Environmental-Terrestrial-Soil-Natur...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOLDTERMS:Environmental-Terrestrial-Soil-Pasture</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>GOLDTERMS:4203</td>\n",
       "      <td>4203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>GOLDTERMS:Environmental-Terrestrial-Soil-Uncla...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>GOLDTERMS:5804</td>\n",
       "      <td>5804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>GOLDTERMS:4241</td>\n",
       "      <td>4241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>GOLDTERMS:4209</td>\n",
       "      <td>4209</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86 rows Ã— 2 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 93
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:10.282962Z",
     "start_time": "2024-11-14T20:48:10.277451Z"
    }
   },
   "cell_type": "code",
   "source": [
    "soil_path_ids = goldterms_soil_subjects['path_id'].dropna().unique().tolist()\n",
    "soil_path_ids = [int(id) for id in soil_path_ids]\n"
   ],
   "id": "9e70094c5c28b5ae",
   "outputs": [],
   "execution_count": 94
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:10.509943Z",
     "start_time": "2024-11-14T20:48:10.472086Z"
    }
   },
   "cell_type": "code",
   "source": [
    "gold_soil_biosamples_frame = gold_biosamples_frame[\n",
    "    gold_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'].isin(soil_path_ids)]\n"
   ],
   "id": "d294ca1680c31685",
   "outputs": [],
   "execution_count": 95
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:19.712985Z",
     "start_time": "2024-11-14T20:48:19.652421Z"
    }
   },
   "cell_type": "code",
   "source": "gold_soil_biosamples_frame",
   "id": "ea40acff447df533",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "       BIOSAMPLE GOLD ID                                     BIOSAMPLE NAME  \\\n",
       "11             Gb0050971  Soil ecosystem from  different sites within th...   \n",
       "12             Gb0050972  Soil ecosystem from  different sites within th...   \n",
       "13             Gb0050973  Soil ecosystem from  different sites within th...   \n",
       "14             Gb0050974  Soil ecosystem from  different sites within th...   \n",
       "56             Gb0051017  Mammuthus primigenius fossil ecosystem from Bo...   \n",
       "...                  ...                                                ...   \n",
       "210832         Gb0405261  Soil microbial communities from San Joaquin Ex...   \n",
       "210833         Gb0405262  Soil microbial communities from San Joaquin Ex...   \n",
       "210834         Gb0405263  Soil microbial communities from San Joaquin Ex...   \n",
       "210835         Gb0405264  Soil microbial communities from San Joaquin Ex...   \n",
       "210836         Gb0405265  Soil microbial communities from San Joaquin Ex...   \n",
       "\n",
       "        BIOSAMPLE NCBI TAX ID BIOSAMPLE NCBI TAX NAME  \\\n",
       "11                   410658.0         soil metagenome   \n",
       "12                   410658.0         soil metagenome   \n",
       "13                   410658.0         soil metagenome   \n",
       "14                   410658.0         soil metagenome   \n",
       "56                   444079.0       fossil metagenome   \n",
       "...                       ...                     ...   \n",
       "210832               410658.0         soil metagenome   \n",
       "210833               410658.0         soil metagenome   \n",
       "210834               410658.0         soil metagenome   \n",
       "210835               410658.0         soil metagenome   \n",
       "210836               410658.0         soil metagenome   \n",
       "\n",
       "                         BIOSAMPLE SAMPLE COLLECTION SITE  \\\n",
       "11              soil classification Mesic aquic argiudoll   \n",
       "12                   soil classification Dystric brunisol   \n",
       "13                  soil classificaiton Distrophic oxisol   \n",
       "14      soil classification euic, hyperthermic lithic ...   \n",
       "56                     Bolshaya Kolopatkaya river, Russia   \n",
       "...                                                   ...   \n",
       "210832                                               soil   \n",
       "210833                                               soil   \n",
       "210834                                               soil   \n",
       "210835                                               soil   \n",
       "210836                                               soil   \n",
       "\n",
       "       BIOSAMPLE SAMPLE COLLECTION DATE  \\\n",
       "11                                  NaN   \n",
       "12                                  NaN   \n",
       "13                                  NaN   \n",
       "14                                  NaN   \n",
       "56                                  NaN   \n",
       "...                                 ...   \n",
       "210832                       2023-02-27   \n",
       "210833                       2023-02-27   \n",
       "210834                       2023-02-27   \n",
       "210835                       2023-02-27   \n",
       "210836                       2023-02-27   \n",
       "\n",
       "                            BIOSAMPLE GEOGRAPHIC LOCATION  BIOSAMPLE LATITUDE  \\\n",
       "11      Soils collected from different sites within th...           40.104616   \n",
       "12      Soils collected from different sites within th...           52.743203   \n",
       "13      Soils collected from different sites within th...          -29.539671   \n",
       "14      Soils collected from different sites within th...           26.663199   \n",
       "56                                 Russia: Sakha Republic           70.000000   \n",
       "...                                                   ...                 ...   \n",
       "210832                    USA: Yosemite Lakes, California           37.112820   \n",
       "210833                    USA: Yosemite Lakes, California           37.112820   \n",
       "210834                    USA: Yosemite Lakes, California           37.111270   \n",
       "210835                    USA: Yosemite Lakes, California           37.111270   \n",
       "210836                    USA: Yosemite Lakes, California           37.111270   \n",
       "\n",
       "        BIOSAMPLE LONGITUDE  BIOSAMPLE ECOSYSTEM PATH ID BIOSAMPLE ECOSYSTEM  \\\n",
       "11               -88.226517                         4212       Environmental   \n",
       "12               -91.718433                         4212       Environmental   \n",
       "13               -55.107556                         4212       Environmental   \n",
       "14               -80.628500                         4212       Environmental   \n",
       "56               151.000000                         4418       Environmental   \n",
       "...                     ...                          ...                 ...   \n",
       "210832          -119.731615                         4212       Environmental   \n",
       "210833          -119.731615                         4212       Environmental   \n",
       "210834          -119.728168                         4212       Environmental   \n",
       "210835          -119.728168                         4212       Environmental   \n",
       "210836          -119.728168                         4212       Environmental   \n",
       "\n",
       "       BIOSAMPLE ECOSYSTEM CATEGORY BIOSAMPLE ECOSYSTEM TYPE  \\\n",
       "11                      Terrestrial                     Soil   \n",
       "12                      Terrestrial                     Soil   \n",
       "13                      Terrestrial                     Soil   \n",
       "14                      Terrestrial                     Soil   \n",
       "56                      Terrestrial                     Soil   \n",
       "...                             ...                      ...   \n",
       "210832                  Terrestrial                     Soil   \n",
       "210833                  Terrestrial                     Soil   \n",
       "210834                  Terrestrial                     Soil   \n",
       "210835                  Terrestrial                     Soil   \n",
       "210836                  Terrestrial                     Soil   \n",
       "\n",
       "       BIOSAMPLE ECOSYSTEM SUBTYPE BIOSAMPLE SPECIFIC ECOSYSTEM  \n",
       "11                    Unclassified                 Unclassified  \n",
       "12                    Unclassified                 Unclassified  \n",
       "13                    Unclassified                 Unclassified  \n",
       "14                    Unclassified                 Unclassified  \n",
       "56                          Fossil                 Unclassified  \n",
       "...                            ...                          ...  \n",
       "210832                Unclassified                 Unclassified  \n",
       "210833                Unclassified                 Unclassified  \n",
       "210834                Unclassified                 Unclassified  \n",
       "210835                Unclassified                 Unclassified  \n",
       "210836                Unclassified                 Unclassified  \n",
       "\n",
       "[15370 rows x 15 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOSAMPLE GOLD ID</th>\n",
       "      <th>BIOSAMPLE NAME</th>\n",
       "      <th>BIOSAMPLE NCBI TAX ID</th>\n",
       "      <th>BIOSAMPLE NCBI TAX NAME</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION SITE</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION DATE</th>\n",
       "      <th>BIOSAMPLE GEOGRAPHIC LOCATION</th>\n",
       "      <th>BIOSAMPLE LATITUDE</th>\n",
       "      <th>BIOSAMPLE LONGITUDE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM PATH ID</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM CATEGORY</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM TYPE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM SUBTYPE</th>\n",
       "      <th>BIOSAMPLE SPECIFIC ECOSYSTEM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Gb0050971</td>\n",
       "      <td>Soil ecosystem from  different sites within th...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil classification Mesic aquic argiudoll</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soils collected from different sites within th...</td>\n",
       "      <td>40.104616</td>\n",
       "      <td>-88.226517</td>\n",
       "      <td>4212</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Gb0050972</td>\n",
       "      <td>Soil ecosystem from  different sites within th...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil classification Dystric brunisol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soils collected from different sites within th...</td>\n",
       "      <td>52.743203</td>\n",
       "      <td>-91.718433</td>\n",
       "      <td>4212</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Gb0050973</td>\n",
       "      <td>Soil ecosystem from  different sites within th...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil classificaiton Distrophic oxisol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soils collected from different sites within th...</td>\n",
       "      <td>-29.539671</td>\n",
       "      <td>-55.107556</td>\n",
       "      <td>4212</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Gb0050974</td>\n",
       "      <td>Soil ecosystem from  different sites within th...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil classification euic, hyperthermic lithic ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soils collected from different sites within th...</td>\n",
       "      <td>26.663199</td>\n",
       "      <td>-80.628500</td>\n",
       "      <td>4212</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Gb0051017</td>\n",
       "      <td>Mammuthus primigenius fossil ecosystem from Bo...</td>\n",
       "      <td>444079.0</td>\n",
       "      <td>fossil metagenome</td>\n",
       "      <td>Bolshaya Kolopatkaya river, Russia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Russia: Sakha Republic</td>\n",
       "      <td>70.000000</td>\n",
       "      <td>151.000000</td>\n",
       "      <td>4418</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Fossil</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210832</th>\n",
       "      <td>Gb0405261</td>\n",
       "      <td>Soil microbial communities from San Joaquin Ex...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil</td>\n",
       "      <td>2023-02-27</td>\n",
       "      <td>USA: Yosemite Lakes, California</td>\n",
       "      <td>37.112820</td>\n",
       "      <td>-119.731615</td>\n",
       "      <td>4212</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210833</th>\n",
       "      <td>Gb0405262</td>\n",
       "      <td>Soil microbial communities from San Joaquin Ex...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil</td>\n",
       "      <td>2023-02-27</td>\n",
       "      <td>USA: Yosemite Lakes, California</td>\n",
       "      <td>37.112820</td>\n",
       "      <td>-119.731615</td>\n",
       "      <td>4212</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210834</th>\n",
       "      <td>Gb0405263</td>\n",
       "      <td>Soil microbial communities from San Joaquin Ex...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil</td>\n",
       "      <td>2023-02-27</td>\n",
       "      <td>USA: Yosemite Lakes, California</td>\n",
       "      <td>37.111270</td>\n",
       "      <td>-119.728168</td>\n",
       "      <td>4212</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210835</th>\n",
       "      <td>Gb0405264</td>\n",
       "      <td>Soil microbial communities from San Joaquin Ex...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil</td>\n",
       "      <td>2023-02-27</td>\n",
       "      <td>USA: Yosemite Lakes, California</td>\n",
       "      <td>37.111270</td>\n",
       "      <td>-119.728168</td>\n",
       "      <td>4212</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210836</th>\n",
       "      <td>Gb0405265</td>\n",
       "      <td>Soil microbial communities from San Joaquin Ex...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil</td>\n",
       "      <td>2023-02-27</td>\n",
       "      <td>USA: Yosemite Lakes, California</td>\n",
       "      <td>37.111270</td>\n",
       "      <td>-119.728168</td>\n",
       "      <td>4212</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15370 rows Ã— 15 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 96
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:19.959762Z",
     "start_time": "2024-11-14T20:48:19.936490Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_mixs_broad_frame = pd.read_sql_query(goldterms_envo_query, goldterms_conn)",
   "id": "4b4ed63d71bf1e65",
   "outputs": [],
   "execution_count": 97
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:20.569709Z",
     "start_time": "2024-11-14T20:48:20.189150Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_mixs_broad_frame['mixs_broad_label'] = goldterms_mixs_broad_frame['object'].apply(envo_adapter.label)",
   "id": "e736fe9475ea3d",
   "outputs": [],
   "execution_count": 98
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:20.584630Z",
     "start_time": "2024-11-14T20:48:20.577346Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_mixs_broad_frame['path_id'] = goldterms_mixs_broad_frame['subject'].str.extract(r'GOLDTERMS:(\\d+)')",
   "id": "ce93f9d692231ed8",
   "outputs": [],
   "execution_count": 99
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:29.471278Z",
     "start_time": "2024-11-14T20:48:29.456022Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_mixs_broad_frame",
   "id": "f0e775fd25205bcd",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                                stanza  \\\n",
       "0            GOLDTERMS:Engineered-Artificial-ecosystem   \n",
       "1    GOLDTERMS:Engineered-Artificial-ecosystem-Plan...   \n",
       "2    GOLDTERMS:Engineered-Artificial-ecosystem-Plan...   \n",
       "3    GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...   \n",
       "4          GOLDTERMS:Engineered-Bioreactor-DHS-reactor   \n",
       "..                                                 ...   \n",
       "643                                     GOLDTERMS:5841   \n",
       "644                                     GOLDTERMS:5843   \n",
       "645                                     GOLDTERMS:5846   \n",
       "646                                     GOLDTERMS:5849   \n",
       "647                                     GOLDTERMS:5850   \n",
       "\n",
       "                                               subject       predicate  \\\n",
       "0            GOLDTERMS:Engineered-Artificial-ecosystem  mixs:env_broad   \n",
       "1    GOLDTERMS:Engineered-Artificial-ecosystem-Plan...  mixs:env_broad   \n",
       "2    GOLDTERMS:Engineered-Artificial-ecosystem-Plan...  mixs:env_broad   \n",
       "3    GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...  mixs:env_broad   \n",
       "4          GOLDTERMS:Engineered-Bioreactor-DHS-reactor  mixs:env_broad   \n",
       "..                                                 ...             ...   \n",
       "643                                     GOLDTERMS:5841  mixs:env_broad   \n",
       "644                                     GOLDTERMS:5843  mixs:env_broad   \n",
       "645                                     GOLDTERMS:5846  mixs:env_broad   \n",
       "646                                     GOLDTERMS:5849  mixs:env_broad   \n",
       "647                                     GOLDTERMS:5850  mixs:env_broad   \n",
       "\n",
       "            object value datatype language graph           mixs_broad_label  \\\n",
       "0    ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "1    ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "2    ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "3    ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "4    ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "..             ...   ...      ...      ...   ...                        ...   \n",
       "643  ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "644  ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "645  ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "646  ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "647  ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "\n",
       "    path_id  \n",
       "0       NaN  \n",
       "1       NaN  \n",
       "2       NaN  \n",
       "3       NaN  \n",
       "4       NaN  \n",
       "..      ...  \n",
       "643    5841  \n",
       "644    5843  \n",
       "645    5846  \n",
       "646    5849  \n",
       "647    5850  \n",
       "\n",
       "[648 rows x 10 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stanza</th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "      <th>value</th>\n",
       "      <th>datatype</th>\n",
       "      <th>language</th>\n",
       "      <th>graph</th>\n",
       "      <th>mixs_broad_label</th>\n",
       "      <th>path_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem</td>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem-Plan...</td>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem-Plan...</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem-Plan...</td>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem-Plan...</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...</td>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-DHS-reactor</td>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-DHS-reactor</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>643</th>\n",
       "      <td>GOLDTERMS:5841</td>\n",
       "      <td>GOLDTERMS:5841</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>5841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>GOLDTERMS:5843</td>\n",
       "      <td>GOLDTERMS:5843</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>5843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>GOLDTERMS:5846</td>\n",
       "      <td>GOLDTERMS:5846</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>5846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>GOLDTERMS:5849</td>\n",
       "      <td>GOLDTERMS:5849</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>5849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>GOLDTERMS:5850</td>\n",
       "      <td>GOLDTERMS:5850</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>5850</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>648 rows Ã— 10 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 100
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:35.552197Z",
     "start_time": "2024-11-14T20:48:35.540959Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Fill NaN values in 'BIOSAMPLE ECOSYSTEM PATH ID' with 0 and convert to int\n",
    "gold_soil_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'] = gold_soil_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'].fillna(0).astype(int)\n",
    "\n",
    "# Drop rows with NaN in 'path_id' in goldterms_mixs_broad_frame\n",
    "goldterms_mixs_broad_frame = goldterms_mixs_broad_frame.dropna(subset=['path_id'])\n",
    "\n",
    "# Convert 'path_id' to int\n",
    "goldterms_mixs_broad_frame['path_id'] = goldterms_mixs_broad_frame['path_id'].astype(int)\n",
    "\n",
    "# Perform the left merge\n",
    "gold_soil_biosamples_inferred_broad = gold_soil_biosamples_frame.merge(\n",
    "    goldterms_mixs_broad_frame,\n",
    "    left_on='BIOSAMPLE ECOSYSTEM PATH ID',\n",
    "    right_on='path_id',\n",
    "    how='left'\n",
    ")\n"
   ],
   "id": "7a3f4269e0736579",
   "outputs": [],
   "execution_count": 101
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:35.637942Z",
     "start_time": "2024-11-14T20:48:35.610109Z"
    }
   },
   "cell_type": "code",
   "source": "gold_soil_biosamples_inferred_broad",
   "id": "27e616ea4cbd27ab",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      BIOSAMPLE GOLD ID                                     BIOSAMPLE NAME  \\\n",
       "0             Gb0050971  Soil ecosystem from  different sites within th...   \n",
       "1             Gb0050972  Soil ecosystem from  different sites within th...   \n",
       "2             Gb0050973  Soil ecosystem from  different sites within th...   \n",
       "3             Gb0050974  Soil ecosystem from  different sites within th...   \n",
       "4             Gb0051017  Mammuthus primigenius fossil ecosystem from Bo...   \n",
       "...                 ...                                                ...   \n",
       "15365         Gb0405261  Soil microbial communities from San Joaquin Ex...   \n",
       "15366         Gb0405262  Soil microbial communities from San Joaquin Ex...   \n",
       "15367         Gb0405263  Soil microbial communities from San Joaquin Ex...   \n",
       "15368         Gb0405264  Soil microbial communities from San Joaquin Ex...   \n",
       "15369         Gb0405265  Soil microbial communities from San Joaquin Ex...   \n",
       "\n",
       "       BIOSAMPLE NCBI TAX ID BIOSAMPLE NCBI TAX NAME  \\\n",
       "0                   410658.0         soil metagenome   \n",
       "1                   410658.0         soil metagenome   \n",
       "2                   410658.0         soil metagenome   \n",
       "3                   410658.0         soil metagenome   \n",
       "4                   444079.0       fossil metagenome   \n",
       "...                      ...                     ...   \n",
       "15365               410658.0         soil metagenome   \n",
       "15366               410658.0         soil metagenome   \n",
       "15367               410658.0         soil metagenome   \n",
       "15368               410658.0         soil metagenome   \n",
       "15369               410658.0         soil metagenome   \n",
       "\n",
       "                        BIOSAMPLE SAMPLE COLLECTION SITE  \\\n",
       "0              soil classification Mesic aquic argiudoll   \n",
       "1                   soil classification Dystric brunisol   \n",
       "2                  soil classificaiton Distrophic oxisol   \n",
       "3      soil classification euic, hyperthermic lithic ...   \n",
       "4                     Bolshaya Kolopatkaya river, Russia   \n",
       "...                                                  ...   \n",
       "15365                                               soil   \n",
       "15366                                               soil   \n",
       "15367                                               soil   \n",
       "15368                                               soil   \n",
       "15369                                               soil   \n",
       "\n",
       "      BIOSAMPLE SAMPLE COLLECTION DATE  \\\n",
       "0                                  NaN   \n",
       "1                                  NaN   \n",
       "2                                  NaN   \n",
       "3                                  NaN   \n",
       "4                                  NaN   \n",
       "...                                ...   \n",
       "15365                       2023-02-27   \n",
       "15366                       2023-02-27   \n",
       "15367                       2023-02-27   \n",
       "15368                       2023-02-27   \n",
       "15369                       2023-02-27   \n",
       "\n",
       "                           BIOSAMPLE GEOGRAPHIC LOCATION  BIOSAMPLE LATITUDE  \\\n",
       "0      Soils collected from different sites within th...           40.104616   \n",
       "1      Soils collected from different sites within th...           52.743203   \n",
       "2      Soils collected from different sites within th...          -29.539671   \n",
       "3      Soils collected from different sites within th...           26.663199   \n",
       "4                                 Russia: Sakha Republic           70.000000   \n",
       "...                                                  ...                 ...   \n",
       "15365                    USA: Yosemite Lakes, California           37.112820   \n",
       "15366                    USA: Yosemite Lakes, California           37.112820   \n",
       "15367                    USA: Yosemite Lakes, California           37.111270   \n",
       "15368                    USA: Yosemite Lakes, California           37.111270   \n",
       "15369                    USA: Yosemite Lakes, California           37.111270   \n",
       "\n",
       "       BIOSAMPLE LONGITUDE  BIOSAMPLE ECOSYSTEM PATH ID  ...          stanza  \\\n",
       "0               -88.226517                         4212  ...  GOLDTERMS:4212   \n",
       "1               -91.718433                         4212  ...  GOLDTERMS:4212   \n",
       "2               -55.107556                         4212  ...  GOLDTERMS:4212   \n",
       "3               -80.628500                         4212  ...  GOLDTERMS:4212   \n",
       "4               151.000000                         4418  ...  GOLDTERMS:4418   \n",
       "...                    ...                          ...  ...             ...   \n",
       "15365          -119.731615                         4212  ...  GOLDTERMS:4212   \n",
       "15366          -119.731615                         4212  ...  GOLDTERMS:4212   \n",
       "15367          -119.728168                         4212  ...  GOLDTERMS:4212   \n",
       "15368          -119.728168                         4212  ...  GOLDTERMS:4212   \n",
       "15369          -119.728168                         4212  ...  GOLDTERMS:4212   \n",
       "\n",
       "              subject       predicate         object value datatype language  \\\n",
       "0      GOLDTERMS:4212  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "1      GOLDTERMS:4212  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "2      GOLDTERMS:4212  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "3      GOLDTERMS:4212  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "4      GOLDTERMS:4418  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "...               ...             ...            ...   ...      ...      ...   \n",
       "15365  GOLDTERMS:4212  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "15366  GOLDTERMS:4212  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "15367  GOLDTERMS:4212  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "15368  GOLDTERMS:4212  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "15369  GOLDTERMS:4212  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "\n",
       "      graph      mixs_broad_label path_id  \n",
       "0      None  environmental system    4212  \n",
       "1      None  environmental system    4212  \n",
       "2      None  environmental system    4212  \n",
       "3      None  environmental system    4212  \n",
       "4      None  environmental system    4418  \n",
       "...     ...                   ...     ...  \n",
       "15365  None  environmental system    4212  \n",
       "15366  None  environmental system    4212  \n",
       "15367  None  environmental system    4212  \n",
       "15368  None  environmental system    4212  \n",
       "15369  None  environmental system    4212  \n",
       "\n",
       "[15370 rows x 25 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOSAMPLE GOLD ID</th>\n",
       "      <th>BIOSAMPLE NAME</th>\n",
       "      <th>BIOSAMPLE NCBI TAX ID</th>\n",
       "      <th>BIOSAMPLE NCBI TAX NAME</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION SITE</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION DATE</th>\n",
       "      <th>BIOSAMPLE GEOGRAPHIC LOCATION</th>\n",
       "      <th>BIOSAMPLE LATITUDE</th>\n",
       "      <th>BIOSAMPLE LONGITUDE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM PATH ID</th>\n",
       "      <th>...</th>\n",
       "      <th>stanza</th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "      <th>value</th>\n",
       "      <th>datatype</th>\n",
       "      <th>language</th>\n",
       "      <th>graph</th>\n",
       "      <th>mixs_broad_label</th>\n",
       "      <th>path_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gb0050971</td>\n",
       "      <td>Soil ecosystem from  different sites within th...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil classification Mesic aquic argiudoll</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soils collected from different sites within th...</td>\n",
       "      <td>40.104616</td>\n",
       "      <td>-88.226517</td>\n",
       "      <td>4212</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>4212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gb0050972</td>\n",
       "      <td>Soil ecosystem from  different sites within th...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil classification Dystric brunisol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soils collected from different sites within th...</td>\n",
       "      <td>52.743203</td>\n",
       "      <td>-91.718433</td>\n",
       "      <td>4212</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>4212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gb0050973</td>\n",
       "      <td>Soil ecosystem from  different sites within th...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil classificaiton Distrophic oxisol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soils collected from different sites within th...</td>\n",
       "      <td>-29.539671</td>\n",
       "      <td>-55.107556</td>\n",
       "      <td>4212</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>4212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gb0050974</td>\n",
       "      <td>Soil ecosystem from  different sites within th...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil classification euic, hyperthermic lithic ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Soils collected from different sites within th...</td>\n",
       "      <td>26.663199</td>\n",
       "      <td>-80.628500</td>\n",
       "      <td>4212</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>4212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gb0051017</td>\n",
       "      <td>Mammuthus primigenius fossil ecosystem from Bo...</td>\n",
       "      <td>444079.0</td>\n",
       "      <td>fossil metagenome</td>\n",
       "      <td>Bolshaya Kolopatkaya river, Russia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Russia: Sakha Republic</td>\n",
       "      <td>70.000000</td>\n",
       "      <td>151.000000</td>\n",
       "      <td>4418</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:4418</td>\n",
       "      <td>GOLDTERMS:4418</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>4418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15365</th>\n",
       "      <td>Gb0405261</td>\n",
       "      <td>Soil microbial communities from San Joaquin Ex...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil</td>\n",
       "      <td>2023-02-27</td>\n",
       "      <td>USA: Yosemite Lakes, California</td>\n",
       "      <td>37.112820</td>\n",
       "      <td>-119.731615</td>\n",
       "      <td>4212</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>4212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15366</th>\n",
       "      <td>Gb0405262</td>\n",
       "      <td>Soil microbial communities from San Joaquin Ex...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil</td>\n",
       "      <td>2023-02-27</td>\n",
       "      <td>USA: Yosemite Lakes, California</td>\n",
       "      <td>37.112820</td>\n",
       "      <td>-119.731615</td>\n",
       "      <td>4212</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>4212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15367</th>\n",
       "      <td>Gb0405263</td>\n",
       "      <td>Soil microbial communities from San Joaquin Ex...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil</td>\n",
       "      <td>2023-02-27</td>\n",
       "      <td>USA: Yosemite Lakes, California</td>\n",
       "      <td>37.111270</td>\n",
       "      <td>-119.728168</td>\n",
       "      <td>4212</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>4212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15368</th>\n",
       "      <td>Gb0405264</td>\n",
       "      <td>Soil microbial communities from San Joaquin Ex...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil</td>\n",
       "      <td>2023-02-27</td>\n",
       "      <td>USA: Yosemite Lakes, California</td>\n",
       "      <td>37.111270</td>\n",
       "      <td>-119.728168</td>\n",
       "      <td>4212</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>4212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15369</th>\n",
       "      <td>Gb0405265</td>\n",
       "      <td>Soil microbial communities from San Joaquin Ex...</td>\n",
       "      <td>410658.0</td>\n",
       "      <td>soil metagenome</td>\n",
       "      <td>soil</td>\n",
       "      <td>2023-02-27</td>\n",
       "      <td>USA: Yosemite Lakes, California</td>\n",
       "      <td>37.111270</td>\n",
       "      <td>-119.728168</td>\n",
       "      <td>4212</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>GOLDTERMS:4212</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>4212</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15370 rows Ã— 25 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 102
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "**Use `gold_soil_biosamples_inferred_broad` as an approximation of `local/goldData_biosamples-inferred-soil-env_broad_scale-counts.tsv`**",
   "id": "7e6055a76a17daf0"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:35.680595Z",
     "start_time": "2024-11-14T20:48:35.677825Z"
    }
   },
   "cell_type": "code",
   "source": "",
   "id": "559b36b44158c43a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "Can this be reconstructed in a traceable and reusable manner quickly? \n",
    "\n",
    "If so, then reuse this for generating water and sediment broad scale voting sheets.\n",
    "\n",
    "See https://github.com/microbiomedata/external-metadata-awareness/blob/main/Makefiles/soil-env_broad_scale.Makefile\n",
    "\n",
    "and all of it's dependencies\n",
    "\n",
    "**or this newer, un-merged version**\n",
    "\n",
    "```makefile\n",
    "local/soil-env-broad-scale-evidence-table.tsv: config/soil-env_broad_scale-evidence-config.yaml \\\n",
    "local/biome-ids.tsv \\\n",
    "local/EnvBroadScaleSoilEnum-pvs-keys-parsed-unique.csv \\\n",
    "local/nmdc-production-biosamples-soil-env_broad_scale.tsv \\\n",
    "local/ncbi-mims-soil-biosamples-env_broad_scale-annotated.tsv \\\n",
    "local/goldData_biosamples-inferred-soil-env_broad_scale-counts.tsv\n",
    "\t$(RUN) python external_metadata_awareness/extract_value_set_evidence.py \\\n",
    "\t\t--config $< \\\n",
    "\t\t--downsample-uncounted \\\n",
    "\t\t--output-file $@\n",
    "```\n",
    "\n"
   ],
   "id": "e98ee08462e02a05"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "```config/soil-env_broad_scale-evidence-config.yaml``` from the same unmerged branch:\n",
    "\n",
    "```yaml\n",
    "- filename: local/biome-ids.tsv # biome_descendants_frame\n",
    "  output_prefix: all_biomes_oak\n",
    "  header: false\n",
    "  data_column_number: 1\n",
    "- filename: local/EnvBroadScaleSoilEnum-pvs-keys-parsed-unique.csv # pv_validation_results['valids']\n",
    "  output_prefix: historical_permissible_values\n",
    "  header: true\n",
    "  data_column_name: normalized_curie\n",
    "- filename: local/nmdc-production-biosamples-soil-env_broad_scale.tsv # biosample_contexts_frame\n",
    "  output_prefix: NMDC_soil\n",
    "  header: false\n",
    "  data_column_number: 1\n",
    "  count_column_number: 2\n",
    "- filename: local/ncbi-mims-soil-biosamples-env_broad_scale-annotated.tsv # ncbi_frame\n",
    "  output_prefix: NCBI_mims_soil_trusting_CURIe\n",
    "  header: true\n",
    "  data_column_name: normalized_curie\n",
    "  count_column_name: count\n",
    "- filename: local/ncbi-mims-soil-biosamples-env_broad_scale-annotated.tsv\n",
    "  output_prefix: NCBI_mims_soil_trusting_labels\n",
    "  header: true\n",
    "  data_column_name: matched_id\n",
    "  count_column_name: count\n",
    "- filename: local/goldData_biosamples-inferred-soil-env_broad_scale-counts.tsv # gold_soil_biosamples_inferred_broad\n",
    "  output_prefix: GOLD_env_terr_soil\n",
    "  header: false\n",
    "  data_column_number: 1\n",
    "  count_column_number: 2\n",
    "```"
   ],
   "id": "b9aa0126f0fecc23"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:35.765597Z",
     "start_time": "2024-11-14T20:48:35.762472Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows = set()",
   "id": "74d28ee43bc2df13",
   "outputs": [],
   "execution_count": 103
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:35.898520Z",
     "start_time": "2024-11-14T20:48:35.895192Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update(biome_descendants_frame['curie'])",
   "id": "4a5858d51d55c1f3",
   "outputs": [],
   "execution_count": 104
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:35.959111Z",
     "start_time": "2024-11-14T20:48:35.954338Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update([ i['curie'] for i in pv_validation_results['valids'] ])",
   "id": "3a206d349955474b",
   "outputs": [],
   "execution_count": 105
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:42.249184Z",
     "start_time": "2024-11-14T20:48:42.241385Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update(nmdc_biosample_contexts_frame['env_broad_scale_id'])",
   "id": "6612c4f580111c69",
   "outputs": [],
   "execution_count": 106
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:48.899255Z",
     "start_time": "2024-11-14T20:48:48.895524Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update(ncbi_frame['extracted_curie'])",
   "id": "4c9a01436eb4376",
   "outputs": [],
   "execution_count": 107
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:48:57.113414Z",
     "start_time": "2024-11-14T20:48:57.108674Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update(ncbi_frame['longest_annotation_curie'])",
   "id": "79fc8410c9ad2cb1",
   "outputs": [],
   "execution_count": 108
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:49:14.756580Z",
     "start_time": "2024-11-14T20:49:14.751601Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update(gold_soil_biosamples_inferred_broad['object'])",
   "id": "3a349c2bc9dc144d",
   "outputs": [],
   "execution_count": 109
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "`include_in_rows` should not include a unique list of CURIes for any of our theoretical, legacy and empirical sources. May include NA. May include string that aren't really CURIes for a class in EnvO.\n",
    "\n",
    "Now: create a dataframe and add labels and counts or justification for their presence."
   ],
   "id": "4b229f48e264fbb3"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:49:14.773873Z",
     "start_time": "2024-11-14T20:49:14.763706Z"
    }
   },
   "cell_type": "code",
   "source": "rows_lod = []",
   "id": "8cc32502bceef8ea",
   "outputs": [],
   "execution_count": 110
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:49:22.016676Z",
     "start_time": "2024-11-14T20:49:21.007041Z"
    }
   },
   "cell_type": "code",
   "source": [
    "biome_curies = list(biome_descendants_frame['curie'])\n",
    "legacy_pv_curies = [i['curie'] for i in pv_validation_results['valids']]\n",
    "terrestrial_biome_curies = list(envo_adapter.descendants('ENVO:00000446', predicates=[IS_A]))\n",
    "aquatic_biome_curies = list(envo_adapter.descendants('ENVO:00002030', predicates=[IS_A]))\n",
    "abp_curies = list(envo_adapter.descendants('ENVO:01000813', predicates=[IS_A]))\n",
    "env_sys_curies = list(envo_adapter.descendants('ENVO:01000254', predicates=[IS_A]))\n",
    "env_mat_curies = list(envo_adapter.descendants('ENVO:00010483', predicates=[IS_A]))\n",
    "obsoletes_curies = list(envo_adapter.obsoletes())\n",
    "\n",
    "for curie in include_in_rows:\n",
    "    if curie is None:\n",
    "        continue\n",
    "    row = {\n",
    "        'curie': curie,\n",
    "        'label': envo_adapter.label(curie),\n",
    "        'envo_native': False,\n",
    "        'obsolete': False,\n",
    "        'legacy_pv': False,\n",
    "        'abp': False,\n",
    "        'env_sys': False,\n",
    "        'biome': False,\n",
    "        'terrestrial_biome': False,\n",
    "        'aquatic_biome': False,\n",
    "        'env_mat': False,\n",
    "    }\n",
    "    prefix, local_id = curie.split(':')\n",
    "    if prefix and prefix == 'ENVO' and row['label'] is not None:\n",
    "        row['envo_native'] = True\n",
    "    if curie in biome_curies:\n",
    "        row['biome'] = True\n",
    "    if curie in terrestrial_biome_curies:\n",
    "        row['terrestrial_biome'] = True\n",
    "    if curie in aquatic_biome_curies:\n",
    "        row['aquatic_biome'] = True\n",
    "    if curie in abp_curies:\n",
    "        row['abp'] = True\n",
    "    if curie in env_sys_curies:\n",
    "        row['env_sys'] = True\n",
    "    if curie in env_mat_curies:\n",
    "        row['env_mat'] = True\n",
    "    if curie in legacy_pv_curies:\n",
    "        row['legacy_pv'] = True\n",
    "    if curie in obsoletes_curies:\n",
    "        row['obsolete'] = True\n",
    "    rows_lod.append(row)\n",
    "\n",
    "# todo terrestrial biome, aquatic biome, ABP, environmental material\n",
    "    "
   ],
   "id": "2a88bc28334675b9",
   "outputs": [],
   "execution_count": 111
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:49:28.851372Z",
     "start_time": "2024-11-14T20:49:28.845167Z"
    }
   },
   "cell_type": "code",
   "source": "rows_frame = pd.DataFrame(rows_lod)",
   "id": "e3c1c8acd0d86bd8",
   "outputs": [],
   "execution_count": 112
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:49:37.311796Z",
     "start_time": "2024-11-14T20:49:37.260472Z"
    }
   },
   "cell_type": "code",
   "source": "rows_frame",
   "id": "1b6a1abbfd349989",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "             curie                           label  envo_native  obsolete  \\\n",
       "0    ENVO:00002188  obsolete powdered milk formula         True      True   \n",
       "1    ENVO:00002174                obsolete hot dog         True      True   \n",
       "2      NCBITaxon:1                            root        False     False   \n",
       "3     PATO:0001323                            area        False     False   \n",
       "4    ENVO:00002336                            None        False     False   \n",
       "..             ...                             ...          ...       ...   \n",
       "987  ENVO:00002033                         carcass         True     False   \n",
       "988  ENVO:00005764                       pond soil         True     False   \n",
       "989  ENVO:00002102                            None        False     False   \n",
       "990  ENVO:01000222      subtropical woodland biome         True     False   \n",
       "991  ENVO:00002379                            None        False     False   \n",
       "\n",
       "     legacy_pv    abp  env_sys  biome  terrestrial_biome  aquatic_biome  \\\n",
       "0        False  False    False  False              False          False   \n",
       "1        False  False    False  False              False          False   \n",
       "2        False  False    False  False              False          False   \n",
       "3        False  False    False  False              False          False   \n",
       "4        False  False    False  False              False          False   \n",
       "..         ...    ...      ...    ...                ...            ...   \n",
       "987      False  False    False  False              False          False   \n",
       "988      False   True    False  False              False          False   \n",
       "989      False  False    False  False              False          False   \n",
       "990       True   True     True   True               True          False   \n",
       "991      False  False    False  False              False          False   \n",
       "\n",
       "     env_mat  \n",
       "0      False  \n",
       "1      False  \n",
       "2      False  \n",
       "3      False  \n",
       "4      False  \n",
       "..       ...  \n",
       "987    False  \n",
       "988     True  \n",
       "989    False  \n",
       "990    False  \n",
       "991    False  \n",
       "\n",
       "[992 rows x 11 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>label</th>\n",
       "      <th>envo_native</th>\n",
       "      <th>obsolete</th>\n",
       "      <th>legacy_pv</th>\n",
       "      <th>abp</th>\n",
       "      <th>env_sys</th>\n",
       "      <th>biome</th>\n",
       "      <th>terrestrial_biome</th>\n",
       "      <th>aquatic_biome</th>\n",
       "      <th>env_mat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:00002188</td>\n",
       "      <td>obsolete powdered milk formula</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:00002174</td>\n",
       "      <td>obsolete hot dog</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:1</td>\n",
       "      <td>root</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PATO:0001323</td>\n",
       "      <td>area</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:00002336</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>ENVO:00002033</td>\n",
       "      <td>carcass</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988</th>\n",
       "      <td>ENVO:00005764</td>\n",
       "      <td>pond soil</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>ENVO:00002102</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>ENVO:01000222</td>\n",
       "      <td>subtropical woodland biome</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>ENVO:00002379</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>992 rows Ã— 11 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 113
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:49:46.225816Z",
     "start_time": "2024-11-14T20:49:46.219855Z"
    }
   },
   "cell_type": "code",
   "source": [
    "nmdc_biosample_ebs_counts = nmdc_soil_biosample_contexts_frame['env_broad_scale_id'].value_counts().reset_index()\n",
    "nmdc_biosample_ebs_counts.columns = ['curie', 'nmdc_ebs_count']\n"
   ],
   "id": "8392ef84a3254d56",
   "outputs": [],
   "execution_count": 114
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:49:46.312035Z",
     "start_time": "2024-11-14T20:49:46.299416Z"
    }
   },
   "cell_type": "code",
   "source": "nmdc_biosample_ebs_counts",
   "id": "fd4a155b6cb06235",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "            curie  nmdc_ebs_count\n",
       "0   ENVO:00000446            5400\n",
       "1   ENVO:01000174             192\n",
       "2   ENVO:01000177             113\n",
       "3   ENVO:01000250              31\n",
       "4   ENVO:01000221              27\n",
       "5   ENVO:01000219              22\n",
       "6   ENVO:01001837              18\n",
       "7   ENVO:01000249              15\n",
       "8   ENVO:01000245               9\n",
       "9   ENVO:01000215               7\n",
       "10  ENVO:01000179               6\n",
       "11  ENVO:01000216               6\n",
       "12  ENVO:01000247               5\n",
       "13  ENVO:01000189               4\n",
       "14  ENVO:01000211               3\n",
       "15  ENVO:01000183               2"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>nmdc_ebs_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>5400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:01000174</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:01000177</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENVO:01000250</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:01000221</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENVO:01000219</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENVO:01001837</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENVO:01000249</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENVO:01000245</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENVO:01000215</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENVO:01000179</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENVO:01000216</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENVO:01000247</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENVO:01000189</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENVO:01000211</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENVO:01000183</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 115
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:49:46.391058Z",
     "start_time": "2024-11-14T20:49:46.382785Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Perform the left merge\n",
    "rows_frame = rows_frame.merge(\n",
    "    nmdc_biosample_ebs_counts,\n",
    "    left_on='curie',\n",
    "    right_on='curie',\n",
    "    how='left'\n",
    ")"
   ],
   "id": "3bd7ce9b1a42f70a",
   "outputs": [],
   "execution_count": 116
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:49:46.516425Z",
     "start_time": "2024-11-14T20:49:46.487905Z"
    }
   },
   "cell_type": "code",
   "source": "rows_frame",
   "id": "cdde35e2977eaf9a",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "             curie                           label  envo_native  obsolete  \\\n",
       "0    ENVO:00002188  obsolete powdered milk formula         True      True   \n",
       "1    ENVO:00002174                obsolete hot dog         True      True   \n",
       "2      NCBITaxon:1                            root        False     False   \n",
       "3     PATO:0001323                            area        False     False   \n",
       "4    ENVO:00002336                            None        False     False   \n",
       "..             ...                             ...          ...       ...   \n",
       "987  ENVO:00002033                         carcass         True     False   \n",
       "988  ENVO:00005764                       pond soil         True     False   \n",
       "989  ENVO:00002102                            None        False     False   \n",
       "990  ENVO:01000222      subtropical woodland biome         True     False   \n",
       "991  ENVO:00002379                            None        False     False   \n",
       "\n",
       "     legacy_pv    abp  env_sys  biome  terrestrial_biome  aquatic_biome  \\\n",
       "0        False  False    False  False              False          False   \n",
       "1        False  False    False  False              False          False   \n",
       "2        False  False    False  False              False          False   \n",
       "3        False  False    False  False              False          False   \n",
       "4        False  False    False  False              False          False   \n",
       "..         ...    ...      ...    ...                ...            ...   \n",
       "987      False  False    False  False              False          False   \n",
       "988      False   True    False  False              False          False   \n",
       "989      False  False    False  False              False          False   \n",
       "990       True   True     True   True               True          False   \n",
       "991      False  False    False  False              False          False   \n",
       "\n",
       "     env_mat  nmdc_ebs_count  \n",
       "0      False             NaN  \n",
       "1      False             NaN  \n",
       "2      False             NaN  \n",
       "3      False             NaN  \n",
       "4      False             NaN  \n",
       "..       ...             ...  \n",
       "987    False             NaN  \n",
       "988     True             NaN  \n",
       "989    False             NaN  \n",
       "990    False             NaN  \n",
       "991    False             NaN  \n",
       "\n",
       "[992 rows x 12 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>label</th>\n",
       "      <th>envo_native</th>\n",
       "      <th>obsolete</th>\n",
       "      <th>legacy_pv</th>\n",
       "      <th>abp</th>\n",
       "      <th>env_sys</th>\n",
       "      <th>biome</th>\n",
       "      <th>terrestrial_biome</th>\n",
       "      <th>aquatic_biome</th>\n",
       "      <th>env_mat</th>\n",
       "      <th>nmdc_ebs_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:00002188</td>\n",
       "      <td>obsolete powdered milk formula</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:00002174</td>\n",
       "      <td>obsolete hot dog</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:1</td>\n",
       "      <td>root</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PATO:0001323</td>\n",
       "      <td>area</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:00002336</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>ENVO:00002033</td>\n",
       "      <td>carcass</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988</th>\n",
       "      <td>ENVO:00005764</td>\n",
       "      <td>pond soil</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>ENVO:00002102</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>ENVO:01000222</td>\n",
       "      <td>subtropical woodland biome</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>ENVO:00002379</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>992 rows Ã— 12 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 117
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T20:49:46.698236Z",
     "start_time": "2024-11-14T20:49:46.694317Z"
    }
   },
   "cell_type": "code",
   "source": "",
   "id": "460de8efcd09385d",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
