{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "630126a7790cc3a7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:37.647045Z",
     "start_time": "2024-11-20T15:52:34.296723Z"
    }
   },
   "outputs": [],
   "source": [
    "from common import *\n",
    "\n",
    "import gzip\n",
    "import os\n",
    "import shutil\n",
    "import sqlite3\n",
    "from urllib.parse import urlparse\n",
    "\n",
    "import duckdb\n",
    "import pandas as pd\n",
    "import requests\n",
    "from oaklib import get_adapter\n",
    "from oaklib.datamodels.vocabulary import IS_A\n",
    "from scipy.sparse import hstack\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7b0ae730412c28ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hello\n"
     ]
    }
   ],
   "source": [
    "print(\"hello\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f952b5888d2c5f0d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:37.654974Z",
     "start_time": "2024-11-20T15:52:37.651703Z"
    }
   },
   "outputs": [],
   "source": [
    "# Initialize cache dictionaries for predict_from_normalized_env_packages\n",
    "# todo how to move the definitions for function that use these globals?\n",
    "ancestor_cache = {}\n",
    "descendant_cache = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5001cbd978bbdd28",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:37.722398Z",
     "start_time": "2024-11-20T15:52:37.719788Z"
    }
   },
   "outputs": [],
   "source": [
    "# todo deal with circularity in env package prediction -> env triad reporting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ef284816f7ce9d6",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "95f7efc06f20987a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:37.769831Z",
     "start_time": "2024-11-20T15:52:37.765065Z"
    }
   },
   "outputs": [],
   "source": [
    "# ANCHOR_CURIE = 'ENVO:00000428' # biome\n",
    "ANCHOR_CURIE = 'ENVO:01000813' # astronomical body part \"abp\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "be00ed63d6472ead",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:37.817868Z",
     "start_time": "2024-11-20T15:52:37.814303Z"
    }
   },
   "outputs": [],
   "source": [
    "gold_scope = 'mixs:env_local'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eda591e14a91cd01",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:37.866949Z",
     "start_time": "2024-11-20T15:52:37.862102Z"
    }
   },
   "outputs": [],
   "source": [
    "GOLDTERMS_SOIL = 'GOLDTERMS:4212'\n",
    "GOLDTERMS_WATER = 'GOLDTERMS:3984'\n",
    "GOLDTERMS_SEDIMENT = 'GOLDTERMS:3985' #  doesn't have any subclasses\n",
    "GOLDTERMS_PLANT_ASSOCIATED = '' #  doesn't have any subclasses\n",
    "# \n",
    "# GOLDTERMS:4180, 'Environmental > Aquatic > Freshwater > Pond > Sediment' and ~64 more don't share a common root\n",
    "# poetry run runoak -i sqlite:obo:goldterms info 't~sediment'\n",
    "\n",
    "# GOLDTERMS plant associated?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bba8bf1e75430ef",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:37.914207Z",
     "start_time": "2024-11-20T15:52:37.911137Z"
    }
   },
   "outputs": [],
   "source": [
    "# GOLDTERMS_ROOT = GOLDTERMS_SOIL\n",
    "GOLDTERMS_ROOT = GOLDTERMS_PLANT_ASSOCIATED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7aa7b42c3f20ce44",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:37.960988Z",
     "start_time": "2024-11-20T15:52:37.957913Z"
    }
   },
   "outputs": [],
   "source": [
    "ncbi_scope = 'env_local_scale'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f5654d6f9319576b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:38.008705Z",
     "start_time": "2024-11-20T15:52:38.004521Z"
    }
   },
   "outputs": [],
   "source": [
    "# todo new since soil: why are we only considering MIMS.me for discovering appropriate env triad values?\n",
    "\n",
    "# NCBI_PACKAGE = 'MIMS.me.soil.6.0'\n",
    "# NCBI_PACKAGE = 'MIMS.me.water.6.0'\n",
    "# NCBI_PACKAGE = 'MIMS.me.sediment.6.0'\n",
    "NCBI_PACKAGE = 'MIMS.me.plant-associated.6.0'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ce80b2d7e64e5710",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:38.057580Z",
     "start_time": "2024-11-20T15:52:38.053004Z"
    }
   },
   "outputs": [],
   "source": [
    "nmdc_scope= 'env_local_scale_id'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5e358e42ad885188",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:38.111800Z",
     "start_time": "2024-11-20T15:52:38.108146Z"
    }
   },
   "outputs": [],
   "source": [
    "# TARGET_ENV_PACKAGE = 'soil'\n",
    "# TARGET_ENV_PACKAGE = 'water'\n",
    "# TARGET_ENV_PACKAGE = 'sediment'\n",
    "TARGET_ENV_PACKAGE = 'plant-associated'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "20750df32ebb2033",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:38.462627Z",
     "start_time": "2024-11-20T15:52:38.458879Z"
    }
   },
   "outputs": [],
   "source": [
    "# CONTEXT_ENUM = \"EnvBroadScaleSoilEnum\"\n",
    "# CONTEXT_ENUM = \"EnvBroadScaleWaterEnum\"\n",
    "# CONTEXT_ENUM = \"EnvBroadScaleSedimentEnum\"\n",
    "# CONTEXT_ENUM = \"EnvBroadScalePlantAssociatedEnum\"\n",
    "CONTEXT_ENUM = \"EnvLocalScalePlantAssociatedEnum\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a14267f45e841207",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:39.043773Z",
     "start_time": "2024-11-20T15:52:39.040435Z"
    }
   },
   "outputs": [],
   "source": [
    "output_file_name = \"plant_associated_env_local_scale_rows_frame.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adc728101c3beedb",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "914fdae0d02dad58",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:40.186056Z",
     "start_time": "2024-11-20T15:52:40.183262Z"
    }
   },
   "outputs": [],
   "source": [
    "# Approved prefixes (case-insensitive)\n",
    "approved_prefixes = ['ENVO']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e673d16992993c5a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:40.726518Z",
     "start_time": "2024-11-20T15:52:40.723459Z"
    }
   },
   "outputs": [],
   "source": [
    "MIN_ANNOTATION_LEN = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a1e9e99c5ac41ef4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:41.263005Z",
     "start_time": "2024-11-20T15:52:41.259884Z"
    }
   },
   "outputs": [],
   "source": [
    "previous_submission_schema_url = \"https://raw.githubusercontent.com/microbiomedata/submission-schema/v10.7.0/src/nmdc_submission_schema/schema/nmdc_submission_schema.yaml\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cde9a57dca22a7c1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:41.754301Z",
     "start_time": "2024-11-20T15:52:41.750557Z"
    }
   },
   "outputs": [],
   "source": [
    "NMDC_RUNTIME_BASE_URL = 'https://api.microbiomedata.org/nmdcschema/'\n",
    "STUDY_SET_COLLECTION = 'study_set'\n",
    "BIOSAMPLE_SET_COLLECTION = 'biosample_set'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8240c0de4cbcec24",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:42.231013Z",
     "start_time": "2024-11-20T15:52:42.227806Z"
    }
   },
   "outputs": [],
   "source": [
    "envo_adapter_string = \"sqlite:obo:envo\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "eb27b996bb24645a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:42.721214Z",
     "start_time": "2024-11-20T15:52:42.718287Z"
    }
   },
   "outputs": [],
   "source": [
    "# goldterms_adapter_string = \"sqlite:obo:envo\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "375bac6b84ab6198",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:43.199252Z",
     "start_time": "2024-11-20T15:52:43.196336Z"
    }
   },
   "outputs": [],
   "source": [
    "env_package_override_file = '../../mam-env-package-overrides.tsv'\n",
    "override_column = 'mam_inferred_env_package'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8a8be3a9e6123df5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:43.697234Z",
     "start_time": "2024-11-20T15:52:43.694461Z"
    }
   },
   "outputs": [],
   "source": [
    "ncbi_duckdb_url = 'https://portal.nersc.gov/project/m3408/biosamples_duckdb/ncbi_biosamples_2024-09-23.duckdb.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fb9c800afcc86146",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:44.187774Z",
     "start_time": "2024-11-20T15:52:44.184786Z"
    }
   },
   "outputs": [],
   "source": [
    "gold_data_url = \"https://gold.jgi.doe.gov/download?mode=site_excel\"\n",
    "BIOSAMPLES_SHEET = \"Biosample\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2ae64976b48e359",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:44.731242Z",
     "start_time": "2024-11-20T15:52:44.727800Z"
    }
   },
   "outputs": [],
   "source": [
    "goldterms_semsql_url = \"https://s3.amazonaws.com/bbop-sqlite/goldterms.db.gz\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6b672d93e1f5c53",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:45.323791Z",
     "start_time": "2024-11-20T15:52:45.320479Z"
    }
   },
   "outputs": [],
   "source": [
    "goldterms_subclass_query = f\"\"\"\n",
    "select\n",
    "\tsubject\n",
    "from\n",
    "\tentailed_edge ee\n",
    "where\n",
    "\tpredicate = 'rdfs:subClassOf'\n",
    "\tand object = '{GOLDTERMS_ROOT}'\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ae751e002f65be71",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:45.964172Z",
     "start_time": "2024-11-20T15:52:45.961141Z"
    }
   },
   "outputs": [],
   "source": [
    "# todo could this have been done with a OAK query, eliminating the need to explicitly download the file?\n",
    "\n",
    "goldterms_envo_query = f\"\"\"\n",
    "SELECT\n",
    "\t*\n",
    "FROM\n",
    "\tstatements s\n",
    "WHERE\n",
    "\tpredicate = '{gold_scope}'\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d33879ec4c6ca1df",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:47.119750Z",
     "start_time": "2024-11-20T15:52:47.116357Z"
    }
   },
   "outputs": [],
   "source": [
    "ncbi_query = f\"\"\"\n",
    "SELECT content, COUNT(1) AS sample_count \n",
    "FROM attributes \n",
    "WHERE harmonized_name = '{ncbi_scope}' AND package_content = '{NCBI_PACKAGE}'\n",
    "GROUP BY content\n",
    "ORDER BY COUNT(1) DESC\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f03bb583dfde6d48",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:47.860607Z",
     "start_time": "2024-11-20T15:52:47.848428Z"
    }
   },
   "outputs": [],
   "source": [
    "def predict_from_normalized_env_packages(df_raw, adapter):\n",
    "    # Apply the function to the relevant columns\n",
    "\n",
    "    df = df_raw.copy()\n",
    "\n",
    "    print(df.shape)\n",
    "    for column in ['env_broad_scale_id', 'env_local_scale_id', 'env_medium_id']:\n",
    "        df[f'{column}_ancestors'] = df[column].apply(lambda x: get_hierarchy_terms(x, adapter)['ancestors'])\n",
    "        df[f'{column}_descendants'] = df[column].apply(lambda x: get_hierarchy_terms(x, adapter)['descendants'])\n",
    "\n",
    "    # Vectorize each set of terms separately\n",
    "    broad_scale_ancestors = vectorize_terms(df, 'env_broad_scale_id_ancestors')\n",
    "    broad_scale_descendants = vectorize_terms(df, 'env_broad_scale_id_descendants')\n",
    "\n",
    "    local_scale_ancestors = vectorize_terms(df, 'env_local_scale_id_ancestors')\n",
    "    local_scale_descendants = vectorize_terms(df, 'env_local_scale_id_descendants')\n",
    "\n",
    "    medium_ancestors = vectorize_terms(df, 'env_medium_id_ancestors')\n",
    "    medium_descendants = vectorize_terms(df, 'env_medium_id_descendants')\n",
    "\n",
    "    # Combine all feature matrices\n",
    "    X = hstack([\n",
    "        broad_scale_ancestors,\n",
    "        broad_scale_descendants,\n",
    "        local_scale_ancestors,\n",
    "        local_scale_descendants,\n",
    "        medium_ancestors,\n",
    "        medium_descendants\n",
    "    ])\n",
    "\n",
    "    # Filter the DataFrame to only include non-null rows for the target column\n",
    "    df_filtered = df[df['normalized_env_package'].notnull() & (df['normalized_env_package'] != \"\")]\n",
    "\n",
    "    # Extract the target variable\n",
    "    y = df_filtered['normalized_env_package']\n",
    "\n",
    "    # Ensure X corresponds to the filtered rows\n",
    "    X_filtered = X[df_filtered.index]\n",
    "\n",
    "    # Split the data into training and testing sets\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X_filtered, y, test_size=0.3, random_state=42)\n",
    "\n",
    "    # Train a Random Forest Classifier\n",
    "    clf = RandomForestClassifier(n_estimators=100, random_state=42)\n",
    "    clf.fit(X_train, y_train)\n",
    "\n",
    "    # Make predictions on the test set\n",
    "    y_pred = clf.predict(X_test)\n",
    "\n",
    "    # Evaluate the model\n",
    "    print(classification_report(y_test, y_pred))\n",
    "\n",
    "    # not determining confidence for each class nor saving any diagnostics any more\n",
    "\n",
    "    return clf.predict(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2cc9b8ce20346804",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:52:48.474049Z",
     "start_time": "2024-11-20T15:52:48.468257Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_hierarchy_terms(my_curie: str, adapter) -> dict:\n",
    "    \"\"\"\n",
    "    Extract ancestor and descendant terms from the ontology for a given CURIE,\n",
    "    using caching to improve performance and filtering by 'is_a' relationships.\n",
    "\n",
    "    Args:\n",
    "        my_curie (str): CURIE identifier for the ontology term.\n",
    "        adapter: Ontology adapter.\n",
    "\n",
    "    Returns:\n",
    "        dict: Dictionary containing lists of ancestor and descendant terms.\n",
    "    \"\"\"\n",
    "    if my_curie not in ancestor_cache:\n",
    "        try:\n",
    "            ancestors = list(adapter.ancestors(my_curie, predicates=[IS_A]))\n",
    "            ancestor_cache[my_curie] = [adapter.label(ancestor) for ancestor in ancestors if ancestor]\n",
    "        except Exception as my_e:\n",
    "            print(f\"Error retrieving ancestors for {my_curie}: {my_e}\")\n",
    "            ancestor_cache[my_curie] = []\n",
    "\n",
    "    if my_curie not in descendant_cache:\n",
    "        try:\n",
    "            descendants = list(adapter.descendants(my_curie, predicates=[IS_A]))\n",
    "            descendant_cache[my_curie] = [adapter.label(descendant) for descendant in descendants if descendant]\n",
    "        except Exception as my_e:\n",
    "            print(f\"Error retrieving descendants for {my_curie}: {my_e}\")\n",
    "            descendant_cache[my_curie] = []\n",
    "\n",
    "    return {\n",
    "        'ancestors': ancestor_cache[my_curie],\n",
    "        'descendants': descendant_cache[my_curie],\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6b934c793818f5bc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:54:05.245002Z",
     "start_time": "2024-11-20T15:54:05.240893Z"
    }
   },
   "outputs": [],
   "source": [
    "# Determine the filenames and target directory\n",
    "ncbi_compressed_filename = urlparse(ncbi_duckdb_url).path.split('/')[-1]\n",
    "ncbi_filename = os.path.splitext(ncbi_compressed_filename)[0]\n",
    "target_dir = os.path.join(\"..\", \"..\")  # Two levels up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b10963f1795ca86a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:54:05.868950Z",
     "start_time": "2024-11-20T15:54:05.865978Z"
    }
   },
   "outputs": [],
   "source": [
    "# Fetch the contents from the URL and save compressed file in target directory\n",
    "ncbi_compressed_file_path = os.path.join(target_dir, ncbi_compressed_filename)\n",
    "\n",
    "# ncbi_response = requests.get(ncbi_duckdb_url)\n",
    "# with open(ncbi_compressed_file_path, \"wb\") as f:\n",
    "#     f.write(ncbi_response.content)\n",
    "\n",
    "# ~ 2 minutes @ 250 Mbps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a65971b89486ffe8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:54:06.755126Z",
     "start_time": "2024-11-20T15:54:06.752030Z"
    }
   },
   "outputs": [],
   "source": [
    "# Unzip the compressed file and save the extracted file in target directory\n",
    "ncbi_uncompressed_file_path = os.path.join(target_dir, ncbi_filename)\n",
    "# with gzip.open(ncbi_compressed_file_path, \"rb\") as f_in:\n",
    "#     with open(ncbi_uncompressed_file_path, \"wb\") as f_out:\n",
    "#         shutil.copyfileobj(f_in, f_out)\n",
    "#\n",
    "# # ~ 2 minutes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a3c0ec1db5a4f847",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:54:07.955470Z",
     "start_time": "2024-11-20T15:54:07.909804Z"
    }
   },
   "outputs": [],
   "source": [
    "ncbi_conn = duckdb.connect(database=ncbi_uncompressed_file_path, read_only=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2f5ec46c2396082c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:54:09.899779Z",
     "start_time": "2024-11-20T15:54:09.867278Z"
    }
   },
   "outputs": [],
   "source": [
    "envo_adapter = get_adapter(envo_adapter_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "67808c8134cd1f9a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:54:12.206564Z",
     "start_time": "2024-11-20T15:54:11.554451Z"
    }
   },
   "outputs": [],
   "source": [
    "anchor_descendants = get_curie_descendants_label_dict(ANCHOR_CURIE, [IS_A], envo_adapter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "97f53975986137a2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:54:38.955684Z",
     "start_time": "2024-11-20T15:54:38.951427Z"
    }
   },
   "outputs": [],
   "source": [
    "anchor_descendants_lod = curie_descendants_label_dict_to_lod(anchor_descendants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6ff81c86ce7c4064",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:54:39.901272Z",
     "start_time": "2024-11-20T15:54:39.896836Z"
    }
   },
   "outputs": [],
   "source": [
    "anchor_descendants_frame = curie_descendants_label_lod_to_df(anchor_descendants_lod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "7d9f5e51cb21844",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-20T15:55:09.269125Z",
     "start_time": "2024-11-20T15:55:09.248141Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:00001999</td>\n",
       "      <td>marine water body</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:01000188</td>\n",
       "      <td>tropical savanna biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:00000487</td>\n",
       "      <td>paternoster lake</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENVO:01000860</td>\n",
       "      <td>temperate marine upwelling biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:01000199</td>\n",
       "      <td>mediterranean forest biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1731</th>\n",
       "      <td>ENVO:01000429</td>\n",
       "      <td>burrow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1732</th>\n",
       "      <td>ENVO:01000431</td>\n",
       "      <td>mixed forest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733</th>\n",
       "      <td>ENVO:01000536</td>\n",
       "      <td>factory</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1734</th>\n",
       "      <td>ENVO:00000873</td>\n",
       "      <td>freshwater biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1735</th>\n",
       "      <td>ENVO:03500002</td>\n",
       "      <td>public park</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1736 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              curie                             label\n",
       "0     ENVO:00001999                 marine water body\n",
       "1     ENVO:01000188            tropical savanna biome\n",
       "2     ENVO:00000487                  paternoster lake\n",
       "3     ENVO:01000860  temperate marine upwelling biome\n",
       "4     ENVO:01000199        mediterranean forest biome\n",
       "...             ...                               ...\n",
       "1731  ENVO:01000429                            burrow\n",
       "1732  ENVO:01000431                      mixed forest\n",
       "1733  ENVO:01000536                           factory\n",
       "1734  ENVO:00000873                  freshwater biome\n",
       "1735  ENVO:03500002                       public park\n",
       "\n",
       "[1736 rows x 2 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anchor_descendants_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c418e6f7c75da52",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3f73cf88902b5ca6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv = get_schemaview_from_source(previous_submission_schema_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c6355aee16d6a340",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An AttributeError occurred: 'NoneType' object has no attribute 'permissible_values'\n"
     ]
    }
   ],
   "source": [
    "# todo break out slow steps into its own cell\n",
    "\n",
    "try:\n",
    "    CONTEXT_ENUM_def = sv.get_enum(CONTEXT_ENUM)\n",
    "    context_pvs_keys = list(CONTEXT_ENUM_def.permissible_values.keys())\n",
    "except AttributeError as e:\n",
    "    # Handle the AttributeError\n",
    "    print(f\"An AttributeError occurred: {e}\")\n",
    "    context_pvs_keys =[]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "f0631ea631f6c807",
   "metadata": {},
   "outputs": [],
   "source": [
    "initially_parsed_context_pvs = parse_hierarchically_underscored_strings(context_pvs_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "4f9efbd6e1e2c834",
   "metadata": {},
   "outputs": [],
   "source": [
    "deduped_context_pvs = dedupe_underscoreless_pvs(initially_parsed_context_pvs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "eb81442383577618",
   "metadata": {},
   "outputs": [],
   "source": [
    "pv_validation_results = validate_curie_label_list_dict(deduped_context_pvs, envo_adapter, print_flag=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "aa0d1bdc45bb3f3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'problems': [], 'valids': []}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pv_validation_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd05c55ab2a84b0d",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "53d8c3995e597527",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched page 1 with 1000 documents. Total fetched: 1000\n",
      "Fetched page 2 with 1000 documents. Total fetched: 2000\n",
      "Fetched page 3 with 1000 documents. Total fetched: 3000\n",
      "Fetched page 4 with 1000 documents. Total fetched: 4000\n",
      "Fetched page 5 with 1000 documents. Total fetched: 5000\n",
      "Fetched page 6 with 1000 documents. Total fetched: 6000\n",
      "Fetched page 7 with 1000 documents. Total fetched: 7000\n",
      "Fetched page 8 with 1000 documents. Total fetched: 8000\n",
      "Fetched page 9 with 320 documents. Total fetched: 8320\n",
      "All documents fetched.\n"
     ]
    }
   ],
   "source": [
    "# todo rename to all_nmdc_samples etc\n",
    "all_nmdc_biosamples = get_docs_from_nmdc_collection(NMDC_RUNTIME_BASE_URL,\n",
    "                                               BIOSAMPLE_SET_COLLECTION)  # Example with stop_after"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "95a21f388c30f596",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # todo I don't think we're actually using this\n",
    "# all_studies = get_docs_from_nmdc_collection(NMDC_RUNTIME_BASE_URL, STUDY_SET_COLLECTION)  # Example with stop_after"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c8f6e276e7e29c15",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_pacakge_overrides = tsv_to_dict_of_dicts(env_package_override_file, 'id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d3bd77c4549f2db3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# env_pacakge_overrides\n",
    "# todo or show as frame\n",
    "# todo include some other columns for context?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "cdd0699ed7bfa22d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overriding env_package for biosample nmdc:bsm-11-0k8nkx16 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-19v98823 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-1yvac190 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-28kgw077 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-2hswww54 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-34przm31 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-35m0rm03 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-3636w778 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3nffqc45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3nhng665 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3r4g4610 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-3tr7f795 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3xc5ft48 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3xthwx28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-42c9kj17 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-43rwvm93 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-4838sk73 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5ak8ja54 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5g2kjf44 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5j742c82 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5qq80d09 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5rgay328 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5w4ba396 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-61xfks69 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-62fxv990 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-674m9194 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-678ndj04 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-69g1rw32 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6k8cmy53 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6m5dma62 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6qgwc522 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6xa4vj59 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6y26bz06 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6ya0cs71 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6z11g163 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-73zqmh40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-766dwa91 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7b566r47 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-7cfwd964 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7pyrek52 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7yg2ne22 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7z1qvf63 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-82qw5v38 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-899bm454 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8aktkq28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8bewbp90 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8cgpqn03 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8e43d898 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8g7xbe66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-91bdea61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-94c2xj88 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-9c0hk014 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-9f2ahd94 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-a6p0kd54 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-acehgb61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-acfvxq24 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-aesqd647 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-af8kqq09 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-ahfq0n74 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-b2w6cr26 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-b7pv2a25 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-b9cjnj29 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bcns6e03 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bgar5122 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bja22k98 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bm9haf23 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bqp0m764 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-byzpm179 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-c12hyq86 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-c52npq37 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-c756vg97 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-can3rz84 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cjp6kd42 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ck98jg55 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cqbpqf62 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cv5vxb38 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cvqsze66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cxprmv88 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cz68sy30 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-d8cgz835 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-djqze645 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dppkv536 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dqnfvd06 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dtyb1w45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dva76848 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dz264p50 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-e0rz6p82 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-e747ch69 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-epzqzg45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-evfbfp31 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-f8tzvk82 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fbm8fa50 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-feynq214 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fgjpsy56 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fm4p4558 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fqchmc77 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-fy3j8w19 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fyey4107 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fzvzb345 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g35dnf26 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g3r2dm95 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g4my3419 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g4rxzh95 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-gch9pp73 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-gd31q856 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-gnbsqd30 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-gvp5zp51 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-gwx12029 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-h2k7nv93 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-h3e54f51 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-h96d6881 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ht2c1k71 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ht2v0s47 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-hv1hv533 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j1bay453 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j1hqcz15 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j2pfz414 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j5y27f16 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j6h8pw46 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j6j1p639 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jjchgw36 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jp678560 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jq2e3v76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jx6ar740 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-jxazjs92 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jy6ece20 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-k0vz1j87 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-k1twxb65 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-k9192006 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kdja3v28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kesw6757 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kmpwgz16 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-kqbbyb62 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-krcjqt76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kv966g71 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-m0h9cb45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-m4e83773 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-m4p1ck60 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-m5prkp97 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-m69k6p77 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-m9k8tx16 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ma7ry826 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-mb9s8c17 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-mbcv4b48 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n2a7p974 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n6vwc542 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n8t7zs15 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n9sx1a80 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nher1m25 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-nm187g36 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nr7vje40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nrkha596 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nwh9ez71 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nx7zwq61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-p4wss361 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-p9j7rw40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pebqj065 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pgmhxj86 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pgxc4f08 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ppp4fk41 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pq105b36 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pvr0ez88 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pygdem51 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-q3qsx507 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-q3sa0t28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-q94v7385 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qam3hk28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qe13q836 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qg6kxr61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qhq2kh40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qms4zn61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qqc8e352 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qvm0vt03 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qy2fef59 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qzncy838 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r3b0ak56 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r3tm7493 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r4kwpw90 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r7k97q10 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-rk62cv25 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-rpj4zm44 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-rzz0z710 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-s6r1q327 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-s70cv630 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sa4z4f71 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-sdjt4464 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sdjyhj66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-shjn2p67 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sjtd0y05 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-spav5m37 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sv771k17 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-syb42s66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-t7xb8h63 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ta8dt754 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tcxb3w32 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-te5fxe60 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-th3sr643 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tkwn9650 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tn8q1269 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-tnbabs47 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tpatza67 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-trbf6m66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-trpbtt49 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ttg2t449 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-tz96dg48 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-v0p16d89 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-v9k6q078 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-veffrc25 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-vme3w877 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-vp7n5531 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-vpe6j733 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-vrsw1c02 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-w1t8bv92 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wh35sj59 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wnrp0p76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wqy7mc31 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wr7h1t97 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-x0sc5x81 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-xy5ern23 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-xy8t6808 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-y31mxv90 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-y5w1j489 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-y80cxc26 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-yd4dqe75 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-yfc2at70 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-yvzvbr38 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-z4hspp43 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-z66zfk92 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-z75j2v52 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-zf5qxz83 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-zsdrmw76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-zvh2hc82 from  to plant-associated\n"
     ]
    }
   ],
   "source": [
    "biosample_contexts_lod = biosamples_lod_context_extractor(all_nmdc_biosamples, envo_adapter,\n",
    "                                                          my_env_pacakge_overrides=env_pacakge_overrides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4b93dbbb4912a825",
   "metadata": {},
   "outputs": [],
   "source": [
    "nmdc_biosample_contexts_frame = pd.DataFrame(biosample_contexts_lod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "f7cc947dcc620637",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Value counts for normalized_env_package column:\n",
      "normalized_env_package\n",
      "                                                   5838\n",
      "soil                                               1665\n",
      "plant-associated                                    401\n",
      "water                                               192\n",
      "miscellaneous natural or artificial environment     140\n",
      "host-associated                                      61\n",
      "hydrocarbon resources-fluids_swabs                   23\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# print a value count for the normalized_env_package column\n",
    "print(\"Value counts for normalized_env_package column:\")\n",
    "print(nmdc_biosample_contexts_frame['normalized_env_package'].value_counts(dropna=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "c09274a258eaa131",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8320, 14)\n",
      "                                                 precision    recall  f1-score   support\n",
      "\n",
      "                                host-associated       1.00      1.00      1.00        21\n",
      "             hydrocarbon resources-fluids_swabs       1.00      0.83      0.91         6\n",
      "miscellaneous natural or artificial environment       1.00      1.00      1.00        44\n",
      "                               plant-associated       1.00      1.00      1.00       132\n",
      "                                           soil       1.00      1.00      1.00       489\n",
      "                                          water       0.98      1.00      0.99        53\n",
      "\n",
      "                                       accuracy                           1.00       745\n",
      "                                      macro avg       1.00      0.97      0.98       745\n",
      "                                   weighted avg       1.00      1.00      1.00       745\n",
      "\n"
     ]
    }
   ],
   "source": [
    "package_predictions = predict_from_normalized_env_packages(nmdc_biosample_contexts_frame, envo_adapter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "8db3619186f07681",
   "metadata": {},
   "outputs": [],
   "source": [
    "nmdc_biosample_contexts_frame['predicted_env_package'] = package_predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "764a152a8c9fb9ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8320, 15)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nmdc_biosample_contexts_frame.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f8e447afbe4734f",
   "metadata": {},
   "outputs": [],
   "source": [
    "nmdc_biosample_contexts_frame = nmdc_biosample_contexts_frame[\n",
    "    nmdc_biosample_contexts_frame['predicted_env_package'] == TARGET_ENV_PACKAGE]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "8c10b65e05f9df26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(401, 15)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nmdc_biosample_contexts_frame.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38d09a3bb6bfb3c3",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "5789ecb82c53fd3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame = ncbi_conn.execute(ncbi_query).fetchdf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "dd689a5d175c9fd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame.insert(0, 'serial_number', range(1, len(ncbi_frame) + 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c33fd385877863f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes values with counts of one... useful for discovering drag-down submissions?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "a2492565d3934896",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame['content_list'] = ncbi_frame['content'].str.split('|')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "20aa3fa342102bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame['content_count'] = ncbi_frame['content_list'].apply(len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "34682934e6d4b38d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1955, 5)"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "d708e65b1ef28344",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame = ncbi_frame.explode('content_list').reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "3feabb64849b8402",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1984, 5)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "a76934c8555f4923",
   "metadata": {},
   "outputs": [],
   "source": [
    "# how many content_list strings contain envo multiple times now?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "ec3f672a6982e836",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame['envo_count'] = ncbi_frame['content_list'].str.lower().str.count(\"envo\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "927b2041141d82ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "envo_count\n",
       "0    1742\n",
       "1     227\n",
       "3      14\n",
       "2       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame['envo_count'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36d4d1c696c1a5d5",
   "metadata": {},
   "source": [
    "doesn't account for multiple label strings delimited with something other than '|'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "393d2f52a7b7155c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame[['extracted_label', 'extracted_curie']] = ncbi_frame['content_list'].apply(parse_curie_label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "2b310df2655ee870",
   "metadata": {},
   "outputs": [],
   "source": [
    "parse_failures = ncbi_frame[\n",
    "    (ncbi_frame['envo_count'] > 0) & (ncbi_frame['extracted_curie'].isna() | (ncbi_frame['extracted_curie'] == ''))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "f0ec687bf5e3843c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>serial_number</th>\n",
       "      <th>content</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>content_list</th>\n",
       "      <th>content_count</th>\n",
       "      <th>envo_count</th>\n",
       "      <th>extracted_label</th>\n",
       "      <th>extracted_curie</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>63</td>\n",
       "      <td>Tropical woodland biome [ENVO01000220]</td>\n",
       "      <td>223</td>\n",
       "      <td>Tropical woodland biome [ENVO01000220]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Tropical woodland biome [ENVO01000220]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>153</td>\n",
       "      <td>Tropical shrubland biome [ENVO01000214]</td>\n",
       "      <td>77</td>\n",
       "      <td>Tropical shrubland biome [ENVO01000214]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Tropical shrubland biome [ENVO01000214]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     serial_number                                  content  sample_count  \\\n",
       "63              63   Tropical woodland biome [ENVO01000220]           223   \n",
       "157            153  Tropical shrubland biome [ENVO01000214]            77   \n",
       "\n",
       "                                content_list  content_count  envo_count  \\\n",
       "63    Tropical woodland biome [ENVO01000220]              1           1   \n",
       "157  Tropical shrubland biome [ENVO01000214]              1           1   \n",
       "\n",
       "                             extracted_label extracted_curie  \n",
       "63    Tropical woodland biome [ENVO01000220]            None  \n",
       "157  Tropical shrubland biome [ENVO01000214]            None  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_failures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "6be66d60a992144b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame['real_label'] = ncbi_frame['extracted_curie'].apply(envo_adapter.label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "8bc7d8d00aeab56",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Skipping statements(subject=ENVO:00000112,predicate=oio:hasDbXref,object=<http://www.eionet.europa.eu/gemet/concept/8704>,value=None,datatype=None,language=None,); ValueError: <http://www.eionet.europa.eu/gemet/concept/8704> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:00001996,predicate=oio:hasDbXref,object=<https://en.wikipedia.org/wiki/Acid_mine_drainage>,value=None,datatype=None,language=None,); ValueError: <https://en.wikipedia.org/wiki/Acid_mine_drainage> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01000225,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01000227,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001495,predicate=oio:hasDbXref,object=<https://www.goosocean.org/components/com_oe/oe.php?task=download&id=35904&version=2.0&lang=1&format=1>,value=None,datatype=None,language=None,); ValueError: <https://www.goosocean.org/components/com_oe/oe.php?task=download&id=35904&version=2.0&lang=1&format=1> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001621,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001626,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/deserts-and-xeric-shrublands>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/deserts-and-xeric-shrublands> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001627,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/ecoregions/im1304>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/ecoregions/im1304> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001628,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/ecoregions/im1303>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/ecoregions/im1303> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001644,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C021/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C021/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001644,predicate=oio:hasDbXref,object=None,value=Carbonate which is formed as the result of some biological process.,datatype=None,language=None,); ValueError: Carbonate which is formed as the result of some biological process. is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001645,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C013/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C013/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001646,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C014/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C014/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001647,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C004/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C004/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001649,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C011/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C011/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001650,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C012/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C012/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001651,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C003/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C003/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001652,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C005/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C005/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001685,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001692,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001743,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001799,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001801,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001823,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001878,predicate=oio:hasDbXref,object=<https://www.goosocean.org/index.php?option=com_oe&task=viewDocumentRecord&docID=17476>,value=None,datatype=None,language=None,); ValueError: <https://www.goosocean.org/index.php?option=com_oe&task=viewDocumentRecord&docID=17476> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001879,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001880,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001881,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:02000093,predicate=oio:hasDbXref,object=<https://en.wikipedia.org/wiki/Uranium_mining>,value=None,datatype=None,language=None,); ValueError: <https://en.wikipedia.org/wiki/Uranium_mining> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:02000094,predicate=oio:hasDbXref,object=<https://en.wikipedia.org/wiki/Ore>,value=None,datatype=None,language=None,); ValueError: <https://en.wikipedia.org/wiki/Ore> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:03501363,predicate=oio:hasDbXref,object=<https://www.conveyorsystemsltd.co.uk/news/168/126/Roller-Conveyor-vs-Belt-Conveyor.htm>,value=None,datatype=None,language=None,); ValueError: <https://www.conveyorsystemsltd.co.uk/news/168/126/Roller-Conveyor-vs-Belt-Conveyor.htm> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:06105112,predicate=oio:hasDbXref,object=<http://eurovoc.europa.eu/1837>,value=None,datatype=None,language=None,); ValueError: <http://eurovoc.europa.eu/1837> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:06105112,predicate=oio:hasDbXref,object=<http://aims.fao.org/aos/agrovoc/c_5998>,value=None,datatype=None,language=None,); ValueError: <http://aims.fao.org/aos/agrovoc/c_5998> is not a valid URI or CURIE\n",
      "WARNING:root:Skipping <http://geneontology.org/foo/applies-pattern> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <http://schema.org/image> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <http://swrl.stanford.edu/ontologies/3.3/swrla.owl#isRuleEnabled> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q2306597> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q2> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q525> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q715269> as it is not a valid CURIE\n"
     ]
    }
   ],
   "source": [
    "# Apply the function to each row in the 'label' column\n",
    "ncbi_frame['longest_annotation_curie'] = ncbi_frame['extracted_label'].apply(\n",
    "    lambda x: get_longest_annotation_curie(x, envo_adapter, MIN_ANNOTATION_LEN))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "38e6a45ffdafc058",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame['longest_annotation_label'] = ncbi_frame['longest_annotation_curie'].apply(envo_adapter.label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "d393c31729bdfb34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>serial_number</th>\n",
       "      <th>content</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>content_list</th>\n",
       "      <th>content_count</th>\n",
       "      <th>envo_count</th>\n",
       "      <th>extracted_label</th>\n",
       "      <th>extracted_curie</th>\n",
       "      <th>real_label</th>\n",
       "      <th>longest_annotation_curie</th>\n",
       "      <th>longest_annotation_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>woodland</td>\n",
       "      <td>6385</td>\n",
       "      <td>woodland</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>woodland</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000057</td>\n",
       "      <td>mangrove swamp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>rhizosphere environment [ENVO_01000999]</td>\n",
       "      <td>5325</td>\n",
       "      <td>rhizosphere environment [ENVO_01000999]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere environment</td>\n",
       "      <td>ENVO:01000999</td>\n",
       "      <td>rhizosphere environment</td>\n",
       "      <td>ENVO:01000999</td>\n",
       "      <td>rhizosphere environment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>2924</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>laboratory environment [ENVO:01001405]</td>\n",
       "      <td>2683</td>\n",
       "      <td>laboratory environment [ENVO:01001405]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>laboratory environment</td>\n",
       "      <td>ENVO:01001405</td>\n",
       "      <td>laboratory environment</td>\n",
       "      <td>ENVO:01001405</td>\n",
       "      <td>laboratory environment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>forest</td>\n",
       "      <td>1768</td>\n",
       "      <td>forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>forest</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000111</td>\n",
       "      <td>forested area</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979</th>\n",
       "      <td>1951</td>\n",
       "      <td>Vineyard-SMDRTd1</td>\n",
       "      <td>1</td>\n",
       "      <td>Vineyard-SMDRTd1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Vineyard-SMDRTd1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1980</th>\n",
       "      <td>1952</td>\n",
       "      <td>Vineyard-VARTc2</td>\n",
       "      <td>1</td>\n",
       "      <td>Vineyard-VARTc2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Vineyard-VARTc2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1981</th>\n",
       "      <td>1953</td>\n",
       "      <td>Vineyard-VVRTc3</td>\n",
       "      <td>1</td>\n",
       "      <td>Vineyard-VVRTc3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Vineyard-VVRTc3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1982</th>\n",
       "      <td>1954</td>\n",
       "      <td>Vineyard-VVRTe1</td>\n",
       "      <td>1</td>\n",
       "      <td>Vineyard-VVRTe1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Vineyard-VVRTe1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1983</th>\n",
       "      <td>1955</td>\n",
       "      <td>Female Maize Seed</td>\n",
       "      <td>1</td>\n",
       "      <td>Female Maize Seed</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Female Maize Seed</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1984 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      serial_number                                  content  sample_count  \\\n",
       "0                 1                                 woodland          6385   \n",
       "1                 2  rhizosphere environment [ENVO_01000999]          5325   \n",
       "2                 3                           not applicable          2924   \n",
       "3                 4   laboratory environment [ENVO:01001405]          2683   \n",
       "4                 5                                   forest          1768   \n",
       "...             ...                                      ...           ...   \n",
       "1979           1951                         Vineyard-SMDRTd1             1   \n",
       "1980           1952                          Vineyard-VARTc2             1   \n",
       "1981           1953                          Vineyard-VVRTc3             1   \n",
       "1982           1954                          Vineyard-VVRTe1             1   \n",
       "1983           1955                        Female Maize Seed             1   \n",
       "\n",
       "                                 content_list  content_count  envo_count  \\\n",
       "0                                    woodland              1           0   \n",
       "1     rhizosphere environment [ENVO_01000999]              1           1   \n",
       "2                              not applicable              1           0   \n",
       "3      laboratory environment [ENVO:01001405]              1           1   \n",
       "4                                      forest              1           0   \n",
       "...                                       ...            ...         ...   \n",
       "1979                         Vineyard-SMDRTd1              1           0   \n",
       "1980                          Vineyard-VARTc2              1           0   \n",
       "1981                          Vineyard-VVRTc3              1           0   \n",
       "1982                          Vineyard-VVRTe1              1           0   \n",
       "1983                        Female Maize Seed              1           0   \n",
       "\n",
       "              extracted_label extracted_curie               real_label  \\\n",
       "0                    woodland            None                     None   \n",
       "1     rhizosphere environment   ENVO:01000999  rhizosphere environment   \n",
       "2              not applicable            None                     None   \n",
       "3      laboratory environment   ENVO:01001405   laboratory environment   \n",
       "4                      forest            None                     None   \n",
       "...                       ...             ...                      ...   \n",
       "1979         Vineyard-SMDRTd1            None                     None   \n",
       "1980          Vineyard-VARTc2            None                     None   \n",
       "1981          Vineyard-VVRTc3            None                     None   \n",
       "1982          Vineyard-VVRTe1            None                     None   \n",
       "1983        Female Maize Seed            None                     None   \n",
       "\n",
       "     longest_annotation_curie longest_annotation_label  \n",
       "0               ENVO:00000057           mangrove swamp  \n",
       "1               ENVO:01000999  rhizosphere environment  \n",
       "2                        None                     None  \n",
       "3               ENVO:01001405   laboratory environment  \n",
       "4               ENVO:00000111            forested area  \n",
       "...                       ...                      ...  \n",
       "1979            ENVO:00000116                 vineyard  \n",
       "1980            ENVO:00000116                 vineyard  \n",
       "1981            ENVO:00000116                 vineyard  \n",
       "1982            ENVO:00000116                 vineyard  \n",
       "1983                     None                     None  \n",
       "\n",
       "[1984 rows x 11 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a34f812b53b3b945",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "a66b83e1abdbe5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mark/.cache/pypoetry/virtualenvs/nmdc-submission-schema-DC6HKp4p-py3.10/lib/python3.10/site-packages/openpyxl/styles/stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default\n",
      "  warn(\"Workbook contains no default style, apply openpyxl's default\")\n"
     ]
    }
   ],
   "source": [
    "gold_biosamples_frame = pd.read_excel(gold_data_url, sheet_name=BIOSAMPLES_SHEET)\n",
    "# 2 minutes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "e56753c932a76fc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gold_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'] = gold_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'].fillna(\n",
    "    0).astype(int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "6a4bc761aa25612e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOSAMPLE GOLD ID</th>\n",
       "      <th>BIOSAMPLE NAME</th>\n",
       "      <th>BIOSAMPLE NCBI TAX ID</th>\n",
       "      <th>BIOSAMPLE NCBI TAX NAME</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION SITE</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION DATE</th>\n",
       "      <th>BIOSAMPLE GEOGRAPHIC LOCATION</th>\n",
       "      <th>BIOSAMPLE LATITUDE</th>\n",
       "      <th>BIOSAMPLE LONGITUDE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM PATH ID</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM CATEGORY</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM TYPE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM SUBTYPE</th>\n",
       "      <th>BIOSAMPLE SPECIFIC ECOSYSTEM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gb0011929</td>\n",
       "      <td>GEBA_MDM Biosample from Great Boiling Spring, ...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Great Boiling Spring (GBS), Nevada</td>\n",
       "      <td>40.661433</td>\n",
       "      <td>-119.366250</td>\n",
       "      <td>3992</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal springs</td>\n",
       "      <td>Hot (42-90C)</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gb0035601</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gb0035602</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gb0035635</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gb0035638</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210932</th>\n",
       "      <td>Gb0405291</td>\n",
       "      <td>Freshwater biofilm microbial communities from ...</td>\n",
       "      <td>718308.0</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "      <td>creek biofilm</td>\n",
       "      <td>2023-07-26</td>\n",
       "      <td>USA: Lewis Run NEON Field Site, Briggs, Virginia</td>\n",
       "      <td>39.095630</td>\n",
       "      <td>-77.983216</td>\n",
       "      <td>8389</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Creek</td>\n",
       "      <td>Biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210933</th>\n",
       "      <td>Gb0405292</td>\n",
       "      <td>Freshwater biofilm microbial communities from ...</td>\n",
       "      <td>718308.0</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "      <td>creek biofilm</td>\n",
       "      <td>2023-07-26</td>\n",
       "      <td>USA: Lewis Run NEON Field Site, Briggs, Virginia</td>\n",
       "      <td>39.095630</td>\n",
       "      <td>-77.983216</td>\n",
       "      <td>8389</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Creek</td>\n",
       "      <td>Biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210934</th>\n",
       "      <td>Gb0405293</td>\n",
       "      <td>Freshwater microbial communities from Lake Sug...</td>\n",
       "      <td>449393.0</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>lake water</td>\n",
       "      <td>2023-08-01</td>\n",
       "      <td>USA: Lake Suggs NEON Field Site, Melrose, Florida</td>\n",
       "      <td>29.688140</td>\n",
       "      <td>-82.017318</td>\n",
       "      <td>4179</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Lake</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210935</th>\n",
       "      <td>Gb0405294</td>\n",
       "      <td>Freshwater microbial communities from Prairie ...</td>\n",
       "      <td>449393.0</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>lake water</td>\n",
       "      <td>2023-08-08</td>\n",
       "      <td>USA: Prairie Lake NEON Field Site, Vashti, Nor...</td>\n",
       "      <td>47.159710</td>\n",
       "      <td>-99.118723</td>\n",
       "      <td>4179</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Lake</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210936</th>\n",
       "      <td>Gb0405295</td>\n",
       "      <td>Freshwater biofilm microbial communities from ...</td>\n",
       "      <td>718308.0</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "      <td>creek biofilm</td>\n",
       "      <td>2023-07-24</td>\n",
       "      <td>USA: Kings Creek NEON Field Site, Manhattan, K...</td>\n",
       "      <td>39.105060</td>\n",
       "      <td>-96.603829</td>\n",
       "      <td>8389</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Creek</td>\n",
       "      <td>Biofilm</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210937 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BIOSAMPLE GOLD ID                                     BIOSAMPLE NAME  \\\n",
       "0              Gb0011929  GEBA_MDM Biosample from Great Boiling Spring, ...   \n",
       "1              Gb0035601  Compost enrichment cellulose adapted microbial...   \n",
       "2              Gb0035602  Compost enrichment cellulose adapted microbial...   \n",
       "3              Gb0035635  Compost enrichment cellulose adapted microbial...   \n",
       "4              Gb0035638  Compost enrichment cellulose adapted microbial...   \n",
       "...                  ...                                                ...   \n",
       "210932         Gb0405291  Freshwater biofilm microbial communities from ...   \n",
       "210933         Gb0405292  Freshwater biofilm microbial communities from ...   \n",
       "210934         Gb0405293  Freshwater microbial communities from Lake Sug...   \n",
       "210935         Gb0405294  Freshwater microbial communities from Prairie ...   \n",
       "210936         Gb0405295  Freshwater biofilm microbial communities from ...   \n",
       "\n",
       "        BIOSAMPLE NCBI TAX ID BIOSAMPLE NCBI TAX NAME  \\\n",
       "0                    749907.0     sediment metagenome   \n",
       "1                    702656.0      compost metagenome   \n",
       "2                    702656.0      compost metagenome   \n",
       "3                    702656.0      compost metagenome   \n",
       "4                    702656.0      compost metagenome   \n",
       "...                       ...                     ...   \n",
       "210932               718308.0      biofilm metagenome   \n",
       "210933               718308.0      biofilm metagenome   \n",
       "210934               449393.0   freshwater metagenome   \n",
       "210935               449393.0   freshwater metagenome   \n",
       "210936               718308.0      biofilm metagenome   \n",
       "\n",
       "       BIOSAMPLE SAMPLE COLLECTION SITE BIOSAMPLE SAMPLE COLLECTION DATE  \\\n",
       "0                              Sediment                              NaN   \n",
       "1                           single cell                       2012-10-31   \n",
       "2                           single cell                       2012-10-31   \n",
       "3                           single cell                       2012-10-31   \n",
       "4                           single cell                       2012-10-31   \n",
       "...                                 ...                              ...   \n",
       "210932                    creek biofilm                       2023-07-26   \n",
       "210933                    creek biofilm                       2023-07-26   \n",
       "210934                       lake water                       2023-08-01   \n",
       "210935                       lake water                       2023-08-08   \n",
       "210936                    creek biofilm                       2023-07-24   \n",
       "\n",
       "                            BIOSAMPLE GEOGRAPHIC LOCATION  BIOSAMPLE LATITUDE  \\\n",
       "0                      Great Boiling Spring (GBS), Nevada           40.661433   \n",
       "1       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "2       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "3       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "4       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "...                                                   ...                 ...   \n",
       "210932   USA: Lewis Run NEON Field Site, Briggs, Virginia           39.095630   \n",
       "210933   USA: Lewis Run NEON Field Site, Briggs, Virginia           39.095630   \n",
       "210934  USA: Lake Suggs NEON Field Site, Melrose, Florida           29.688140   \n",
       "210935  USA: Prairie Lake NEON Field Site, Vashti, Nor...           47.159710   \n",
       "210936  USA: Kings Creek NEON Field Site, Manhattan, K...           39.105060   \n",
       "\n",
       "        BIOSAMPLE LONGITUDE  BIOSAMPLE ECOSYSTEM PATH ID BIOSAMPLE ECOSYSTEM  \\\n",
       "0               -119.366250                         3992       Environmental   \n",
       "1               -122.289600                         6039          Engineered   \n",
       "2               -122.289600                         6039          Engineered   \n",
       "3               -122.289600                         6039          Engineered   \n",
       "4               -122.289600                         6039          Engineered   \n",
       "...                     ...                          ...                 ...   \n",
       "210932           -77.983216                         8389       Environmental   \n",
       "210933           -77.983216                         8389       Environmental   \n",
       "210934           -82.017318                         4179       Environmental   \n",
       "210935           -99.118723                         4179       Environmental   \n",
       "210936           -96.603829                         8389       Environmental   \n",
       "\n",
       "       BIOSAMPLE ECOSYSTEM CATEGORY BIOSAMPLE ECOSYSTEM TYPE  \\\n",
       "0                           Aquatic          Thermal springs   \n",
       "1                       Solid waste              Green waste   \n",
       "2                       Solid waste              Green waste   \n",
       "3                       Solid waste              Green waste   \n",
       "4                       Solid waste              Green waste   \n",
       "...                             ...                      ...   \n",
       "210932                      Aquatic               Freshwater   \n",
       "210933                      Aquatic               Freshwater   \n",
       "210934                      Aquatic               Freshwater   \n",
       "210935                      Aquatic               Freshwater   \n",
       "210936                      Aquatic               Freshwater   \n",
       "\n",
       "       BIOSAMPLE ECOSYSTEM SUBTYPE BIOSAMPLE SPECIFIC ECOSYSTEM  \n",
       "0                     Hot (42-90C)                 Unclassified  \n",
       "1                       Composting                 Unclassified  \n",
       "2                       Composting                 Unclassified  \n",
       "3                       Composting                 Unclassified  \n",
       "4                       Composting                 Unclassified  \n",
       "...                            ...                          ...  \n",
       "210932                       Creek                      Biofilm  \n",
       "210933                       Creek                      Biofilm  \n",
       "210934                        Lake                 Unclassified  \n",
       "210935                        Lake                 Unclassified  \n",
       "210936                       Creek                      Biofilm  \n",
       "\n",
       "[210937 rows x 15 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gold_biosamples_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "900a1ccb2e55da85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "goldterms.db\n"
     ]
    }
   ],
   "source": [
    "# Determine the filenames and target directory\n",
    "goldterms_compressed_filename = urlparse(goldterms_semsql_url).path.split('/')[-1]\n",
    "goldterms_filename = os.path.splitext(goldterms_compressed_filename)[0]\n",
    "target_dir = os.path.join(\"..\", \"..\")  # Two levels up\n",
    "\n",
    "# Print to confirm the filenames\n",
    "print(goldterms_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "ba5fb12553bf4825",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch the contents from the URL and save compressed file in target directory\n",
    "goldterms_response = requests.get(goldterms_semsql_url)\n",
    "goldterms_compressed_file_path = os.path.join(target_dir, goldterms_compressed_filename)\n",
    "with open(goldterms_compressed_file_path, \"wb\") as f:\n",
    "    f.write(goldterms_response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "f60c064a587dd407",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unzip the compressed file and save the extracted file in target directory\n",
    "goldterms_uncompressed_file_path = os.path.join(target_dir, goldterms_filename)\n",
    "with gzip.open(goldterms_compressed_file_path, \"rb\") as f_in:\n",
    "    with open(goldterms_uncompressed_file_path, \"wb\") as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "1f663ead96b88e75",
   "metadata": {},
   "outputs": [],
   "source": [
    "goldterms_conn = sqlite3.connect(goldterms_uncompressed_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "bcedfe9ddb7254fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "goldterms_subjects = pd.read_sql_query(goldterms_subclass_query, goldterms_conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "82a240e419eb635d",
   "metadata": {},
   "outputs": [],
   "source": [
    "goldterms_subjects['path_id'] = goldterms_subjects['subject'].str.extract(r'GOLDTERMS:(\\d+)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "8e90c8dfb6b820b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>path_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [subject, path_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "goldterms_subjects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "9e70094c5c28b5ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "gold_path_ids = goldterms_subjects['path_id'].dropna().unique().tolist()\n",
    "gold_path_ids = [int(my_id) for my_id in gold_path_ids]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "d294ca1680c31685",
   "metadata": {},
   "outputs": [],
   "source": [
    "gold_env_filtered_biosamples_frame = gold_biosamples_frame[\n",
    "    gold_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'].isin(gold_path_ids)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "ea40acff447df533",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOSAMPLE GOLD ID</th>\n",
       "      <th>BIOSAMPLE NAME</th>\n",
       "      <th>BIOSAMPLE NCBI TAX ID</th>\n",
       "      <th>BIOSAMPLE NCBI TAX NAME</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION SITE</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION DATE</th>\n",
       "      <th>BIOSAMPLE GEOGRAPHIC LOCATION</th>\n",
       "      <th>BIOSAMPLE LATITUDE</th>\n",
       "      <th>BIOSAMPLE LONGITUDE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM PATH ID</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM CATEGORY</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM TYPE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM SUBTYPE</th>\n",
       "      <th>BIOSAMPLE SPECIFIC ECOSYSTEM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [BIOSAMPLE GOLD ID, BIOSAMPLE NAME, BIOSAMPLE NCBI TAX ID, BIOSAMPLE NCBI TAX NAME, BIOSAMPLE SAMPLE COLLECTION SITE, BIOSAMPLE SAMPLE COLLECTION DATE, BIOSAMPLE GEOGRAPHIC LOCATION, BIOSAMPLE LATITUDE, BIOSAMPLE LONGITUDE, BIOSAMPLE ECOSYSTEM PATH ID, BIOSAMPLE ECOSYSTEM, BIOSAMPLE ECOSYSTEM CATEGORY, BIOSAMPLE ECOSYSTEM TYPE, BIOSAMPLE ECOSYSTEM SUBTYPE, BIOSAMPLE SPECIFIC ECOSYSTEM]\n",
       "Index: []"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gold_env_filtered_biosamples_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "4b4ed63d71bf1e65",
   "metadata": {},
   "outputs": [],
   "source": [
    "goldterms_context_frame = pd.read_sql_query(goldterms_envo_query, goldterms_conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "e736fe9475ea3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "goldterms_context_frame['object_label'] = goldterms_context_frame['object'].apply(envo_adapter.label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "ce93f9d692231ed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "goldterms_context_frame['path_id'] = goldterms_context_frame['subject'].str.extract(r'GOLDTERMS:(\\d+)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "f0e775fd25205bcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stanza</th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "      <th>value</th>\n",
       "      <th>datatype</th>\n",
       "      <th>language</th>\n",
       "      <th>graph</th>\n",
       "      <th>object_label</th>\n",
       "      <th>path_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...</td>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...</td>\n",
       "      <td>mixs:env_local</td>\n",
       "      <td>OBI:0001046</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-DHS-reactor</td>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-DHS-reactor</td>\n",
       "      <td>mixs:env_local</td>\n",
       "      <td>OBI:0001046</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Membrane-biore...</td>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Membrane-biore...</td>\n",
       "      <td>mixs:env_local</td>\n",
       "      <td>OBI:0001046</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Membrane-biore...</td>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Membrane-biore...</td>\n",
       "      <td>mixs:env_local</td>\n",
       "      <td>OBI:0001046</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Passive-biorea...</td>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Passive-biorea...</td>\n",
       "      <td>mixs:env_local</td>\n",
       "      <td>OBI:0001046</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>GOLDTERMS:5838</td>\n",
       "      <td>GOLDTERMS:5838</td>\n",
       "      <td>mixs:env_local</td>\n",
       "      <td>OBI:0001046</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>GOLDTERMS:5841</td>\n",
       "      <td>GOLDTERMS:5841</td>\n",
       "      <td>mixs:env_local</td>\n",
       "      <td>ENVO:00000076</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>mine</td>\n",
       "      <td>5841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>GOLDTERMS:5843</td>\n",
       "      <td>GOLDTERMS:5843</td>\n",
       "      <td>mixs:env_local</td>\n",
       "      <td>OBI:0001046</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>GOLDTERMS:5846</td>\n",
       "      <td>GOLDTERMS:5846</td>\n",
       "      <td>mixs:env_local</td>\n",
       "      <td>OBI:0001046</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>GOLDTERMS:5850</td>\n",
       "      <td>GOLDTERMS:5850</td>\n",
       "      <td>mixs:env_local</td>\n",
       "      <td>ENVO:03600021</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>tailings pond</td>\n",
       "      <td>5850</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>458 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                stanza  \\\n",
       "0    GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...   \n",
       "1          GOLDTERMS:Engineered-Bioreactor-DHS-reactor   \n",
       "2    GOLDTERMS:Engineered-Bioreactor-Membrane-biore...   \n",
       "3    GOLDTERMS:Engineered-Bioreactor-Membrane-biore...   \n",
       "4    GOLDTERMS:Engineered-Bioreactor-Passive-biorea...   \n",
       "..                                                 ...   \n",
       "453                                     GOLDTERMS:5838   \n",
       "454                                     GOLDTERMS:5841   \n",
       "455                                     GOLDTERMS:5843   \n",
       "456                                     GOLDTERMS:5846   \n",
       "457                                     GOLDTERMS:5850   \n",
       "\n",
       "                                               subject       predicate  \\\n",
       "0    GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...  mixs:env_local   \n",
       "1          GOLDTERMS:Engineered-Bioreactor-DHS-reactor  mixs:env_local   \n",
       "2    GOLDTERMS:Engineered-Bioreactor-Membrane-biore...  mixs:env_local   \n",
       "3    GOLDTERMS:Engineered-Bioreactor-Membrane-biore...  mixs:env_local   \n",
       "4    GOLDTERMS:Engineered-Bioreactor-Passive-biorea...  mixs:env_local   \n",
       "..                                                 ...             ...   \n",
       "453                                     GOLDTERMS:5838  mixs:env_local   \n",
       "454                                     GOLDTERMS:5841  mixs:env_local   \n",
       "455                                     GOLDTERMS:5843  mixs:env_local   \n",
       "456                                     GOLDTERMS:5846  mixs:env_local   \n",
       "457                                     GOLDTERMS:5850  mixs:env_local   \n",
       "\n",
       "            object value datatype language graph   object_label path_id  \n",
       "0      OBI:0001046  None     None     None  None           None     NaN  \n",
       "1      OBI:0001046  None     None     None  None           None     NaN  \n",
       "2      OBI:0001046  None     None     None  None           None     NaN  \n",
       "3      OBI:0001046  None     None     None  None           None     NaN  \n",
       "4      OBI:0001046  None     None     None  None           None     NaN  \n",
       "..             ...   ...      ...      ...   ...            ...     ...  \n",
       "453    OBI:0001046  None     None     None  None           None    5838  \n",
       "454  ENVO:00000076  None     None     None  None           mine    5841  \n",
       "455    OBI:0001046  None     None     None  None           None    5843  \n",
       "456    OBI:0001046  None     None     None  None           None    5846  \n",
       "457  ENVO:03600021  None     None     None  None  tailings pond    5850  \n",
       "\n",
       "[458 rows x 10 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "goldterms_context_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "7a3f4269e0736579",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill NaN values in 'BIOSAMPLE ECOSYSTEM PATH ID' with 0 and convert to int\n",
    "gold_env_filtered_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'] = gold_env_filtered_biosamples_frame[\n",
    "    'BIOSAMPLE ECOSYSTEM PATH ID'].fillna(0).astype(int)\n",
    "\n",
    "# Drop rows with NaN in 'path_id' in goldterms_context_frame\n",
    "goldterms_context_frame = goldterms_context_frame.dropna(subset=['path_id'])\n",
    "\n",
    "# Convert 'path_id' to int\n",
    "goldterms_context_frame['path_id'] = goldterms_context_frame['path_id'].astype(int)\n",
    "\n",
    "# Perform the left merge\n",
    "gold_env_filtered_biosamples_with_inferred = gold_env_filtered_biosamples_frame.merge(\n",
    "    goldterms_context_frame,\n",
    "    left_on='BIOSAMPLE ECOSYSTEM PATH ID',\n",
    "    right_on='path_id',\n",
    "    how='left'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "27e616ea4cbd27ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOSAMPLE GOLD ID</th>\n",
       "      <th>BIOSAMPLE NAME</th>\n",
       "      <th>BIOSAMPLE NCBI TAX ID</th>\n",
       "      <th>BIOSAMPLE NCBI TAX NAME</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION SITE</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION DATE</th>\n",
       "      <th>BIOSAMPLE GEOGRAPHIC LOCATION</th>\n",
       "      <th>BIOSAMPLE LATITUDE</th>\n",
       "      <th>BIOSAMPLE LONGITUDE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM PATH ID</th>\n",
       "      <th>...</th>\n",
       "      <th>stanza</th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "      <th>value</th>\n",
       "      <th>datatype</th>\n",
       "      <th>language</th>\n",
       "      <th>graph</th>\n",
       "      <th>object_label</th>\n",
       "      <th>path_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [BIOSAMPLE GOLD ID, BIOSAMPLE NAME, BIOSAMPLE NCBI TAX ID, BIOSAMPLE NCBI TAX NAME, BIOSAMPLE SAMPLE COLLECTION SITE, BIOSAMPLE SAMPLE COLLECTION DATE, BIOSAMPLE GEOGRAPHIC LOCATION, BIOSAMPLE LATITUDE, BIOSAMPLE LONGITUDE, BIOSAMPLE ECOSYSTEM PATH ID, BIOSAMPLE ECOSYSTEM, BIOSAMPLE ECOSYSTEM CATEGORY, BIOSAMPLE ECOSYSTEM TYPE, BIOSAMPLE ECOSYSTEM SUBTYPE, BIOSAMPLE SPECIFIC ECOSYSTEM, stanza, subject, predicate, object, value, datatype, language, graph, object_label, path_id]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 25 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gold_env_filtered_biosamples_with_inferred"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e6055a76a17daf0",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "74d28ee43bc2df13",
   "metadata": {},
   "outputs": [],
   "source": [
    "include_in_rows = set()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "4a5858d51d55c1f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "include_in_rows.update(anchor_descendants_frame['curie'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "3a206d349955474b",
   "metadata": {},
   "outputs": [],
   "source": [
    "include_in_rows.update([i['curie'] for i in pv_validation_results['valids']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "6612c4f580111c69",
   "metadata": {},
   "outputs": [],
   "source": [
    "include_in_rows.update(nmdc_biosample_contexts_frame[nmdc_scope])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "4c9a01436eb4376",
   "metadata": {},
   "outputs": [],
   "source": [
    "include_in_rows.update(ncbi_frame['extracted_curie'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "79fc8410c9ad2cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "include_in_rows.update(ncbi_frame['longest_annotation_curie'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "3a349c2bc9dc144d",
   "metadata": {},
   "outputs": [],
   "source": [
    "include_in_rows.update(gold_env_filtered_biosamples_with_inferred['object'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "8cc32502bceef8ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "rows_lod = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "2a88bc28334675b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO MOVE THESE UP, because the expressions are already being used above\n",
    "\n",
    "anchor_curies = list(anchor_descendants_frame['curie'])\n",
    "legacy_pv_curies = [i['curie'] for i in pv_validation_results['valids']]\n",
    "biome_curies = list(envo_adapter.descendants('ENVO:00000428', predicates=[IS_A])) # \n",
    "terrestrial_biome_curies = list(envo_adapter.descendants('ENVO:00000446', predicates=[IS_A]))\n",
    "aquatic_biome_curies = list(envo_adapter.descendants('ENVO:00002030', predicates=[IS_A]))\n",
    "abp_curies = list(envo_adapter.descendants('ENVO:01000813', predicates=[IS_A]))\n",
    "env_sys_curies = list(envo_adapter.descendants('ENVO:01000254', predicates=[IS_A]))\n",
    "env_mat_curies = list(envo_adapter.descendants('ENVO:00010483', predicates=[IS_A]))\n",
    "obsoletes_curies = list(envo_adapter.obsoletes())\n",
    "\n",
    "for curie in include_in_rows:\n",
    "    if curie is None:\n",
    "        continue\n",
    "    row = {\n",
    "        'curie': curie,\n",
    "        'label': envo_adapter.label(curie),\n",
    "        'envo_native': False,\n",
    "        'obsolete': False,\n",
    "        'legacy_pv': False,\n",
    "        'abp': False,\n",
    "        'env_sys': False,\n",
    "        'biome': False,\n",
    "        'terrestrial_biome': False,\n",
    "        'aquatic_biome': False,\n",
    "        'env_mat': False,\n",
    "    }\n",
    "    prefix, local_id = curie.split(':')\n",
    "    if prefix and prefix == 'ENVO' and row['label'] is not None:\n",
    "        row['envo_native'] = True\n",
    "    if curie in biome_curies:\n",
    "        row['biome'] = True\n",
    "    if curie in terrestrial_biome_curies:\n",
    "        row['terrestrial_biome'] = True\n",
    "    if curie in aquatic_biome_curies:\n",
    "        row['aquatic_biome'] = True\n",
    "    if curie in abp_curies:\n",
    "        row['abp'] = True\n",
    "    if curie in env_sys_curies:\n",
    "        row['env_sys'] = True\n",
    "    if curie in env_mat_curies:\n",
    "        row['env_mat'] = True\n",
    "    if curie in legacy_pv_curies:\n",
    "        row['legacy_pv'] = True\n",
    "    if curie in obsoletes_curies:\n",
    "        row['obsolete'] = True\n",
    "    rows_lod.append(row)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "e3c1c8acd0d86bd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "rows_frame = pd.DataFrame(rows_lod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "1b64138f2a698553",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>label</th>\n",
       "      <th>envo_native</th>\n",
       "      <th>obsolete</th>\n",
       "      <th>legacy_pv</th>\n",
       "      <th>abp</th>\n",
       "      <th>env_sys</th>\n",
       "      <th>biome</th>\n",
       "      <th>terrestrial_biome</th>\n",
       "      <th>aquatic_biome</th>\n",
       "      <th>env_mat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:00000111</td>\n",
       "      <td>forested area</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:03501282</td>\n",
       "      <td>hangar</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:00000146</td>\n",
       "      <td>snow field</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENVO:01000672</td>\n",
       "      <td>geological joint</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:01000710</td>\n",
       "      <td>flood</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1910</th>\n",
       "      <td>ENVO:00000165</td>\n",
       "      <td>asphalt lake</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1911</th>\n",
       "      <td>ENVO:03501273</td>\n",
       "      <td>listed building</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1912</th>\n",
       "      <td>ENVO:01000362</td>\n",
       "      <td>inflationary cave</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1913</th>\n",
       "      <td>ENVO:01000943</td>\n",
       "      <td>vegetated hill</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1914</th>\n",
       "      <td>ENVO:03600000</td>\n",
       "      <td>cleanroom</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1915 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              curie              label  envo_native  obsolete  legacy_pv  \\\n",
       "0     ENVO:00000111      forested area         True     False      False   \n",
       "1     ENVO:03501282             hangar         True     False      False   \n",
       "2     ENVO:00000146         snow field         True     False      False   \n",
       "3     ENVO:01000672   geological joint         True     False      False   \n",
       "4     ENVO:01000710              flood         True     False      False   \n",
       "...             ...                ...          ...       ...        ...   \n",
       "1910  ENVO:00000165       asphalt lake         True     False      False   \n",
       "1911  ENVO:03501273    listed building         True     False      False   \n",
       "1912  ENVO:01000362  inflationary cave         True     False      False   \n",
       "1913  ENVO:01000943     vegetated hill         True     False      False   \n",
       "1914  ENVO:03600000          cleanroom         True     False      False   \n",
       "\n",
       "       abp  env_sys  biome  terrestrial_biome  aquatic_biome  env_mat  \n",
       "0     True    False  False              False          False    False  \n",
       "1     True    False  False              False          False    False  \n",
       "2     True    False  False              False          False    False  \n",
       "3     True    False  False              False          False    False  \n",
       "4     True    False  False              False          False    False  \n",
       "...    ...      ...    ...                ...            ...      ...  \n",
       "1910  True    False  False              False          False    False  \n",
       "1911  True    False  False              False          False    False  \n",
       "1912  True    False  False              False          False    False  \n",
       "1913  True    False  False              False          False    False  \n",
       "1914  True    False  False              False          False    False  \n",
       "\n",
       "[1915 rows x 11 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "8392ef84a3254d56",
   "metadata": {},
   "outputs": [],
   "source": [
    "nmdc_biosample_scoped_counts = nmdc_biosample_contexts_frame[nmdc_scope].value_counts().reset_index()\n",
    "nmdc_biosample_scoped_counts.columns = ['curie', 'nmdc_scoped_count']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "fd4a155b6cb06235",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>nmdc_scoped_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:01001057</td>\n",
       "      <td>199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:01001442</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PO:0025025</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           curie  nmdc_scoped_count\n",
       "0  ENVO:01001057                199\n",
       "1  ENVO:01001442                192\n",
       "2     PO:0025025                 10"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nmdc_biosample_scoped_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "3bd7ce9b1a42f70a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform the left merge\n",
    "rows_frame = rows_frame.merge(\n",
    "    nmdc_biosample_scoped_counts,\n",
    "    left_on='curie',\n",
    "    right_on='curie',\n",
    "    how='left'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "335b3ab91d029ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOSAMPLE GOLD ID</th>\n",
       "      <th>BIOSAMPLE NAME</th>\n",
       "      <th>BIOSAMPLE NCBI TAX ID</th>\n",
       "      <th>BIOSAMPLE NCBI TAX NAME</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION SITE</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION DATE</th>\n",
       "      <th>BIOSAMPLE GEOGRAPHIC LOCATION</th>\n",
       "      <th>BIOSAMPLE LATITUDE</th>\n",
       "      <th>BIOSAMPLE LONGITUDE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM PATH ID</th>\n",
       "      <th>...</th>\n",
       "      <th>stanza</th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "      <th>value</th>\n",
       "      <th>datatype</th>\n",
       "      <th>language</th>\n",
       "      <th>graph</th>\n",
       "      <th>object_label</th>\n",
       "      <th>path_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [BIOSAMPLE GOLD ID, BIOSAMPLE NAME, BIOSAMPLE NCBI TAX ID, BIOSAMPLE NCBI TAX NAME, BIOSAMPLE SAMPLE COLLECTION SITE, BIOSAMPLE SAMPLE COLLECTION DATE, BIOSAMPLE GEOGRAPHIC LOCATION, BIOSAMPLE LATITUDE, BIOSAMPLE LONGITUDE, BIOSAMPLE ECOSYSTEM PATH ID, BIOSAMPLE ECOSYSTEM, BIOSAMPLE ECOSYSTEM CATEGORY, BIOSAMPLE ECOSYSTEM TYPE, BIOSAMPLE ECOSYSTEM SUBTYPE, BIOSAMPLE SPECIFIC ECOSYSTEM, stanza, subject, predicate, object, value, datatype, language, graph, object_label, path_id]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 25 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gold_env_filtered_biosamples_with_inferred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "c7a4d5fd716efd3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gold_biosample_scoped_counts = gold_env_filtered_biosamples_with_inferred['object'].value_counts().reset_index()\n",
    "gold_biosample_scoped_counts.columns = ['curie', 'gold_scoped_count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "da2433c6d19907a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>gold_scoped_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [curie, gold_scoped_count]\n",
       "Index: []"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gold_biosample_scoped_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "f5c673de7e5fd32a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform the left merge\n",
    "rows_frame = rows_frame.merge(\n",
    "    gold_biosample_scoped_counts,\n",
    "    left_on='curie',\n",
    "    right_on='curie',\n",
    "    how='left'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "cdde35e2977eaf9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>label</th>\n",
       "      <th>envo_native</th>\n",
       "      <th>obsolete</th>\n",
       "      <th>legacy_pv</th>\n",
       "      <th>abp</th>\n",
       "      <th>env_sys</th>\n",
       "      <th>biome</th>\n",
       "      <th>terrestrial_biome</th>\n",
       "      <th>aquatic_biome</th>\n",
       "      <th>env_mat</th>\n",
       "      <th>nmdc_scoped_count</th>\n",
       "      <th>gold_scoped_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:00000111</td>\n",
       "      <td>forested area</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:03501282</td>\n",
       "      <td>hangar</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:00000146</td>\n",
       "      <td>snow field</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENVO:01000672</td>\n",
       "      <td>geological joint</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:01000710</td>\n",
       "      <td>flood</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1910</th>\n",
       "      <td>ENVO:00000165</td>\n",
       "      <td>asphalt lake</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1911</th>\n",
       "      <td>ENVO:03501273</td>\n",
       "      <td>listed building</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1912</th>\n",
       "      <td>ENVO:01000362</td>\n",
       "      <td>inflationary cave</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1913</th>\n",
       "      <td>ENVO:01000943</td>\n",
       "      <td>vegetated hill</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1914</th>\n",
       "      <td>ENVO:03600000</td>\n",
       "      <td>cleanroom</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1915 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              curie              label  envo_native  obsolete  legacy_pv  \\\n",
       "0     ENVO:00000111      forested area         True     False      False   \n",
       "1     ENVO:03501282             hangar         True     False      False   \n",
       "2     ENVO:00000146         snow field         True     False      False   \n",
       "3     ENVO:01000672   geological joint         True     False      False   \n",
       "4     ENVO:01000710              flood         True     False      False   \n",
       "...             ...                ...          ...       ...        ...   \n",
       "1910  ENVO:00000165       asphalt lake         True     False      False   \n",
       "1911  ENVO:03501273    listed building         True     False      False   \n",
       "1912  ENVO:01000362  inflationary cave         True     False      False   \n",
       "1913  ENVO:01000943     vegetated hill         True     False      False   \n",
       "1914  ENVO:03600000          cleanroom         True     False      False   \n",
       "\n",
       "       abp  env_sys  biome  terrestrial_biome  aquatic_biome  env_mat  \\\n",
       "0     True    False  False              False          False    False   \n",
       "1     True    False  False              False          False    False   \n",
       "2     True    False  False              False          False    False   \n",
       "3     True    False  False              False          False    False   \n",
       "4     True    False  False              False          False    False   \n",
       "...    ...      ...    ...                ...            ...      ...   \n",
       "1910  True    False  False              False          False    False   \n",
       "1911  True    False  False              False          False    False   \n",
       "1912  True    False  False              False          False    False   \n",
       "1913  True    False  False              False          False    False   \n",
       "1914  True    False  False              False          False    False   \n",
       "\n",
       "      nmdc_scoped_count  gold_scoped_count  \n",
       "0                   NaN                NaN  \n",
       "1                   NaN                NaN  \n",
       "2                   NaN                NaN  \n",
       "3                   NaN                NaN  \n",
       "4                   NaN                NaN  \n",
       "...                 ...                ...  \n",
       "1910                NaN                NaN  \n",
       "1911                NaN                NaN  \n",
       "1912                NaN                NaN  \n",
       "1913                NaN                NaN  \n",
       "1914                NaN                NaN  \n",
       "\n",
       "[1915 rows x 13 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "9baa9d73af8159ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 990 rows in https://docs.google.com/spreadsheets/d/12WH3eduBq2qSTy9zVF3n7fyajn6ssLZL/edit?gid=546570706#gid=546570706"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "460de8efcd09385d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gold and ncbi counts are slightly trickier\n",
    "# for gold may want to include presence or mapping in goldterms in addition to biosamples counts\n",
    "# ncbi: we have extracted curies and annotated curies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "f416a6fe17c57df8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# todo move this stuff up to immediately after the creation of ncbi_frame ?\n",
    "\n",
    "# todo don't accept extracted curie if no real label?\n",
    "# any kind of string similarity checking for label of annotated curie vs extracted label ?\n",
    "# look for long runs of curies?\n",
    "# can we measure the beneficial impact of any of this? current crux: how to distribute counts\n",
    "\n",
    "ncbi_frame['curie_list'] = ncbi_frame.apply(\n",
    "    lambda my_row: list({my_row['extracted_curie'], my_row['longest_annotation_curie']} - {None}),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "ncbi_frame['unique_curie_count'] = ncbi_frame['curie_list'].apply(len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "8e61714bd08518e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>serial_number</th>\n",
       "      <th>content</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>content_list</th>\n",
       "      <th>content_count</th>\n",
       "      <th>envo_count</th>\n",
       "      <th>extracted_label</th>\n",
       "      <th>extracted_curie</th>\n",
       "      <th>real_label</th>\n",
       "      <th>longest_annotation_curie</th>\n",
       "      <th>longest_annotation_label</th>\n",
       "      <th>curie_list</th>\n",
       "      <th>unique_curie_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>woodland</td>\n",
       "      <td>6385</td>\n",
       "      <td>woodland</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>woodland</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000057</td>\n",
       "      <td>mangrove swamp</td>\n",
       "      <td>[ENVO:00000057]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>rhizosphere environment [ENVO_01000999]</td>\n",
       "      <td>5325</td>\n",
       "      <td>rhizosphere environment [ENVO_01000999]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere environment</td>\n",
       "      <td>ENVO:01000999</td>\n",
       "      <td>rhizosphere environment</td>\n",
       "      <td>ENVO:01000999</td>\n",
       "      <td>rhizosphere environment</td>\n",
       "      <td>[ENVO:01000999]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>2924</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>laboratory environment [ENVO:01001405]</td>\n",
       "      <td>2683</td>\n",
       "      <td>laboratory environment [ENVO:01001405]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>laboratory environment</td>\n",
       "      <td>ENVO:01001405</td>\n",
       "      <td>laboratory environment</td>\n",
       "      <td>ENVO:01001405</td>\n",
       "      <td>laboratory environment</td>\n",
       "      <td>[ENVO:01001405]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>forest</td>\n",
       "      <td>1768</td>\n",
       "      <td>forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>forest</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000111</td>\n",
       "      <td>forested area</td>\n",
       "      <td>[ENVO:00000111]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979</th>\n",
       "      <td>1951</td>\n",
       "      <td>Vineyard-SMDRTd1</td>\n",
       "      <td>1</td>\n",
       "      <td>Vineyard-SMDRTd1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Vineyard-SMDRTd1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "      <td>[ENVO:00000116]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1980</th>\n",
       "      <td>1952</td>\n",
       "      <td>Vineyard-VARTc2</td>\n",
       "      <td>1</td>\n",
       "      <td>Vineyard-VARTc2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Vineyard-VARTc2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "      <td>[ENVO:00000116]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1981</th>\n",
       "      <td>1953</td>\n",
       "      <td>Vineyard-VVRTc3</td>\n",
       "      <td>1</td>\n",
       "      <td>Vineyard-VVRTc3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Vineyard-VVRTc3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "      <td>[ENVO:00000116]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1982</th>\n",
       "      <td>1954</td>\n",
       "      <td>Vineyard-VVRTe1</td>\n",
       "      <td>1</td>\n",
       "      <td>Vineyard-VVRTe1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Vineyard-VVRTe1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "      <td>[ENVO:00000116]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1983</th>\n",
       "      <td>1955</td>\n",
       "      <td>Female Maize Seed</td>\n",
       "      <td>1</td>\n",
       "      <td>Female Maize Seed</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Female Maize Seed</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1984 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      serial_number                                  content  sample_count  \\\n",
       "0                 1                                 woodland          6385   \n",
       "1                 2  rhizosphere environment [ENVO_01000999]          5325   \n",
       "2                 3                           not applicable          2924   \n",
       "3                 4   laboratory environment [ENVO:01001405]          2683   \n",
       "4                 5                                   forest          1768   \n",
       "...             ...                                      ...           ...   \n",
       "1979           1951                         Vineyard-SMDRTd1             1   \n",
       "1980           1952                          Vineyard-VARTc2             1   \n",
       "1981           1953                          Vineyard-VVRTc3             1   \n",
       "1982           1954                          Vineyard-VVRTe1             1   \n",
       "1983           1955                        Female Maize Seed             1   \n",
       "\n",
       "                                 content_list  content_count  envo_count  \\\n",
       "0                                    woodland              1           0   \n",
       "1     rhizosphere environment [ENVO_01000999]              1           1   \n",
       "2                              not applicable              1           0   \n",
       "3      laboratory environment [ENVO:01001405]              1           1   \n",
       "4                                      forest              1           0   \n",
       "...                                       ...            ...         ...   \n",
       "1979                         Vineyard-SMDRTd1              1           0   \n",
       "1980                          Vineyard-VARTc2              1           0   \n",
       "1981                          Vineyard-VVRTc3              1           0   \n",
       "1982                          Vineyard-VVRTe1              1           0   \n",
       "1983                        Female Maize Seed              1           0   \n",
       "\n",
       "              extracted_label extracted_curie               real_label  \\\n",
       "0                    woodland            None                     None   \n",
       "1     rhizosphere environment   ENVO:01000999  rhizosphere environment   \n",
       "2              not applicable            None                     None   \n",
       "3      laboratory environment   ENVO:01001405   laboratory environment   \n",
       "4                      forest            None                     None   \n",
       "...                       ...             ...                      ...   \n",
       "1979         Vineyard-SMDRTd1            None                     None   \n",
       "1980          Vineyard-VARTc2            None                     None   \n",
       "1981          Vineyard-VVRTc3            None                     None   \n",
       "1982          Vineyard-VVRTe1            None                     None   \n",
       "1983        Female Maize Seed            None                     None   \n",
       "\n",
       "     longest_annotation_curie longest_annotation_label       curie_list  \\\n",
       "0               ENVO:00000057           mangrove swamp  [ENVO:00000057]   \n",
       "1               ENVO:01000999  rhizosphere environment  [ENVO:01000999]   \n",
       "2                        None                     None               []   \n",
       "3               ENVO:01001405   laboratory environment  [ENVO:01001405]   \n",
       "4               ENVO:00000111            forested area  [ENVO:00000111]   \n",
       "...                       ...                      ...              ...   \n",
       "1979            ENVO:00000116                 vineyard  [ENVO:00000116]   \n",
       "1980            ENVO:00000116                 vineyard  [ENVO:00000116]   \n",
       "1981            ENVO:00000116                 vineyard  [ENVO:00000116]   \n",
       "1982            ENVO:00000116                 vineyard  [ENVO:00000116]   \n",
       "1983                     None                     None               []   \n",
       "\n",
       "      unique_curie_count  \n",
       "0                      1  \n",
       "1                      1  \n",
       "2                      0  \n",
       "3                      1  \n",
       "4                      1  \n",
       "...                  ...  \n",
       "1979                   1  \n",
       "1980                   1  \n",
       "1981                   1  \n",
       "1982                   1  \n",
       "1983                   0  \n",
       "\n",
       "[1984 rows x 13 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "7a69fc0bd5e16703",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "unique_curie_count\n",
       "1    1299\n",
       "0     645\n",
       "2      40\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame['unique_curie_count'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "2b43b2226a208cf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "double_curie_frame = ncbi_frame[ncbi_frame['unique_curie_count'] > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "9e74fd42072e1242",
   "metadata": {},
   "outputs": [],
   "source": [
    "double_curie_frame = double_curie_frame[['extracted_curie', 'longest_annotation_curie']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "3406bb4668dc4ecc",
   "metadata": {},
   "outputs": [],
   "source": [
    "double_curie_frame = double_curie_frame.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "79a954a6c86b49c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "double_curie_frame[['extracted_prefix', 'extracted_local_id']] = double_curie_frame['extracted_curie'].str.split(':', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "9822b8635cb794c",
   "metadata": {},
   "outputs": [],
   "source": [
    "double_curie_frame['extracted_local_id_int'] = pd.to_numeric(double_curie_frame['extracted_local_id'], errors='coerce').astype('Int64')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "7cd8a092fdd31038",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure extracted_local_id_int is unique and sorted\n",
    "unique_sorted_series = double_curie_frame['extracted_local_id_int'].dropna().drop_duplicates().sort_values()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "97c87784915ccb88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find stretches\n",
    "stretches_dict = find_consecutive_stretches_dict(unique_sorted_series)\n",
    "\n",
    "# pprint.pprint(stretches_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "4dc6b1dc0c012b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the stretches dictionary into a DataFrame\n",
    "stretches_df = stretches_dict_to_long_dataframe(stretches_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "64f939c9fdeadd27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stretch_id</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1001431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1001432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1001433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1001434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1001435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>1001436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>1001437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>1001438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>1001439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>1001440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>1001441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>1001442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>1001443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>1001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>1001445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>1001446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>1001447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>1001448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>1001449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>1001450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>1001451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>1001452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>1001453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>1001454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>1001455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>1001456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1</td>\n",
       "      <td>1001457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>1001458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>1001459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1</td>\n",
       "      <td>1001460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1</td>\n",
       "      <td>1001461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1</td>\n",
       "      <td>1001462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1</td>\n",
       "      <td>1001463</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    stretch_id    value\n",
       "0            1  1001431\n",
       "1            1  1001432\n",
       "2            1  1001433\n",
       "3            1  1001434\n",
       "4            1  1001435\n",
       "5            1  1001436\n",
       "6            1  1001437\n",
       "7            1  1001438\n",
       "8            1  1001439\n",
       "9            1  1001440\n",
       "10           1  1001441\n",
       "11           1  1001442\n",
       "12           1  1001443\n",
       "13           1  1001444\n",
       "14           1  1001445\n",
       "15           1  1001446\n",
       "16           1  1001447\n",
       "17           1  1001448\n",
       "18           1  1001449\n",
       "19           1  1001450\n",
       "20           1  1001451\n",
       "21           1  1001452\n",
       "22           1  1001453\n",
       "23           1  1001454\n",
       "24           1  1001455\n",
       "25           1  1001456\n",
       "26           1  1001457\n",
       "27           1  1001458\n",
       "28           1  1001459\n",
       "29           1  1001460\n",
       "30           1  1001461\n",
       "31           1  1001462\n",
       "32           1  1001463"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stretches_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "712d1158df0d01c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform the left merge\n",
    "double_curie_frame = double_curie_frame.merge(\n",
    "    stretches_df,\n",
    "    left_on='extracted_local_id_int',\n",
    "    right_on='value',\n",
    "    how='left'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "1e339dc7cb20377d",
   "metadata": {},
   "outputs": [],
   "source": [
    "stretch_summary_df = summarize_stretch_groups(double_curie_frame)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "dfd6a50c15372421",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stretch_id</th>\n",
       "      <th>most_common_longest_annotation_curie</th>\n",
       "      <th>fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   stretch_id most_common_longest_annotation_curie  fraction\n",
       "0         1.0                        ENVO:01001430       1.0"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stretch_summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "33d066472e719724",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform the left merge\n",
    "double_curie_frame = double_curie_frame.merge(\n",
    "    stretch_summary_df,\n",
    "    left_on='stretch_id',\n",
    "    right_on='stretch_id',\n",
    "    how='left'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "4ee4a3b5e3d57324",
   "metadata": {},
   "outputs": [],
   "source": [
    "drag_evidence_frame = double_curie_frame[double_curie_frame['stretch_id'] >= 1]\n",
    "drag_evidence_frame = drag_evidence_frame[['extracted_curie', 'longest_annotation_curie']]\n",
    "drag_evidence_frame['drag_evidence'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "b538401f9eba6f23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>extracted_curie</th>\n",
       "      <th>longest_annotation_curie</th>\n",
       "      <th>drag_evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENVO:01001432</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENVO:01001435</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENVO:01001436</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENVO:01001433</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENVO:01001434</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENVO:01001431</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENVO:01001442</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENVO:01001462</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENVO:01001463</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENVO:01001443</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENVO:01001450</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENVO:01001451</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENVO:01001455</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENVO:01001458</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENVO:01001444</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENVO:01001447</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENVO:01001449</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENVO:01001461</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENVO:01001439</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENVO:01001441</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENVO:01001452</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENVO:01001453</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENVO:01001456</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENVO:01001459</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENVO:01001460</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ENVO:01001440</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ENVO:01001448</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENVO:01001437</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENVO:01001446</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENVO:01001454</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENVO:01001438</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ENVO:01001445</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ENVO:01001457</td>\n",
       "      <td>ENVO:01001430</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   extracted_curie longest_annotation_curie  drag_evidence\n",
       "6    ENVO:01001432            ENVO:01001430           True\n",
       "7    ENVO:01001435            ENVO:01001430           True\n",
       "8    ENVO:01001436            ENVO:01001430           True\n",
       "9    ENVO:01001433            ENVO:01001430           True\n",
       "10   ENVO:01001434            ENVO:01001430           True\n",
       "11   ENVO:01001431            ENVO:01001430           True\n",
       "12   ENVO:01001442            ENVO:01001430           True\n",
       "13   ENVO:01001462            ENVO:01001430           True\n",
       "14   ENVO:01001463            ENVO:01001430           True\n",
       "15   ENVO:01001443            ENVO:01001430           True\n",
       "16   ENVO:01001450            ENVO:01001430           True\n",
       "17   ENVO:01001451            ENVO:01001430           True\n",
       "18   ENVO:01001455            ENVO:01001430           True\n",
       "19   ENVO:01001458            ENVO:01001430           True\n",
       "20   ENVO:01001444            ENVO:01001430           True\n",
       "21   ENVO:01001447            ENVO:01001430           True\n",
       "22   ENVO:01001449            ENVO:01001430           True\n",
       "23   ENVO:01001461            ENVO:01001430           True\n",
       "24   ENVO:01001439            ENVO:01001430           True\n",
       "25   ENVO:01001441            ENVO:01001430           True\n",
       "26   ENVO:01001452            ENVO:01001430           True\n",
       "27   ENVO:01001453            ENVO:01001430           True\n",
       "28   ENVO:01001456            ENVO:01001430           True\n",
       "29   ENVO:01001459            ENVO:01001430           True\n",
       "30   ENVO:01001460            ENVO:01001430           True\n",
       "31   ENVO:01001440            ENVO:01001430           True\n",
       "32   ENVO:01001448            ENVO:01001430           True\n",
       "33   ENVO:01001437            ENVO:01001430           True\n",
       "34   ENVO:01001446            ENVO:01001430           True\n",
       "35   ENVO:01001454            ENVO:01001430           True\n",
       "36   ENVO:01001438            ENVO:01001430           True\n",
       "37   ENVO:01001445            ENVO:01001430           True\n",
       "38   ENVO:01001457            ENVO:01001430           True"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drag_evidence_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "6c204f542b99cc8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame = ncbi_frame.merge(\n",
    "    drag_evidence_frame,\n",
    "    left_on=['extracted_curie', 'longest_annotation_curie'],\n",
    "    right_on=['extracted_curie', 'longest_annotation_curie'],\n",
    "    how='left'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "60cb2591f94be794",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize dragless_curie_list with curie_list values\n",
    "ncbi_frame[\"dragless_curie_list\"] = ncbi_frame[\"curie_list\"]\n",
    "\n",
    "# Update dragless_curie_list based on the condition\n",
    "for index, row in ncbi_frame.iterrows():\n",
    "    if row[\"drag_evidence\"] is True:\n",
    "        if row[\"longest_annotation_curie\"] is not None:\n",
    "            ncbi_frame.at[index, \"dragless_curie_list\"] = [row[\"longest_annotation_curie\"]]\n",
    "        else:\n",
    "            ncbi_frame.at[index, \"dragless_curie_list\"] = []\n",
    "\n",
    "ncbi_frame['dragless_curie_count'] = ncbi_frame['dragless_curie_list'].apply(len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "ad797f2c47ad4d2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "unique_curie_count\n",
       "1    1299\n",
       "0     645\n",
       "2      40\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame['unique_curie_count'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "db64af727a390bde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dragless_curie_count\n",
       "1    1332\n",
       "0     645\n",
       "2       7\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame['dragless_curie_count'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "e7369d5d29ddd57e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1984, 16)"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "f43289b34183795d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame_undisputed = ncbi_frame[ncbi_frame['dragless_curie_count'] <= 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "abc66f266ab36d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1977, 16)"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame_undisputed.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "df155f2a8e6f043e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame_disputed = ncbi_frame[ncbi_frame['dragless_curie_count'] > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "1e255b1b388190ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7, 16)"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame_disputed.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "f572b9083b2c0183",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame_disputed = ncbi_frame_disputed.explode(\"dragless_curie_list\", ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "b5166d98cf8b8c35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14, 16)"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_frame_disputed.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "dbbf498f0dc1e3e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_frame_disputed[\"dragless_curie_list\"] = ncbi_frame_disputed[\"dragless_curie_list\"].apply(lambda x: [x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "cffbde835766f6fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine the rows of ncbi_frame_undisputed and ncbi_frame_disputed into a new DataFrame\n",
    "ncbi_disputes_exploded_frame = pd.concat([ncbi_frame_undisputed, ncbi_frame_disputed], ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "be0bc9e01b0223b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1991, 16)"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_disputes_exploded_frame.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "30054792f2183507",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>serial_number</th>\n",
       "      <th>content</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>content_list</th>\n",
       "      <th>content_count</th>\n",
       "      <th>envo_count</th>\n",
       "      <th>extracted_label</th>\n",
       "      <th>extracted_curie</th>\n",
       "      <th>real_label</th>\n",
       "      <th>longest_annotation_curie</th>\n",
       "      <th>longest_annotation_label</th>\n",
       "      <th>curie_list</th>\n",
       "      <th>unique_curie_count</th>\n",
       "      <th>drag_evidence</th>\n",
       "      <th>dragless_curie_list</th>\n",
       "      <th>dragless_curie_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>woodland</td>\n",
       "      <td>6385</td>\n",
       "      <td>woodland</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>woodland</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000057</td>\n",
       "      <td>mangrove swamp</td>\n",
       "      <td>[ENVO:00000057]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:00000057]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>rhizosphere environment [ENVO_01000999]</td>\n",
       "      <td>5325</td>\n",
       "      <td>rhizosphere environment [ENVO_01000999]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>rhizosphere environment</td>\n",
       "      <td>ENVO:01000999</td>\n",
       "      <td>rhizosphere environment</td>\n",
       "      <td>ENVO:01000999</td>\n",
       "      <td>rhizosphere environment</td>\n",
       "      <td>[ENVO:01000999]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:01000999]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>2924</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>laboratory environment [ENVO:01001405]</td>\n",
       "      <td>2683</td>\n",
       "      <td>laboratory environment [ENVO:01001405]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>laboratory environment</td>\n",
       "      <td>ENVO:01001405</td>\n",
       "      <td>laboratory environment</td>\n",
       "      <td>ENVO:01001405</td>\n",
       "      <td>laboratory environment</td>\n",
       "      <td>[ENVO:01001405]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:01001405]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>forest</td>\n",
       "      <td>1768</td>\n",
       "      <td>forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>forest</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000111</td>\n",
       "      <td>forested area</td>\n",
       "      <td>[ENVO:00000111]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:00000111]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1986</th>\n",
       "      <td>231</td>\n",
       "      <td>NY_fields[ENVO:00000040]</td>\n",
       "      <td>36</td>\n",
       "      <td>NY_fields[ENVO:00000040]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NY_fields</td>\n",
       "      <td>ENVO:00000040</td>\n",
       "      <td>waterfall</td>\n",
       "      <td>ENVO:01000352</td>\n",
       "      <td>field</td>\n",
       "      <td>[ENVO:00000040, ENVO:01000352]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:01000352]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1987</th>\n",
       "      <td>425</td>\n",
       "      <td>intertidal zone [ENVO:0000316]</td>\n",
       "      <td>8</td>\n",
       "      <td>intertidal zone [ENVO:0000316]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>intertidal zone</td>\n",
       "      <td>ENVO:0000316</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000316</td>\n",
       "      <td>intertidal zone</td>\n",
       "      <td>[ENVO:00000316, ENVO:0000316]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:00000316]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1988</th>\n",
       "      <td>425</td>\n",
       "      <td>intertidal zone [ENVO:0000316]</td>\n",
       "      <td>8</td>\n",
       "      <td>intertidal zone [ENVO:0000316]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>intertidal zone</td>\n",
       "      <td>ENVO:0000316</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000316</td>\n",
       "      <td>intertidal zone</td>\n",
       "      <td>[ENVO:00000316, ENVO:0000316]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:0000316]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1989</th>\n",
       "      <td>1018</td>\n",
       "      <td>green_house[ENVO:00000040]</td>\n",
       "      <td>1</td>\n",
       "      <td>green_house[ENVO:00000040]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>green_house</td>\n",
       "      <td>ENVO:00000040</td>\n",
       "      <td>waterfall</td>\n",
       "      <td>ENVO:01000417</td>\n",
       "      <td>house</td>\n",
       "      <td>[ENVO:00000040, ENVO:01000417]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:00000040]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1990</th>\n",
       "      <td>1018</td>\n",
       "      <td>green_house[ENVO:00000040]</td>\n",
       "      <td>1</td>\n",
       "      <td>green_house[ENVO:00000040]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>green_house</td>\n",
       "      <td>ENVO:00000040</td>\n",
       "      <td>waterfall</td>\n",
       "      <td>ENVO:01000417</td>\n",
       "      <td>house</td>\n",
       "      <td>[ENVO:00000040, ENVO:01000417]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:01000417]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1991 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      serial_number                                  content  sample_count  \\\n",
       "0                 1                                 woodland          6385   \n",
       "1                 2  rhizosphere environment [ENVO_01000999]          5325   \n",
       "2                 3                           not applicable          2924   \n",
       "3                 4   laboratory environment [ENVO:01001405]          2683   \n",
       "4                 5                                   forest          1768   \n",
       "...             ...                                      ...           ...   \n",
       "1986            231                 NY_fields[ENVO:00000040]            36   \n",
       "1987            425           intertidal zone [ENVO:0000316]             8   \n",
       "1988            425           intertidal zone [ENVO:0000316]             8   \n",
       "1989           1018               green_house[ENVO:00000040]             1   \n",
       "1990           1018               green_house[ENVO:00000040]             1   \n",
       "\n",
       "                                 content_list  content_count  envo_count  \\\n",
       "0                                    woodland              1           0   \n",
       "1     rhizosphere environment [ENVO_01000999]              1           1   \n",
       "2                              not applicable              1           0   \n",
       "3      laboratory environment [ENVO:01001405]              1           1   \n",
       "4                                      forest              1           0   \n",
       "...                                       ...            ...         ...   \n",
       "1986                 NY_fields[ENVO:00000040]              1           1   \n",
       "1987           intertidal zone [ENVO:0000316]              1           1   \n",
       "1988           intertidal zone [ENVO:0000316]              1           1   \n",
       "1989               green_house[ENVO:00000040]              1           1   \n",
       "1990               green_house[ENVO:00000040]              1           1   \n",
       "\n",
       "              extracted_label extracted_curie               real_label  \\\n",
       "0                    woodland            None                     None   \n",
       "1     rhizosphere environment   ENVO:01000999  rhizosphere environment   \n",
       "2              not applicable            None                     None   \n",
       "3      laboratory environment   ENVO:01001405   laboratory environment   \n",
       "4                      forest            None                     None   \n",
       "...                       ...             ...                      ...   \n",
       "1986                NY_fields   ENVO:00000040                waterfall   \n",
       "1987          intertidal zone    ENVO:0000316                     None   \n",
       "1988          intertidal zone    ENVO:0000316                     None   \n",
       "1989              green_house   ENVO:00000040                waterfall   \n",
       "1990              green_house   ENVO:00000040                waterfall   \n",
       "\n",
       "     longest_annotation_curie longest_annotation_label  \\\n",
       "0               ENVO:00000057           mangrove swamp   \n",
       "1               ENVO:01000999  rhizosphere environment   \n",
       "2                        None                     None   \n",
       "3               ENVO:01001405   laboratory environment   \n",
       "4               ENVO:00000111            forested area   \n",
       "...                       ...                      ...   \n",
       "1986            ENVO:01000352                    field   \n",
       "1987            ENVO:00000316          intertidal zone   \n",
       "1988            ENVO:00000316          intertidal zone   \n",
       "1989            ENVO:01000417                    house   \n",
       "1990            ENVO:01000417                    house   \n",
       "\n",
       "                          curie_list  unique_curie_count drag_evidence  \\\n",
       "0                    [ENVO:00000057]                   1           NaN   \n",
       "1                    [ENVO:01000999]                   1           NaN   \n",
       "2                                 []                   0           NaN   \n",
       "3                    [ENVO:01001405]                   1           NaN   \n",
       "4                    [ENVO:00000111]                   1           NaN   \n",
       "...                              ...                 ...           ...   \n",
       "1986  [ENVO:00000040, ENVO:01000352]                   2           NaN   \n",
       "1987   [ENVO:00000316, ENVO:0000316]                   2           NaN   \n",
       "1988   [ENVO:00000316, ENVO:0000316]                   2           NaN   \n",
       "1989  [ENVO:00000040, ENVO:01000417]                   2           NaN   \n",
       "1990  [ENVO:00000040, ENVO:01000417]                   2           NaN   \n",
       "\n",
       "     dragless_curie_list  dragless_curie_count  \n",
       "0        [ENVO:00000057]                     1  \n",
       "1        [ENVO:01000999]                     1  \n",
       "2                     []                     0  \n",
       "3        [ENVO:01001405]                     1  \n",
       "4        [ENVO:00000111]                     1  \n",
       "...                  ...                   ...  \n",
       "1986     [ENVO:01000352]                     2  \n",
       "1987     [ENVO:00000316]                     2  \n",
       "1988      [ENVO:0000316]                     2  \n",
       "1989     [ENVO:00000040]                     2  \n",
       "1990     [ENVO:01000417]                     2  \n",
       "\n",
       "[1991 rows x 16 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_disputes_exploded_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "6c4bcc094860c1ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_disputes_exploded_frame['post_explode_curie_count'] = ncbi_disputes_exploded_frame['dragless_curie_list'].apply(len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "2c017b4a72fea50d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "post_explode_curie_count\n",
       "1    1346\n",
       "0     645\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_disputes_exploded_frame['post_explode_curie_count'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "53565dbf2f899386",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set 'post_explode_curie' to the 0th item in 'dragless_curie_list'\n",
    "ncbi_disputes_exploded_frame[\"post_explode_curie\"] = ncbi_disputes_exploded_frame[\"dragless_curie_list\"].apply(\n",
    "    lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "b7aad78ffe14089a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ncbi_biosample_scoped_counts = (\n",
    "    ncbi_disputes_exploded_frame.groupby(\"post_explode_curie\")[\"sample_count\"].sum().reset_index()\n",
    ")\n",
    "\n",
    "ncbi_biosample_scoped_counts.columns = ['curie', 'ncbi_scoped_count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "3fdf963c16419e30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>ncbi_scoped_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BFO:0000015</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BFO:0000029</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEBI:22695</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEBI:24632</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEBI:33290</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>RO:0002577</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>UBERON:0000178</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>UBERON:0001988</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>UBERON:0002100</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>UBERON:0002416</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>308 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              curie  ncbi_scoped_count\n",
       "0       BFO:0000015                 18\n",
       "1       BFO:0000029                  5\n",
       "2       CHEBI:22695                 11\n",
       "3       CHEBI:24632                  1\n",
       "4       CHEBI:33290                209\n",
       "..              ...                ...\n",
       "303      RO:0002577                148\n",
       "304  UBERON:0000178                 48\n",
       "305  UBERON:0001988                  3\n",
       "306  UBERON:0002100                  6\n",
       "307  UBERON:0002416                224\n",
       "\n",
       "[308 rows x 2 columns]"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_biosample_scoped_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "a27bbaecf2e606f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform the left merge\n",
    "rows_frame = rows_frame.merge(\n",
    "    ncbi_biosample_scoped_counts,\n",
    "    left_on='curie',\n",
    "    right_on='curie',\n",
    "    how='left'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "f684c7a677faf979",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>label</th>\n",
       "      <th>envo_native</th>\n",
       "      <th>obsolete</th>\n",
       "      <th>legacy_pv</th>\n",
       "      <th>abp</th>\n",
       "      <th>env_sys</th>\n",
       "      <th>biome</th>\n",
       "      <th>terrestrial_biome</th>\n",
       "      <th>aquatic_biome</th>\n",
       "      <th>env_mat</th>\n",
       "      <th>nmdc_scoped_count</th>\n",
       "      <th>gold_scoped_count</th>\n",
       "      <th>ncbi_scoped_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:00000111</td>\n",
       "      <td>forested area</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4927.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:03501282</td>\n",
       "      <td>hangar</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:00000146</td>\n",
       "      <td>snow field</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENVO:01000672</td>\n",
       "      <td>geological joint</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:01000710</td>\n",
       "      <td>flood</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1910</th>\n",
       "      <td>ENVO:00000165</td>\n",
       "      <td>asphalt lake</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1911</th>\n",
       "      <td>ENVO:03501273</td>\n",
       "      <td>listed building</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1912</th>\n",
       "      <td>ENVO:01000362</td>\n",
       "      <td>inflationary cave</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1913</th>\n",
       "      <td>ENVO:01000943</td>\n",
       "      <td>vegetated hill</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1914</th>\n",
       "      <td>ENVO:03600000</td>\n",
       "      <td>cleanroom</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1915 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              curie              label  envo_native  obsolete  legacy_pv  \\\n",
       "0     ENVO:00000111      forested area         True     False      False   \n",
       "1     ENVO:03501282             hangar         True     False      False   \n",
       "2     ENVO:00000146         snow field         True     False      False   \n",
       "3     ENVO:01000672   geological joint         True     False      False   \n",
       "4     ENVO:01000710              flood         True     False      False   \n",
       "...             ...                ...          ...       ...        ...   \n",
       "1910  ENVO:00000165       asphalt lake         True     False      False   \n",
       "1911  ENVO:03501273    listed building         True     False      False   \n",
       "1912  ENVO:01000362  inflationary cave         True     False      False   \n",
       "1913  ENVO:01000943     vegetated hill         True     False      False   \n",
       "1914  ENVO:03600000          cleanroom         True     False      False   \n",
       "\n",
       "       abp  env_sys  biome  terrestrial_biome  aquatic_biome  env_mat  \\\n",
       "0     True    False  False              False          False    False   \n",
       "1     True    False  False              False          False    False   \n",
       "2     True    False  False              False          False    False   \n",
       "3     True    False  False              False          False    False   \n",
       "4     True    False  False              False          False    False   \n",
       "...    ...      ...    ...                ...            ...      ...   \n",
       "1910  True    False  False              False          False    False   \n",
       "1911  True    False  False              False          False    False   \n",
       "1912  True    False  False              False          False    False   \n",
       "1913  True    False  False              False          False    False   \n",
       "1914  True    False  False              False          False    False   \n",
       "\n",
       "      nmdc_scoped_count  gold_scoped_count  ncbi_scoped_count  \n",
       "0                   NaN                NaN             4927.0  \n",
       "1                   NaN                NaN                NaN  \n",
       "2                   NaN                NaN                NaN  \n",
       "3                   NaN                NaN                NaN  \n",
       "4                   NaN                NaN               45.0  \n",
       "...                 ...                ...                ...  \n",
       "1910                NaN                NaN                NaN  \n",
       "1911                NaN                NaN                NaN  \n",
       "1912                NaN                NaN                NaN  \n",
       "1913                NaN                NaN               21.0  \n",
       "1914                NaN                NaN                NaN  \n",
       "\n",
       "[1915 rows x 14 columns]"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "a2422c5de4c58386",
   "metadata": {},
   "outputs": [],
   "source": [
    "rows_frame.to_csv(output_file_name, sep=\"\\t\", index=false)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cef24b1-d24e-4920-9f9e-fe04ae733e50",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
