
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<link href="https://microbiomedata.github.io/submission-schema/HostAssociatedInterface/" rel="canonical"/>
<link href="../assets/images/favicon.png" rel="icon"/>
<meta content="mkdocs-1.6.1, mkdocs-material-9.5.48" name="generator"/>
<title>Class: Host Associated (HostAssociatedInterface) - submission-schema</title>
<link href="../assets/stylesheets/main.6f8fc17f.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
</head>
<body dir="ltr">
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#class-host-associated-hostassociatedinterface">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="submission-schema" class="md-header__button md-logo" data-md-component="logo" href=".." title="submission-schema">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"></path></svg>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            submission-schema
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Class: Host Associated (HostAssociatedInterface)
            
          </span>
</div>
</div>
</div>
<script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" title="Clear" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="" tabindex="0">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list" role="presentation"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/submission-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="submission-schema" class="md-nav__button md-logo" data-md-component="logo" href=".." title="submission-schema">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"></path></svg>
</a>
    submission-schema
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/submission-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="..">
<span class="md-ellipsis">
    Index
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../about/">
<span class="md-ellipsis">
    About
  </span>
</a>
</li>
<li class="md-nav__item md-nav__item--section md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
<span class="md-ellipsis">
    DataHarmonizer Playground
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
            DataHarmonizer Playground
          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="https://microbiomedata.github.io/submission-schema/_playground/?template=nmdc_submission_schema/SoilInterface">
<span class="md-ellipsis">
    SoilInterface
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="https://microbiomedata.github.io/submission-schema/_playground/?template=nmdc_submission_schema/WaterInterface">
<span class="md-ellipsis">
    WaterInterface
  </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#inheritance">
<span class="md-ellipsis">
      Inheritance
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#slots">
<span class="md-ellipsis">
      Slots
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#usages">
<span class="md-ellipsis">
      Usages
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#identifier-and-mapping-information">
<span class="md-ellipsis">
      Identifier and Mapping Information
    </span>
</a>
<nav aria-label="Identifier and Mapping Information" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#annotations">
<span class="md-ellipsis">
      Annotations
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#schema-source">
<span class="md-ellipsis">
      Schema Source
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#mappings">
<span class="md-ellipsis">
      Mappings
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#linkml-source">
<span class="md-ellipsis">
      LinkML Source
    </span>
</a>
<nav aria-label="LinkML Source" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#direct">
<span class="md-ellipsis">
      Direct
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#induced">
<span class="md-ellipsis">
      Induced
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1 id="class-host-associated-hostassociatedinterface">Class: Host Associated (HostAssociatedInterface)</h1>
<p><em>host_associated dh_interface</em></p>
<p>URI: <a href="https://example.com/nmdc_sub_schema/HostAssociatedInterface">nmdc_sub_schema:HostAssociatedInterface</a></p>
<div class="mermaid"> classDiagram
    class HostAssociatedInterface
    click HostAssociatedInterface href "../HostAssociatedInterface/"
      DhMultiviewCommonColumnsMixin &lt;|-- HostAssociatedInterface
        click DhMultiviewCommonColumnsMixin href "../DhMultiviewCommonColumnsMixin/"
      SampIdNewTermsMixin &lt;|-- HostAssociatedInterface
        click SampIdNewTermsMixin href "../SampIdNewTermsMixin/"
      DhInterface &lt;|-- HostAssociatedInterface
        click DhInterface href "../DhInterface/"

      HostAssociatedInterface : alt

      HostAssociatedInterface : analysis_type





        HostAssociatedInterface --&gt; "1..*" AnalysisTypeEnum : analysis_type
        click AnalysisTypeEnum href "../AnalysisTypeEnum/"



      HostAssociatedInterface : ances_data

      HostAssociatedInterface : biol_stat





        HostAssociatedInterface --&gt; "0..1" BiolStatEnum : biol_stat
        click BiolStatEnum href "../BiolStatEnum/"



      HostAssociatedInterface : blood_press_diast

      HostAssociatedInterface : blood_press_syst

      HostAssociatedInterface : chem_administration

      HostAssociatedInterface : collection_date

      HostAssociatedInterface : depth

      HostAssociatedInterface : ecosystem





        HostAssociatedInterface --&gt; "0..1 _recommended_" EcosystemEnum : ecosystem
        click EcosystemEnum href "../EcosystemEnum/"



      HostAssociatedInterface : ecosystem_category





        HostAssociatedInterface --&gt; "0..1 _recommended_" EcosystemCategoryEnum : ecosystem_category
        click EcosystemCategoryEnum href "../EcosystemCategoryEnum/"



      HostAssociatedInterface : ecosystem_subtype





        HostAssociatedInterface --&gt; "0..1 _recommended_" EcosystemSubtypeEnum : ecosystem_subtype
        click EcosystemSubtypeEnum href "../EcosystemSubtypeEnum/"



      HostAssociatedInterface : ecosystem_type





        HostAssociatedInterface --&gt; "0..1 _recommended_" EcosystemTypeEnum : ecosystem_type
        click EcosystemTypeEnum href "../EcosystemTypeEnum/"



      HostAssociatedInterface : elev

      HostAssociatedInterface : env_broad_scale

      HostAssociatedInterface : env_local_scale

      HostAssociatedInterface : env_medium

      HostAssociatedInterface : experimental_factor

      HostAssociatedInterface : genetic_mod

      HostAssociatedInterface : geo_loc_name

      HostAssociatedInterface : gravidity

      HostAssociatedInterface : host_age

      HostAssociatedInterface : host_body_habitat

      HostAssociatedInterface : host_body_product

      HostAssociatedInterface : host_body_site

      HostAssociatedInterface : host_body_temp

      HostAssociatedInterface : host_color

      HostAssociatedInterface : host_common_name

      HostAssociatedInterface : host_diet

      HostAssociatedInterface : host_disease_stat

      HostAssociatedInterface : host_dry_mass

      HostAssociatedInterface : host_family_relation

      HostAssociatedInterface : host_genotype

      HostAssociatedInterface : host_growth_cond

      HostAssociatedInterface : host_height

      HostAssociatedInterface : host_last_meal

      HostAssociatedInterface : host_length

      HostAssociatedInterface : host_life_stage

      HostAssociatedInterface : host_phenotype

      HostAssociatedInterface : host_sex

      HostAssociatedInterface : host_shape

      HostAssociatedInterface : host_subject_id

      HostAssociatedInterface : host_subspecf_genlin

      HostAssociatedInterface : host_substrate

      HostAssociatedInterface : host_symbiont

      HostAssociatedInterface : host_taxid

      HostAssociatedInterface : host_tot_mass

      HostAssociatedInterface : lat_lon

      HostAssociatedInterface : organism_count

      HostAssociatedInterface : oxy_stat_samp





        HostAssociatedInterface --&gt; "0..1" OxyStatSampEnum : oxy_stat_samp
        click OxyStatSampEnum href "../OxyStatSampEnum/"



      HostAssociatedInterface : perturbation

      HostAssociatedInterface : salinity

      HostAssociatedInterface : samp_capt_status





        HostAssociatedInterface --&gt; "0..1" SampCaptStatusEnum : samp_capt_status
        click SampCaptStatusEnum href "../SampCaptStatusEnum/"



      HostAssociatedInterface : samp_collec_device

      HostAssociatedInterface : samp_collec_method

      HostAssociatedInterface : samp_dis_stage





        HostAssociatedInterface --&gt; "0..1" SampDisStageEnum : samp_dis_stage
        click SampDisStageEnum href "../SampDisStageEnum/"



      HostAssociatedInterface : samp_mat_process

      HostAssociatedInterface : samp_name

      HostAssociatedInterface : samp_size

      HostAssociatedInterface : samp_store_dur

      HostAssociatedInterface : samp_store_loc

      HostAssociatedInterface : samp_store_temp

      HostAssociatedInterface : sample_link

      HostAssociatedInterface : size_frac

      HostAssociatedInterface : source_mat_id

      HostAssociatedInterface : specific_ecosystem





        HostAssociatedInterface --&gt; "0..1 _recommended_" SpecificEcosystemEnum : specific_ecosystem
        click SpecificEcosystemEnum href "../SpecificEcosystemEnum/"



      HostAssociatedInterface : temp


</div>
<h2 id="inheritance">Inheritance</h2>
<ul>
<li><a href="../DhInterface/">DhInterface</a><ul>
<li><strong>HostAssociatedInterface</strong> [ <a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a> <a href="../SampIdNewTermsMixin/">SampIdNewTermsMixin</a>]</li>
</ul>
</li>
</ul>
<h2 id="slots">Slots</h2>
<table>
<thead>
<tr>
<th>Name</th>
<th>Cardinality and Range</th>
<th>Description</th>
<th>Inheritance</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../alt/">alt</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../Float/">Float</a></td>
<td>Heights of objects such as airplanes, space shuttles, rockets, atmospheric ba...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ances_data/">ances_data</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about either pedigree or other ancestral information description ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../biol_stat/">biol_stat</a></td>
<td>0..1 <br/> <a href="../BiolStatEnum/">BiolStatEnum</a></td>
<td>The level of genome modification</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../blood_press_diast/">blood_press_diast</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Resting diastolic blood pressure, measured as mm mercury</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../blood_press_syst/">blood_press_syst</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Resting systolic blood pressure, measured as mm mercury</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../chem_administration/">chem_administration</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>List of chemical compounds administered to the host or site where sampling oc...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../collection_date/">collection_date</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>The date of sampling</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../depth/">depth</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>The vertical distance below local surface</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ecosystem/">ecosystem</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../EcosystemEnum/">EcosystemEnum</a></td>
<td>An ecosystem is a combination of a physical environment (abiotic factors) and...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ecosystem_category/">ecosystem_category</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../EcosystemCategoryEnum/">EcosystemCategoryEnum</a></td>
<td>Ecosystem categories represent divisions within the ecosystem based on specif...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ecosystem_subtype/">ecosystem_subtype</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../EcosystemSubtypeEnum/">EcosystemSubtypeEnum</a></td>
<td>Ecosystem subtypes represent further subdivision of Ecosystem types into more...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ecosystem_type/">ecosystem_type</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../EcosystemTypeEnum/">EcosystemTypeEnum</a></td>
<td>Ecosystem types represent things having common characteristics within the Eco...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../elev/">elev</a></td>
<td>1 <br/> <a href="../Float/">Float</a></td>
<td>Elevation of the sampling site is its height above a fixed reference point, m...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../env_broad_scale/">env_broad_scale</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>In this field, report which major environmental system your sample or specime...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../env_local_scale/">env_local_scale</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>Report the entity or entities which are in the sample or specimen s local vic...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../env_medium/">env_medium</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>Report the environmental material(s) immediately surrounding the sample or sp...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../experimental_factor/">experimental_factor</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Variable aspects of an experiment design that can be used to describe an expe...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../genetic_mod/">genetic_mod</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Genetic modifications of the genome of an organism, which may occur naturally...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../geo_loc_name/">geo_loc_name</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>The geographical origin of the sample as defined by the country or sea name f...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../gravidity/">gravidity</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Whether or not subject is gravid, and if yes date due or date post-conception...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_age/">host_age</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Age of host at the time of sampling; relevant scale depends on species and st...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_body_habitat/">host_body_habitat</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Original body habitat where the sample was obtained from</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_body_product/">host_body_product</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Substance produced by the body, e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_body_site/">host_body_site</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Name of body site where the sample was obtained from, such as a specific orga...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_body_temp/">host_body_temp</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Core body temperature of the host when sample was collected</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_color/">host_color</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The color of host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_common_name/">host_common_name</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Common name of the host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_diet/">host_diet</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Type of diet depending on the host, for animals omnivore, herbivore etc</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_disease_stat/">host_disease_stat</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>List of diseases with which the host has been diagnosed; can include multiple...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_dry_mass/">host_dry_mass</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Measurement of dry mass</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_family_relation/">host_family_relation</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Familial relationships to other hosts in the same study; can include multiple...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_genotype/">host_genotype</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Observed genotype</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_growth_cond/">host_growth_cond</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Literature reference giving growth conditions of the host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_height/">host_height</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The height of subject</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_last_meal/">host_last_meal</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Content of last meal and time since feeding; can include multiple values</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_length/">host_length</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The length of subject</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_life_stage/">host_life_stage</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Description of life stage of host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_phenotype/">host_phenotype</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Phenotype of human or other host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_sex/">host_sex</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Gender or physical sex of the host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_shape/">host_shape</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Morphological shape of host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_subject_id/">host_subject_id</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>A unique identifier by which each subject can be referred to, de-identified</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_subspecf_genlin/">host_subspecf_genlin</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about the genetic distinctness of the host organism below the sub...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_substrate/">host_substrate</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The growth substrate of the host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_symbiont/">host_symbiont</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The taxonomic name of the organism(s) found living in mutualistic, commensali...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_taxid/">host_taxid</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>NCBI taxon id of the host, e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_tot_mass/">host_tot_mass</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Total mass of the host at collection, the unit depends on host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../lat_lon/">lat_lon</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>The geographical origin of the sample as defined by latitude and longitude</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../organism_count/">organism_count</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Total cell count of any organism (or group of organisms) per gram, volume or ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../oxy_stat_samp/">oxy_stat_samp</a></td>
<td>0..1 <br/> <a href="../OxyStatSampEnum/">OxyStatSampEnum</a></td>
<td>Oxygenation status of sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../perturbation/">perturbation</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Type of perturbation, e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../salinity/">salinity</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The total concentration of all dissolved salts in a liquid or solid sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_capt_status/">samp_capt_status</a></td>
<td>0..1 <br/> <a href="../SampCaptStatusEnum/">SampCaptStatusEnum</a></td>
<td>Reason for the sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_collec_device/">samp_collec_device</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>The device used to collect an environmental sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_collec_method/">samp_collec_method</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The method employed for collecting the sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_dis_stage/">samp_dis_stage</a></td>
<td>0..1 <br/> <a href="../SampDisStageEnum/">SampDisStageEnum</a></td>
<td>Stage of the disease at the time of sample collection, e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_mat_process/">samp_mat_process</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>A brief description of any processing applied to the sample during or after r...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_size/">samp_size</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The total amount (volume or mass) of sample collected</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_store_dur/">samp_store_dur</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Duration for which the sample was stored</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_store_loc/">samp_store_loc</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Location at which sample was stored, usually name of a specific freezer/room</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_store_temp/">samp_store_temp</a></td>
<td>1 <br/> <a href="../Float/">Float</a></td>
<td>Temperature at which the sample was stored (degrees are assumed)</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../sample_link/">sample_link</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>A unique identifier to assign parent-child, subsample, or sibling samples</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../size_frac/">size_frac</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Filtering pore size used in sample preparation</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../specific_ecosystem/">specific_ecosystem</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../SpecificEcosystemEnum/">SpecificEcosystemEnum</a></td>
<td>Specific ecosystems represent specific features of the environment like aphot...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../temp/">temp</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Temperature of the sample at the time of sampling</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../analysis_type/">analysis_type</a></td>
<td>1..* <br/> <a href="../AnalysisTypeEnum/">AnalysisTypeEnum</a></td>
<td>Select all the data types associated or available for this biosample</td>
<td><a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a></td>
</tr>
<tr>
<td><a href="../samp_name/">samp_name</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>A local identifier or name that for the material sample collected</td>
<td><a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a></td>
</tr>
<tr>
<td><a href="../source_mat_id/">source_mat_id</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>A globally unique identifier assigned to the parent sample or sample that is ...</td>
<td><a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a></td>
</tr>
</tbody>
</table>
<h2 id="usages">Usages</h2>
<table>
<thead>
<tr>
<th>used by</th>
<th>used in</th>
<th>type</th>
<th>used</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../SampleData/">SampleData</a></td>
<td><a href="../host_associated_data/">host_associated_data</a></td>
<td>range</td>
<td><a href="./">HostAssociatedInterface</a></td>
</tr>
</tbody>
</table>
<h2 id="identifier-and-mapping-information">Identifier and Mapping Information</h2>
<h3 id="annotations">Annotations</h3>
<table>
<thead>
<tr>
<th>property</th>
<th>value</th>
</tr>
</thead>
<tbody>
<tr>
<td>excel_worksheet_name</td>
<td>host-associated</td>
</tr>
</tbody>
</table>
<h3 id="schema-source">Schema Source</h3>
<ul>
<li>from schema: <a href="https://example.com/nmdc_submission_schema">https://example.com/nmdc_submission_schema</a></li>
</ul>
<h2 id="mappings">Mappings</h2>
<table>
<thead>
<tr>
<th>Mapping Type</th>
<th>Mapped Value</th>
</tr>
</thead>
<tbody>
<tr>
<td>self</td>
<td>nmdc_sub_schema:HostAssociatedInterface</td>
</tr>
<tr>
<td>native</td>
<td>nmdc_sub_schema:HostAssociatedInterface</td>
</tr>
</tbody>
</table>
<h2 id="linkml-source">LinkML Source</h2>
<!-- TODO: investigate https://stackoverflow.com/questions/37606292/how-to-create-tabbed-code-blocks-in-mkdocs-or-sphinx -->
<h3 id="direct">Direct</h3>
<details>
<pre><code class="language-yaml">name: HostAssociatedInterface
annotations:
  excel_worksheet_name:
    tag: excel_worksheet_name
    value: host-associated
description: host_associated dh_interface
title: Host Associated
from_schema: https://example.com/nmdc_submission_schema
is_a: DhInterface
mixins:
- DhMultiviewCommonColumnsMixin
- SampIdNewTermsMixin
slots:
- alt
- ances_data
- biol_stat
- blood_press_diast
- blood_press_syst
- chem_administration
- collection_date
- depth
- ecosystem
- ecosystem_category
- ecosystem_subtype
- ecosystem_type
- elev
- env_broad_scale
- env_local_scale
- env_medium
- experimental_factor
- genetic_mod
- geo_loc_name
- gravidity
- host_age
- host_body_habitat
- host_body_product
- host_body_site
- host_body_temp
- host_color
- host_common_name
- host_diet
- host_disease_stat
- host_dry_mass
- host_family_relation
- host_genotype
- host_growth_cond
- host_height
- host_last_meal
- host_length
- host_life_stage
- host_phenotype
- host_sex
- host_shape
- host_subject_id
- host_subspecf_genlin
- host_substrate
- host_symbiont
- host_taxid
- host_tot_mass
- lat_lon
- organism_count
- oxy_stat_samp
- perturbation
- salinity
- samp_capt_status
- samp_collec_device
- samp_collec_method
- samp_dis_stage
- samp_mat_process
- samp_size
- samp_store_dur
- samp_store_loc
- samp_store_temp
- sample_link
- size_frac
- specific_ecosystem
- temp
slot_usage:
  depth:
    name: depth
    examples:
    - value: 0 - 0.1
    - value: '1'
    recommended: true
    pattern: ^[-+]?[0-9]*\.?[0-9]+(\s*-\s*[-+]?[0-9]*\.?[0-9]+)?$
    structured_pattern:
      syntax: ^{float}(\s*-\s*{float})?$
      interpolated: true
  ecosystem:
    name: ecosystem
    range: EcosystemEnum
  ecosystem_category:
    name: ecosystem_category
    range: EcosystemCategoryEnum
  ecosystem_subtype:
    name: ecosystem_subtype
    range: EcosystemSubtypeEnum
  ecosystem_type:
    name: ecosystem_type
    range: EcosystemTypeEnum
  elev:
    name: elev
    required: true
  env_local_scale:
    name: env_local_scale
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
      interpolated: true
  env_medium:
    name: env_medium
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
      interpolated: true
  lat_lon:
    name: lat_lon
    required: true
  samp_store_temp:
    name: samp_store_temp
    required: true
  specific_ecosystem:
    name: specific_ecosystem
    range: SpecificEcosystemEnum

</code></pre>
</details>
<h3 id="induced">Induced</h3>
<details>
<pre><code class="language-yaml">name: HostAssociatedInterface
annotations:
  excel_worksheet_name:
    tag: excel_worksheet_name
    value: host-associated
description: host_associated dh_interface
title: Host Associated
from_schema: https://example.com/nmdc_submission_schema
is_a: DhInterface
mixins:
- DhMultiviewCommonColumnsMixin
- SampIdNewTermsMixin
slot_usage:
  depth:
    name: depth
    examples:
    - value: 0 - 0.1
    - value: '1'
    recommended: true
    pattern: ^[-+]?[0-9]*\.?[0-9]+(\s*-\s*[-+]?[0-9]*\.?[0-9]+)?$
    structured_pattern:
      syntax: ^{float}(\s*-\s*{float})?$
      interpolated: true
  ecosystem:
    name: ecosystem
    range: EcosystemEnum
  ecosystem_category:
    name: ecosystem_category
    range: EcosystemCategoryEnum
  ecosystem_subtype:
    name: ecosystem_subtype
    range: EcosystemSubtypeEnum
  ecosystem_type:
    name: ecosystem_type
    range: EcosystemTypeEnum
  elev:
    name: elev
    required: true
  env_local_scale:
    name: env_local_scale
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
      interpolated: true
  env_medium:
    name: env_medium
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
      interpolated: true
  lat_lon:
    name: lat_lon
    required: true
  samp_store_temp:
    name: samp_store_temp
    required: true
  specific_ecosystem:
    name: specific_ecosystem
    range: SpecificEcosystemEnum
attributes:
  alt:
    name: alt
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: meter
      storage_units:
        tag: storage_units
        value: m
    description: Heights of objects such as airplanes, space shuttles, rockets, atmospheric
      balloons and heights of places such as atmospheric layers and clouds. It is
      used to measure the height of an object which is above the earth's surface.
      In this context, the altitude measurement is the vertical distance between the
      earth's surface above sea level and the sampled position in the air
    title: altitude
    comments:
    - Value must be reported in m. Provide the numeric portion only.
    examples:
    - value: '100'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 26
    slot_uri: MIXS:0000094
    alias: alt
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    slot_group: mixs_modified_section
    range: float
    recommended: true
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
    unit:
      ucum_code: m
  ances_data:
    name: ances_data
    description: Information about either pedigree or other ancestral information
      description (e.g. parental variety in case of mutant or selection), e.g. A/3*B
      (meaning [(A x B) x B] x B)
    title: ancestral data
    examples:
    - value: A/3*B
    from_schema: https://example.com/nmdc_submission_schema
    rank: 237
    slot_uri: MIXS:0000247
    alias: ances_data
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  biol_stat:
    name: biol_stat
    description: The level of genome modification
    title: biological status
    examples:
    - value: natural
    from_schema: https://example.com/nmdc_submission_schema
    rank: 239
    keywords:
    - status
    slot_uri: MIXS:0000858
    alias: biol_stat
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: BiolStatEnum
  blood_press_diast:
    name: blood_press_diast
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: millimeter mercury
      storage_units:
        tag: storage_units
        value: mm[Hg]
    description: Resting diastolic blood pressure, measured as mm mercury
    title: host blood pressure diastolic
    comments:
    - Value must be reported in mm[Hg]. Provide the numeric portion only.
    from_schema: https://example.com/nmdc_submission_schema
    rank: 299
    keywords:
    - host
    - host.
    - pressure
    slot_uri: MIXS:0000258
    alias: blood_press_diast
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: float
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
    unit:
      ucum_code: mm[Hg]
  blood_press_syst:
    name: blood_press_syst
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: millimeter mercury
      storage_units:
        tag: storage_units
        value: mm[Hg]
    description: Resting systolic blood pressure, measured as mm mercury
    title: host blood pressure systolic
    comments:
    - Value must be reported in mm[Hg]. Provide the numeric portion only.
    from_schema: https://example.com/nmdc_submission_schema
    rank: 300
    keywords:
    - host
    - host.
    - pressure
    slot_uri: MIXS:0000259
    alias: blood_press_syst
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: float
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
    unit:
      ucum_code: mm[Hg]
  chem_administration:
    name: chem_administration
    annotations:
      Expected_value:
        tag: Expected_value
        value: CHEBI;timestamp
    description: List of chemical compounds administered to the host or site where
      sampling occurred, and when (e.g. Antibiotics, n fertilizer, air filter); can
      include multiple compounds. For chemical entities of biological interest ontology
      (chebi) (v 163), http://purl.bioontology.org/ontology/chebi
    title: chemical administration
    examples:
    - value: agar [CHEBI:2509];2018-05-11|agar [CHEBI:2509];2018-05-22
    - value: agar [CHEBI:2509];2018-05
    from_schema: https://example.com/nmdc_submission_schema
    rank: 17
    keywords:
    - administration
    string_serialization: '{termLabel} [{termID}];{timestamp}'
    slot_uri: MIXS:0000751
    alias: chem_administration
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
    multivalued: false
    pattern: ^(\S+.*\S+ \[[A-za-z]+:\d+\];[12]\d{3}(?:(?:-(?:0[1-9]|1[0-2]))(?:-(?:0[1-9]|[12]\d|3[01]))?)?\|)*(\S+.*\S+
      \[[A-za-z]+:\d+\];[12]\d{3}(?:(?:-(?:0[1-9]|1[0-2]))(?:-(?:0[1-9]|[12]\d|3[01]))?)?)$
  collection_date:
    name: collection_date
    description: The date of sampling
    title: collection date
    examples:
    - value: '2021-04-15'
    - value: 2021-04
    - value: '2021'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 3
    keywords:
    - date
    slot_uri: MIXS:0000011
    alias: collection_date
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    required: true
    pattern: ^[12]\d{3}(?:(?:-(?:0[1-9]|1[0-2]))(?:-(?:0[1-9]|[12]\d|3[01]))?)?$
  depth:
    name: depth
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: meter
      storage_units:
        tag: storage_units
        value: m
    description: The vertical distance below local surface. For sediment or soil samples
      depth is measured from sediment or soil surface, respectively. Depth can be
      reported as an interval for subsurface samples
    title: depth, meters
    comments:
    - All depths must be reported in meters. Provide the numerical portion only.
    examples:
    - value: 0 - 0.1
    - value: '1'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 9
    keywords:
    - depth
    slot_uri: MIXS:0000018
    alias: depth
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
    pattern: ^[-+]?[0-9]*\.?[0-9]+(\s*-\s*[-+]?[0-9]*\.?[0-9]+)?$
    structured_pattern:
      syntax: ^{float}(\s*-\s*{float})?$
      interpolated: true
    unit:
      ucum_code: m
  ecosystem:
    name: ecosystem
    description: An ecosystem is a combination of a physical environment (abiotic
      factors) and all the organisms (biotic factors) that interact with this environment.
      Ecosystem is in position 1/5 in a GOLD path.
    comments:
    - The abiotic factors play a profound role on the type and composition of organisms
      in a given environment. The GOLD Ecosystem at the top of the five-level classification
      system is aimed at capturing the broader environment from which an organism
      or environmental sample is collected. The three broad groups under Ecosystem
      are Environmental, Host-associated, and Engineered. They represent samples collected
      from a natural environment or from another organism or from engineered environments
      like bioreactors respectively.
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - https://gold.jgi.doe.gov/help
    rank: 9
    is_a: gold_path_field
    alias: ecosystem
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: EcosystemEnum
    recommended: true
  ecosystem_category:
    name: ecosystem_category
    description: Ecosystem categories represent divisions within the ecosystem based
      on specific characteristics of the environment from where an organism or sample
      is isolated. Ecosystem category is in position 2/5 in a GOLD path.
    comments:
    - The Environmental ecosystem (for example) is divided into Air, Aquatic and Terrestrial.
      Ecosystem categories for Host-associated samples can be individual hosts or
      phyla and for engineered samples it may be manipulated environments like bioreactors,
      solid waste etc.
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - https://gold.jgi.doe.gov/help
    rank: 10
    is_a: gold_path_field
    alias: ecosystem_category
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: EcosystemCategoryEnum
    recommended: true
  ecosystem_subtype:
    name: ecosystem_subtype
    description: Ecosystem subtypes represent further subdivision of Ecosystem types
      into more distinct subtypes. Ecosystem subtype is in position 4/5 in a GOLD
      path.
    comments:
    - Ecosystem Type Marine (Environmental -&gt; Aquatic -&gt; Marine) is further divided
      (for example) into Intertidal zone, Coastal, Pelagic, Intertidal zone etc. in
      the Ecosystem subtype category.
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - https://gold.jgi.doe.gov/help
    rank: 12
    is_a: gold_path_field
    alias: ecosystem_subtype
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: EcosystemSubtypeEnum
    recommended: true
  ecosystem_type:
    name: ecosystem_type
    description: Ecosystem types represent things having common characteristics within
      the Ecosystem Category. These common characteristics based grouping is still
      broad but specific to the characteristics of a given environment. Ecosystem
      type is in position 3/5 in a GOLD path.
    comments:
    - The Aquatic ecosystem category (for example) may have ecosystem types like Marine
      or Thermal springs etc. Ecosystem category Air may have Indoor air or Outdoor
      air as different Ecosystem Types. In the case of Host-associated samples, ecosystem
      type can represent Respiratory system, Digestive system, Roots etc.
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - https://gold.jgi.doe.gov/help
    rank: 11
    is_a: gold_path_field
    alias: ecosystem_type
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: EcosystemTypeEnum
    recommended: true
  elev:
    name: elev
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: meter
    description: Elevation of the sampling site is its height above a fixed reference
      point, most commonly the mean sea level. Elevation is mainly used when referring
      to points on the earth's surface, while altitude is used for points above the
      surface, such as an aircraft in flight or a spacecraft in orbit
    title: elevation, meters
    comments:
    - All elevations must be reported in meters. Provide the numerical portion only.
    - Please use https://www.advancedconverter.com/map-tools/find-altitude-by-coordinates,
      if needed, to help estimate the elevation based on latitude and longitude coordinates.
    examples:
    - value: '225'
    - value: '0'
    - value: '1250'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 6
    keywords:
    - elevation
    slot_uri: MIXS:0000093
    alias: elev
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: float
    required: true
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
    unit:
      ucum_code: m
  env_broad_scale:
    name: env_broad_scale
    annotations:
      tooltip:
        tag: tooltip
        value: The biome or major environmental system where the sample or specimen
          originated. Choose values from subclasses of the 'biome' class [ENVO:00000428]
          in the Environment Ontology (ENVO). For host-associated or plant-associated
          samples, use terms from the UBERON or Plant Ontology to describe the broad
          anatomical or morphological context
    description: 'In this field, report which major environmental system your sample
      or specimen came from. The systems identified should have a coarse spatial grain,
      to provide the general environmental context of where the sampling was done
      (e.g. were you in the desert or a rainforest?). We recommend using subclasses
      of ENVO''s biome class: http://purl.obolibrary.org/obo/ENVO_00000428. Format
      (one term): termLabel [termID], Format (multiple terms): termLabel [termID]|termLabel
      [termID]|termLabel [termID]. Example: Annotating a water sample from the photic
      zone in middle of the Atlantic Ocean, consider: oceanic epipelagic zone biome
      [ENVO:01000033]. Example: Annotating a sample from the Amazon rainforest consider:
      tropical moist broadleaf forest biome [ENVO:01000228]. If needed, request new
      terms on the ENVO tracker, identified here: http://www.obofoundry.org/ontology/envo.html'
    title: broad-scale environmental context
    examples:
    - value: oceanic epipelagic zone biome [ENVO:01000035]
    from_schema: https://example.com/nmdc_submission_schema
    rank: 6
    keywords:
    - context
    - environmental
    is_a: mixs_env_triad_field
    slot_uri: MIXS:0000012
    alias: env_broad_scale
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: string
    required: true
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[ENVO:\d{7,8}\]$
    structured_pattern:
      syntax: ^{termLabel} \[ENVO:\d{7,8}\]$
      interpolated: true
  env_local_scale:
    name: env_local_scale
    annotations:
      Expected_value:
        tag: Expected_value
        value: Environmental entities having causal influences upon the entity at
          time of sampling
      tooltip:
        tag: tooltip
        value: The specific environmental entities or features near the sample or
          specimen that significantly influence its characteristics or composition.
          These entities are typically smaller in scale than the broad environmental
          context. Values for this field should be countable, material nouns and must
          be chosen from subclasses of BFO:0000040 (material entity) that appear in
          the Environment Ontology (ENVO). For host-associated or plant-associated
          samples, use terms from the UBERON or Plant Ontology to describe specific
          anatomical structures or plant parts.
    description: 'Report the entity or entities which are in the sample or specimen
      s local vicinity and which you believe have significant causal influences on
      your sample or specimen. We recommend using EnvO terms which are of smaller
      spatial grain than your entry for env_broad_scale. Terms, such as anatomical
      sites, from other OBO Library ontologies which interoperate with EnvO (e.g.
      UBERON) are accepted in this field. EnvO documentation about how to use the
      field: https://github.com/EnvironmentOntology/envo/wiki/Using-ENVO-with-MIxS'
    title: local environmental context
    examples:
    - value: canopy [ENVO:00000047]
    - value: herb and fern layer [ENVO:01000337]
    - value: understory [ENVO:01000335]
    - value: shrub layer [ENVO:01000336]
    from_schema: https://example.com/nmdc_submission_schema
    rank: 7
    keywords:
    - context
    - environmental
    is_a: mixs_env_triad_field
    slot_uri: MIXS:0000013
    alias: env_local_scale
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: string
    required: true
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
      interpolated: true
  env_medium:
    name: env_medium
    annotations:
      tooltip:
        tag: tooltip
        value: The predominant environmental material or substrate that directly surrounds
          or hosts the sample or specimen at the time of sampling. Choose values from
          subclasses of the 'environmental material' class [ENVO:00010483] in the
          Environment Ontology (ENVO). Values for this field should be measurable
          or mass material nouns, representing continuous environmental materials.
          For host-associated or plant-associated samples, use terms from the UBERON
          or Plant Ontology to indicate a tissue, organ, or plant structure
    description: 'Report the environmental material(s) immediately surrounding the
      sample or specimen at the time of sampling. We recommend using subclasses of
      ''environmental material'' (http://purl.obolibrary.org/obo/ENVO_00010483). EnvO
      documentation about how to use the field: https://github.com/EnvironmentOntology/envo/wiki/Using-ENVO-with-MIxS
      . Terms from other OBO ontologies are permissible as long as they reference
      mass/volume nouns (e.g. air, water, blood) and not discrete, countable entities
      (e.g. a tree, a leaf, a table top)'
    title: environmental medium
    examples:
    - value: soil [ENVO:00001998]
    from_schema: https://example.com/nmdc_submission_schema
    rank: 8
    keywords:
    - environmental
    is_a: mixs_env_triad_field
    slot_uri: MIXS:0000014
    alias: env_medium
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: string
    required: true
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|UBERON:\d{7})\]$
      interpolated: true
  experimental_factor:
    name: experimental_factor
    annotations:
      Expected_value:
        tag: Expected_value
        value: text or EFO and/or OBI
    description: Variable aspects of an experiment design that can be used to describe
      an experiment, or set of experiments, in an increasingly detailed manner. This
      field accepts ontology terms from Experimental Factor Ontology (EFO) and/or
      Ontology for Biomedical Investigations (OBI)
    title: experimental factor
    examples:
    - value: time series design [EFO:0001779]
    from_schema: https://example.com/nmdc_submission_schema
    rank: 12
    keywords:
    - experimental
    - factor
    string_serialization: '{termLabel} [{termID}]|{text}'
    slot_uri: MIXS:0000008
    alias: experimental_factor
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_section
    range: string
    multivalued: false
    pattern: ^\S+.*\S+ \[[a-zA-Z]{2,}:\d+\]$
  genetic_mod:
    name: genetic_mod
    description: Genetic modifications of the genome of an organism, which may occur
      naturally by spontaneous mutation, or be introduced by some experimental means,
      e.g. specification of a transgene or the gene knocked-out or details of transient
      transfection
    title: genetic modification
    examples:
    - value: PMID:19497774
    from_schema: https://example.com/nmdc_submission_schema
    rank: 244
    slot_uri: MIXS:0000859
    alias: genetic_mod
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^(PMID:\d+|doi:10.\d{2,9}/.*|https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_\+.~#?&amp;\/=]*))$
    structured_pattern:
      syntax: ^({PMID}|{DOI}|{URL})$
      interpolated: true
  geo_loc_name:
    name: geo_loc_name
    description: The geographical origin of the sample as defined by the country or
      sea name followed by specific region name.
    title: geographic location (country and/or sea,region)
    examples:
    - value: 'USA: Maryland, Bethesda'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 4
    keywords:
    - geographic
    - location
    slot_uri: MIXS:0000010
    alias: geo_loc_name
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    required: true
    pattern: '^([^\s-]{1,2}|[^\s-]+.+[^\s-]+): ([^\s-]{1,2}|[^\s-]+.+[^\s-]+), ([^\s-]{1,2}|[^\s-]+.+[^\s-]+)$'
    structured_pattern:
      syntax: '^{country}: {region}, {specific_location}$'
      interpolated: true
      partial_match: true
  gravidity:
    name: gravidity
    description: Whether or not subject is gravid, and if yes date due or date post-conception,
      specifying which is used
    title: gravidity
    examples:
    - value: yes;due date:2018-05-11
    from_schema: https://example.com/nmdc_submission_schema
    rank: 312
    string_serialization: '{boolean};{timestamp}'
    slot_uri: MIXS:0000875
    alias: gravidity
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_age:
    name: host_age
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: year, day, hour
      storage_units:
        tag: storage_units
        value: a|d|h
    description: Age of host at the time of sampling; relevant scale depends on species
      and study, e.g. Could be seconds for amoebae or centuries for trees
    title: host age
    comments:
    - 'Value must be reported in one of the following units: a, d, h. Provide the
      numeric portion followed by a space and the unit.'
    examples:
    - value: 10 d
    from_schema: https://example.com/nmdc_submission_schema
    rank: 249
    keywords:
    - age
    - host
    - host.
    slot_uri: MIXS:0000255
    alias: host_age
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  host_body_habitat:
    name: host_body_habitat
    description: Original body habitat where the sample was obtained from
    title: host body habitat
    from_schema: https://example.com/nmdc_submission_schema
    rank: 319
    keywords:
    - body
    - habitat
    - host
    - host.
    slot_uri: MIXS:0000866
    alias: host_body_habitat
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_body_product:
    name: host_body_product
    annotations:
      Expected_value:
        tag: Expected_value
        value: FMA or UBERON
    description: Substance produced by the body, e.g. Stool, mucus, where the sample
      was obtained from. Use terms from the foundational model of anatomy ontology
      (fma) or Uber-anatomy ontology (UBERON)
    title: host body product
    examples:
    - value: mucus [UBERON:0000912]
    from_schema: https://example.com/nmdc_submission_schema
    rank: 320
    keywords:
    - body
    - host
    - host.
    - product
    string_serialization: '{termLabel} [{termID}]'
    slot_uri: MIXS:0000888
    alias: host_body_product
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^\S+.*\S+ \[[a-zA-Z]{2,}:\d+\]$
  host_body_site:
    name: host_body_site
    annotations:
      Expected_value:
        tag: Expected_value
        value: FMA or UBERON
    description: Name of body site where the sample was obtained from, such as a specific
      organ or tissue (tongue, lung etc...). Use terms from the foundational model
      of anatomy ontology (fma) or the Uber-anatomy ontology (UBERON)
    title: host body site
    from_schema: https://example.com/nmdc_submission_schema
    rank: 321
    keywords:
    - body
    - host
    - site
    string_serialization: '{termLabel} [{termID}]'
    slot_uri: MIXS:0000867
    alias: host_body_site
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^\S+.*\S+ \[[a-zA-Z]{2,}:\d+\]$
  host_body_temp:
    name: host_body_temp
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: degree Celsius
      storage_units:
        tag: storage_units
        value: Cel
    description: Core body temperature of the host when sample was collected
    title: host body temperature
    comments:
    - Value must be reported in degrees Celsius. Provide the numeric portion only.
    examples:
    - value: '15'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 322
    keywords:
    - body
    - host
    - host.
    - temperature
    slot_uri: MIXS:0000274
    alias: host_body_temp
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: float
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
    unit:
      descriptive_name: degrees Celsius
      ucum_code: Cel
  host_color:
    name: host_color
    description: The color of host
    title: host color
    from_schema: https://example.com/nmdc_submission_schema
    rank: 323
    keywords:
    - host
    - host.
    slot_uri: MIXS:0000260
    alias: host_color
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_common_name:
    name: host_common_name
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: ''
    description: Common name of the host
    title: host common name
    from_schema: https://example.com/nmdc_submission_schema
    rank: 250
    keywords:
    - host
    - host.
    slot_uri: MIXS:0000248
    alias: host_common_name
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_diet:
    name: host_diet
    description: Type of diet depending on the host, for animals omnivore, herbivore
      etc., for humans high-fat, meditteranean etc.; can include multiple diet types
    title: host diet
    from_schema: https://example.com/nmdc_submission_schema
    rank: 324
    keywords:
    - diet
    - host
    - host.
    slot_uri: MIXS:0000869
    alias: host_diet
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  host_disease_stat:
    name: host_disease_stat
    annotations:
      Expected_value:
        tag: Expected_value
        value: disease name or Disease Ontology term
    description: List of diseases with which the host has been diagnosed; can include
      multiple diagnoses. The value of the field depends on host; for humans the terms
      should be chosen from the DO (Human Disease Ontology) at https://www.disease-ontology.org,
      non-human host diseases are free text
    title: host disease status
    from_schema: https://example.com/nmdc_submission_schema
    rank: 390
    keywords:
    - disease
    - host
    - host.
    - status
    string_serialization: '{termLabel} [{termID}]|{text}'
    slot_uri: MIXS:0000031
    alias: host_disease_stat
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_dry_mass:
    name: host_dry_mass
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: kilogram, gram
      storage_units:
        tag: storage_units
        value: g|kg
    description: Measurement of dry mass
    title: host dry mass
    comments:
    - 'Value must be reported in one of the following units: g, kg. Provide the numeric
      portion followed by a space and the unit.'
    examples:
    - value: 500 g
    from_schema: https://example.com/nmdc_submission_schema
    rank: 251
    keywords:
    - dry
    - host
    - host.
    - mass
    slot_uri: MIXS:0000257
    alias: host_dry_mass
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  host_family_relation:
    name: host_family_relation
    annotations:
      Expected_value:
        tag: Expected_value
        value: relationship type;arbitrary identifier
    description: Familial relationships to other hosts in the same study; can include
      multiple relationships.
    title: host family relationship
    comments:
    - Legacy MIxS slot removed from GSC MIxS v6.2.2. Retained for submission-schema
      compatibility.
    examples:
    - value: offspring;Mussel25
    from_schema: https://example.com/nmdc_submission_schema
    rank: 325
    slot_uri: MIXS:0000872
    alias: host_family_relation
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  host_genotype:
    name: host_genotype
    description: Observed genotype
    title: host genotype
    from_schema: https://example.com/nmdc_submission_schema
    rank: 252
    keywords:
    - host
    - host.
    slot_uri: MIXS:0000365
    alias: host_genotype
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_growth_cond:
    name: host_growth_cond
    description: Literature reference giving growth conditions of the host
    title: host growth conditions
    examples:
    - value: https://academic.oup.com/icesjms/article/68/2/349/617247
    from_schema: https://example.com/nmdc_submission_schema
    rank: 326
    keywords:
    - condition
    - growth
    - host
    - host.
    slot_uri: MIXS:0000871
    alias: host_growth_cond
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^(PMID:\d+|doi:10.\d{2,9}/.*|https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_\+.~#?&amp;\/=]*)|.*)$
    structured_pattern:
      syntax: ^({PMID}|{DOI}|{URL}|{text})$
      interpolated: true
      partial_match: true
  host_height:
    name: host_height
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: centimeter, millimeter, meter
      storage_units:
        tag: storage_units
        value: cm|m|mm
    description: The height of subject
    title: host height
    comments:
    - 'Value must be reported in one of the following units: cm, m, mm. Provide the
      numeric portion followed by a space and the unit.'
    examples:
    - value: 0.1 m
    from_schema: https://example.com/nmdc_submission_schema
    rank: 253
    keywords:
    - height
    - host
    - host.
    slot_uri: MIXS:0000264
    alias: host_height
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  host_last_meal:
    name: host_last_meal
    annotations:
      Expected_value:
        tag: Expected_value
        value: content;duration
    description: Content of last meal and time since feeding; can include multiple
      values
    title: host last meal
    from_schema: https://example.com/nmdc_submission_schema
    rank: 327
    keywords:
    - host
    - host.
    string_serialization: '{text};{duration}'
    slot_uri: MIXS:0000870
    alias: host_last_meal
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  host_length:
    name: host_length
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: centimeter, millimeter, meter
      storage_units:
        tag: storage_units
        value: cm|m|mm
    description: The length of subject
    title: host length
    comments:
    - 'Value must be reported in one of the following units: cm, m, mm. Provide the
      numeric portion followed by a space and the unit.'
    examples:
    - value: 1 m
    from_schema: https://example.com/nmdc_submission_schema
    rank: 254
    keywords:
    - host
    - host.
    - length
    slot_uri: MIXS:0000256
    alias: host_length
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  host_life_stage:
    name: host_life_stage
    annotations:
      Expected_value:
        tag: Expected_value
        value: stage
    description: Description of life stage of host
    title: host life stage
    from_schema: https://example.com/nmdc_submission_schema
    rank: 255
    keywords:
    - host
    - host.
    - life
    slot_uri: MIXS:0000251
    alias: host_life_stage
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_phenotype:
    name: host_phenotype
    annotations:
      Expected_value:
        tag: Expected_value
        value: PATO or HP
    description: Phenotype of human or other host. Use terms from the phenotypic quality
      ontology (pato) or the Human Phenotype Ontology (HP)
    title: host phenotype
    from_schema: https://example.com/nmdc_submission_schema
    rank: 256
    keywords:
    - host
    - host.
    string_serialization: '{termLabel} [{termID}]'
    slot_uri: MIXS:0000874
    alias: host_phenotype
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^\S+.*\S+ \[[a-zA-Z]{2,}:\d+\]$
  host_sex:
    name: host_sex
    annotations:
      Expected_value:
        tag: Expected_value
        value: enumeration
    description: Gender or physical sex of the host
    title: host sex
    comments:
    - example of non-binary from Excel sheets does not match any of the enumerated
      values
    from_schema: https://example.com/nmdc_submission_schema
    rank: 328
    keywords:
    - host
    - host.
    string_serialization: '[female|hermaphrodite|non-binary|male|transgender|transgender
      (female to male)|transgender (male to female) |undeclared]'
    slot_uri: MIXS:0000811
    alias: host_sex
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_shape:
    name: host_shape
    description: Morphological shape of host
    title: host shape
    examples:
    - value: round
    from_schema: https://example.com/nmdc_submission_schema
    rank: 329
    keywords:
    - host
    - host.
    slot_uri: MIXS:0000261
    alias: host_shape
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_subject_id:
    name: host_subject_id
    description: A unique identifier by which each subject can be referred to, de-identified
    title: host subject id
    from_schema: https://example.com/nmdc_submission_schema
    rank: 330
    keywords:
    - host
    - host.
    - identifier
    slot_uri: MIXS:0000861
    alias: host_subject_id
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_subspecf_genlin:
    name: host_subspecf_genlin
    annotations:
      Expected_value:
        tag: Expected_value
        value: Genetic lineage below lowest rank of NCBI taxonomy, which is subspecies,
          e.g. serovar, biotype, ecotype, variety, cultivar
    description: Information about the genetic distinctness of the host organism below
      the subspecies level e.g., serovar, serotype, biotype, ecotype, variety, cultivar,
      or any relevant genetic typing schemes like Group I plasmid. Subspecies should
      not be recorded in this term, but in the NCBI taxonomy. Supply both the lineage
      name and the lineage rank separated by a colon, e.g., biovar:abc123
    title: host subspecific genetic lineage
    examples:
    - value: 'serovar:Newport, variety:glabrum, cultivar: Red Delicious'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 257
    keywords:
    - host
    - host.
    - lineage
    string_serialization: '{rank name}:{text}'
    slot_uri: MIXS:0001318
    alias: host_subspecf_genlin
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  host_substrate:
    name: host_substrate
    description: The growth substrate of the host
    title: host substrate
    examples:
    - value: rock
    from_schema: https://example.com/nmdc_submission_schema
    rank: 331
    keywords:
    - host
    - host.
    slot_uri: MIXS:0000252
    alias: host_substrate
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_symbiont:
    name: host_symbiont
    annotations:
      Expected_value:
        tag: Expected_value
        value: species name or common name
    description: The taxonomic name of the organism(s) found living in mutualistic,
      commensalistic, or parasitic symbiosis with the specific host. The sampled symbiont
      can have its own symbionts. For example, parasites may have hyperparasites (=parasites
      of the parasite)
    title: observed host symbionts
    from_schema: https://example.com/nmdc_submission_schema
    rank: 258
    keywords:
    - host
    - host.
    - observed
    - symbiosis
    slot_uri: MIXS:0001298
    alias: host_symbiont
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  host_taxid:
    name: host_taxid
    annotations:
      Expected_value:
        tag: Expected_value
        value: NCBI taxon identifier
    description: NCBI taxon id of the host, e.g. 9606
    title: host taxid
    comments:
    - Homo sapiens [NCBITaxon:9606] would be a reasonable has_raw_value
    from_schema: https://example.com/nmdc_submission_schema
    rank: 259
    keywords:
    - host
    - host.
    - taxon
    slot_uri: MIXS:0000250
    alias: host_taxid
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_tot_mass:
    name: host_tot_mass
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: kilogram, gram
      storage_units:
        tag: storage_units
        value: g|kg
    description: Total mass of the host at collection, the unit depends on host
    title: host total mass
    comments:
    - 'Value must be reported in one of the following units: g, kg. Provide the numeric
      portion followed by a space and the unit.'
    examples:
    - value: 2500 g
    from_schema: https://example.com/nmdc_submission_schema
    rank: 260
    keywords:
    - host
    - host.
    - mass
    - total
    slot_uri: MIXS:0000263
    alias: host_tot_mass
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  lat_lon:
    name: lat_lon
    description: The geographical origin of the sample as defined by latitude and
      longitude. The values should be reported in decimal degrees, limited to 8 decimal
      points, and in WGS84 system
    title: geographic location (latitude and longitude)
    notes:
    - This is currently a required field but it's not clear if this should be required
      for human hosts
    examples:
    - value: 50.586825 6.408977
    from_schema: https://example.com/nmdc_submission_schema
    rank: 5
    keywords:
    - geographic
    - location
    slot_uri: MIXS:0000009
    alias: lat_lon
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    required: true
    pattern: ^(-?((?:[0-8]?[0-9](?:\.\d{0,8})?)|90)) -?[0-9]+(?:\.[0-9]{0,8})?$|^-?(1[0-7]{1,2})$
    structured_pattern:
      syntax: ^{lat} {lon}$
      interpolated: true
      partial_match: true
  organism_count:
    name: organism_count
    annotations:
      Expected_value:
        tag: Expected_value
        value: organism name;measurement value;enumeration
      storage_units:
        tag: storage_units
        value: '1'
    description: 'Total cell count of any organism (or group of organisms) per gram,
      volume or area of sample, should include name of organism followed by count.
      The method that was used for the enumeration (e.g. qPCR, atp, mpn, etc.) Should
      also be provided. (example: total prokaryotes; 3.5e7 cells per ml; qpcr)'
    title: organism count
    examples:
    - value: total prokaryotes;3.5e7 cells/mL;qPCR
    from_schema: https://example.com/nmdc_submission_schema
    rank: 30
    keywords:
    - count
    - organism
    string_serialization: '{text};{float} {unit};[ATP|MPN|qPCR|other]'
    slot_uri: MIXS:0000103
    alias: organism_count
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
    pattern: ^(\S+.*\S+;[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)? \S+.*\S+;(qPCR|ATP|MPN|other)\|)*(\S+.*\S+;[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?
      \S+.*\S+;(qPCR|ATP|MPN|other))$
  oxy_stat_samp:
    name: oxy_stat_samp
    description: Oxygenation status of sample
    title: oxygenation status of sample
    examples:
    - value: aerobic
    from_schema: https://example.com/nmdc_submission_schema
    rank: 25
    keywords:
    - oxygen
    - sample
    - status
    slot_uri: MIXS:0000753
    alias: oxy_stat_samp
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_section
    range: OxyStatSampEnum
  perturbation:
    name: perturbation
    annotations:
      Expected_value:
        tag: Expected_value
        value: perturbation type name;perturbation interval and duration
    description: Type of perturbation, e.g. chemical administration, physical disturbance,
      etc., coupled with perturbation regimen including how many times the perturbation
      was repeated, how long each perturbation lasted, and the start and end time
      of the entire perturbation period; can include multiple perturbation types
    title: perturbation
    examples:
    - value: antibiotic addition;R2/2018-05-11T14:30Z/2018-05-11T19:30Z/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 33
    keywords:
    - perturbation
    string_serialization: '{text};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000754
    alias: perturbation
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  salinity:
    name: salinity
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: practical salinity unit, percentage
      storage_units:
        tag: storage_units
        value: '%|mg/L'
    description: The total concentration of all dissolved salts in a liquid or solid
      sample. While salinity can be measured by a complete chemical analysis, this
      method is difficult and time consuming. More often, it is instead derived from
      the conductivity measurement. This is known as practical salinity. These derivations
      compare the specific conductance of the sample to a salinity standard such as
      seawater
    title: salinity
    comments:
    - 'Value must be reported in one of the following units: %, mg/L. Provide the
      numeric portion followed by a space and the unit.'
    examples:
    - value: 70 mg/L
    from_schema: https://example.com/nmdc_submission_schema
    rank: 54
    keywords:
    - salinity
    slot_uri: MIXS:0000183
    alias: salinity
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  samp_capt_status:
    name: samp_capt_status
    description: Reason for the sample
    title: sample capture status
    examples:
    - value: farm sample
    from_schema: https://example.com/nmdc_submission_schema
    rank: 282
    keywords:
    - sample
    - status
    slot_uri: MIXS:0000860
    alias: samp_capt_status
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: SampCaptStatusEnum
  samp_collec_device:
    name: samp_collec_device
    annotations:
      Expected_value:
        tag: Expected_value
        value: device name
    description: The device used to collect an environmental sample. This field accepts
      terms listed under environmental sampling device (http://purl.obolibrary.org/obo/ENVO).
      This field also accepts terms listed under specimen collection device (http://purl.obolibrary.org/obo/GENEPIO_0002094)
    title: sample collection device
    comments:
    - Report dimensions and details when applicable
    from_schema: https://example.com/nmdc_submission_schema
    structured_aliases:
      samp_collect_device:
        literal_form: samp_collect_device
        contexts:
        - https://github.com/GenomicsStandardsConsortium/mixs/releases/tag/v6.2.2
    rank: 14
    keywords:
    - device
    - sample
    string_serialization: '{termLabel} [{termID}]|{text}'
    slot_uri: MIXS:0000002
    alias: samp_collec_device
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
  samp_collec_method:
    name: samp_collec_method
    description: The method employed for collecting the sample
    title: sample collection method
    comments:
    - This can be a citation or description
    from_schema: https://example.com/nmdc_submission_schema
    structured_aliases:
      samp_collect_method:
        literal_form: samp_collect_method
        contexts:
        - https://github.com/GenomicsStandardsConsortium/mixs/releases/tag/v6.2.2
    rank: 15
    keywords:
    - method
    - sample
    slot_uri: MIXS:0001225
    alias: samp_collec_method
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
  samp_dis_stage:
    name: samp_dis_stage
    description: Stage of the disease at the time of sample collection, e.g. inoculation,
      penetration, infection, growth and reproduction, dissemination of pathogen
    title: sample disease stage
    examples:
    - value: infection
    from_schema: https://example.com/nmdc_submission_schema
    rank: 283
    keywords:
    - disease
    - sample
    slot_uri: MIXS:0000249
    alias: samp_dis_stage
    owner: HostAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: SampDisStageEnum
  samp_mat_process:
    name: samp_mat_process
    description: A brief description of any processing applied to the sample during
      or after retrieving the sample from environment, or a link to the relevant protocol(s)
      performed
    title: sample material processing
    examples:
    - value: filtering of seawater
    - value: storing samples in ethanol
    from_schema: https://example.com/nmdc_submission_schema
    rank: 12
    keywords:
    - material
    - process
    - sample
    slot_uri: MIXS:0000016
    alias: samp_mat_process
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
  samp_size:
    name: samp_size
    annotations:
      storage_units:
        tag: storage_units
        value: L|g|mL|mg
    description: The total amount (volume or mass) of sample collected.
    title: amount or size of sample collected
    comments:
    - 'Value must be reported in one of the following units: L, g, mL, mg. Provide
      the numeric portion followed by a space and the unit.'
    - This refers to the TOTAL amount of sample collected from the experiment. NOT
      the amount sent to each institution or collected for a specific analysis.
    examples:
    - value: 5 g
    - value: 10 mL
    from_schema: https://example.com/nmdc_submission_schema
    rank: 18
    keywords:
    - sample
    - size
    slot_uri: MIXS:0000001
    alias: samp_size
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  samp_store_dur:
    name: samp_store_dur
    description: Duration for which the sample was stored. Indicate the duration for
      which the sample was stored written in ISO 8601 format
    title: sample storage duration
    examples:
    - value: P1Y6M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 353
    keywords:
    - duration
    - period
    - sample
    - storage
    slot_uri: MIXS:0000116
    alias: samp_store_dur
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^P(?:(?:\d+D|\d+M(?:\d+D)?|\d+Y(?:\d+M(?:\d+D)?)?)(?:T(?:\d+H(?:\d+M(?:\d+S)?)?|\d+M(?:\d+S)?|\d+S))?|T(?:\d+H(?:\d+M(?:\d+S)?)?|\d+M(?:\d+S)?|\d+S)|\d+W)$
    structured_pattern:
      syntax: ^{duration}$
      interpolated: true
      partial_match: true
  samp_store_loc:
    name: samp_store_loc
    annotations:
      Expected_value:
        tag: Expected_value
        value: location name
    description: Location at which sample was stored, usually name of a specific freezer/room
    title: sample storage location
    examples:
    - value: Freezer no:5
    from_schema: https://example.com/nmdc_submission_schema
    rank: 41
    keywords:
    - location
    - sample
    - storage
    slot_uri: MIXS:0000755
    alias: samp_store_loc
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_core_section
    range: string
  samp_store_temp:
    name: samp_store_temp
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: degree Celsius
      storage_units:
        tag: storage_units
        value: Cel
    description: Temperature at which the sample was stored (degrees are assumed)
    title: sample storage temperature
    comments:
    - Value must be reported in degrees Celsius. Provide the numeric portion only.
    examples:
    - value: '-80'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 7
    keywords:
    - sample
    - storage
    - temperature
    slot_uri: MIXS:0000110
    alias: samp_store_temp
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: float
    required: true
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
    unit:
      descriptive_name: degrees Celsius
      ucum_code: Cel
  sample_link:
    name: sample_link
    description: A unique identifier to assign parent-child, subsample, or sibling
      samples. This is relevant when a sample or other material was used to generate
      the new sample.
    title: sample linkage
    comments:
    - 'This field allows multiple entries separated by ; (Examples: Soil collected
      from the field will link with the soil used in an incubation. The soil a plant
      was grown in links to the plant sample. An original culture sample was transferred
      to a new vial and generated a new sample)'
    examples:
    - value: igsn:DSJ0284
    from_schema: https://example.com/nmdc_submission_schema
    rank: 5
    string_serialization: '{text}:{text}'
    alias: sample_link
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    - SampIdNewTermsMixin
    slot_group: sample_id_section
    range: string
    recommended: true
    multivalued: false
    pattern: '[^\:\n\r]+\:[^\:\n\r]+'
  size_frac:
    name: size_frac
    annotations:
      Expected_value:
        tag: Expected_value
        value: filter size value range
    description: Filtering pore size used in sample preparation
    title: size fraction selected
    examples:
    - value: 0-0.22 micrometer
    from_schema: https://example.com/nmdc_submission_schema
    rank: 285
    keywords:
    - fraction
    - size
    string_serialization: '{float}-{float} {unit}'
    slot_uri: MIXS:0000017
    alias: size_frac
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_core_section
    range: string
  specific_ecosystem:
    name: specific_ecosystem
    description: Specific ecosystems represent specific features of the environment
      like aphotic zone in an ocean or gastric mucosa within a host digestive system.
      Specific ecosystem is in position 5/5 in a GOLD path.
    comments:
    - Specific ecosystems help to define samples based on very specific characteristics
      of an environment under the five-level classification system.
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - https://gold.jgi.doe.gov/help
    rank: 13
    is_a: gold_path_field
    alias: specific_ecosystem
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: SpecificEcosystemEnum
    recommended: true
  temp:
    name: temp
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: degree Celsius
      storage_units:
        tag: storage_units
        value: Cel
    description: Temperature of the sample at the time of sampling
    title: temperature
    comments:
    - Value must be reported in degrees Celsius. Provide the numeric portion only.
    examples:
    - value: '25'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 37
    keywords:
    - temperature
    slot_uri: MIXS:0000113
    alias: temp
    owner: HostAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_section
    range: float
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
    unit:
      descriptive_name: degrees Celsius
      ucum_code: Cel
  analysis_type:
    name: analysis_type
    description: Select all the data types associated or available for this biosample
    title: analysis/data type
    comments:
    - MIxS:investigation_type was included as a `see_also` but that term doesn't resolve
      any more
    examples:
    - value: metagenomics; metabolomics; metaproteomics
    from_schema: https://example.com/nmdc_submission_schema
    rank: 3
    alias: analysis_type
    owner: HostAssociatedInterface
    domain_of:
    - DhMultiviewCommonColumnsMixin
    slot_group: sample_id_section
    range: AnalysisTypeEnum
    required: true
    multivalued: true
  samp_name:
    name: samp_name
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: ''
    description: A local identifier or name that for the material sample collected.
      Refers to the original material collected or to any derived sub-samples.
    title: sample name
    comments:
    - It can have any format, but we suggest that you make it concise, unique and
      consistent within your lab, and as informative as possible.
    examples:
    - value: Rock core CB1178(5-6) from NSW
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1
    keywords:
    - sample
    slot_uri: MIXS:0001107
    identifier: true
    alias: samp_name
    owner: HostAssociatedInterface
    domain_of:
    - DhMultiviewCommonColumnsMixin
    slot_group: sample_id_section
    range: string
    required: true
  source_mat_id:
    name: source_mat_id
    annotations:
      Expected_value:
        tag: Expected_value
        value: 'for cultures of microorganisms: identifiers for two culture collections;
          for other material a unique arbitrary identifer'
    description: A globally unique identifier assigned to the parent sample or sample
      that is the source  of this sample.
    title: source material identifier
    todos:
    - Currently, the comments say to use UUIDs. However, if we implement assigning
      NMDC identifiers with the minter we dont need to require a GUID. It can be an
      optional field to fill out only if they already have a resolvable ID.
    comments:
    - Identifiers must be prefixed. Possible FAIR prefixes are IGSNs (http://www.geosamples.org/getigsn),
      NCBI biosample accession numbers, ARK identifiers (https://arks.org/). These
      IDs enable linking to derived analytes and subsamples. If you have not assigned
      FAIR identifiers to your samples, you can generate UUIDs (https://www.uuidgenerator.net/).
    - 'Identifiers must be prefixed. Possible FAIR prefixes are: `igsn` for International  Generic
      Sample Numbers (http://www.geosamples.org/getigsn), `biosample` for NCBI  biosample
      accession IDs, `gold` for GOLD identifiers.'
    examples:
    - value: igsn:AU1243
    - value: biosample:SAMEA2397676
    from_schema: https://example.com/nmdc_submission_schema
    rank: 2
    keywords:
    - identifier
    - material
    - source
    slot_uri: MIXS:0000026
    alias: source_mat_id
    owner: HostAssociatedInterface
    domain_of:
    - DhMultiviewCommonColumnsMixin
    slot_group: sample_id_section
    range: string
    multivalued: false
    pattern: ^igsn:[a-zA-Z0-9]+|biosample:SAMN[a-zA-Z0-9]+|biosample:SAME[a-zA-Z0-9]+|biosample:SAMJ[a-zA-Z0-9]+|gold:Gb[0-9]+$

</code></pre>
</details>
</article>
</div>
<script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "..", "features": ["content.tabs.link", "navigation.sections"], "search": "../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../assets/javascripts/bundle.83f73b43.min.js"></script>
<script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
<script src="../js/tablesort.js"></script>
<script type="module">import mermaid from "https://unpkg.com/mermaid@10.9.0/dist/mermaid.esm.min.mjs";
mermaid.initialize({});</script></body>
</html>