
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<link href="https://microbiomedata.github.io/submission-schema/SampleData/" rel="canonical"/>
<link href="../assets/images/favicon.png" rel="icon"/>
<meta content="mkdocs-1.6.1, mkdocs-material-9.5.48" name="generator"/>
<title>Class: SampleData (SampleData) - submission-schema</title>
<link href="../assets/stylesheets/main.6f8fc17f.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
</head>
<body dir="ltr">
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#class-sampledata-sampledata">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="submission-schema" class="md-header__button md-logo" data-md-component="logo" href=".." title="submission-schema">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"></path></svg>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            submission-schema
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Class: SampleData (SampleData)
            
          </span>
</div>
</div>
</div>
<script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" title="Clear" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="" tabindex="0">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list" role="presentation"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/submission-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="submission-schema" class="md-nav__button md-logo" data-md-component="logo" href=".." title="submission-schema">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"></path></svg>
</a>
    submission-schema
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/submission-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="..">
<span class="md-ellipsis">
    Index
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../about/">
<span class="md-ellipsis">
    About
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="https://microbiomedata.github.io/submission-schema/_playground/">
<span class="md-ellipsis">
    Submission Schema DH Playground
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="https://microbiomedata.github.io/submission-schema/_playground/?template=nmdc_submission_schema/SoilInterface">
<span class="md-ellipsis">
    Direct submission of soil samples
  </span>
</a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#slots">
<span class="md-ellipsis">
      Slots
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#identifier-and-mapping-information">
<span class="md-ellipsis">
      Identifier and Mapping Information
    </span>
</a>
<nav aria-label="Identifier and Mapping Information" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#schema-source">
<span class="md-ellipsis">
      Schema Source
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#mappings">
<span class="md-ellipsis">
      Mappings
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#linkml-source">
<span class="md-ellipsis">
      LinkML Source
    </span>
</a>
<nav aria-label="LinkML Source" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#direct">
<span class="md-ellipsis">
      Direct
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#induced">
<span class="md-ellipsis">
      Induced
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1 id="class-sampledata-sampledata">Class: SampleData (SampleData)</h1>
<p><em>represents data produced by the DataHarmonizer tabs of the submission portal</em></p>
<p>URI: <a href="https://example.com/nmdc_sub_schema/SampleData">nmdc_sub_schema:SampleData</a></p>
<div class="mermaid"> classDiagram
    class SampleData
    click SampleData href "../SampleData"
      SampleData : air_data




    SampleData --&gt; "*" AirInterface : air_data
    click AirInterface href "../AirInterface"


      SampleData : biofilm_data




    SampleData --&gt; "*" BiofilmInterface : biofilm_data
    click BiofilmInterface href "../BiofilmInterface"


      SampleData : built_env_data




    SampleData --&gt; "*" BuiltEnvInterface : built_env_data
    click BuiltEnvInterface href "../BuiltEnvInterface"


      SampleData : emsl_data




    SampleData --&gt; "*" EmslInterface : emsl_data
    click EmslInterface href "../EmslInterface"


      SampleData : host_associated_data




    SampleData --&gt; "*" HostAssociatedInterface : host_associated_data
    click HostAssociatedInterface href "../HostAssociatedInterface"


      SampleData : jgi_mg_data




    SampleData --&gt; "*" JgiMgInterface : jgi_mg_data
    click JgiMgInterface href "../JgiMgInterface"


      SampleData : jgi_mg_lr_data




    SampleData --&gt; "*" JgiMgLrInterface : jgi_mg_lr_data
    click JgiMgLrInterface href "../JgiMgLrInterface"


      SampleData : jgi_mt_data




    SampleData --&gt; "*" JgiMtInterface : jgi_mt_data
    click JgiMtInterface href "../JgiMtInterface"


      SampleData : metagenome_sequencing_interleaved_data




    SampleData --&gt; "*" MetagenomeSequencingInterleavedDataInterface : metagenome_sequencing_interleaved_data
    click MetagenomeSequencingInterleavedDataInterface href "../MetagenomeSequencingInterleavedDataInterface"


      SampleData : metagenome_sequencing_non_interleaved_data




    SampleData --&gt; "*" MetagenomeSequencingNonInterleavedDataInterface : metagenome_sequencing_non_interleaved_data
    click MetagenomeSequencingNonInterleavedDataInterface href "../MetagenomeSequencingNonInterleavedDataInterface"


      SampleData : metatranscriptome_sequencing_interleaved_data




    SampleData --&gt; "*" MetatranscriptomeSequencingInterleavedDataInterface : metatranscriptome_sequencing_interleaved_data
    click MetatranscriptomeSequencingInterleavedDataInterface href "../MetatranscriptomeSequencingInterleavedDataInterface"


      SampleData : metatranscriptome_sequencing_non_interleaved_data




    SampleData --&gt; "*" MetatranscriptomeSequencingNonInterleavedDataInterface : metatranscriptome_sequencing_non_interleaved_data
    click MetatranscriptomeSequencingNonInterleavedDataInterface href "../MetatranscriptomeSequencingNonInterleavedDataInterface"


      SampleData : plant_associated_data




    SampleData --&gt; "*" PlantAssociatedInterface : plant_associated_data
    click PlantAssociatedInterface href "../PlantAssociatedInterface"


      SampleData : sediment_data




    SampleData --&gt; "*" SedimentInterface : sediment_data
    click SedimentInterface href "../SedimentInterface"


      SampleData : soil_data




    SampleData --&gt; "*" SoilInterface : soil_data
    click SoilInterface href "../SoilInterface"


      SampleData : wastewater_sludge_data




    SampleData --&gt; "*" WastewaterSludgeInterface : wastewater_sludge_data
    click WastewaterSludgeInterface href "../WastewaterSludgeInterface"


      SampleData : water_data




    SampleData --&gt; "*" WaterInterface : water_data
    click WaterInterface href "../WaterInterface"



</div>
<!-- no inheritance hierarchy -->
<h2 id="slots">Slots</h2>
<table>
<thead>
<tr>
<th>Name</th>
<th>Cardinality and Range</th>
<th>Description</th>
<th>Inheritance</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../air_data/">air_data</a></td>
<td>* <br/> <a href="../AirInterface/">AirInterface</a></td>
<td>aggregation slot relating air data collections to a SampleData container</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../biofilm_data/">biofilm_data</a></td>
<td>* <br/> <a href="../BiofilmInterface/">BiofilmInterface</a></td>
<td>aggregation slot relating biofilm data collections to a SampleData container</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../built_env_data/">built_env_data</a></td>
<td>* <br/> <a href="../BuiltEnvInterface/">BuiltEnvInterface</a></td>
<td>aggregation slot relating built_env data collections to a SampleData containe...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_associated_data/">host_associated_data</a></td>
<td>* <br/> <a href="../HostAssociatedInterface/">HostAssociatedInterface</a></td>
<td>aggregation slot relating host_associated data collections to a SampleData co...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../plant_associated_data/">plant_associated_data</a></td>
<td>* <br/> <a href="../PlantAssociatedInterface/">PlantAssociatedInterface</a></td>
<td>aggregation slot relating plant_associated data collections to a SampleData c...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../sediment_data/">sediment_data</a></td>
<td>* <br/> <a href="../SedimentInterface/">SedimentInterface</a></td>
<td>aggregation slot relating sediment data collections to a SampleData container</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../soil_data/">soil_data</a></td>
<td>* <br/> <a href="../SoilInterface/">SoilInterface</a></td>
<td>aggregation slot relating soil data collections to a SampleData container</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../wastewater_sludge_data/">wastewater_sludge_data</a></td>
<td>* <br/> <a href="../WastewaterSludgeInterface/">WastewaterSludgeInterface</a></td>
<td>aggregation slot relating wastewater_sludge data collections to a SampleData ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../water_data/">water_data</a></td>
<td>* <br/> <a href="../WaterInterface/">WaterInterface</a></td>
<td>aggregation slot relating water data collections to a SampleData container</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../emsl_data/">emsl_data</a></td>
<td>* <br/> <a href="../EmslInterface/">EmslInterface</a></td>
<td>aggregation slot relating emsl data collections to a SampleData container</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../jgi_mg_lr_data/">jgi_mg_lr_data</a></td>
<td>* <br/> <a href="../JgiMgLrInterface/">JgiMgLrInterface</a></td>
<td>aggregation slot relating jgi_mg_lr data collections to a SampleData containe...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../jgi_mg_data/">jgi_mg_data</a></td>
<td>* <br/> <a href="../JgiMgInterface/">JgiMgInterface</a></td>
<td>aggregation slot relating jgi_mg data collections to a SampleData container</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../jgi_mt_data/">jgi_mt_data</a></td>
<td>* <br/> <a href="../JgiMtInterface/">JgiMtInterface</a></td>
<td>aggregation slot relating jgi_mt data collections to a SampleData container</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../metagenome_sequencing_non_interleaved_data/">metagenome_sequencing_non_interleaved_data</a></td>
<td>* <br/> <a href="../MetagenomeSequencingNonInterleavedDataInterface/">MetagenomeSequencingNonInterleavedDataInterface</a></td>
<td>aggregation slot relating non-interleaved metagenome sequencing data collecti...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../metagenome_sequencing_interleaved_data/">metagenome_sequencing_interleaved_data</a></td>
<td>* <br/> <a href="../MetagenomeSequencingInterleavedDataInterface/">MetagenomeSequencingInterleavedDataInterface</a></td>
<td>aggregation slot relating interleaved metagenome sequencing data collections ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../metatranscriptome_sequencing_non_interleaved_data/">metatranscriptome_sequencing_non_interleaved_data</a></td>
<td>* <br/> <a href="../MetatranscriptomeSequencingNonInterleavedDataInterface/">MetatranscriptomeSequencingNonInterleavedDataInterface</a></td>
<td>aggregation slot relating non-interleaved metatranscriptome sequencing data c...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../metatranscriptome_sequencing_interleaved_data/">metatranscriptome_sequencing_interleaved_data</a></td>
<td>* <br/> <a href="../MetatranscriptomeSequencingInterleavedDataInterface/">MetatranscriptomeSequencingInterleavedDataInterface</a></td>
<td>aggregation slot relating interleaved metatranscriptome sequencing data colle...</td>
<td>direct</td>
</tr>
</tbody>
</table>
<h2 id="identifier-and-mapping-information">Identifier and Mapping Information</h2>
<h3 id="schema-source">Schema Source</h3>
<ul>
<li>from schema: <a href="https://example.com/nmdc_submission_schema">https://example.com/nmdc_submission_schema</a></li>
</ul>
<h2 id="mappings">Mappings</h2>
<table>
<thead>
<tr>
<th>Mapping Type</th>
<th>Mapped Value</th>
</tr>
</thead>
<tbody>
<tr>
<td>self</td>
<td>nmdc_sub_schema:SampleData</td>
</tr>
<tr>
<td>native</td>
<td>nmdc_sub_schema:SampleData</td>
</tr>
</tbody>
</table>
<h2 id="linkml-source">LinkML Source</h2>
<!-- TODO: investigate https://stackoverflow.com/questions/37606292/how-to-create-tabbed-code-blocks-in-mkdocs-or-sphinx -->
<h3 id="direct">Direct</h3>
<details>
<pre><code class="language-yaml">name: SampleData
description: represents data produced by the DataHarmonizer tabs of the submission
  portal
title: SampleData
from_schema: https://example.com/nmdc_submission_schema
slots:
- air_data
- biofilm_data
- built_env_data
- host_associated_data
- plant_associated_data
- sediment_data
- soil_data
- wastewater_sludge_data
- water_data
- emsl_data
- jgi_mg_lr_data
- jgi_mg_data
- jgi_mt_data
- metagenome_sequencing_non_interleaved_data
- metagenome_sequencing_interleaved_data
- metatranscriptome_sequencing_non_interleaved_data
- metatranscriptome_sequencing_interleaved_data
tree_root: true

</code></pre>
</details>
<h3 id="induced">Induced</h3>
<details>
<pre><code class="language-yaml">name: SampleData
description: represents data produced by the DataHarmonizer tabs of the submission
  portal
title: SampleData
from_schema: https://example.com/nmdc_submission_schema
attributes:
  air_data:
    name: air_data
    description: aggregation slot relating air data collections to a SampleData container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: air_data
    owner: SampleData
    domain_of:
    - SampleData
    range: AirInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  biofilm_data:
    name: biofilm_data
    description: aggregation slot relating biofilm data collections to a SampleData
      container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: biofilm_data
    owner: SampleData
    domain_of:
    - SampleData
    range: BiofilmInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  built_env_data:
    name: built_env_data
    description: aggregation slot relating built_env data collections to a SampleData
      container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: built_env_data
    owner: SampleData
    domain_of:
    - SampleData
    range: BuiltEnvInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  host_associated_data:
    name: host_associated_data
    description: aggregation slot relating host_associated data collections to a SampleData
      container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: host_associated_data
    owner: SampleData
    domain_of:
    - SampleData
    range: HostAssociatedInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  plant_associated_data:
    name: plant_associated_data
    description: aggregation slot relating plant_associated data collections to a
      SampleData container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: plant_associated_data
    owner: SampleData
    domain_of:
    - SampleData
    range: PlantAssociatedInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  sediment_data:
    name: sediment_data
    description: aggregation slot relating sediment data collections to a SampleData
      container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: sediment_data
    owner: SampleData
    domain_of:
    - SampleData
    range: SedimentInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  soil_data:
    name: soil_data
    description: aggregation slot relating soil data collections to a SampleData container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: soil_data
    owner: SampleData
    domain_of:
    - SampleData
    range: SoilInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  wastewater_sludge_data:
    name: wastewater_sludge_data
    description: aggregation slot relating wastewater_sludge data collections to a
      SampleData container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: wastewater_sludge_data
    owner: SampleData
    domain_of:
    - SampleData
    range: WastewaterSludgeInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  water_data:
    name: water_data
    description: aggregation slot relating water data collections to a SampleData
      container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: water_data
    owner: SampleData
    domain_of:
    - SampleData
    range: WaterInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  emsl_data:
    name: emsl_data
    description: aggregation slot relating emsl data collections to a SampleData container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: emsl_data
    owner: SampleData
    domain_of:
    - SampleData
    range: EmslInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  jgi_mg_lr_data:
    name: jgi_mg_lr_data
    description: aggregation slot relating jgi_mg_lr data collections to a SampleData
      container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: jgi_mg_lr_data
    owner: SampleData
    domain_of:
    - SampleData
    range: JgiMgLrInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  jgi_mg_data:
    name: jgi_mg_data
    description: aggregation slot relating jgi_mg data collections to a SampleData
      container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: jgi_mg_data
    owner: SampleData
    domain_of:
    - SampleData
    range: JgiMgInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  jgi_mt_data:
    name: jgi_mt_data
    description: aggregation slot relating jgi_mt data collections to a SampleData
      container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: jgi_mt_data
    owner: SampleData
    domain_of:
    - SampleData
    range: JgiMtInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  metagenome_sequencing_non_interleaved_data:
    name: metagenome_sequencing_non_interleaved_data
    description: aggregation slot relating non-interleaved metagenome sequencing data
      collections to a  SampleData container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: metagenome_sequencing_non_interleaved_data
    owner: SampleData
    domain_of:
    - SampleData
    range: MetagenomeSequencingNonInterleavedDataInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  metagenome_sequencing_interleaved_data:
    name: metagenome_sequencing_interleaved_data
    description: aggregation slot relating interleaved metagenome sequencing data
      collections to a SampleData  container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: metagenome_sequencing_interleaved_data
    owner: SampleData
    domain_of:
    - SampleData
    range: MetagenomeSequencingInterleavedDataInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  metatranscriptome_sequencing_non_interleaved_data:
    name: metatranscriptome_sequencing_non_interleaved_data
    description: aggregation slot relating non-interleaved metatranscriptome sequencing
      data collections to a  SampleData container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: metatranscriptome_sequencing_non_interleaved_data
    owner: SampleData
    domain_of:
    - SampleData
    range: MetatranscriptomeSequencingNonInterleavedDataInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
  metatranscriptome_sequencing_interleaved_data:
    name: metatranscriptome_sequencing_interleaved_data
    description: aggregation slot relating interleaved metatranscriptome sequencing
      data collections to a  SampleData container
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1000
    alias: metatranscriptome_sequencing_interleaved_data
    owner: SampleData
    domain_of:
    - SampleData
    range: MetatranscriptomeSequencingInterleavedDataInterface
    multivalued: true
    inlined: true
    inlined_as_list: true
tree_root: true

</code></pre>
</details>
</article>
</div>
<script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "..", "features": ["content.tabs.link"], "search": "../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../assets/javascripts/bundle.83f73b43.min.js"></script>
<script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
<script src="../javascripts/tablesort.js"></script>
<script type="module">import mermaid from "https://unpkg.com/mermaid@10.9.0/dist/mermaid.esm.min.mjs";
mermaid.initialize({});</script></body>
</html>