
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<link href="https://microbiomedata.github.io/submission-schema/PlantAssociatedInterface/" rel="canonical"/>
<link href="../assets/images/favicon.png" rel="icon"/>
<meta content="mkdocs-1.6.1, mkdocs-material-9.5.48" name="generator"/>
<title>Class: Plant Associated (PlantAssociatedInterface) - submission-schema</title>
<link href="../assets/stylesheets/main.6f8fc17f.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
</head>
<body dir="ltr">
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#class-plant-associated-plantassociatedinterface">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="submission-schema" class="md-header__button md-logo" data-md-component="logo" href=".." title="submission-schema">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"></path></svg>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            submission-schema
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Class: Plant Associated (PlantAssociatedInterface)
            
          </span>
</div>
</div>
</div>
<script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" title="Clear" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="" tabindex="0">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list" role="presentation"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/submission-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="submission-schema" class="md-nav__button md-logo" data-md-component="logo" href=".." title="submission-schema">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"></path></svg>
</a>
    submission-schema
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/submission-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="..">
<span class="md-ellipsis">
    Index
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../about/">
<span class="md-ellipsis">
    About
  </span>
</a>
</li>
<li class="md-nav__item md-nav__item--section md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
<span class="md-ellipsis">
    DataHarmonizer Playground
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
            DataHarmonizer Playground
          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="https://microbiomedata.github.io/submission-schema/_playground/?template=nmdc_submission_schema/SoilInterface">
<span class="md-ellipsis">
    SoilInterface
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="https://microbiomedata.github.io/submission-schema/_playground/?template=nmdc_submission_schema/WaterInterface">
<span class="md-ellipsis">
    WaterInterface
  </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#inheritance">
<span class="md-ellipsis">
      Inheritance
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#slots">
<span class="md-ellipsis">
      Slots
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#usages">
<span class="md-ellipsis">
      Usages
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#identifier-and-mapping-information">
<span class="md-ellipsis">
      Identifier and Mapping Information
    </span>
</a>
<nav aria-label="Identifier and Mapping Information" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#annotations">
<span class="md-ellipsis">
      Annotations
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#schema-source">
<span class="md-ellipsis">
      Schema Source
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#mappings">
<span class="md-ellipsis">
      Mappings
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#linkml-source">
<span class="md-ellipsis">
      LinkML Source
    </span>
</a>
<nav aria-label="LinkML Source" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#direct">
<span class="md-ellipsis">
      Direct
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#induced">
<span class="md-ellipsis">
      Induced
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1 id="class-plant-associated-plantassociatedinterface">Class: Plant Associated (PlantAssociatedInterface)</h1>
<p><em>plant_associated dh_interface</em></p>
<p>URI: <a href="https://example.com/nmdc_sub_schema/PlantAssociatedInterface">nmdc_sub_schema:PlantAssociatedInterface</a></p>
<div class="mermaid"> classDiagram
    class PlantAssociatedInterface
    click PlantAssociatedInterface href "../PlantAssociatedInterface/"
      DhMultiviewCommonColumnsMixin &lt;|-- PlantAssociatedInterface
        click DhMultiviewCommonColumnsMixin href "../DhMultiviewCommonColumnsMixin/"
      SampIdNewTermsMixin &lt;|-- PlantAssociatedInterface
        click SampIdNewTermsMixin href "../SampIdNewTermsMixin/"
      DhInterface &lt;|-- PlantAssociatedInterface
        click DhInterface href "../DhInterface/"

      PlantAssociatedInterface : air_temp_regm

      PlantAssociatedInterface : analysis_type





        PlantAssociatedInterface --&gt; "1..*" AnalysisTypeEnum : analysis_type
        click AnalysisTypeEnum href "../AnalysisTypeEnum/"



      PlantAssociatedInterface : ances_data

      PlantAssociatedInterface : antibiotic_regm

      PlantAssociatedInterface : biol_stat





        PlantAssociatedInterface --&gt; "0..1" BiolStatEnum : biol_stat
        click BiolStatEnum href "../BiolStatEnum/"



      PlantAssociatedInterface : biotic_regm

      PlantAssociatedInterface : biotic_relationship





        PlantAssociatedInterface --&gt; "0..1" BioticRelationshipEnum : biotic_relationship
        click BioticRelationshipEnum href "../BioticRelationshipEnum/"



      PlantAssociatedInterface : chem_administration

      PlantAssociatedInterface : chem_mutagen

      PlantAssociatedInterface : climate_environment

      PlantAssociatedInterface : collection_date

      PlantAssociatedInterface : collection_date_inc

      PlantAssociatedInterface : collection_time

      PlantAssociatedInterface : cult_root_med

      PlantAssociatedInterface : depth

      PlantAssociatedInterface : ecosystem





        PlantAssociatedInterface --&gt; "0..1 _recommended_" EcosystemEnum : ecosystem
        click EcosystemEnum href "../EcosystemEnum/"



      PlantAssociatedInterface : ecosystem_category





        PlantAssociatedInterface --&gt; "0..1 _recommended_" EcosystemCategoryEnum : ecosystem_category
        click EcosystemCategoryEnum href "../EcosystemCategoryEnum/"



      PlantAssociatedInterface : ecosystem_subtype





        PlantAssociatedInterface --&gt; "0..1 _recommended_" EcosystemSubtypeEnum : ecosystem_subtype
        click EcosystemSubtypeEnum href "../EcosystemSubtypeEnum/"



      PlantAssociatedInterface : ecosystem_type





        PlantAssociatedInterface --&gt; "0..1 _recommended_" EcosystemTypeEnum : ecosystem_type
        click EcosystemTypeEnum href "../EcosystemTypeEnum/"



      PlantAssociatedInterface : elev

      PlantAssociatedInterface : env_broad_scale

      PlantAssociatedInterface : env_local_scale

      PlantAssociatedInterface : env_medium

      PlantAssociatedInterface : experimental_factor

      PlantAssociatedInterface : fertilizer_regm

      PlantAssociatedInterface : fungicide_regm

      PlantAssociatedInterface : gaseous_environment

      PlantAssociatedInterface : genetic_mod

      PlantAssociatedInterface : geo_loc_name

      PlantAssociatedInterface : gravity

      PlantAssociatedInterface : growth_facil





        PlantAssociatedInterface --&gt; "1" GrowthFacilEnum : growth_facil
        click GrowthFacilEnum href "../GrowthFacilEnum/"



      PlantAssociatedInterface : growth_habit





        PlantAssociatedInterface --&gt; "0..1" GrowthHabitEnum : growth_habit
        click GrowthHabitEnum href "../GrowthHabitEnum/"



      PlantAssociatedInterface : growth_hormone_regm

      PlantAssociatedInterface : herbicide_regm

      PlantAssociatedInterface : host_age

      PlantAssociatedInterface : host_common_name

      PlantAssociatedInterface : host_disease_stat

      PlantAssociatedInterface : host_dry_mass

      PlantAssociatedInterface : host_genotype

      PlantAssociatedInterface : host_height

      PlantAssociatedInterface : host_length

      PlantAssociatedInterface : host_life_stage

      PlantAssociatedInterface : host_phenotype

      PlantAssociatedInterface : host_subspecf_genlin

      PlantAssociatedInterface : host_symbiont

      PlantAssociatedInterface : host_taxid

      PlantAssociatedInterface : host_tot_mass

      PlantAssociatedInterface : host_wet_mass

      PlantAssociatedInterface : humidity_regm

      PlantAssociatedInterface : isotope_exposure

      PlantAssociatedInterface : lat_lon

      PlantAssociatedInterface : light_regm

      PlantAssociatedInterface : mechanical_damage

      PlantAssociatedInterface : mineral_nutr_regm

      PlantAssociatedInterface : non_min_nutr_regm

      PlantAssociatedInterface : organism_count

      PlantAssociatedInterface : oxy_stat_samp





        PlantAssociatedInterface --&gt; "0..1" OxyStatSampEnum : oxy_stat_samp
        click OxyStatSampEnum href "../OxyStatSampEnum/"



      PlantAssociatedInterface : perturbation

      PlantAssociatedInterface : pesticide_regm

      PlantAssociatedInterface : ph_regm

      PlantAssociatedInterface : plant_growth_med

      PlantAssociatedInterface : plant_product

      PlantAssociatedInterface : plant_sex





        PlantAssociatedInterface --&gt; "0..1" PlantSexEnum : plant_sex
        click PlantSexEnum href "../PlantSexEnum/"



      PlantAssociatedInterface : plant_struc

      PlantAssociatedInterface : radiation_regm

      PlantAssociatedInterface : rainfall_regm

      PlantAssociatedInterface : root_cond

      PlantAssociatedInterface : root_med_carbon

      PlantAssociatedInterface : root_med_macronutr

      PlantAssociatedInterface : root_med_micronutr

      PlantAssociatedInterface : root_med_ph

      PlantAssociatedInterface : root_med_regl

      PlantAssociatedInterface : root_med_solid

      PlantAssociatedInterface : root_med_suppl

      PlantAssociatedInterface : salinity

      PlantAssociatedInterface : salinity_meth

      PlantAssociatedInterface : salt_regm

      PlantAssociatedInterface : samp_capt_status





        PlantAssociatedInterface --&gt; "0..1" SampCaptStatusEnum : samp_capt_status
        click SampCaptStatusEnum href "../SampCaptStatusEnum/"



      PlantAssociatedInterface : samp_collec_device

      PlantAssociatedInterface : samp_collec_method

      PlantAssociatedInterface : samp_dis_stage





        PlantAssociatedInterface --&gt; "0..1" SampDisStageEnum : samp_dis_stage
        click SampDisStageEnum href "../SampDisStageEnum/"



      PlantAssociatedInterface : samp_mat_process

      PlantAssociatedInterface : samp_name

      PlantAssociatedInterface : samp_size

      PlantAssociatedInterface : samp_store_dur

      PlantAssociatedInterface : samp_store_loc

      PlantAssociatedInterface : samp_store_temp

      PlantAssociatedInterface : sample_link

      PlantAssociatedInterface : season_environment

      PlantAssociatedInterface : size_frac

      PlantAssociatedInterface : source_mat_id

      PlantAssociatedInterface : specific_ecosystem





        PlantAssociatedInterface --&gt; "0..1 _recommended_" SpecificEcosystemEnum : specific_ecosystem
        click SpecificEcosystemEnum href "../SpecificEcosystemEnum/"



      PlantAssociatedInterface : standing_water_regm

      PlantAssociatedInterface : start_date_inc

      PlantAssociatedInterface : temp

      PlantAssociatedInterface : tiss_cult_growth_med

      PlantAssociatedInterface : water_temp_regm

      PlantAssociatedInterface : watering_regm


</div>
<h2 id="inheritance">Inheritance</h2>
<ul>
<li><a href="../DhInterface/">DhInterface</a><ul>
<li><strong>PlantAssociatedInterface</strong> [ <a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a> <a href="../SampIdNewTermsMixin/">SampIdNewTermsMixin</a>]</li>
</ul>
</li>
</ul>
<h2 id="slots">Slots</h2>
<table>
<thead>
<tr>
<th>Name</th>
<th>Cardinality and Range</th>
<th>Description</th>
<th>Inheritance</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../air_temp_regm/">air_temp_regm</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving an exposure to varying temperatures; sh...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ances_data/">ances_data</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about either pedigree or other ancestral information description ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../antibiotic_regm/">antibiotic_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving antibiotic administration; should inclu...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../biol_stat/">biol_stat</a></td>
<td>0..1 <br/> <a href="../BiolStatEnum/">BiolStatEnum</a></td>
<td>The level of genome modification</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../biotic_regm/">biotic_regm</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>Information about treatment(s) involving use of biotic factors, such as bacte...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../biotic_relationship/">biotic_relationship</a></td>
<td>0..1 <br/> <a href="../BioticRelationshipEnum/">BioticRelationshipEnum</a></td>
<td>Description of relationship(s) between the subject organism and other organis...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../chem_administration/">chem_administration</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>List of chemical compounds administered to the host or site where sampling oc...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../chem_mutagen/">chem_mutagen</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Treatment involving use of mutagens; should include the name of mutagen, amou...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../climate_environment/">climate_environment</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>Treatment involving an exposure to a particular climate; treatment regimen in...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../collection_date/">collection_date</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>The date of sampling</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../collection_date_inc/">collection_date_inc</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>Date the incubation was harvested/collected/ended</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../collection_time/">collection_time</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>The time of sampling, either as an instance (single point) or interval</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../cult_root_med/">cult_root_med</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Name or reference for the hydroponic or in vitro culture rooting medium; can ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../depth/">depth</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>The vertical distance below local surface</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ecosystem/">ecosystem</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../EcosystemEnum/">EcosystemEnum</a></td>
<td>An ecosystem is a combination of a physical environment (abiotic factors) and...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ecosystem_category/">ecosystem_category</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../EcosystemCategoryEnum/">EcosystemCategoryEnum</a></td>
<td>Ecosystem categories represent divisions within the ecosystem based on specif...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ecosystem_subtype/">ecosystem_subtype</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../EcosystemSubtypeEnum/">EcosystemSubtypeEnum</a></td>
<td>Ecosystem subtypes represent further subdivision of Ecosystem types into more...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ecosystem_type/">ecosystem_type</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../EcosystemTypeEnum/">EcosystemTypeEnum</a></td>
<td>Ecosystem types represent things having common characteristics within the Eco...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../elev/">elev</a></td>
<td>1 <br/> <a href="../Float/">Float</a></td>
<td>Elevation of the sampling site is its height above a fixed reference point, m...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../env_broad_scale/">env_broad_scale</a></td>
<td>1 <br/> <a href="../String/">String</a> or <br/><a href="../EnvBroadScalePlantAssociatedEnum/">EnvBroadScalePlantAssociatedEnum</a> or <br/><a href="../String/">String</a></td>
<td>In this field, report which major environmental system your sample or specime...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../env_local_scale/">env_local_scale</a></td>
<td>1 <br/> <a href="../String/">String</a> or <br/><a href="../EnvLocalScalePlantAssociatedEnum/">EnvLocalScalePlantAssociatedEnum</a> or <br/><a href="../String/">String</a></td>
<td>Report the entity or entities which are in the sample or specimen s local vic...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../env_medium/">env_medium</a></td>
<td>1 <br/> <a href="../String/">String</a> or <br/><a href="../EnvMediumPlantAssociatedEnum/">EnvMediumPlantAssociatedEnum</a> or <br/><a href="../String/">String</a></td>
<td>Report the environmental material(s) immediately surrounding the sample or sp...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../experimental_factor/">experimental_factor</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Variable aspects of an experiment design that can be used to describe an expe...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../fertilizer_regm/">fertilizer_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving the use of fertilizers; should include ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../fungicide_regm/">fungicide_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving use of fungicides; should include the n...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../gaseous_environment/">gaseous_environment</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>Use of conditions with differing gaseous environments; should include the nam...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../genetic_mod/">genetic_mod</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Genetic modifications of the genome of an organism, which may occur naturally...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../geo_loc_name/">geo_loc_name</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>The geographical origin of the sample as defined by the country or sea name f...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../gravity/">gravity</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving use of gravity factor to study various ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../growth_facil/">growth_facil</a></td>
<td>1 <br/> <a href="../GrowthFacilEnum/">GrowthFacilEnum</a></td>
<td>Type of facility/location where the sample was harvested; controlled vocabula...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../growth_habit/">growth_habit</a></td>
<td>0..1 <br/> <a href="../GrowthHabitEnum/">GrowthHabitEnum</a></td>
<td>Characteristic shape, appearance or growth form of a plant species</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../growth_hormone_regm/">growth_hormone_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving use of growth hormones; should include ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../herbicide_regm/">herbicide_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving use of herbicides; information about tr...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_age/">host_age</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Age of host at the time of sampling; relevant scale depends on species and st...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_common_name/">host_common_name</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>Common name of the host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_disease_stat/">host_disease_stat</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>List of diseases with which the host has been diagnosed; can include multiple...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_dry_mass/">host_dry_mass</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Measurement of dry mass</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_genotype/">host_genotype</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Observed genotype</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_height/">host_height</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The height of subject</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_length/">host_length</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The length of subject</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_life_stage/">host_life_stage</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Description of life stage of host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_phenotype/">host_phenotype</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Phenotype of human or other host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_subspecf_genlin/">host_subspecf_genlin</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about the genetic distinctness of the host organism below the sub...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_symbiont/">host_symbiont</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The taxonomic name of the organism(s) found living in mutualistic, commensali...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_taxid/">host_taxid</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>NCBI taxon id of the host, e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_tot_mass/">host_tot_mass</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Total mass of the host at collection, the unit depends on host</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../host_wet_mass/">host_wet_mass</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Measurement of wet mass</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../humidity_regm/">humidity_regm</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving an exposure to varying degree of humidi...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../isotope_exposure/">isotope_exposure</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>List isotope exposure or addition applied to your sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../lat_lon/">lat_lon</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>The geographical origin of the sample as defined by latitude and longitude</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../light_regm/">light_regm</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>Information about treatment(s) involving exposure to light, including both li...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../mechanical_damage/">mechanical_damage</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about any mechanical damage exerted on the plant; can include mul...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../mineral_nutr_regm/">mineral_nutr_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving the use of mineral supplements; should ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../non_min_nutr_regm/">non_min_nutr_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving the exposure of plant to non-mineral nu...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../organism_count/">organism_count</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Total cell count of any organism (or group of organisms) per gram, volume or ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../oxy_stat_samp/">oxy_stat_samp</a></td>
<td>0..1 <br/> <a href="../OxyStatSampEnum/">OxyStatSampEnum</a></td>
<td>Oxygenation status of sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../perturbation/">perturbation</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Type of perturbation, e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../pesticide_regm/">pesticide_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving use of insecticides; should include the...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../ph_regm/">ph_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving exposure of plants to varying levels of...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../plant_growth_med/">plant_growth_med</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Specification of the media for growing the plants or tissue cultured samples,...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../plant_product/">plant_product</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Substance produced by the plant, where the sample was obtained from</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../plant_sex/">plant_sex</a></td>
<td>0..1 <br/> <a href="../PlantSexEnum/">PlantSexEnum</a></td>
<td>Sex of the reproductive parts on the whole plant, e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../plant_struc/">plant_struc</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>Name of plant structure the sample was obtained from; for Plant Ontology (PO)...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../radiation_regm/">radiation_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving exposure of plant or a plant part to a ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../rainfall_regm/">rainfall_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving an exposure to a given amount of rainfa...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../root_cond/">root_cond</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Relevant rooting conditions such as field plot size, sowing density, containe...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../root_med_carbon/">root_med_carbon</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Source of organic carbon in the culture rooting medium; e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../root_med_macronutr/">root_med_macronutr</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Measurement of the culture rooting medium macronutrients (N,P, K, Ca, Mg, S);...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../root_med_micronutr/">root_med_micronutr</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Measurement of the culture rooting medium micronutrients (Fe, Mn, Zn, B, Cu, ...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../root_med_ph/">root_med_ph</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>pH measurement of the culture rooting medium; e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../root_med_regl/">root_med_regl</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Growth regulators in the culture rooting medium such as cytokinins, auxins, g...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../root_med_solid/">root_med_solid</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Specification of the solidifying agent in the culture rooting medium; e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../root_med_suppl/">root_med_suppl</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Organic supplements of the culture rooting medium, such as vitamins, amino ac...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../salinity/">salinity</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The total concentration of all dissolved salts in a liquid or solid sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../salinity_meth/">salinity_meth</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Reference or method used in determining salinity</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../salt_regm/">salt_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving use of salts as supplement to liquid an...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_capt_status/">samp_capt_status</a></td>
<td>0..1 <br/> <a href="../SampCaptStatusEnum/">SampCaptStatusEnum</a></td>
<td>Reason for the sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_collec_device/">samp_collec_device</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>The device used to collect an environmental sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_collec_method/">samp_collec_method</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The method employed for collecting the sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_dis_stage/">samp_dis_stage</a></td>
<td>0..1 <br/> <a href="../SampDisStageEnum/">SampDisStageEnum</a></td>
<td>Stage of the disease at the time of sample collection, e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_mat_process/">samp_mat_process</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>A brief description of any processing applied to the sample during or after r...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_size/">samp_size</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>The total amount (volume or mass) of sample collected</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_store_dur/">samp_store_dur</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Duration for which the sample was stored</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_store_loc/">samp_store_loc</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Location at which sample was stored, usually name of a specific freezer/room</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../samp_store_temp/">samp_store_temp</a></td>
<td>1 <br/> <a href="../Float/">Float</a></td>
<td>Temperature at which the sample was stored (degrees are assumed)</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../sample_link/">sample_link</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>A unique identifier to assign parent-child, subsample, or sibling samples</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../season_environment/">season_environment</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Treatment involving an exposure to a particular season (e</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../size_frac/">size_frac</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Filtering pore size used in sample preparation</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../specific_ecosystem/">specific_ecosystem</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../SpecificEcosystemEnum/">SpecificEcosystemEnum</a></td>
<td>Specific ecosystems represent specific features of the environment like aphot...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../standing_water_regm/">standing_water_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Treatment involving an exposure to standing water during a plant's life span,...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../start_date_inc/">start_date_inc</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>Date the incubation was started</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../temp/">temp</a></td>
<td>0..1 <br/> <a href="../Float/">Float</a></td>
<td>Temperature of the sample at the time of sampling</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../tiss_cult_growth_med/">tiss_cult_growth_med</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Description of plant tissue culture growth media used</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../water_temp_regm/">water_temp_regm</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving an exposure to water with varying degre...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../watering_regm/">watering_regm</a></td>
<td>0..1 <em>recommended</em> <br/> <a href="../String/">String</a></td>
<td>Information about treatment involving an exposure to watering frequencies, tr...</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../analysis_type/">analysis_type</a></td>
<td>1..* <br/> <a href="../AnalysisTypeEnum/">AnalysisTypeEnum</a></td>
<td>Select all the data types associated or available for this biosample</td>
<td><a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a></td>
</tr>
<tr>
<td><a href="../samp_name/">samp_name</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>A local identifier or name that for the material sample collected</td>
<td><a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a></td>
</tr>
<tr>
<td><a href="../source_mat_id/">source_mat_id</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>A globally unique identifier assigned to the parent sample or sample that is ...</td>
<td><a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a></td>
</tr>
</tbody>
</table>
<h2 id="usages">Usages</h2>
<table>
<thead>
<tr>
<th>used by</th>
<th>used in</th>
<th>type</th>
<th>used</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../SampleData/">SampleData</a></td>
<td><a href="../plant_associated_data/">plant_associated_data</a></td>
<td>range</td>
<td><a href="./">PlantAssociatedInterface</a></td>
</tr>
</tbody>
</table>
<h2 id="identifier-and-mapping-information">Identifier and Mapping Information</h2>
<h3 id="annotations">Annotations</h3>
<table>
<thead>
<tr>
<th>property</th>
<th>value</th>
</tr>
</thead>
<tbody>
<tr>
<td>excel_worksheet_name</td>
<td>plant-associated</td>
</tr>
</tbody>
</table>
<h3 id="schema-source">Schema Source</h3>
<ul>
<li>from schema: <a href="https://example.com/nmdc_submission_schema">https://example.com/nmdc_submission_schema</a></li>
</ul>
<h2 id="mappings">Mappings</h2>
<table>
<thead>
<tr>
<th>Mapping Type</th>
<th>Mapped Value</th>
</tr>
</thead>
<tbody>
<tr>
<td>self</td>
<td>nmdc_sub_schema:PlantAssociatedInterface</td>
</tr>
<tr>
<td>native</td>
<td>nmdc_sub_schema:PlantAssociatedInterface</td>
</tr>
</tbody>
</table>
<h2 id="linkml-source">LinkML Source</h2>
<!-- TODO: investigate https://stackoverflow.com/questions/37606292/how-to-create-tabbed-code-blocks-in-mkdocs-or-sphinx -->
<h3 id="direct">Direct</h3>
<details>
<pre><code class="language-yaml">name: PlantAssociatedInterface
annotations:
  excel_worksheet_name:
    tag: excel_worksheet_name
    value: plant-associated
description: plant_associated dh_interface
title: Plant Associated
from_schema: https://example.com/nmdc_submission_schema
is_a: DhInterface
mixins:
- DhMultiviewCommonColumnsMixin
- SampIdNewTermsMixin
slots:
- air_temp_regm
- ances_data
- antibiotic_regm
- biol_stat
- biotic_regm
- biotic_relationship
- chem_administration
- chem_mutagen
- climate_environment
- collection_date
- collection_date_inc
- collection_time
- cult_root_med
- depth
- ecosystem
- ecosystem_category
- ecosystem_subtype
- ecosystem_type
- elev
- env_broad_scale
- env_local_scale
- env_medium
- experimental_factor
- fertilizer_regm
- fungicide_regm
- gaseous_environment
- genetic_mod
- geo_loc_name
- gravity
- growth_facil
- growth_habit
- growth_hormone_regm
- herbicide_regm
- host_age
- host_common_name
- host_disease_stat
- host_dry_mass
- host_genotype
- host_height
- host_length
- host_life_stage
- host_phenotype
- host_subspecf_genlin
- host_symbiont
- host_taxid
- host_tot_mass
- host_wet_mass
- humidity_regm
- isotope_exposure
- lat_lon
- light_regm
- mechanical_damage
- mineral_nutr_regm
- non_min_nutr_regm
- organism_count
- oxy_stat_samp
- perturbation
- pesticide_regm
- ph_regm
- plant_growth_med
- plant_product
- plant_sex
- plant_struc
- radiation_regm
- rainfall_regm
- root_cond
- root_med_carbon
- root_med_macronutr
- root_med_micronutr
- root_med_ph
- root_med_regl
- root_med_solid
- root_med_suppl
- salinity
- salinity_meth
- salt_regm
- samp_capt_status
- samp_collec_device
- samp_collec_method
- samp_dis_stage
- samp_mat_process
- samp_size
- samp_store_dur
- samp_store_loc
- samp_store_temp
- sample_link
- season_environment
- size_frac
- specific_ecosystem
- standing_water_regm
- start_date_inc
- temp
- tiss_cult_growth_med
- water_temp_regm
- watering_regm
slot_usage:
  air_temp_regm:
    name: air_temp_regm
    recommended: true
  biotic_regm:
    name: biotic_regm
    recommended: true
  climate_environment:
    name: climate_environment
    recommended: true
  depth:
    name: depth
    examples:
    - value: 0 - 0.1
    - value: '1'
    recommended: true
    pattern: ^[-+]?[0-9]*\.?[0-9]+(\s*-\s*[-+]?[0-9]*\.?[0-9]+)?$
    structured_pattern:
      syntax: ^{float}(\s*-\s*{float})?$
      interpolated: true
  ecosystem:
    name: ecosystem
    range: EcosystemEnum
  ecosystem_category:
    name: ecosystem_category
    range: EcosystemCategoryEnum
  ecosystem_subtype:
    name: ecosystem_subtype
    range: EcosystemSubtypeEnum
  ecosystem_type:
    name: ecosystem_type
    range: EcosystemTypeEnum
  elev:
    name: elev
    required: true
  env_local_scale:
    name: env_local_scale
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|PO:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|PO:\d{7})\]$
      interpolated: true
    any_of:
    - range: EnvLocalScalePlantAssociatedEnum
    - range: string
  env_medium:
    name: env_medium
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|PO:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|PO:\d{7})\]$
      interpolated: true
    any_of:
    - range: EnvMediumPlantAssociatedEnum
    - range: string
  gaseous_environment:
    name: gaseous_environment
    recommended: true
  host_common_name:
    name: host_common_name
    required: true
  host_taxid:
    name: host_taxid
    required: true
  humidity_regm:
    name: humidity_regm
    recommended: true
  lat_lon:
    name: lat_lon
    required: true
  light_regm:
    name: light_regm
    recommended: true
  samp_store_temp:
    name: samp_store_temp
    required: true
  specific_ecosystem:
    name: specific_ecosystem
    range: SpecificEcosystemEnum
  watering_regm:
    name: watering_regm
    recommended: true
  env_broad_scale:
    name: env_broad_scale
    any_of:
    - range: EnvBroadScalePlantAssociatedEnum
    - range: string

</code></pre>
</details>
<h3 id="induced">Induced</h3>
<details>
<pre><code class="language-yaml">name: PlantAssociatedInterface
annotations:
  excel_worksheet_name:
    tag: excel_worksheet_name
    value: plant-associated
description: plant_associated dh_interface
title: Plant Associated
from_schema: https://example.com/nmdc_submission_schema
is_a: DhInterface
mixins:
- DhMultiviewCommonColumnsMixin
- SampIdNewTermsMixin
slot_usage:
  air_temp_regm:
    name: air_temp_regm
    recommended: true
  biotic_regm:
    name: biotic_regm
    recommended: true
  climate_environment:
    name: climate_environment
    recommended: true
  depth:
    name: depth
    examples:
    - value: 0 - 0.1
    - value: '1'
    recommended: true
    pattern: ^[-+]?[0-9]*\.?[0-9]+(\s*-\s*[-+]?[0-9]*\.?[0-9]+)?$
    structured_pattern:
      syntax: ^{float}(\s*-\s*{float})?$
      interpolated: true
  ecosystem:
    name: ecosystem
    range: EcosystemEnum
  ecosystem_category:
    name: ecosystem_category
    range: EcosystemCategoryEnum
  ecosystem_subtype:
    name: ecosystem_subtype
    range: EcosystemSubtypeEnum
  ecosystem_type:
    name: ecosystem_type
    range: EcosystemTypeEnum
  elev:
    name: elev
    required: true
  env_local_scale:
    name: env_local_scale
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|PO:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|PO:\d{7})\]$
      interpolated: true
    any_of:
    - range: EnvLocalScalePlantAssociatedEnum
    - range: string
  env_medium:
    name: env_medium
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|PO:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|PO:\d{7})\]$
      interpolated: true
    any_of:
    - range: EnvMediumPlantAssociatedEnum
    - range: string
  gaseous_environment:
    name: gaseous_environment
    recommended: true
  host_common_name:
    name: host_common_name
    required: true
  host_taxid:
    name: host_taxid
    required: true
  humidity_regm:
    name: humidity_regm
    recommended: true
  lat_lon:
    name: lat_lon
    required: true
  light_regm:
    name: light_regm
    recommended: true
  samp_store_temp:
    name: samp_store_temp
    required: true
  specific_ecosystem:
    name: specific_ecosystem
    range: SpecificEcosystemEnum
  watering_regm:
    name: watering_regm
    recommended: true
  env_broad_scale:
    name: env_broad_scale
    any_of:
    - range: EnvBroadScalePlantAssociatedEnum
    - range: string
attributes:
  air_temp_regm:
    name: air_temp_regm
    annotations:
      Expected_value:
        tag: Expected_value
        value: temperature value;treatment interval and duration
      Preferred_unit:
        tag: Preferred_unit
        value: degree Celsius
    description: Information about treatment involving an exposure to varying temperatures;
      should include the temperature, treatment regimen including how many times the
      treatment was repeated, how long each treatment lasted, and the start and end
      time of the entire treatment; can include different temperature regimens
    title: air temperature regimen
    examples:
    - value: 25 degree Celsius;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 16
    keywords:
    - air
    - regimen
    - temperature
    string_serialization: '{float} {unit};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000551
    alias: air_temp_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
    multivalued: false
  ances_data:
    name: ances_data
    description: Information about either pedigree or other ancestral information
      description (e.g. parental variety in case of mutant or selection), e.g. A/3*B
      (meaning [(A x B) x B] x B)
    title: ancestral data
    examples:
    - value: A/3*B
    from_schema: https://example.com/nmdc_submission_schema
    rank: 237
    slot_uri: MIXS:0000247
    alias: ances_data
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  antibiotic_regm:
    name: antibiotic_regm
    annotations:
      Expected_value:
        tag: Expected_value
        value: antibiotic name;antibiotic amount;treatment interval and duration
      Preferred_unit:
        tag: Preferred_unit
        value: milligram
    description: Information about treatment involving antibiotic administration;
      should include the name of antibiotic, amount administered, treatment regimen
      including how many times the treatment was repeated, how long each treatment
      lasted, and the start and end time of the entire treatment; can include multiple
      antibiotic regimens
    title: antibiotic regimen
    examples:
    - value: penicillin;5 milligram;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 238
    keywords:
    - regimen
    string_serialization: '{text};{float} {unit};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000553
    alias: antibiotic_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  biol_stat:
    name: biol_stat
    description: The level of genome modification
    title: biological status
    examples:
    - value: natural
    from_schema: https://example.com/nmdc_submission_schema
    rank: 239
    keywords:
    - status
    slot_uri: MIXS:0000858
    alias: biol_stat
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: BiolStatEnum
  biotic_regm:
    name: biotic_regm
    description: Information about treatment(s) involving use of biotic factors, such
      as bacteria, viruses or fungi
    title: biotic regimen
    examples:
    - value: sample inoculated with Rhizobium spp. Culture
    from_schema: https://example.com/nmdc_submission_schema
    rank: 13
    keywords:
    - regimen
    slot_uri: MIXS:0001038
    alias: biotic_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
    multivalued: false
  biotic_relationship:
    name: biotic_relationship
    description: Description of relationship(s) between the subject organism and other
      organism(s) it is associated with. E.g., parasite on species X; mutualist with
      species Y. The target organism is the subject of the relationship, and the other
      organism(s) is the object
    title: observed biotic relationship
    examples:
    - value: free living
    from_schema: https://example.com/nmdc_submission_schema
    rank: 22
    keywords:
    - observed
    - relationship
    slot_uri: MIXS:0000028
    alias: biotic_relationship
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    slot_group: mixs_modified_section
    range: BioticRelationshipEnum
  chem_administration:
    name: chem_administration
    annotations:
      Expected_value:
        tag: Expected_value
        value: CHEBI;timestamp
    description: List of chemical compounds administered to the host or site where
      sampling occurred, and when (e.g. Antibiotics, n fertilizer, air filter); can
      include multiple compounds. For chemical entities of biological interest ontology
      (chebi) (v 163), http://purl.bioontology.org/ontology/chebi
    title: chemical administration
    examples:
    - value: agar [CHEBI:2509];2018-05-11|agar [CHEBI:2509];2018-05-22
    - value: agar [CHEBI:2509];2018-05
    from_schema: https://example.com/nmdc_submission_schema
    rank: 17
    keywords:
    - administration
    string_serialization: '{termLabel} [{termID}];{timestamp}'
    slot_uri: MIXS:0000751
    alias: chem_administration
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
    multivalued: false
    pattern: ^(\S+.*\S+ \[[A-za-z]+:\d+\];[12]\d{3}(?:(?:-(?:0[1-9]|1[0-2]))(?:-(?:0[1-9]|[12]\d|3[01]))?)?\|)*(\S+.*\S+
      \[[A-za-z]+:\d+\];[12]\d{3}(?:(?:-(?:0[1-9]|1[0-2]))(?:-(?:0[1-9]|[12]\d|3[01]))?)?)$
  chem_mutagen:
    name: chem_mutagen
    annotations:
      Expected_value:
        tag: Expected_value
        value: mutagen name;mutagen amount;treatment interval and duration
      Preferred_unit:
        tag: Preferred_unit
        value: milligram per liter
    description: Treatment involving use of mutagens; should include the name of mutagen,
      amount administered, treatment regimen including how many times the treatment
      was repeated, how long each treatment lasted, and the start and end time of
      the entire treatment; can include multiple mutagen regimens
    title: chemical mutagen
    examples:
    - value: nitrous acid;0.5 milligram per liter;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 240
    string_serialization: '{text};{float} {unit};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000555
    alias: chem_mutagen
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  climate_environment:
    name: climate_environment
    description: Treatment involving an exposure to a particular climate; treatment
      regimen including how many times the treatment was repeated, how long each treatment
      lasted, and the start and end time of the entire treatment; can include multiple
      climates
    title: climate environment
    deprecated: true, slot is inconsistently used and provides redundant information
      to other slots, https://github.com/GenomicsStandardsConsortium/mixs/issues/591
      and https://github.com/microbiomedata/nmdc-schema/issues/586
    todos:
    - description says "can include multiple climates" but multivalued is set to false
    - add examples, i need to see some examples to add correctly formatted example.
    examples:
    - value: tropical climate;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 19
    keywords:
    - environment
    slot_uri: MIXS:0001040
    alias: climate_environment
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
    multivalued: false
  collection_date:
    name: collection_date
    description: The date of sampling
    title: collection date
    examples:
    - value: '2021-04-15'
    - value: 2021-04
    - value: '2021'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 3
    keywords:
    - date
    slot_uri: MIXS:0000011
    alias: collection_date
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    required: true
    pattern: ^[12]\d{3}(?:(?:-(?:0[1-9]|1[0-2]))(?:-(?:0[1-9]|[12]\d|3[01]))?)?$
  collection_date_inc:
    name: collection_date_inc
    description: Date the incubation was harvested/collected/ended. Only relevant
      for incubation samples.
    title: incubation collection date
    notes:
    - MIxS collection_date accepts (truncated) ISO8601. DH taking arbitrary precision
      date only
    comments:
    - Date should be formatted as YYYY(-MM(-DD)). Ie, 2021-04-15, 2021-04 and 2021
      are all acceptable.
    examples:
    - value: '2021-04-15'
    - value: 2021-04
    - value: '2021'
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - MIXS:0000011
    rank: 2
    string_serialization: '{date, arbitrary precision}'
    alias: collection_date_inc
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    - SoilMixsInspiredMixin
    slot_group: mixs_inspired_section
    range: string
    recommended: true
    pattern: ^[12]\d{3}(?:(?:-(?:0[1-9]|1[0-2]))(?:-(?:0[1-9]|[12]\d|3[01]))?)?$
  collection_time:
    name: collection_time
    description: The time of sampling, either as an instance (single point) or interval.
    title: collection time, GMT
    notes:
    - MIxS collection_date accepts (truncated) ISO8601. DH taking seconds optional
      time only
    comments:
    - 'Time should be entered as HH:MM(:SS) in GMT. See here for a converter: https://www.worldtimebuddy.com/pst-to-gmt-converter'
    examples:
    - value: '13:33'
    - value: '13:33:55'
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - MIXS:0000011
    rank: 1
    string_serialization: '{time, seconds optional}'
    alias: collection_time
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    - SoilMixsInspiredMixin
    slot_group: mixs_inspired_section
    range: string
    recommended: true
    pattern: ^([01]?\d|2[0-3]|24(?=:00?:00?$)):([0-5]\d)(:([0-5]\d))?$
  cult_root_med:
    name: cult_root_med
    description: Name or reference for the hydroponic or in vitro culture rooting
      medium; can be the name of a commonly used medium or reference to a specific
      medium, e.g. Murashige and Skoog medium. If the medium has not been formally
      published, use the rooting medium descriptors
    title: culture rooting medium
    examples:
    - value: http://himedialabs.com/TD/PT158.pdf
    from_schema: https://example.com/nmdc_submission_schema
    rank: 241
    keywords:
    - culture
    slot_uri: MIXS:0001041
    alias: cult_root_med
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^(.*|PMID:\d+|doi:10.\d{2,9}/.*|https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_\+.~#?&amp;\/=]*))$
    structured_pattern:
      syntax: ^({text}|{PMID}|{DOI}|{URL})$
      interpolated: true
  depth:
    name: depth
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: meter
      storage_units:
        tag: storage_units
        value: m
    description: The vertical distance below local surface. For sediment or soil samples
      depth is measured from sediment or soil surface, respectively. Depth can be
      reported as an interval for subsurface samples
    title: depth, meters
    comments:
    - All depths must be reported in meters. Provide the numerical portion only.
    examples:
    - value: 0 - 0.1
    - value: '1'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 9
    keywords:
    - depth
    slot_uri: MIXS:0000018
    alias: depth
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
    pattern: ^[-+]?[0-9]*\.?[0-9]+(\s*-\s*[-+]?[0-9]*\.?[0-9]+)?$
    structured_pattern:
      syntax: ^{float}(\s*-\s*{float})?$
      interpolated: true
    unit:
      ucum_code: m
  ecosystem:
    name: ecosystem
    description: An ecosystem is a combination of a physical environment (abiotic
      factors) and all the organisms (biotic factors) that interact with this environment.
      Ecosystem is in position 1/5 in a GOLD path.
    comments:
    - The abiotic factors play a profound role on the type and composition of organisms
      in a given environment. The GOLD Ecosystem at the top of the five-level classification
      system is aimed at capturing the broader environment from which an organism
      or environmental sample is collected. The three broad groups under Ecosystem
      are Environmental, Host-associated, and Engineered. They represent samples collected
      from a natural environment or from another organism or from engineered environments
      like bioreactors respectively.
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - https://gold.jgi.doe.gov/help
    rank: 9
    is_a: gold_path_field
    alias: ecosystem
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: EcosystemEnum
    recommended: true
  ecosystem_category:
    name: ecosystem_category
    description: Ecosystem categories represent divisions within the ecosystem based
      on specific characteristics of the environment from where an organism or sample
      is isolated. Ecosystem category is in position 2/5 in a GOLD path.
    comments:
    - The Environmental ecosystem (for example) is divided into Air, Aquatic and Terrestrial.
      Ecosystem categories for Host-associated samples can be individual hosts or
      phyla and for engineered samples it may be manipulated environments like bioreactors,
      solid waste etc.
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - https://gold.jgi.doe.gov/help
    rank: 10
    is_a: gold_path_field
    alias: ecosystem_category
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: EcosystemCategoryEnum
    recommended: true
  ecosystem_subtype:
    name: ecosystem_subtype
    description: Ecosystem subtypes represent further subdivision of Ecosystem types
      into more distinct subtypes. Ecosystem subtype is in position 4/5 in a GOLD
      path.
    comments:
    - Ecosystem Type Marine (Environmental -&gt; Aquatic -&gt; Marine) is further divided
      (for example) into Intertidal zone, Coastal, Pelagic, Intertidal zone etc. in
      the Ecosystem subtype category.
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - https://gold.jgi.doe.gov/help
    rank: 12
    is_a: gold_path_field
    alias: ecosystem_subtype
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: EcosystemSubtypeEnum
    recommended: true
  ecosystem_type:
    name: ecosystem_type
    description: Ecosystem types represent things having common characteristics within
      the Ecosystem Category. These common characteristics based grouping is still
      broad but specific to the characteristics of a given environment. Ecosystem
      type is in position 3/5 in a GOLD path.
    comments:
    - The Aquatic ecosystem category (for example) may have ecosystem types like Marine
      or Thermal springs etc. Ecosystem category Air may have Indoor air or Outdoor
      air as different Ecosystem Types. In the case of Host-associated samples, ecosystem
      type can represent Respiratory system, Digestive system, Roots etc.
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - https://gold.jgi.doe.gov/help
    rank: 11
    is_a: gold_path_field
    alias: ecosystem_type
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: EcosystemTypeEnum
    recommended: true
  elev:
    name: elev
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: meter
    description: Elevation of the sampling site is its height above a fixed reference
      point, most commonly the mean sea level. Elevation is mainly used when referring
      to points on the earth's surface, while altitude is used for points above the
      surface, such as an aircraft in flight or a spacecraft in orbit
    title: elevation, meters
    comments:
    - All elevations must be reported in meters. Provide the numerical portion only.
    - Please use https://www.advancedconverter.com/map-tools/find-altitude-by-coordinates,
      if needed, to help estimate the elevation based on latitude and longitude coordinates.
    examples:
    - value: '225'
    - value: '0'
    - value: '1250'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 6
    keywords:
    - elevation
    slot_uri: MIXS:0000093
    alias: elev
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: float
    required: true
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
    unit:
      ucum_code: m
  env_broad_scale:
    name: env_broad_scale
    annotations:
      tooltip:
        tag: tooltip
        value: The biome or major environmental system where the sample or specimen
          originated. Choose values from subclasses of the 'biome' class [ENVO:00000428]
          in the Environment Ontology (ENVO). For host-associated or plant-associated
          samples, use terms from the UBERON or Plant Ontology to describe the broad
          anatomical or morphological context
    description: 'In this field, report which major environmental system your sample
      or specimen came from. The systems identified should have a coarse spatial grain,
      to provide the general environmental context of where the sampling was done
      (e.g. were you in the desert or a rainforest?). We recommend using subclasses
      of ENVO''s biome class: http://purl.obolibrary.org/obo/ENVO_00000428. Format
      (one term): termLabel [termID], Format (multiple terms): termLabel [termID]|termLabel
      [termID]|termLabel [termID]. Example: Annotating a water sample from the photic
      zone in middle of the Atlantic Ocean, consider: oceanic epipelagic zone biome
      [ENVO:01000033]. Example: Annotating a sample from the Amazon rainforest consider:
      tropical moist broadleaf forest biome [ENVO:01000228]. If needed, request new
      terms on the ENVO tracker, identified here: http://www.obofoundry.org/ontology/envo.html'
    title: broad-scale environmental context
    examples:
    - value: oceanic epipelagic zone biome [ENVO:01000035]
    from_schema: https://example.com/nmdc_submission_schema
    rank: 6
    keywords:
    - context
    - environmental
    is_a: mixs_env_triad_field
    slot_uri: MIXS:0000012
    alias: env_broad_scale
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: string
    required: true
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[ENVO:\d{7,8}\]$
    structured_pattern:
      syntax: ^{termLabel} \[ENVO:\d{7,8}\]$
      interpolated: true
    any_of:
    - range: EnvBroadScalePlantAssociatedEnum
    - range: string
  env_local_scale:
    name: env_local_scale
    annotations:
      Expected_value:
        tag: Expected_value
        value: Environmental entities having causal influences upon the entity at
          time of sampling
      tooltip:
        tag: tooltip
        value: The specific environmental entities or features near the sample or
          specimen that significantly influence its characteristics or composition.
          These entities are typically smaller in scale than the broad environmental
          context. Values for this field should be countable, material nouns and must
          be chosen from subclasses of BFO:0000040 (material entity) that appear in
          the Environment Ontology (ENVO). For host-associated or plant-associated
          samples, use terms from the UBERON or Plant Ontology to describe specific
          anatomical structures or plant parts.
    description: 'Report the entity or entities which are in the sample or specimen
      s local vicinity and which you believe have significant causal influences on
      your sample or specimen. We recommend using EnvO terms which are of smaller
      spatial grain than your entry for env_broad_scale. Terms, such as anatomical
      sites, from other OBO Library ontologies which interoperate with EnvO (e.g.
      UBERON) are accepted in this field. EnvO documentation about how to use the
      field: https://github.com/EnvironmentOntology/envo/wiki/Using-ENVO-with-MIxS'
    title: local environmental context
    examples:
    - value: canopy [ENVO:00000047]
    - value: herb and fern layer [ENVO:01000337]
    - value: understory [ENVO:01000335]
    - value: shrub layer [ENVO:01000336]
    from_schema: https://example.com/nmdc_submission_schema
    rank: 7
    keywords:
    - context
    - environmental
    is_a: mixs_env_triad_field
    slot_uri: MIXS:0000013
    alias: env_local_scale
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: string
    required: true
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|PO:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|PO:\d{7})\]$
      interpolated: true
    any_of:
    - range: EnvLocalScalePlantAssociatedEnum
    - range: string
  env_medium:
    name: env_medium
    annotations:
      tooltip:
        tag: tooltip
        value: The predominant environmental material or substrate that directly surrounds
          or hosts the sample or specimen at the time of sampling. Choose values from
          subclasses of the 'environmental material' class [ENVO:00010483] in the
          Environment Ontology (ENVO). Values for this field should be measurable
          or mass material nouns, representing continuous environmental materials.
          For host-associated or plant-associated samples, use terms from the UBERON
          or Plant Ontology to indicate a tissue, organ, or plant structure
    description: 'Report the environmental material(s) immediately surrounding the
      sample or specimen at the time of sampling. We recommend using subclasses of
      ''environmental material'' (http://purl.obolibrary.org/obo/ENVO_00010483). EnvO
      documentation about how to use the field: https://github.com/EnvironmentOntology/envo/wiki/Using-ENVO-with-MIxS
      . Terms from other OBO ontologies are permissible as long as they reference
      mass/volume nouns (e.g. air, water, blood) and not discrete, countable entities
      (e.g. a tree, a leaf, a table top)'
    title: environmental medium
    examples:
    - value: soil [ENVO:00001998]
    from_schema: https://example.com/nmdc_submission_schema
    rank: 8
    keywords:
    - environmental
    is_a: mixs_env_triad_field
    slot_uri: MIXS:0000014
    alias: env_medium
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: string
    required: true
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[(ENVO:\d{7,8}|PO:\d{7})\]$
    structured_pattern:
      syntax: ^{termLabel} \[(ENVO:\d{7,8}|PO:\d{7})\]$
      interpolated: true
    any_of:
    - range: EnvMediumPlantAssociatedEnum
    - range: string
  experimental_factor:
    name: experimental_factor
    annotations:
      Expected_value:
        tag: Expected_value
        value: text or EFO and/or OBI
    description: Variable aspects of an experiment design that can be used to describe
      an experiment, or set of experiments, in an increasingly detailed manner. This
      field accepts ontology terms from Experimental Factor Ontology (EFO) and/or
      Ontology for Biomedical Investigations (OBI)
    title: experimental factor
    examples:
    - value: time series design [EFO:0001779]
    from_schema: https://example.com/nmdc_submission_schema
    rank: 12
    keywords:
    - experimental
    - factor
    string_serialization: '{termLabel} [{termID}]|{text}'
    slot_uri: MIXS:0000008
    alias: experimental_factor
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_section
    range: string
    multivalued: false
    pattern: ^\S+.*\S+ \[[a-zA-Z]{2,}:\d+\]$
  fertilizer_regm:
    name: fertilizer_regm
    annotations:
      Expected_value:
        tag: Expected_value
        value: fertilizer name;fertilizer amount;treatment interval and duration
      Preferred_unit:
        tag: Preferred_unit
        value: gram, mole per liter, milligram per liter
    description: Information about treatment involving the use of fertilizers; should
      include the name of fertilizer, amount administered, treatment regimen including
      how many times the treatment was repeated, how long each treatment lasted, and
      the start and end time of the entire treatment; can include multiple fertilizer
      regimens
    title: fertilizer regimen
    examples:
    - value: urea;0.6 milligram per liter;R2/2018-05-11:T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 242
    keywords:
    - regimen
    string_serialization: '{text};{float} {unit};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000556
    alias: fertilizer_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  fungicide_regm:
    name: fungicide_regm
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: gram, mole per liter, milligram per liter
    description: Information about treatment involving use of fungicides; should include
      the name of fungicide, amount administered, treatment regimen including how
      many times the treatment was repeated, how long each treatment lasted, and the
      start and end time of the entire treatment; can include multiple fungicide regimens
    title: fungicide regimen
    examples:
    - value: bifonazole;1 mole per liter;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 243
    keywords:
    - regimen
    slot_uri: MIXS:0000557
    alias: fungicide_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  gaseous_environment:
    name: gaseous_environment
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: micromole per liter
    description: Use of conditions with differing gaseous environments; should include
      the name of gaseous compound, amount administered, treatment duration, interval
      and total experimental duration; can include multiple gaseous environment regimens
    title: gaseous environment
    todos:
    - would like to see usage examples for this slot. Requiring micromole/L seems
      too limiting and doesn't match expected_value value
    - did I do this right? keep the example that's provided and add another? so as
      to not override
    examples:
    - value: CO2; 500ppm above ambient; constant
    - value: nitric oxide;0.5 micromole per liter;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 20
    keywords:
    - environment
    slot_uri: MIXS:0000558
    alias: gaseous_environment
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
    multivalued: false
  genetic_mod:
    name: genetic_mod
    description: Genetic modifications of the genome of an organism, which may occur
      naturally by spontaneous mutation, or be introduced by some experimental means,
      e.g. specification of a transgene or the gene knocked-out or details of transient
      transfection
    title: genetic modification
    examples:
    - value: PMID:19497774
    from_schema: https://example.com/nmdc_submission_schema
    rank: 244
    slot_uri: MIXS:0000859
    alias: genetic_mod
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^(PMID:\d+|doi:10.\d{2,9}/.*|https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_\+.~#?&amp;\/=]*))$
    structured_pattern:
      syntax: ^({PMID}|{DOI}|{URL})$
      interpolated: true
  geo_loc_name:
    name: geo_loc_name
    description: The geographical origin of the sample as defined by the country or
      sea name followed by specific region name.
    title: geographic location (country and/or sea,region)
    examples:
    - value: 'USA: Maryland, Bethesda'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 4
    keywords:
    - geographic
    - location
    slot_uri: MIXS:0000010
    alias: geo_loc_name
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    required: true
    pattern: '^([^\s-]{1,2}|[^\s-]+.+[^\s-]+): ([^\s-]{1,2}|[^\s-]+.+[^\s-]+), ([^\s-]{1,2}|[^\s-]+.+[^\s-]+)$'
    structured_pattern:
      syntax: '^{country}: {region}, {specific_location}$'
      interpolated: true
      partial_match: true
  gravity:
    name: gravity
    annotations:
      Expected_value:
        tag: Expected_value
        value: gravity factor value;treatment interval and duration
      Preferred_unit:
        tag: Preferred_unit
        value: meter per square second, g
    description: Information about treatment involving use of gravity factor to study
      various types of responses in presence, absence or modified levels of gravity;
      treatment regimen including how many times the treatment was repeated, how long
      each treatment lasted, and the start and end time of the entire treatment; can
      include multiple treatments
    title: gravity
    examples:
    - value: 12 g;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 245
    string_serialization: '{float} {unit};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000559
    alias: gravity
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  growth_facil:
    name: growth_facil
    annotations:
      Expected_value:
        tag: Expected_value
        value: free text or CO
    description: 'Type of facility/location where the sample was harvested; controlled
      vocabulary: growth chamber, open top chamber, glasshouse, experimental garden,
      field.'
    title: growth facility
    examples:
    - value: growth_chamber
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1
    keywords:
    - facility
    - growth
    string_serialization: '{text}|{termLabel} [{termID}]'
    slot_uri: MIXS:0001043
    alias: growth_facil
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SoilInterface
    slot_group: mixs_modified_section
    range: GrowthFacilEnum
    required: true
  growth_habit:
    name: growth_habit
    description: Characteristic shape, appearance or growth form of a plant species
    title: growth habit
    examples:
    - value: spreading
    from_schema: https://example.com/nmdc_submission_schema
    rank: 246
    keywords:
    - growth
    slot_uri: MIXS:0001044
    alias: growth_habit
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: GrowthHabitEnum
  growth_hormone_regm:
    name: growth_hormone_regm
    annotations:
      Expected_value:
        tag: Expected_value
        value: growth hormone name;growth hormone amount;treatment interval and duration
      Preferred_unit:
        tag: Preferred_unit
        value: gram, mole per liter, milligram per liter
    description: Information about treatment involving use of growth hormones; should
      include the name of growth hormone, amount administered, treatment regimen including
      how many times the treatment was repeated, how long each treatment lasted, and
      the start and end time of the entire treatment; can include multiple growth
      hormone regimens
    title: growth hormone regimen
    examples:
    - value: abscisic acid;0.5 milligram per liter;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 247
    keywords:
    - growth
    - regimen
    string_serialization: '{text};{float} {unit};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000560
    alias: growth_hormone_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  herbicide_regm:
    name: herbicide_regm
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: gram, mole per liter, milligram per liter
    description: Information about treatment involving use of herbicides; information
      about treatment involving use of growth hormones; should include the name of
      herbicide, amount administered, treatment regimen including how many times the
      treatment was repeated, how long each treatment lasted, and the start and end
      time of the entire treatment; can include multiple regimens
    title: herbicide regimen
    examples:
    - value: atrazine;10 milligram per liter;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 248
    keywords:
    - regimen
    slot_uri: MIXS:0000561
    alias: herbicide_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  host_age:
    name: host_age
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: year, day, hour
      storage_units:
        tag: storage_units
        value: a|d|h
    description: Age of host at the time of sampling; relevant scale depends on species
      and study, e.g. Could be seconds for amoebae or centuries for trees
    title: host age
    comments:
    - 'Value must be reported in one of the following units: a, d, h. Provide the
      numeric portion followed by a space and the unit.'
    examples:
    - value: 10 d
    from_schema: https://example.com/nmdc_submission_schema
    rank: 249
    keywords:
    - age
    - host
    - host.
    slot_uri: MIXS:0000255
    alias: host_age
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  host_common_name:
    name: host_common_name
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: ''
    description: Common name of the host
    title: host common name
    from_schema: https://example.com/nmdc_submission_schema
    rank: 250
    keywords:
    - host
    - host.
    slot_uri: MIXS:0000248
    alias: host_common_name
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    required: true
  host_disease_stat:
    name: host_disease_stat
    annotations:
      Expected_value:
        tag: Expected_value
        value: disease name or Disease Ontology term
    description: List of diseases with which the host has been diagnosed; can include
      multiple diagnoses. The value of the field depends on host; for humans the terms
      should be chosen from the DO (Human Disease Ontology) at https://www.disease-ontology.org,
      non-human host diseases are free text
    title: host disease status
    from_schema: https://example.com/nmdc_submission_schema
    rank: 390
    keywords:
    - disease
    - host
    - host.
    - status
    string_serialization: '{termLabel} [{termID}]|{text}'
    slot_uri: MIXS:0000031
    alias: host_disease_stat
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_dry_mass:
    name: host_dry_mass
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: kilogram, gram
      storage_units:
        tag: storage_units
        value: g|kg
    description: Measurement of dry mass
    title: host dry mass
    comments:
    - 'Value must be reported in one of the following units: g, kg. Provide the numeric
      portion followed by a space and the unit.'
    examples:
    - value: 500 g
    from_schema: https://example.com/nmdc_submission_schema
    rank: 251
    keywords:
    - dry
    - host
    - host.
    - mass
    slot_uri: MIXS:0000257
    alias: host_dry_mass
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  host_genotype:
    name: host_genotype
    description: Observed genotype
    title: host genotype
    from_schema: https://example.com/nmdc_submission_schema
    rank: 252
    keywords:
    - host
    - host.
    slot_uri: MIXS:0000365
    alias: host_genotype
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_height:
    name: host_height
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: centimeter, millimeter, meter
      storage_units:
        tag: storage_units
        value: cm|m|mm
    description: The height of subject
    title: host height
    comments:
    - 'Value must be reported in one of the following units: cm, m, mm. Provide the
      numeric portion followed by a space and the unit.'
    examples:
    - value: 0.1 m
    from_schema: https://example.com/nmdc_submission_schema
    rank: 253
    keywords:
    - height
    - host
    - host.
    slot_uri: MIXS:0000264
    alias: host_height
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  host_length:
    name: host_length
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: centimeter, millimeter, meter
      storage_units:
        tag: storage_units
        value: cm|m|mm
    description: The length of subject
    title: host length
    comments:
    - 'Value must be reported in one of the following units: cm, m, mm. Provide the
      numeric portion followed by a space and the unit.'
    examples:
    - value: 1 m
    from_schema: https://example.com/nmdc_submission_schema
    rank: 254
    keywords:
    - host
    - host.
    - length
    slot_uri: MIXS:0000256
    alias: host_length
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  host_life_stage:
    name: host_life_stage
    annotations:
      Expected_value:
        tag: Expected_value
        value: stage
    description: Description of life stage of host
    title: host life stage
    from_schema: https://example.com/nmdc_submission_schema
    rank: 255
    keywords:
    - host
    - host.
    - life
    slot_uri: MIXS:0000251
    alias: host_life_stage
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  host_phenotype:
    name: host_phenotype
    annotations:
      Expected_value:
        tag: Expected_value
        value: PATO or HP
    description: Phenotype of human or other host. Use terms from the phenotypic quality
      ontology (pato) or the Human Phenotype Ontology (HP)
    title: host phenotype
    from_schema: https://example.com/nmdc_submission_schema
    rank: 256
    keywords:
    - host
    - host.
    string_serialization: '{termLabel} [{termID}]'
    slot_uri: MIXS:0000874
    alias: host_phenotype
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^\S+.*\S+ \[[a-zA-Z]{2,}:\d+\]$
  host_subspecf_genlin:
    name: host_subspecf_genlin
    annotations:
      Expected_value:
        tag: Expected_value
        value: Genetic lineage below lowest rank of NCBI taxonomy, which is subspecies,
          e.g. serovar, biotype, ecotype, variety, cultivar
    description: Information about the genetic distinctness of the host organism below
      the subspecies level e.g., serovar, serotype, biotype, ecotype, variety, cultivar,
      or any relevant genetic typing schemes like Group I plasmid. Subspecies should
      not be recorded in this term, but in the NCBI taxonomy. Supply both the lineage
      name and the lineage rank separated by a colon, e.g., biovar:abc123
    title: host subspecific genetic lineage
    examples:
    - value: 'serovar:Newport, variety:glabrum, cultivar: Red Delicious'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 257
    keywords:
    - host
    - host.
    - lineage
    string_serialization: '{rank name}:{text}'
    slot_uri: MIXS:0001318
    alias: host_subspecf_genlin
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  host_symbiont:
    name: host_symbiont
    annotations:
      Expected_value:
        tag: Expected_value
        value: species name or common name
    description: The taxonomic name of the organism(s) found living in mutualistic,
      commensalistic, or parasitic symbiosis with the specific host. The sampled symbiont
      can have its own symbionts. For example, parasites may have hyperparasites (=parasites
      of the parasite)
    title: observed host symbionts
    from_schema: https://example.com/nmdc_submission_schema
    rank: 258
    keywords:
    - host
    - host.
    - observed
    - symbiosis
    slot_uri: MIXS:0001298
    alias: host_symbiont
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  host_taxid:
    name: host_taxid
    annotations:
      Expected_value:
        tag: Expected_value
        value: NCBI taxon identifier
    description: NCBI taxon id of the host, e.g. 9606
    title: host taxid
    comments:
    - Homo sapiens [NCBITaxon:9606] would be a reasonable has_raw_value
    from_schema: https://example.com/nmdc_submission_schema
    rank: 259
    keywords:
    - host
    - host.
    - taxon
    slot_uri: MIXS:0000250
    alias: host_taxid
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    required: true
  host_tot_mass:
    name: host_tot_mass
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: kilogram, gram
      storage_units:
        tag: storage_units
        value: g|kg
    description: Total mass of the host at collection, the unit depends on host
    title: host total mass
    comments:
    - 'Value must be reported in one of the following units: g, kg. Provide the numeric
      portion followed by a space and the unit.'
    examples:
    - value: 2500 g
    from_schema: https://example.com/nmdc_submission_schema
    rank: 260
    keywords:
    - host
    - host.
    - mass
    - total
    slot_uri: MIXS:0000263
    alias: host_tot_mass
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  host_wet_mass:
    name: host_wet_mass
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: kilogram, gram
      storage_units:
        tag: storage_units
        value: g|kg
    description: Measurement of wet mass
    title: host wet mass
    comments:
    - 'Value must be reported in one of the following units: g, kg. Provide the numeric
      portion followed by a space and the unit.'
    examples:
    - value: 1500 g
    from_schema: https://example.com/nmdc_submission_schema
    rank: 261
    keywords:
    - host
    - host.
    - mass
    - wet
    slot_uri: MIXS:0000567
    alias: host_wet_mass
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  humidity_regm:
    name: humidity_regm
    annotations:
      Expected_value:
        tag: Expected_value
        value: humidity value;treatment interval and duration
      Preferred_unit:
        tag: Preferred_unit
        value: gram per cubic meter
    description: Information about treatment involving an exposure to varying degree
      of humidity; information about treatment involving use of growth hormones; should
      include amount of humidity administered, treatment regimen including how many
      times the treatment was repeated, how long each treatment lasted, and the start
      and end time of the entire treatment; can include multiple regimens
    title: humidity regimen
    examples:
    - value: 25 gram per cubic meter;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 21
    keywords:
    - humidity
    - regimen
    string_serialization: '{float} {unit};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000568
    alias: humidity_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
    multivalued: false
  isotope_exposure:
    name: isotope_exposure
    description: List isotope exposure or addition applied to your sample.
    title: isotope exposure/addition
    todos:
    - Can we make the H218O correctly super and subscripted?
    comments:
    - This is required when your experimental design includes the use of isotopically
      labeled compounds
    examples:
    - value: 13C glucose
    - value: 18O water
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - MIXS:0000751
    rank: 16
    alias: isotope_exposure
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    - SoilMixsInspiredMixin
    slot_group: mixs_inspired_section
    range: string
    recommended: true
  lat_lon:
    name: lat_lon
    description: The geographical origin of the sample as defined by latitude and
      longitude. The values should be reported in decimal degrees, limited to 8 decimal
      points, and in WGS84 system
    title: geographic location (latitude and longitude)
    notes:
    - This is currently a required field but it's not clear if this should be required
      for human hosts
    examples:
    - value: 50.586825 6.408977
    from_schema: https://example.com/nmdc_submission_schema
    rank: 5
    keywords:
    - geographic
    - location
    slot_uri: MIXS:0000009
    alias: lat_lon
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    required: true
    pattern: ^(-?((?:[0-8]?[0-9](?:\.\d{0,8})?)|90)) -?[0-9]+(?:\.[0-9]{0,8})?$|^-?(1[0-7]{1,2})$
    structured_pattern:
      syntax: ^{lat} {lon}$
      interpolated: true
      partial_match: true
  light_regm:
    name: light_regm
    annotations:
      Expected_value:
        tag: Expected_value
        value: exposure type;light intensity;light quality
      Preferred_unit:
        tag: Preferred_unit
        value: lux; micrometer, nanometer, angstrom
    description: Information about treatment(s) involving exposure to light, including
      both light intensity and quality
    title: light regimen
    examples:
    - value: incandescant light;10 lux;450 nanometer
    from_schema: https://example.com/nmdc_submission_schema
    rank: 24
    keywords:
    - light
    - regimen
    string_serialization: '{text};{float} {unit};{float} {unit}'
    slot_uri: MIXS:0000569
    alias: light_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
  mechanical_damage:
    name: mechanical_damage
    annotations:
      Expected_value:
        tag: Expected_value
        value: damage type;body site
    description: Information about any mechanical damage exerted on the plant; can
      include multiple damages and sites
    title: mechanical damage
    examples:
    - value: pruning;bark
    from_schema: https://example.com/nmdc_submission_schema
    rank: 262
    string_serialization: '{text};{text}'
    slot_uri: MIXS:0001052
    alias: mechanical_damage
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  mineral_nutr_regm:
    name: mineral_nutr_regm
    annotations:
      Expected_value:
        tag: Expected_value
        value: mineral nutrient name;mineral nutrient amount;treatment interval and
          duration
      Preferred_unit:
        tag: Preferred_unit
        value: gram, mole per liter, milligram per liter
    description: Information about treatment involving the use of mineral supplements;
      should include the name of mineral nutrient, amount administered, treatment
      regimen including how many times the treatment was repeated, how long each treatment
      lasted, and the start and end time of the entire treatment; can include multiple
      mineral nutrient regimens
    title: mineral nutrient regimen
    examples:
    - value: potassium;15 gram;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 263
    keywords:
    - mineral
    - nutrient
    - regimen
    string_serialization: '{text};{float} {unit};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000570
    alias: mineral_nutr_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  non_min_nutr_regm:
    name: non_min_nutr_regm
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: gram, mole per liter, milligram per liter
    description: Information about treatment involving the exposure of plant to non-mineral
      nutrient such as oxygen, hydrogen or carbon; should include the name of non-mineral
      nutrient, amount administered, treatment regimen including how many times the
      treatment was repeated, how long each treatment lasted, and the start and end
      time of the entire treatment; can include multiple non-mineral nutrient regimens
    title: non-mineral nutrient regimen
    examples:
    - value: https://phylogenomics.me/protocols/16s-pcr-protocol/
    from_schema: https://example.com/nmdc_submission_schema
    rank: 264
    keywords:
    - non-mineral
    - nutrient
    - regimen
    slot_uri: MIXS:0000571
    alias: non_min_nutr_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  organism_count:
    name: organism_count
    annotations:
      Expected_value:
        tag: Expected_value
        value: organism name;measurement value;enumeration
      storage_units:
        tag: storage_units
        value: '1'
    description: 'Total cell count of any organism (or group of organisms) per gram,
      volume or area of sample, should include name of organism followed by count.
      The method that was used for the enumeration (e.g. qPCR, atp, mpn, etc.) Should
      also be provided. (example: total prokaryotes; 3.5e7 cells per ml; qpcr)'
    title: organism count
    examples:
    - value: total prokaryotes;3.5e7 cells/mL;qPCR
    from_schema: https://example.com/nmdc_submission_schema
    rank: 30
    keywords:
    - count
    - organism
    string_serialization: '{text};{float} {unit};[ATP|MPN|qPCR|other]'
    slot_uri: MIXS:0000103
    alias: organism_count
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
    pattern: ^(\S+.*\S+;[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)? \S+.*\S+;(qPCR|ATP|MPN|other)\|)*(\S+.*\S+;[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?
      \S+.*\S+;(qPCR|ATP|MPN|other))$
  oxy_stat_samp:
    name: oxy_stat_samp
    description: Oxygenation status of sample
    title: oxygenation status of sample
    examples:
    - value: aerobic
    from_schema: https://example.com/nmdc_submission_schema
    rank: 25
    keywords:
    - oxygen
    - sample
    - status
    slot_uri: MIXS:0000753
    alias: oxy_stat_samp
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_section
    range: OxyStatSampEnum
  perturbation:
    name: perturbation
    annotations:
      Expected_value:
        tag: Expected_value
        value: perturbation type name;perturbation interval and duration
    description: Type of perturbation, e.g. chemical administration, physical disturbance,
      etc., coupled with perturbation regimen including how many times the perturbation
      was repeated, how long each perturbation lasted, and the start and end time
      of the entire perturbation period; can include multiple perturbation types
    title: perturbation
    examples:
    - value: antibiotic addition;R2/2018-05-11T14:30Z/2018-05-11T19:30Z/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 33
    keywords:
    - perturbation
    string_serialization: '{text};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000754
    alias: perturbation
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  pesticide_regm:
    name: pesticide_regm
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: gram, mole per liter, milligram per liter
    description: Information about treatment involving use of insecticides; should
      include the name of pesticide, amount administered, treatment regimen including
      how many times the treatment was repeated, how long each treatment lasted, and
      the start and end time of the entire treatment; can include multiple pesticide
      regimens
    title: pesticide regimen
    examples:
    - value: pyrethrum;0.6 milligram per liter;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 265
    keywords:
    - regimen
    slot_uri: MIXS:0000573
    alias: pesticide_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  ph_regm:
    name: ph_regm
    description: Information about treatment involving exposure of plants to varying
      levels of ph of the growth media, treatment regimen including how many times
      the treatment was repeated, how long each treatment lasted, and the start and
      end time of the entire treatment; can include multiple regimen
    title: pH regimen
    examples:
    - value: 7.6;R2/2018-05-11:T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 266
    keywords:
    - ph
    - regimen
    slot_uri: MIXS:0001056
    alias: ph_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  plant_growth_med:
    name: plant_growth_med
    annotations:
      Expected_value:
        tag: Expected_value
        value: EO or enumeration
    description: Specification of the media for growing the plants or tissue cultured
      samples, e.g. soil, aeroponic, hydroponic, in vitro solid culture medium, in
      vitro liquid culture medium. Recommended value is a specific value from EO:plant
      growth medium (follow this link for terms http://purl.obolibrary.org/obo/EO_0007147)
      or other controlled vocabulary
    title: plant growth medium
    from_schema: https://example.com/nmdc_submission_schema
    rank: 267
    keywords:
    - growth
    - plant
    string_serialization: '{termLabel} [{termID}] or [husk|other artificial liquid
      medium|other artificial solid medium|peat moss|perlite|pumice|sand|soil|vermiculite|water]'
    slot_uri: MIXS:0001057
    alias: plant_growth_med
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^\S+.*\S+ \[[a-zA-Z]{2,}:\d+\]$
  plant_product:
    name: plant_product
    annotations:
      Expected_value:
        tag: Expected_value
        value: product name
    description: Substance produced by the plant, where the sample was obtained from
    title: plant product
    examples:
    - value: xylem sap [PO:0025539]
    from_schema: https://example.com/nmdc_submission_schema
    rank: 268
    keywords:
    - plant
    - product
    slot_uri: MIXS:0001058
    alias: plant_product
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  plant_sex:
    name: plant_sex
    description: Sex of the reproductive parts on the whole plant, e.g. pistillate,
      staminate, monoecieous, hermaphrodite
    title: plant sex
    examples:
    - value: Hermaphroditic
    from_schema: https://example.com/nmdc_submission_schema
    rank: 269
    keywords:
    - plant
    slot_uri: MIXS:0001059
    alias: plant_sex
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: PlantSexEnum
  plant_struc:
    name: plant_struc
    description: Name of plant structure the sample was obtained from; for Plant Ontology
      (PO) (v releases/2017-12-14) terms, see http://purl.bioontology.org/ontology/PO,
      e.g. petiole epidermis (PO_0000051). If an individual flower is sampled, the
      sex of it can be recorded here
    title: plant structure
    examples:
    - value: epidermis [PO:0005679]
    from_schema: https://example.com/nmdc_submission_schema
    rank: 270
    keywords:
    - plant
    slot_uri: MIXS:0001060
    alias: plant_struc
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    required: true
    pattern: ^([^\s-]{1,2}|[^\s-]+.+[^\s-]+) \[[a-zA-Z]{2,}:[a-zA-Z0-9]\d+\]$
    structured_pattern:
      syntax: ^{termLabel} \[{termID}\]$
      interpolated: true
      partial_match: true
  radiation_regm:
    name: radiation_regm
    annotations:
      Expected_value:
        tag: Expected_value
        value: radiation type name;radiation amount;treatment interval and duration
      Preferred_unit:
        tag: Preferred_unit
        value: rad, gray
    description: Information about treatment involving exposure of plant or a plant
      part to a particular radiation regimen; should include the radiation type, amount
      or intensity administered, treatment regimen including how many times the treatment
      was repeated, how long each treatment lasted, and the start and end time of
      the entire treatment; can include multiple radiation regimens
    title: radiation regimen
    examples:
    - value: gamma radiation;60 gray;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 271
    keywords:
    - regimen
    string_serialization: '{text};{float} {unit};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000575
    alias: radiation_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  rainfall_regm:
    name: rainfall_regm
    annotations:
      Expected_value:
        tag: Expected_value
        value: measurement value;treatment interval and duration
      Preferred_unit:
        tag: Preferred_unit
        value: millimeter
    description: Information about treatment involving an exposure to a given amount
      of rainfall, treatment regimen including how many times the treatment was repeated,
      how long each treatment lasted, and the start and end time of the entire treatment;
      can include multiple regimens
    title: rainfall regimen
    examples:
    - value: 15 millimeter;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 272
    keywords:
    - rain
    - regimen
    string_serialization: '{float} {unit};{Rn/start_time/end_time/duration}'
    slot_uri: MIXS:0000576
    alias: rainfall_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  root_cond:
    name: root_cond
    description: Relevant rooting conditions such as field plot size, sowing density,
      container dimensions, number of plants per container
    title: rooting conditions
    examples:
    - value: http://himedialabs.com/TD/PT158.pdf
    from_schema: https://example.com/nmdc_submission_schema
    rank: 273
    keywords:
    - condition
    slot_uri: MIXS:0001061
    alias: root_cond
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^(PMID:\d+|doi:10.\d{2,9}/.*|https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_\+.~#?&amp;\/=]*)|.*)$
    structured_pattern:
      syntax: ^({PMID}|{DOI}|{URL}|{text})$
      interpolated: true
      partial_match: true
  root_med_carbon:
    name: root_med_carbon
    annotations:
      Expected_value:
        tag: Expected_value
        value: carbon source name;measurement value
      Preferred_unit:
        tag: Preferred_unit
        value: milligram per liter
    description: Source of organic carbon in the culture rooting medium; e.g. sucrose
    title: rooting medium carbon
    examples:
    - value: sucrose;50 mg/L
    from_schema: https://example.com/nmdc_submission_schema
    rank: 274
    keywords:
    - carbon
    string_serialization: '{text};{float} {unit}'
    slot_uri: MIXS:0000577
    alias: root_med_carbon
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  root_med_macronutr:
    name: root_med_macronutr
    annotations:
      Expected_value:
        tag: Expected_value
        value: macronutrient name;measurement value
      Preferred_unit:
        tag: Preferred_unit
        value: milligram per liter
    description: Measurement of the culture rooting medium macronutrients (N,P, K,
      Ca, Mg, S); e.g. KH2PO4 (170 mg/L)
    title: rooting medium macronutrients
    examples:
    - value: KH2PO4;170 mg/L
    from_schema: https://example.com/nmdc_submission_schema
    rank: 275
    keywords:
    - macronutrients
    string_serialization: '{text};{float} {unit}'
    slot_uri: MIXS:0000578
    alias: root_med_macronutr
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  root_med_micronutr:
    name: root_med_micronutr
    annotations:
      Expected_value:
        tag: Expected_value
        value: micronutrient name;measurement value
      Preferred_unit:
        tag: Preferred_unit
        value: milligram per liter
    description: Measurement of the culture rooting medium micronutrients (Fe, Mn,
      Zn, B, Cu, Mo); e.g. H3BO3 (6.2 mg/L)
    title: rooting medium micronutrients
    examples:
    - value: H3BO3;6.2 mg/L
    from_schema: https://example.com/nmdc_submission_schema
    rank: 276
    keywords:
    - micronutrients
    string_serialization: '{text};{float} {unit}'
    slot_uri: MIXS:0000579
    alias: root_med_micronutr
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  root_med_ph:
    name: root_med_ph
    annotations:
      storage_units:
        tag: storage_units
        value: '[pH]'
    description: pH measurement of the culture rooting medium; e.g. 5.5
    title: rooting medium pH
    comments:
    - Value must be reported in [pH]. Provide the numeric portion only.
    examples:
    - value: '7.5'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 277
    keywords:
    - ph
    slot_uri: MIXS:0001062
    alias: root_med_ph
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: float
    minimum_value: 0
    maximum_value: 14
    unit:
      ucum_code: '[pH]'
  root_med_regl:
    name: root_med_regl
    annotations:
      Expected_value:
        tag: Expected_value
        value: regulator name;measurement value
      Preferred_unit:
        tag: Preferred_unit
        value: milligram per liter
    description: Growth regulators in the culture rooting medium such as cytokinins,
      auxins, gybberellins, abscisic acid; e.g. 0.5  mg/L NAA
    title: rooting medium regulators
    examples:
    - value: abscisic acid;0.75 milligram per liter
    from_schema: https://example.com/nmdc_submission_schema
    rank: 278
    string_serialization: '{text};{float} {unit}'
    slot_uri: MIXS:0000581
    alias: root_med_regl
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  root_med_solid:
    name: root_med_solid
    description: Specification of the solidifying agent in the culture rooting medium;
      e.g. agar
    title: rooting medium solidifier
    examples:
    - value: agar
    from_schema: https://example.com/nmdc_submission_schema
    rank: 279
    slot_uri: MIXS:0001063
    alias: root_med_solid
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  root_med_suppl:
    name: root_med_suppl
    annotations:
      Expected_value:
        tag: Expected_value
        value: supplement name;measurement value
      Preferred_unit:
        tag: Preferred_unit
        value: milligram per liter
    description: Organic supplements of the culture rooting medium, such as vitamins,
      amino acids, organic acids, antibiotics activated charcoal; e.g. nicotinic acid
      (0.5  mg/L)
    title: rooting medium organic supplements
    examples:
    - value: nicotinic acid;0.5 milligram per liter
    from_schema: https://example.com/nmdc_submission_schema
    rank: 280
    keywords:
    - organic
    string_serialization: '{text};{float} {unit}'
    slot_uri: MIXS:0000580
    alias: root_med_suppl
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
  salinity:
    name: salinity
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: practical salinity unit, percentage
      storage_units:
        tag: storage_units
        value: '%|mg/L'
    description: The total concentration of all dissolved salts in a liquid or solid
      sample. While salinity can be measured by a complete chemical analysis, this
      method is difficult and time consuming. More often, it is instead derived from
      the conductivity measurement. This is known as practical salinity. These derivations
      compare the specific conductance of the sample to a salinity standard such as
      seawater
    title: salinity
    comments:
    - 'Value must be reported in one of the following units: %, mg/L. Provide the
      numeric portion followed by a space and the unit.'
    examples:
    - value: 70 mg/L
    from_schema: https://example.com/nmdc_submission_schema
    rank: 54
    keywords:
    - salinity
    slot_uri: MIXS:0000183
    alias: salinity
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  salinity_meth:
    name: salinity_meth
    annotations:
      Expected_value:
        tag: Expected_value
        value: PMID,DOI or url
    description: Reference or method used in determining salinity.
    title: salinity method
    comments:
    - Legacy MIxS slot removed from GSC MIxS v6.2.2. Retained for submission-schema
      compatibility.
    from_schema: https://example.com/nmdc_submission_schema
    rank: 55
    slot_uri: MIXS:0000341
    alias: salinity_meth
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    slot_group: mixs_section
    range: string
    multivalued: false
    pattern: ^PMID:\d+|doi:10.\d{2,9}/.*|https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_\+.~#?&amp;\/=]*)$
    structured_pattern:
      syntax: ^{PMID}|{DOI}|{URL}$
      interpolated: true
  salt_regm:
    name: salt_regm
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: gram, microgram, mole per liter, gram per liter
    description: Information about treatment involving use of salts as supplement
      to liquid and soil growth media; should include the name of salt, amount administered,
      treatment regimen including how many times the treatment was repeated, how long
      each treatment lasted, and the start and end time of the entire treatment; can
      include multiple salt regimens
    title: salt regimen
    examples:
    - value: NaCl;5 gram per liter;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 281
    keywords:
    - regimen
    - salt
    slot_uri: MIXS:0000582
    alias: salt_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  samp_capt_status:
    name: samp_capt_status
    description: Reason for the sample
    title: sample capture status
    examples:
    - value: farm sample
    from_schema: https://example.com/nmdc_submission_schema
    rank: 282
    keywords:
    - sample
    - status
    slot_uri: MIXS:0000860
    alias: samp_capt_status
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: SampCaptStatusEnum
  samp_collec_device:
    name: samp_collec_device
    annotations:
      Expected_value:
        tag: Expected_value
        value: device name
    description: The device used to collect an environmental sample. This field accepts
      terms listed under environmental sampling device (http://purl.obolibrary.org/obo/ENVO).
      This field also accepts terms listed under specimen collection device (http://purl.obolibrary.org/obo/GENEPIO_0002094)
    title: sample collection device
    comments:
    - Report dimensions and details when applicable
    from_schema: https://example.com/nmdc_submission_schema
    structured_aliases:
      samp_collect_device:
        literal_form: samp_collect_device
        contexts:
        - https://github.com/GenomicsStandardsConsortium/mixs/releases/tag/v6.2.2
    rank: 14
    keywords:
    - device
    - sample
    string_serialization: '{termLabel} [{termID}]|{text}'
    slot_uri: MIXS:0000002
    alias: samp_collec_device
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
  samp_collec_method:
    name: samp_collec_method
    description: The method employed for collecting the sample
    title: sample collection method
    comments:
    - This can be a citation or description
    from_schema: https://example.com/nmdc_submission_schema
    structured_aliases:
      samp_collect_method:
        literal_form: samp_collect_method
        contexts:
        - https://github.com/GenomicsStandardsConsortium/mixs/releases/tag/v6.2.2
    rank: 15
    keywords:
    - method
    - sample
    slot_uri: MIXS:0001225
    alias: samp_collec_method
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
  samp_dis_stage:
    name: samp_dis_stage
    description: Stage of the disease at the time of sample collection, e.g. inoculation,
      penetration, infection, growth and reproduction, dissemination of pathogen
    title: sample disease stage
    examples:
    - value: infection
    from_schema: https://example.com/nmdc_submission_schema
    rank: 283
    keywords:
    - disease
    - sample
    slot_uri: MIXS:0000249
    alias: samp_dis_stage
    owner: PlantAssociatedInterface
    domain_of:
    - HostAssociatedInterface
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: SampDisStageEnum
  samp_mat_process:
    name: samp_mat_process
    description: A brief description of any processing applied to the sample during
      or after retrieving the sample from environment, or a link to the relevant protocol(s)
      performed
    title: sample material processing
    examples:
    - value: filtering of seawater
    - value: storing samples in ethanol
    from_schema: https://example.com/nmdc_submission_schema
    rank: 12
    keywords:
    - material
    - process
    - sample
    slot_uri: MIXS:0000016
    alias: samp_mat_process
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
  samp_size:
    name: samp_size
    annotations:
      storage_units:
        tag: storage_units
        value: L|g|mL|mg
    description: The total amount (volume or mass) of sample collected.
    title: amount or size of sample collected
    comments:
    - 'Value must be reported in one of the following units: L, g, mL, mg. Provide
      the numeric portion followed by a space and the unit.'
    - This refers to the TOTAL amount of sample collected from the experiment. NOT
      the amount sent to each institution or collected for a specific analysis.
    examples:
    - value: 5 g
    - value: 10 mL
    from_schema: https://example.com/nmdc_submission_schema
    rank: 18
    keywords:
    - sample
    - size
    slot_uri: MIXS:0000001
    alias: samp_size
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
  samp_store_dur:
    name: samp_store_dur
    description: Duration for which the sample was stored. Indicate the duration for
      which the sample was stored written in ISO 8601 format
    title: sample storage duration
    examples:
    - value: P1Y6M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 353
    keywords:
    - duration
    - period
    - sample
    - storage
    slot_uri: MIXS:0000116
    alias: samp_store_dur
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^P(?:(?:\d+D|\d+M(?:\d+D)?|\d+Y(?:\d+M(?:\d+D)?)?)(?:T(?:\d+H(?:\d+M(?:\d+S)?)?|\d+M(?:\d+S)?|\d+S))?|T(?:\d+H(?:\d+M(?:\d+S)?)?|\d+M(?:\d+S)?|\d+S)|\d+W)$
    structured_pattern:
      syntax: ^{duration}$
      interpolated: true
      partial_match: true
  samp_store_loc:
    name: samp_store_loc
    annotations:
      Expected_value:
        tag: Expected_value
        value: location name
    description: Location at which sample was stored, usually name of a specific freezer/room
    title: sample storage location
    examples:
    - value: Freezer no:5
    from_schema: https://example.com/nmdc_submission_schema
    rank: 41
    keywords:
    - location
    - sample
    - storage
    slot_uri: MIXS:0000755
    alias: samp_store_loc
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_core_section
    range: string
  samp_store_temp:
    name: samp_store_temp
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: degree Celsius
      storage_units:
        tag: storage_units
        value: Cel
    description: Temperature at which the sample was stored (degrees are assumed)
    title: sample storage temperature
    comments:
    - Value must be reported in degrees Celsius. Provide the numeric portion only.
    examples:
    - value: '-80'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 7
    keywords:
    - sample
    - storage
    - temperature
    slot_uri: MIXS:0000110
    alias: samp_store_temp
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: float
    required: true
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
    unit:
      descriptive_name: degrees Celsius
      ucum_code: Cel
  sample_link:
    name: sample_link
    description: A unique identifier to assign parent-child, subsample, or sibling
      samples. This is relevant when a sample or other material was used to generate
      the new sample.
    title: sample linkage
    comments:
    - 'This field allows multiple entries separated by ; (Examples: Soil collected
      from the field will link with the soil used in an incubation. The soil a plant
      was grown in links to the plant sample. An original culture sample was transferred
      to a new vial and generated a new sample)'
    examples:
    - value: igsn:DSJ0284
    from_schema: https://example.com/nmdc_submission_schema
    rank: 5
    string_serialization: '{text}:{text}'
    alias: sample_link
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    - SampIdNewTermsMixin
    slot_group: sample_id_section
    range: string
    recommended: true
    multivalued: false
    pattern: '[^\:\n\r]+\:[^\:\n\r]+'
  season_environment:
    name: season_environment
    description: Treatment involving an exposure to a particular season (e.g. Winter,
      summer, rabi, rainy etc.), treatment regimen including how many times the treatment
      was repeated, how long each treatment lasted, and the start and end time of
      the entire treatment
    title: seasonal environment
    examples:
    - value: rainy;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 284
    keywords:
    - environment
    - season
    slot_uri: MIXS:0001068
    alias: season_environment
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  size_frac:
    name: size_frac
    annotations:
      Expected_value:
        tag: Expected_value
        value: filter size value range
    description: Filtering pore size used in sample preparation
    title: size fraction selected
    examples:
    - value: 0-0.22 micrometer
    from_schema: https://example.com/nmdc_submission_schema
    rank: 285
    keywords:
    - fraction
    - size
    string_serialization: '{float}-{float} {unit}'
    slot_uri: MIXS:0000017
    alias: size_frac
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_core_section
    range: string
  specific_ecosystem:
    name: specific_ecosystem
    description: Specific ecosystems represent specific features of the environment
      like aphotic zone in an ocean or gastric mucosa within a host digestive system.
      Specific ecosystem is in position 5/5 in a GOLD path.
    comments:
    - Specific ecosystems help to define samples based on very specific characteristics
      of an environment under the five-level classification system.
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - https://gold.jgi.doe.gov/help
    rank: 13
    is_a: gold_path_field
    alias: specific_ecosystem
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: sample_id_section
    range: SpecificEcosystemEnum
    recommended: true
  standing_water_regm:
    name: standing_water_regm
    description: Treatment involving an exposure to standing water during a plant's
      life span, types can be flood water or standing water, treatment regimen including
      how many times the treatment was repeated, how long each treatment lasted, and
      the start and end time of the entire treatment; can include multiple regimens
    title: standing water regimen
    examples:
    - value: standing water;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 286
    keywords:
    - regimen
    - water
    slot_uri: MIXS:0001069
    alias: standing_water_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  start_date_inc:
    name: start_date_inc
    description: Date the incubation was started. Only relevant for incubation samples.
    title: incubation start date
    notes:
    - MIxS collection_date accepts (truncated) ISO8601. DH taking arbitrary precision
      date only
    comments:
    - Date should be formatted as YYYY(-MM(-DD)). Ie, 2021-04-15, 2021-04 and 2021
      are all acceptable.
    examples:
    - value: '2021-04-15'
    - value: 2021-04
    - value: '2021'
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - MIXS:0000011
    rank: 4
    string_serialization: '{date, arbitrary precision}'
    alias: start_date_inc
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    - SoilMixsInspiredMixin
    slot_group: mixs_inspired_section
    range: string
    recommended: true
    pattern: ^[12]\d{3}(?:(?:-(?:0[1-9]|1[0-2]))(?:-(?:0[1-9]|[12]\d|3[01]))?)?$
  temp:
    name: temp
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: degree Celsius
      storage_units:
        tag: storage_units
        value: Cel
    description: Temperature of the sample at the time of sampling
    title: temperature
    comments:
    - Value must be reported in degrees Celsius. Provide the numeric portion only.
    examples:
    - value: '25'
    from_schema: https://example.com/nmdc_submission_schema
    rank: 37
    keywords:
    - temperature
    slot_uri: MIXS:0000113
    alias: temp
    owner: PlantAssociatedInterface
    domain_of:
    - AirInterface
    - BiofilmInterface
    - BuiltEnvInterface
    - HcrCoresInterface
    - HcrFluidsSwabsInterface
    - HostAssociatedInterface
    - MiscEnvsInterface
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WastewaterSludgeInterface
    - WaterInterface
    slot_group: mixs_section
    range: float
    pattern: ^[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?( *- *[-+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)?
      *.*$
    structured_pattern:
      syntax: ^{scientific_float}( *- *{scientific_float})? *{text}$
      interpolated: true
      partial_match: true
    unit:
      descriptive_name: degrees Celsius
      ucum_code: Cel
  tiss_cult_growth_med:
    name: tiss_cult_growth_med
    description: Description of plant tissue culture growth media used
    title: tissue culture growth media
    examples:
    - value: https://link.springer.com/content/pdf/10.1007/BF02796489.pdf
    from_schema: https://example.com/nmdc_submission_schema
    rank: 287
    keywords:
    - culture
    - growth
    slot_uri: MIXS:0001070
    alias: tiss_cult_growth_med
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    pattern: ^(PMID:\d+|doi:10.\d{2,9}/.*|https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_\+.~#?&amp;\/=]*)|.*)$
    structured_pattern:
      syntax: ^({PMID}|{DOI}|{URL}|{text})$
      interpolated: true
      partial_match: true
  water_temp_regm:
    name: water_temp_regm
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: degree Celsius
    description: Information about treatment involving an exposure to water with varying
      degree of temperature, treatment regimen including how many times the treatment
      was repeated, how long each treatment lasted, and the start and end time of
      the entire treatment; can include multiple regimens
    title: water temperature regimen
    examples:
    - value: 15 degree Celsius;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    from_schema: https://example.com/nmdc_submission_schema
    rank: 288
    keywords:
    - regimen
    - temperature
    - water
    slot_uri: MIXS:0000590
    alias: water_temp_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    slot_group: mixs_core_section
    range: string
    multivalued: false
  watering_regm:
    name: watering_regm
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: milliliter, liter
    description: Information about treatment involving an exposure to watering frequencies,
      treatment regimen including how many times the treatment was repeated, how long
      each treatment lasted, and the start and end time of the entire treatment; can
      include multiple regimens
    title: watering regimen
    examples:
    - value: 1 liter;R2/2018-05-11T14:30/2018-05-11T19:30/P1H30M
    - value: 75% water holding capacity; constant
    from_schema: https://example.com/nmdc_submission_schema
    rank: 25
    keywords:
    - regimen
    - water
    slot_uri: MIXS:0000591
    alias: watering_regm
    owner: PlantAssociatedInterface
    domain_of:
    - PlantAssociatedInterface
    - SedimentInterface
    - SoilInterface
    - WaterInterface
    slot_group: mixs_modified_section
    range: string
    recommended: true
    multivalued: false
  analysis_type:
    name: analysis_type
    description: Select all the data types associated or available for this biosample
    title: analysis/data type
    comments:
    - MIxS:investigation_type was included as a `see_also` but that term doesn't resolve
      any more
    examples:
    - value: metagenomics; metabolomics; metaproteomics
    from_schema: https://example.com/nmdc_submission_schema
    rank: 3
    alias: analysis_type
    owner: PlantAssociatedInterface
    domain_of:
    - DhMultiviewCommonColumnsMixin
    slot_group: sample_id_section
    range: AnalysisTypeEnum
    required: true
    multivalued: true
  samp_name:
    name: samp_name
    annotations:
      Preferred_unit:
        tag: Preferred_unit
        value: ''
    description: A local identifier or name that for the material sample collected.
      Refers to the original material collected or to any derived sub-samples.
    title: sample name
    comments:
    - It can have any format, but we suggest that you make it concise, unique and
      consistent within your lab, and as informative as possible.
    examples:
    - value: Rock core CB1178(5-6) from NSW
    from_schema: https://example.com/nmdc_submission_schema
    rank: 1
    keywords:
    - sample
    slot_uri: MIXS:0001107
    identifier: true
    alias: samp_name
    owner: PlantAssociatedInterface
    domain_of:
    - DhMultiviewCommonColumnsMixin
    slot_group: sample_id_section
    range: string
    required: true
  source_mat_id:
    name: source_mat_id
    annotations:
      Expected_value:
        tag: Expected_value
        value: 'for cultures of microorganisms: identifiers for two culture collections;
          for other material a unique arbitrary identifer'
    description: A globally unique identifier assigned to the parent sample or sample
      that is the source  of this sample.
    title: source material identifier
    todos:
    - Currently, the comments say to use UUIDs. However, if we implement assigning
      NMDC identifiers with the minter we dont need to require a GUID. It can be an
      optional field to fill out only if they already have a resolvable ID.
    comments:
    - Identifiers must be prefixed. Possible FAIR prefixes are IGSNs (http://www.geosamples.org/getigsn),
      NCBI biosample accession numbers, ARK identifiers (https://arks.org/). These
      IDs enable linking to derived analytes and subsamples. If you have not assigned
      FAIR identifiers to your samples, you can generate UUIDs (https://www.uuidgenerator.net/).
    - 'Identifiers must be prefixed. Possible FAIR prefixes are: `igsn` for International  Generic
      Sample Numbers (http://www.geosamples.org/getigsn), `biosample` for NCBI  biosample
      accession IDs, `gold` for GOLD identifiers.'
    examples:
    - value: igsn:AU1243
    - value: biosample:SAMEA2397676
    from_schema: https://example.com/nmdc_submission_schema
    rank: 2
    keywords:
    - identifier
    - material
    - source
    slot_uri: MIXS:0000026
    alias: source_mat_id
    owner: PlantAssociatedInterface
    domain_of:
    - DhMultiviewCommonColumnsMixin
    slot_group: sample_id_section
    range: string
    multivalued: false
    pattern: ^igsn:[a-zA-Z0-9]+|biosample:SAMN[a-zA-Z0-9]+|biosample:SAME[a-zA-Z0-9]+|biosample:SAMJ[a-zA-Z0-9]+|gold:Gb[0-9]+$

</code></pre>
</details>
</article>
</div>
<script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "..", "features": ["content.tabs.link", "navigation.sections"], "search": "../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../assets/javascripts/bundle.83f73b43.min.js"></script>
<script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
<script src="../js/tablesort.js"></script>
<script type="module">import mermaid from "https://unpkg.com/mermaid@10.9.0/dist/mermaid.esm.min.mjs";
mermaid.initialize({});</script></body>
</html>