
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<link href="https://microbiomedata.github.io/submission-schema/MetatranscriptomeSequencingNonInterleavedDataInterface/" rel="canonical"/>
<link href="../assets/images/favicon.png" rel="icon"/>
<meta content="mkdocs-1.6.1, mkdocs-material-9.5.48" name="generator"/>
<title>Class: Metatranscriptome Sequence Data (Non-Interleaved) (MetatranscriptomeSequencingNonInterleavedDataInterface) - submission-schema</title>
<link href="../assets/stylesheets/main.6f8fc17f.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
</head>
<body dir="ltr">
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#class-metatranscriptome-sequence-data-non-interleaved-metatranscriptomesequencingnoninterleaveddatainterface">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="submission-schema" class="md-header__button md-logo" data-md-component="logo" href=".." title="submission-schema">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"></path></svg>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            submission-schema
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Class: Metatranscriptome Sequence Data (Non-Interleaved) (MetatranscriptomeSequencingNonInterleavedDataInterface)
            
          </span>
</div>
</div>
</div>
<script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" title="Clear" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="" tabindex="0">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list" role="presentation"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/submission-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="submission-schema" class="md-nav__button md-logo" data-md-component="logo" href=".." title="submission-schema">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"></path></svg>
</a>
    submission-schema
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/microbiomedata/submission-schema" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="..">
<span class="md-ellipsis">
    Index
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../about/">
<span class="md-ellipsis">
    About
  </span>
</a>
</li>
<li class="md-nav__item md-nav__item--section md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
<span class="md-ellipsis">
    DataHarmonizer Playground
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
            DataHarmonizer Playground
          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="https://microbiomedata.github.io/submission-schema/_playground/?template=nmdc_submission_schema/SoilInterface">
<span class="md-ellipsis">
    SoilInterface
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="https://microbiomedata.github.io/submission-schema/_playground/?template=nmdc_submission_schema/WaterInterface">
<span class="md-ellipsis">
    WaterInterface
  </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#inheritance">
<span class="md-ellipsis">
      Inheritance
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#slots">
<span class="md-ellipsis">
      Slots
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#usages">
<span class="md-ellipsis">
      Usages
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#identifier-and-mapping-information">
<span class="md-ellipsis">
      Identifier and Mapping Information
    </span>
</a>
<nav aria-label="Identifier and Mapping Information" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#schema-source">
<span class="md-ellipsis">
      Schema Source
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#mappings">
<span class="md-ellipsis">
      Mappings
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#linkml-source">
<span class="md-ellipsis">
      LinkML Source
    </span>
</a>
<nav aria-label="LinkML Source" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#direct">
<span class="md-ellipsis">
      Direct
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#induced">
<span class="md-ellipsis">
      Induced
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1 id="class-metatranscriptome-sequence-data-non-interleaved-metatranscriptomesequencingnoninterleaveddatainterface">Class: Metatranscriptome Sequence Data (Non-Interleaved) (MetatranscriptomeSequencingNonInterleavedDataInterface)</h1>
<p><em>Interface for non-interleaved metatranscriptome sequencing data</em></p>
<p>URI: <a href="https://example.com/nmdc_sub_schema/MetatranscriptomeSequencingNonInterleavedDataInterface">nmdc_sub_schema:MetatranscriptomeSequencingNonInterleavedDataInterface</a></p>
<div class="mermaid"> classDiagram
    class MetatranscriptomeSequencingNonInterleavedDataInterface
    click MetatranscriptomeSequencingNonInterleavedDataInterface href "../MetatranscriptomeSequencingNonInterleavedDataInterface/"
      DhMultiviewCommonColumnsMixin &lt;|-- MetatranscriptomeSequencingNonInterleavedDataInterface
        click DhMultiviewCommonColumnsMixin href "../DhMultiviewCommonColumnsMixin/"

      MetatranscriptomeSequencingNonInterleavedDataInterface : analysis_type





        MetatranscriptomeSequencingNonInterleavedDataInterface --&gt; "1..*" AnalysisTypeEnum : analysis_type
        click AnalysisTypeEnum href "../AnalysisTypeEnum/"



      MetatranscriptomeSequencingNonInterleavedDataInterface : insdc_bioproject_identifiers

      MetatranscriptomeSequencingNonInterleavedDataInterface : insdc_experiment_identifiers

      MetatranscriptomeSequencingNonInterleavedDataInterface : model





        MetatranscriptomeSequencingNonInterleavedDataInterface --&gt; "1" IlluminaInstrumentModelEnum : model
        click IlluminaInstrumentModelEnum href "../IlluminaInstrumentModelEnum/"



      MetatranscriptomeSequencingNonInterleavedDataInterface : processing_institution





        MetatranscriptomeSequencingNonInterleavedDataInterface --&gt; "0..1" ProcessingInstitutionEnum : processing_institution
        click ProcessingInstitutionEnum href "../ProcessingInstitutionEnum/"



      MetatranscriptomeSequencingNonInterleavedDataInterface : protocol_link

      MetatranscriptomeSequencingNonInterleavedDataInterface : read_1_md5_checksum

      MetatranscriptomeSequencingNonInterleavedDataInterface : read_1_url

      MetatranscriptomeSequencingNonInterleavedDataInterface : read_2_md5_checksum

      MetatranscriptomeSequencingNonInterleavedDataInterface : read_2_url

      MetatranscriptomeSequencingNonInterleavedDataInterface : samp_name

      MetatranscriptomeSequencingNonInterleavedDataInterface : source_mat_id


</div>
<h2 id="inheritance">Inheritance</h2>
<ul>
<li><strong>MetatranscriptomeSequencingNonInterleavedDataInterface</strong> [ <a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a>]</li>
</ul>
<h2 id="slots">Slots</h2>
<table>
<thead>
<tr>
<th>Name</th>
<th>Cardinality and Range</th>
<th>Description</th>
<th>Inheritance</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../read_1_url/">read_1_url</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>URL for FASTQ file of read 1 of a pair of reads</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../read_1_md5_checksum/">read_1_md5_checksum</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>MD5 checksum of file in "read 1 FASTQ"</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../read_2_url/">read_2_url</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>URL for FASTQ file of read 2 of a pair of reads</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../read_2_md5_checksum/">read_2_md5_checksum</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>MD5 checksum of file in "read 2 FASTQ"</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../model/">model</a></td>
<td>1 <br/> <a href="../IlluminaInstrumentModelEnum/">IlluminaInstrumentModelEnum</a></td>
<td>The model of the Illumina sequencing instrument used to generate the data</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../insdc_bioproject_identifiers/">insdc_bioproject_identifiers</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>identifiers for corresponding project in INSDC Bioproject</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../insdc_experiment_identifiers/">insdc_experiment_identifiers</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>If multiple identifiers are provided, separate them with a semicolon</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../processing_institution/">processing_institution</a></td>
<td>0..1 <br/> <a href="../ProcessingInstitutionEnum/">ProcessingInstitutionEnum</a></td>
<td>The organization that processed the sample</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../protocol_link/">protocol_link</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>A URL to a description of the sequencing protocol used to generate the data</td>
<td>direct</td>
</tr>
<tr>
<td><a href="../analysis_type/">analysis_type</a></td>
<td>1..* <br/> <a href="../AnalysisTypeEnum/">AnalysisTypeEnum</a></td>
<td>Select all the data types associated or available for this biosample</td>
<td><a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a></td>
</tr>
<tr>
<td><a href="../samp_name/">samp_name</a></td>
<td>1 <br/> <a href="../String/">String</a></td>
<td>A local identifier or name that for the material sample collected</td>
<td><a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a></td>
</tr>
<tr>
<td><a href="../source_mat_id/">source_mat_id</a></td>
<td>0..1 <br/> <a href="../String/">String</a></td>
<td>A globally unique identifier assigned to the parent sample or sample that is ...</td>
<td><a href="../DhMultiviewCommonColumnsMixin/">DhMultiviewCommonColumnsMixin</a></td>
</tr>
</tbody>
</table>
<h2 id="usages">Usages</h2>
<table>
<thead>
<tr>
<th>used by</th>
<th>used in</th>
<th>type</th>
<th>used</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../SampleData/">SampleData</a></td>
<td><a href="../metatranscriptome_sequencing_non_interleaved_data/">metatranscriptome_sequencing_non_interleaved_data</a></td>
<td>range</td>
<td><a href="./">MetatranscriptomeSequencingNonInterleavedDataInterface</a></td>
</tr>
</tbody>
</table>
<h2 id="identifier-and-mapping-information">Identifier and Mapping Information</h2>
<h3 id="schema-source">Schema Source</h3>
<ul>
<li>from schema: <a href="https://example.com/nmdc_submission_schema">https://example.com/nmdc_submission_schema</a></li>
</ul>
<h2 id="mappings">Mappings</h2>
<table>
<thead>
<tr>
<th>Mapping Type</th>
<th>Mapped Value</th>
</tr>
</thead>
<tbody>
<tr>
<td>self</td>
<td>nmdc_sub_schema:MetatranscriptomeSequencingNonInterleavedDataInterface</td>
</tr>
<tr>
<td>native</td>
<td>nmdc_sub_schema:MetatranscriptomeSequencingNonInterleavedDataInterface</td>
</tr>
</tbody>
</table>
<h2 id="linkml-source">LinkML Source</h2>
<!-- TODO: investigate https://stackoverflow.com/questions/37606292/how-to-create-tabbed-code-blocks-in-mkdocs-or-sphinx -->
<h3 id="direct">Direct</h3>
<details>
<pre><code class="language-yaml">name: MetatranscriptomeSequencingNonInterleavedDataInterface
description: Interface for non-interleaved metatranscriptome sequencing data
title: Metatranscriptome Sequence Data (Non-Interleaved)
from_schema: https://example.com/nmdc_submission_schema
mixins:
- DhMultiviewCommonColumnsMixin
slots:
- read_1_url
- read_1_md5_checksum
- read_2_url
- read_2_md5_checksum
- model
- insdc_bioproject_identifiers
- insdc_experiment_identifiers
- processing_institution
- protocol_link
slot_usage:
  insdc_bioproject_identifiers:
    name: insdc_bioproject_identifiers
    multivalued: false
  insdc_experiment_identifiers:
    name: insdc_experiment_identifiers
    multivalued: false

</code></pre>
</details>
<h3 id="induced">Induced</h3>
<details>
<pre><code class="language-yaml">name: MetatranscriptomeSequencingNonInterleavedDataInterface
description: Interface for non-interleaved metatranscriptome sequencing data
title: Metatranscriptome Sequence Data (Non-Interleaved)
from_schema: https://example.com/nmdc_submission_schema
mixins:
- DhMultiviewCommonColumnsMixin
slot_usage:
  insdc_bioproject_identifiers:
    name: insdc_bioproject_identifiers
    multivalued: false
  insdc_experiment_identifiers:
    name: insdc_experiment_identifiers
    multivalued: false
attributes:
  read_1_url:
    name: read_1_url
    description: URL for FASTQ file of read 1 of a pair of reads.
    title: read 1 FASTQ
    comments:
    - If multiple runs were performed, separate each URL with a semi-colon.
    - External data urls should be available for at least a year. If you would like
      NMDC to submit your data to an appropriate raw data repository on your behalf
      please contact us at microbiomedata.science@gmail.com.
    from_schema: https://example.com/nmdc_submission_schema
    rank: 10
    alias: read_1_url
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - MetagenomeSequencingNonInterleavedDataInterface
    - MetatranscriptomeSequencingNonInterleavedDataInterface
    slot_group: data_files_section
    range: string
    required: true
    pattern: ^https://[^\s;]+(?:\s*;\s*https://[^\s;]+)*$
  read_1_md5_checksum:
    name: read_1_md5_checksum
    description: MD5 checksum of file in "read 1 FASTQ".
    title: read 1 FASTQ MD5
    comments:
    - If multiple runs were performed, separate each checksum with a semi-colon. The
      number of checksums should match the number of URLs in "read 1 FASTQ".
    from_schema: https://example.com/nmdc_submission_schema
    rank: 11
    alias: read_1_md5_checksum
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - MetagenomeSequencingNonInterleavedDataInterface
    - MetatranscriptomeSequencingNonInterleavedDataInterface
    slot_group: data_files_section
    range: string
    pattern: ^[a-fA-F0-9]{32}(?:\s*;\s*[a-fA-F0-9]{32})*$
  read_2_url:
    name: read_2_url
    description: URL for FASTQ file of read 2 of a pair of reads.
    title: read 2 FASTQ
    comments:
    - If multiple runs were performed, separate each URL with a semi-colon.
    - External data urls should be available for at least a year. If you would like
      NMDC to submit your data to an appropriate raw data repository on your behalf
      please contact us at microbiomedata.science@gmail.com.
    from_schema: https://example.com/nmdc_submission_schema
    rank: 12
    alias: read_2_url
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - MetagenomeSequencingNonInterleavedDataInterface
    - MetatranscriptomeSequencingNonInterleavedDataInterface
    slot_group: data_files_section
    range: string
    required: true
    pattern: ^https://[^\s;]+(?:\s*;\s*https://[^\s;]+)*$
  read_2_md5_checksum:
    name: read_2_md5_checksum
    description: MD5 checksum of file in "read 2 FASTQ".
    title: read 2 FASTQ MD5
    comments:
    - If multiple runs were performed, separate each checksum with a semi-colon. The
      number of checksums should match the number of URLs in "read 2 FASTQ".
    from_schema: https://example.com/nmdc_submission_schema
    rank: 13
    alias: read_2_md5_checksum
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - MetagenomeSequencingNonInterleavedDataInterface
    - MetatranscriptomeSequencingNonInterleavedDataInterface
    slot_group: data_files_section
    range: string
    pattern: ^[a-fA-F0-9]{32}(?:\s*;\s*[a-fA-F0-9]{32})*$
  model:
    name: model
    description: The model of the Illumina sequencing instrument used to generate
      the data.
    title: instrument model
    from_schema: https://example.com/nmdc_submission_schema
    rank: 2
    alias: model
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - MetagenomeSequencingNonInterleavedDataInterface
    - MetagenomeSequencingInterleavedDataInterface
    - MetatranscriptomeSequencingNonInterleavedDataInterface
    - MetatranscriptomeSequencingInterleavedDataInterface
    slot_group: sequencing_section
    range: IlluminaInstrumentModelEnum
    required: true
  insdc_bioproject_identifiers:
    name: insdc_bioproject_identifiers
    description: identifiers for corresponding project in INSDC Bioproject
    title: INSDC bioproject identifier
    comments:
    - these are distinct IDs from INSDC SRA/ENA project identifiers, but are usually(?)
      one to one
    examples:
    - value: bioproject:PRJNA366857
    from_schema: https://example.com/nmdc_submission_schema
    see_also:
    - https://www.ncbi.nlm.nih.gov/bioproject/
    - https://www.ddbj.nig.ac.jp/bioproject/index-e.html
    aliases:
    - NCBI bioproject identifiers
    - DDBJ bioproject identifiers
    rank: 5
    is_a: study_identifiers
    mixins:
    - insdc_identifiers
    alias: insdc_bioproject_identifiers
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - MetagenomeSequencingNonInterleavedDataInterface
    - MetagenomeSequencingInterleavedDataInterface
    - MetatranscriptomeSequencingNonInterleavedDataInterface
    - MetatranscriptomeSequencingInterleavedDataInterface
    slot_group: sequencing_section
    range: string
    multivalued: false
    pattern: ^bioproject:PRJ[DEN][A-Z][0-9]+$
  insdc_experiment_identifiers:
    name: insdc_experiment_identifiers
    description: If multiple identifiers are provided, separate them with a semicolon.
      The number of identifiers must match the number of sequencing files.
    title: INSDC experiment identifiers
    from_schema: https://example.com/nmdc_submission_schema
    rank: 6
    is_a: external_database_identifiers
    mixins:
    - insdc_identifiers
    alias: insdc_experiment_identifiers
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - MetagenomeSequencingNonInterleavedDataInterface
    - MetagenomeSequencingInterleavedDataInterface
    - MetatranscriptomeSequencingNonInterleavedDataInterface
    - MetatranscriptomeSequencingInterleavedDataInterface
    slot_group: sequencing_section
    range: string
    multivalued: false
    pattern: ^insdc.sra:(E|D|S)RX[0-9]{6,}$
  processing_institution:
    name: processing_institution
    description: The organization that processed the sample.
    title: processing institution
    from_schema: https://example.com/nmdc_submission_schema
    rank: 3
    alias: processing_institution
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - MetagenomeSequencingNonInterleavedDataInterface
    - MetagenomeSequencingInterleavedDataInterface
    - MetatranscriptomeSequencingNonInterleavedDataInterface
    - MetatranscriptomeSequencingInterleavedDataInterface
    slot_group: sequencing_section
    range: ProcessingInstitutionEnum
  protocol_link:
    name: protocol_link
    description: A URL to a description of the sequencing protocol used to generate
      the data.
    title: protocol
    from_schema: https://example.com/nmdc_submission_schema
    rank: 4
    alias: protocol_link
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - MetagenomeSequencingNonInterleavedDataInterface
    - MetagenomeSequencingInterleavedDataInterface
    - MetatranscriptomeSequencingNonInterleavedDataInterface
    - MetatranscriptomeSequencingInterleavedDataInterface
    slot_group: sequencing_section
    range: string
  analysis_type:
    name: analysis_type
    description: Select all the data types associated or available for this biosample
    title: analysis/data type
    comments:
    - MIxS:investigation_type was included as a `see_also` but that term doesn't resolve
      any more
    examples:
    - value: metagenomics; metabolomics; metaproteomics
    from_schema: https://example.com/nmdc_submission_schema
    rank: 3
    alias: analysis_type
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - DhMultiviewCommonColumnsMixin
    slot_group: sample_id_section
    range: AnalysisTypeEnum
    required: true
    multivalued: true
  samp_name:
    name: samp_name
    annotations:
      expected_value:
        tag: expected_value
        value: text
    description: A local identifier or name that for the material sample collected.
      Refers to the original material collected or to any derived sub-samples.
    title: sample name
    comments:
    - It can have any format, but we suggest that you make it concise, unique and
      consistent within your lab, and as informative as possible.
    examples:
    - value: Rock core CB1178(5-6) from NSW
    from_schema: https://example.com/nmdc_submission_schema
    aliases:
    - sample name
    rank: 1
    is_a: investigation field
    string_serialization: '{text}'
    slot_uri: MIXS:0001107
    identifier: true
    alias: samp_name
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - DhMultiviewCommonColumnsMixin
    slot_group: sample_id_section
    range: string
    required: true
    multivalued: false
  source_mat_id:
    name: source_mat_id
    annotations:
      expected_value:
        tag: expected_value
        value: 'for cultures of microorganisms: identifiers for two culture collections;
          for other material a unique arbitrary identifer'
    description: A globally unique identifier assigned to the parent sample or sample
      that is the source  of this sample.
    title: source material identifier
    todos:
    - Currently, the comments say to use UUIDs. However, if we implement assigning
      NMDC identifiers with the minter we dont need to require a GUID. It can be an
      optional field to fill out only if they already have a resolvable ID.
    comments:
    - Identifiers must be prefixed. Possible FAIR prefixes are IGSNs (http://www.geosamples.org/getigsn),
      NCBI biosample accession numbers, ARK identifiers (https://arks.org/). These
      IDs enable linking to derived analytes and subsamples. If you have not assigned
      FAIR identifiers to your samples, you can generate UUIDs (https://www.uuidgenerator.net/).
    - 'Identifiers must be prefixed. Possible FAIR prefixes are: `igsn` for International  Generic
      Sample Numbers (http://www.geosamples.org/getigsn), `biosample` for NCBI  biosample
      accession IDs, `gold` for GOLD identifiers.'
    examples:
    - value: igsn:AU1243
    - value: biosample:SAMEA2397676
    from_schema: https://example.com/nmdc_submission_schema
    aliases:
    - source material identifiers
    rank: 2
    is_a: nucleic acid sequence source field
    string_serialization: '{text}'
    slot_uri: MIXS:0000026
    alias: source_mat_id
    owner: MetatranscriptomeSequencingNonInterleavedDataInterface
    domain_of:
    - DhMultiviewCommonColumnsMixin
    slot_group: sample_id_section
    range: string
    multivalued: false
    pattern: ^igsn:[a-zA-Z0-9]+|biosample:SAMN[a-zA-Z0-9]+|biosample:SAME[a-zA-Z0-9]+|biosample:SAMJ[a-zA-Z0-9]+|gold:Gb[0-9]+$

</code></pre>
</details>
</article>
</div>
<script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "..", "features": ["content.tabs.link", "navigation.sections"], "search": "../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../assets/javascripts/bundle.83f73b43.min.js"></script>
<script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
<script src="../javascripts/tablesort.js"></script>
<script type="module">import mermaid from "https://unpkg.com/mermaid@10.9.0/dist/mermaid.esm.min.mjs";
mermaid.initialize({});</script></body>
</html>