{
 "cells": [
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.144841Z",
     "start_time": "2024-11-15T19:43:34.645078Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import csv\n",
    "import gzip\n",
    "import os\n",
    "import re\n",
    "import shutil\n",
    "import sqlite3\n",
    "from typing import Dict, Any\n",
    "from urllib.parse import urlparse\n",
    "\n",
    "import duckdb\n",
    "import pandas as pd\n",
    "import requests\n",
    "from linkml_runtime import SchemaView\n",
    "from oaklib import get_adapter\n",
    "from oaklib.datamodels.vocabulary import IS_A\n",
    "from scipy.sparse import hstack\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.feature_extraction.text import CountVectorizer  # from scikit-learn\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "import pprint"
   ],
   "id": "630126a7790cc3a7",
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.152610Z",
     "start_time": "2024-11-15T19:43:39.149551Z"
    }
   },
   "cell_type": "code",
   "source": "# todo deal with circularity in env package prediction -> env triad reporting",
   "id": "5001cbd978bbdd28",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.232897Z",
     "start_time": "2024-11-15T19:43:39.229882Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Approved prefixes (case-insensitive)\n",
    "approved_prefixes = ['ENVO']"
   ],
   "id": "914fdae0d02dad58",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.275638Z",
     "start_time": "2024-11-15T19:43:39.272564Z"
    }
   },
   "cell_type": "code",
   "source": "MIN_ANNOTATION_LEN = 3",
   "id": "e673d16992993c5a",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.323346Z",
     "start_time": "2024-11-15T19:43:39.318994Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# make a biomes curie -> label dict?\n",
    "BIOME_CURIE = 'ENVO:00000428'"
   ],
   "id": "95f7efc06f20987a",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.378573Z",
     "start_time": "2024-11-15T19:43:39.370359Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# ENV_LC = 'soil'\n",
    "# ENV_LC = 'water'\n",
    "ENV_LC = 'sediment'"
   ],
   "id": "5e358e42ad885188",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.427147Z",
     "start_time": "2024-11-15T19:43:39.423699Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# ENV_BROAD_SCALE_ENUM = \"EnvBroadScaleSoilEnum\"\n",
    "# ENV_BROAD_SCALE_ENUM = \"EnvBroadScaleWaterEnum\"\n",
    "ENV_BROAD_SCALE_ENUM = \"EnvBroadScaleSedimentEnum\""
   ],
   "id": "20750df32ebb2033",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.473445Z",
     "start_time": "2024-11-15T19:43:39.470377Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo new since soil: why are we only considering MIMS.me for discovering appropriate env triad values?\n",
    "\n",
    "# NCBI_PACKAGE = 'MIMS.me.soil.6.0'\n",
    "# NCBI_PACKAGE = 'MIMS.me.water.6.0'\n",
    "NCBI_PACKAGE = 'MIMS.me.sediment.6.0'"
   ],
   "id": "f5654d6f9319576b",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.522442Z",
     "start_time": "2024-11-15T19:43:39.518048Z"
    }
   },
   "cell_type": "code",
   "source": [
    "ncbi_query = f\"\"\"\n",
    "SELECT content, COUNT(1) AS sample_count \n",
    "FROM attributes \n",
    "WHERE harmonized_name = 'env_broad_scale' AND package_content = '{NCBI_PACKAGE}'\n",
    "GROUP BY content\n",
    "ORDER BY COUNT(1) DESC\n",
    "\"\"\""
   ],
   "id": "d33879ec4c6ca1df",
   "outputs": [],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.569528Z",
     "start_time": "2024-11-15T19:43:39.566438Z"
    }
   },
   "cell_type": "code",
   "source": "envo_adapter_string = \"sqlite:obo:envo\"",
   "id": "8240c0de4cbcec24",
   "outputs": [],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.617376Z",
     "start_time": "2024-11-15T19:43:39.614288Z"
    }
   },
   "cell_type": "code",
   "source": "# goldterms_adapter_string = \"sqlite:obo:envo\"",
   "id": "eb27b996bb24645a",
   "outputs": [],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.665184Z",
     "start_time": "2024-11-15T19:43:39.661020Z"
    }
   },
   "cell_type": "code",
   "source": [
    "goldterms_semsql_url = \"https://s3.amazonaws.com/bbop-sqlite/goldterms.db.gz\"\n",
    "\n",
    "# https://s3.amazonaws.com/bbop-sqlite/\n",
    "# <Contents>\n",
    "# <Key>goldterms.db.gz</Key>\n",
    "# <LastModified>2024-11-03T17:24:56.000Z</LastModified>\n",
    "# <ETag>\"fe8e35b215786cb9fc347b7fadbe055f\"</ETag>\n",
    "# <Size>2935781</Size>\n",
    "# <StorageClass>STANDARD</StorageClass>\n",
    "# </Contents>\n"
   ],
   "id": "2ae64976b48e359",
   "outputs": [],
   "execution_count": 12
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.712652Z",
     "start_time": "2024-11-15T19:43:39.709477Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo could this have been done with a OAK query, eliminating the need to explicitly download the file?\n",
    "\n",
    "goldterms_envo_query = \"\"\"\n",
    "SELECT\n",
    "\t*\n",
    "FROM\n",
    "\tstatements s\n",
    "WHERE\n",
    "\tpredicate = 'mixs:env_broad'\"\"\""
   ],
   "id": "ae751e002f65be71",
   "outputs": [],
   "execution_count": 13
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.760528Z",
     "start_time": "2024-11-15T19:43:39.756128Z"
    }
   },
   "cell_type": "code",
   "source": [
    "GOLDTERMS_SOIL = 'GOLDTERMS:4212'\n",
    "GOLDTERMS_WATER = 'GOLDTERMS:3984'\n",
    "GOLDTERMS_SEDIMENT = 'GOLDTERMS:3985' #  doesn't have any subclasses\n",
    "# \n",
    "# GOLDTERMS:4180, 'Environmental > Aquatic > Freshwater > Pond > Sediment' and ~64 more don't share a common root\n",
    "# poetry run runoak -i sqlite:obo:goldterms info 't~sediment'\n"
   ],
   "id": "eda591e14a91cd01",
   "outputs": [],
   "execution_count": 14
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.817707Z",
     "start_time": "2024-11-15T19:43:39.808783Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# GOLDTERMS_ROOT = GOLDTERMS_SOIL\n",
    "GOLDTERMS_ROOT = GOLDTERMS_SEDIMENT"
   ],
   "id": "bba8bf1e75430ef",
   "outputs": [],
   "execution_count": 15
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.863778Z",
     "start_time": "2024-11-15T19:43:39.860797Z"
    }
   },
   "cell_type": "code",
   "source": [
    "goldterms_subclass_query = f\"\"\"\n",
    "select\n",
    "\tsubject\n",
    "from\n",
    "\tentailed_edge ee\n",
    "where\n",
    "\tpredicate = 'rdfs:subClassOf'\n",
    "\tand object = '{GOLDTERMS_ROOT}'\n",
    "\"\"\""
   ],
   "id": "6b672d93e1f5c53",
   "outputs": [],
   "execution_count": 16
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.912216Z",
     "start_time": "2024-11-15T19:43:39.908075Z"
    }
   },
   "cell_type": "code",
   "source": "previous_submission_schema_url = \"https://raw.githubusercontent.com/microbiomedata/submission-schema/v10.7.0/src/nmdc_submission_schema/schema/nmdc_submission_schema.yaml\"",
   "id": "a1e9e99c5ac41ef4",
   "outputs": [],
   "execution_count": 17
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:39.958245Z",
     "start_time": "2024-11-15T19:43:39.955168Z"
    }
   },
   "cell_type": "code",
   "source": [
    "NMDC_RUNTIME_BASE_URL = 'https://api.microbiomedata.org/nmdcschema/'\n",
    "STUDY_SET_COLLECTION = 'study_set'\n",
    "BIOSAMPLE_SET_COLLECTION = 'biosample_set'"
   ],
   "id": "cde9a57dca22a7c1",
   "outputs": [],
   "execution_count": 18
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.005595Z",
     "start_time": "2024-11-15T19:43:40.001748Z"
    }
   },
   "cell_type": "code",
   "source": [
    "env_package_override_file = '../../mam-env-package-overrides.tsv'\n",
    "override_column = 'mam_inferred_env_package'"
   ],
   "id": "375bac6b84ab6198",
   "outputs": [],
   "execution_count": 19
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.052367Z",
     "start_time": "2024-11-15T19:43:40.049264Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_duckdb_url = 'https://portal.nersc.gov/project/m3408/biosamples_duckdb/ncbi_biosamples_2024-09-23.duckdb.gz'",
   "id": "8a8be3a9e6123df5",
   "outputs": [],
   "execution_count": 20
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.100453Z",
     "start_time": "2024-11-15T19:43:40.096333Z"
    }
   },
   "cell_type": "code",
   "source": [
    "gold_data_url = \"https://gold.jgi.doe.gov/download?mode=site_excel\"\n",
    "BIOSAMPLES_SHEET = \"Biosample\""
   ],
   "id": "fb9c800afcc86146",
   "outputs": [],
   "execution_count": 21
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.149629Z",
     "start_time": "2024-11-15T19:43:40.145181Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Initialize cache dictionaries for predict_from_normalized_env_packages\n",
    "# todo how to move the definitions for functiosn that use these globals?\n",
    "ancestor_cache = {}\n",
    "descendant_cache = {}"
   ],
   "id": "f952b5888d2c5f0d",
   "outputs": [],
   "execution_count": 22
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.210103Z",
     "start_time": "2024-11-15T19:43:40.198154Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo is filling memory with things like this a good idea? for understandability? or performance?\n",
    "# todo they should be aggregated somewhere, as specified by the config.yaml\n",
    "# todo or should we going straight to data frames? in which case a dlist of dicts might be preferable\n",
    "def get_curie_descendants_label_dict(curie, predicates, adapter):\n",
    "    curie_label_dict = {}\n",
    "    for descendant in adapter.descendants(curie, predicates=predicates):\n",
    "        curie_label_dict[descendant] = adapter.label(descendant)\n",
    "    return curie_label_dict"
   ],
   "id": "342c71e8b7a3366f",
   "outputs": [],
   "execution_count": 23
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.259574Z",
     "start_time": "2024-11-15T19:43:40.256173Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def curie_descendants_label_dict_to_lod(curie_label_dict):\n",
    "    return [{'curie': k, 'label': v} for k, v in curie_label_dict.items()]"
   ],
   "id": "1f4304edb0904f25",
   "outputs": [],
   "execution_count": 24
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.316536Z",
     "start_time": "2024-11-15T19:43:40.306450Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def curie_descendants_label_lod_to_df(curie_label_lod):\n",
    "    return pd.DataFrame(curie_label_lod)"
   ],
   "id": "c43fc53d70ff43e8",
   "outputs": [],
   "execution_count": 25
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.365257Z",
     "start_time": "2024-11-15T19:43:40.362176Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def get_schemaview_from_source(source):\n",
    "    return SchemaView(source)"
   ],
   "id": "28e160b747e7480e",
   "outputs": [],
   "execution_count": 26
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.424359Z",
     "start_time": "2024-11-15T19:43:40.412121Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def parse_hierarchically_underscored_strings(hierarchically_underscored_string_list):\n",
    "    result = []\n",
    "    for item in hierarchically_underscored_string_list:\n",
    "        # Remove leading underscores for label, split on '[' to separate curie\n",
    "        label, curie = item.lstrip('_').split(' [')\n",
    "        # Remove the trailing ']' from curie\n",
    "        curie = curie.rstrip(']')\n",
    "        # Append dictionary with label and curie\n",
    "        result.append({'label': label.strip(), 'curie': curie.strip()})\n",
    "    return result"
   ],
   "id": "8ce34ecfbae82a23",
   "outputs": [],
   "execution_count": 27
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.470466Z",
     "start_time": "2024-11-15T19:43:40.466242Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def dedupe_underscoreless_pvs(underscoreless_pvs):\n",
    "    # Dictionary to store CURIE as key and list of unique labels as values\n",
    "    curie_to_labels = {}\n",
    "\n",
    "    for item in underscoreless_pvs:\n",
    "        curie = item['curie']\n",
    "        label = item['label']\n",
    "\n",
    "        # Initialize the list if curie is not yet a key\n",
    "        if curie not in curie_to_labels:\n",
    "            curie_to_labels[curie] = []\n",
    "\n",
    "        # Add label if it is not already in the list for this curie\n",
    "        if label not in curie_to_labels[curie]:\n",
    "            curie_to_labels[curie].append(label)\n",
    "    return curie_to_labels\n"
   ],
   "id": "ce2b4b74d1fdc0d4",
   "outputs": [],
   "execution_count": 28
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.518913Z",
     "start_time": "2024-11-15T19:43:40.514379Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def validate_curie_label_list_dict(curie_label_dict, adapter, print_flag=False):\n",
    "    problem_curies = []\n",
    "    valid_curies = []\n",
    "    for curie, labels in curie_label_dict.items():\n",
    "        true_label = adapter.label(curie)\n",
    "        if true_label not in labels:\n",
    "            problem_curies.append(curie)\n",
    "            if print_flag:\n",
    "                print(f\"Error: {curie} has true label {true_label} which doesn't appear in {labels}\")\n",
    "        else:\n",
    "            valid_curies.append({\"curie\": curie, \"label\": true_label})\n",
    "    return {\"problems\": problem_curies, \"valids\": valid_curies}"
   ],
   "id": "cbd3ed3aec38ee16",
   "outputs": [],
   "execution_count": 29
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.588907Z",
     "start_time": "2024-11-15T19:43:40.567293Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo could pre-determine the collection sizes\n",
    "# todo could report elapsed time\n",
    "\n",
    "def get_docs_from_nmdc_collection(base_url, collection_name, max_page_size=1000, stop_after=None):\n",
    "    \"\"\"\n",
    "    Fetch all documents from a paginated API. Defaults to fetching a large number of documents per page.\n",
    "    Optionally stop after a specified number of documents.\n",
    "\n",
    "    Parameters:\n",
    "    - base_url: The base URL of the API endpoint (e.g., 'https://api.microbiomedata.org/nmdcschema/').\n",
    "    - collection_name: The name of the collection to fetch (e.g., 'biosample_set').\n",
    "    - max_page_size: The maximum number of items to retrieve per page (default 1000).\n",
    "    - stop_after: Optional parameter to stop fetching after a certain number of documents (default None).\n",
    "\n",
    "    Returns:\n",
    "    - A list of documents fetched from the API.\n",
    "    \"\"\"\n",
    "    documents = []\n",
    "    page_token = None\n",
    "    total_documents = 0\n",
    "    page_count = 0\n",
    "\n",
    "    # Construct the full URL with the collection name\n",
    "    url = f\"{base_url}{collection_name}\"\n",
    "\n",
    "    while True:\n",
    "        page_count += 1\n",
    "        # Prepare the query parameters\n",
    "        params = {\n",
    "            'collection_name': collection_name,\n",
    "            'max_page_size': max_page_size,  # Set large max_page_size to reduce pagination\n",
    "        }\n",
    "\n",
    "        if page_token:\n",
    "            params['page_token'] = page_token  # Add the page token for pagination\n",
    "\n",
    "        # Send the request to the API\n",
    "        response = requests.get(url, params=params)\n",
    "\n",
    "        if response.status_code != 200:\n",
    "            print(f\"Error fetching data: {response.status_code}\")\n",
    "            break\n",
    "\n",
    "        data = response.json()\n",
    "\n",
    "        # Add the current page of documents to the list\n",
    "        num_documents_on_page = len(data['resources'])\n",
    "        documents.extend(data['resources'])\n",
    "        total_documents += num_documents_on_page\n",
    "\n",
    "        # Status reporting\n",
    "        print(f\"Fetched page {page_count} with {num_documents_on_page} documents. Total fetched: {total_documents}\")\n",
    "\n",
    "        # If stop_after is provided, stop fetching after reaching the specified number of documents\n",
    "        if stop_after and total_documents >= stop_after:\n",
    "            documents = documents[:stop_after]  # Trim to the required number\n",
    "            print(f\"Reached stop_after limit of {stop_after} documents.\")\n",
    "            break\n",
    "\n",
    "        # Check if there is a next page\n",
    "        page_token = data.get('next_page_token')\n",
    "        if not page_token:\n",
    "            print(\"All documents fetched.\")\n",
    "            break  # Exit the loop if no more pages are available\n",
    "\n",
    "    return documents"
   ],
   "id": "69ea6069b4c4abbb",
   "outputs": [],
   "execution_count": 30
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.642700Z",
     "start_time": "2024-11-15T19:43:40.637756Z"
    }
   },
   "cell_type": "code",
   "source": [
    "\n",
    "def get_name_or_rawval(env_scale: Dict[str, Any]) -> str:\n",
    "    \"\"\"Safely extract label from environmental scale data.\"\"\"\n",
    "    if env_scale:\n",
    "        term = env_scale.get('term')\n",
    "        if term:\n",
    "            return term.get('name', term.get('has_raw_value', ''))\n",
    "    return ''"
   ],
   "id": "590502abe99de710",
   "outputs": [],
   "execution_count": 31
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.692398Z",
     "start_time": "2024-11-15T19:43:40.686388Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def tsv_to_dict_of_dicts(tsv_file, outer_key_column):\n",
    "    \"\"\"\n",
    "    Reads a TSV file into a dictionary of dictionaries.\n",
    "\n",
    "    :param tsv_file: Path to the TSV file.\n",
    "    :param outer_key_column: The column name or index to be used as the key for the outer dictionary.\n",
    "    :return: A dictionary of dictionaries, with outer keys being the values from the specified column.\n",
    "    \"\"\"\n",
    "    with open(tsv_file, newline='', encoding='utf-8') as f:\n",
    "        reader = csv.DictReader(f, delimiter='\\t')\n",
    "\n",
    "        result = {}\n",
    "\n",
    "        for row in reader:\n",
    "            outer_key = row[outer_key_column]\n",
    "            result[outer_key] = {key: value for key, value in row.items() if key != outer_key_column}\n",
    "\n",
    "    return result"
   ],
   "id": "6668e06afbd16f4",
   "outputs": [],
   "execution_count": 32
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.743342Z",
     "start_time": "2024-11-15T19:43:40.735549Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo only gets authoritative labels from the passed adapter, which is presumably EnvO only\n",
    "# todo would benefit from caching of labels\n",
    "\n",
    "def biosamples_lod_context_extractor(biosamples_lod, adapter, env_pacakge_overrides=None):\n",
    "    new_lod = []\n",
    "    for biosample in biosamples_lod:\n",
    "        insdc_identifiers = biosample.get('insdc_biosample_identifiers', [])\n",
    "\n",
    "        env_broad_scale_label = get_name_or_rawval(biosample.get('env_broad_scale'))\n",
    "        env_local_scale_label = get_name_or_rawval(biosample.get('env_local_scale'))\n",
    "        env_medium_label = get_name_or_rawval(biosample.get('env_medium'))\n",
    "\n",
    "        # Extracting optional scalar env_package.has_raw_value\n",
    "        env_package_has_raw_value = biosample.get('env_package', {}).get('has_raw_value', '')\n",
    "\n",
    "        # Extracting required multivalued part_of\n",
    "        associated_studies = '|'.join(biosample.get('associated_studies', []))  # Assuming part_of is a list of strings\n",
    "\n",
    "        row: Dict[str, str] = {\n",
    "            'id': biosample['id'],\n",
    "            'insdc_biosample_identifiers': '|'.join(insdc_identifiers) if insdc_identifiers else '',\n",
    "\n",
    "            'env_broad_scale_id': biosample['env_broad_scale']['term']['id'],\n",
    "            'env_broad_scale_mongo_label': env_broad_scale_label,\n",
    "            'env_broad_scale_auth_label': adapter.label(biosample['env_broad_scale']['term']['id']),\n",
    "\n",
    "            'env_local_scale_id': biosample['env_local_scale']['term']['id'],\n",
    "            'env_local_scale_mongo_label': env_local_scale_label,\n",
    "            'env_local_scale_auth_label': adapter.label(biosample['env_local_scale']['term']['id']),\n",
    "\n",
    "            'env_medium_id': biosample['env_medium']['term']['id'],\n",
    "            'env_medium_mongo_label': env_medium_label,\n",
    "            'env_medium_auth_label': adapter.label(biosample['env_medium']['term']['id']),\n",
    "\n",
    "            'env_package_has_raw_value': env_package_has_raw_value,\n",
    "            'normalized_env_package': 'soil' if env_package_has_raw_value == 'ENVO:00001998' else env_package_has_raw_value.lower(),\n",
    "            # todo abstract this though label search, or at least providing a lookup structure\n",
    "\n",
    "            'associated_studies': associated_studies\n",
    "        }\n",
    "\n",
    "        if env_pacakge_overrides and biosample['id'] in env_pacakge_overrides:\n",
    "            print(\n",
    "                f\"Overriding env_package for biosample {biosample['id']} from {row['normalized_env_package']} to {env_pacakge_overrides[biosample['id']]['mam_inferred_env_package']}\")\n",
    "            row['normalized_env_package'] = env_pacakge_overrides[biosample['id']]['mam_inferred_env_package']\n",
    "\n",
    "        new_lod.append(row)\n",
    "    return new_lod\n"
   ],
   "id": "53b9c7e7d15845f6",
   "outputs": [],
   "execution_count": 33
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.792709Z",
     "start_time": "2024-11-15T19:43:40.786418Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def get_hierarchy_terms(curie: str, adapter) -> dict:\n",
    "    \"\"\"\n",
    "    Extract ancestor and descendant terms from the ontology for a given CURIE,\n",
    "    using caching to improve performance and filtering by 'is_a' relationships.\n",
    "\n",
    "    Args:\n",
    "        curie (str): CURIE identifier for the ontology term.\n",
    "        adapter: Ontology adapter.\n",
    "\n",
    "    Returns:\n",
    "        dict: Dictionary containing lists of ancestor and descendant terms.\n",
    "    \"\"\"\n",
    "    if curie in ancestor_cache:\n",
    "        ancestors = ancestor_cache[curie]\n",
    "    else:\n",
    "        try:\n",
    "            ancestors = list(adapter.ancestors(curie, predicates=[IS_A]))\n",
    "            ancestor_cache[curie] = [adapter.label(ancestor) for ancestor in ancestors if ancestor]\n",
    "        except Exception as e:\n",
    "            print(f\"Error retrieving ancestors for {curie}: {e}\")\n",
    "            ancestor_cache[curie] = []\n",
    "\n",
    "    if curie in descendant_cache:\n",
    "        descendants = descendant_cache[curie]\n",
    "    else:\n",
    "        try:\n",
    "            descendants = list(adapter.descendants(curie, predicates=[IS_A]))\n",
    "            descendant_cache[curie] = [adapter.label(descendant) for descendant in descendants if descendant]\n",
    "        except Exception as e:\n",
    "            print(f\"Error retrieving descendants for {curie}: {e}\")\n",
    "            descendant_cache[curie] = []\n",
    "\n",
    "    return {\n",
    "        'ancestors': ancestor_cache[curie],\n",
    "        'descendants': descendant_cache[curie],\n",
    "    }"
   ],
   "id": "3efcfc9fcc2fe391",
   "outputs": [],
   "execution_count": 34
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.841334Z",
     "start_time": "2024-11-15T19:43:40.837295Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def vectorize_terms(df, column):\n",
    "    \"\"\"\n",
    "    Vectorize the ancestor or descendant terms for a given column.\n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame): The input dataframe.\n",
    "        column (str): The column name to vectorize.\n",
    "\n",
    "    Returns:\n",
    "        sparse matrix: The vectorized term matrix.\n",
    "    \"\"\"\n",
    "    vectorizer = CountVectorizer()\n",
    "    return vectorizer.fit_transform(\n",
    "        df[column].apply(lambda x: ' '.join([str(term) for term in x if term is not None]) if x is not None else '')\n",
    "    )"
   ],
   "id": "e9be0a2e69ee7409",
   "outputs": [],
   "execution_count": 35
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.891990Z",
     "start_time": "2024-11-15T19:43:40.884435Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def predict_from_normalized_env_packages(df_raw, adapter):\n",
    "    # Apply the function to the relevant columns\n",
    "\n",
    "    df = df_raw.copy()\n",
    "    for column in ['env_broad_scale_id', 'env_local_scale_id', 'env_medium_id']:\n",
    "        df[f'{column}_ancestors'] = df[column].apply(lambda x: get_hierarchy_terms(x, adapter)['ancestors'])\n",
    "        df[f'{column}_descendants'] = df[column].apply(lambda x: get_hierarchy_terms(x, adapter)['descendants'])\n",
    "\n",
    "    # Vectorize each set of terms separately\n",
    "    broad_scale_ancestors = vectorize_terms(df, 'env_broad_scale_id_ancestors')\n",
    "    broad_scale_descendants = vectorize_terms(df, 'env_broad_scale_id_descendants')\n",
    "\n",
    "    local_scale_ancestors = vectorize_terms(df, 'env_local_scale_id_ancestors')\n",
    "    local_scale_descendants = vectorize_terms(df, 'env_local_scale_id_descendants')\n",
    "\n",
    "    medium_ancestors = vectorize_terms(df, 'env_medium_id_ancestors')\n",
    "    medium_descendants = vectorize_terms(df, 'env_medium_id_descendants')\n",
    "\n",
    "    # Combine all feature matrices\n",
    "    X = hstack([\n",
    "        broad_scale_ancestors,\n",
    "        broad_scale_descendants,\n",
    "        local_scale_ancestors,\n",
    "        local_scale_descendants,\n",
    "        medium_ancestors,\n",
    "        medium_descendants\n",
    "    ])\n",
    "\n",
    "    # Filter the DataFrame to only include non-null rows for the target column\n",
    "    df_filtered = df[df['normalized_env_package'].notnull() & (df['normalized_env_package'] != \"\")]\n",
    "\n",
    "    # Extract the target variable\n",
    "    y = df_filtered['normalized_env_package']\n",
    "\n",
    "    # Ensure X corresponds to the filtered rows\n",
    "    X_filtered = X[df_filtered.index]\n",
    "\n",
    "    # Split the data into training and testing sets\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X_filtered, y, test_size=0.3, random_state=42)\n",
    "\n",
    "    # Train a Random Forest Classifier\n",
    "    clf = RandomForestClassifier(n_estimators=100, random_state=42)\n",
    "    clf.fit(X_train, y_train)\n",
    "\n",
    "    # Make predictions on the test set\n",
    "    y_pred = clf.predict(X_test)\n",
    "\n",
    "    # Evaluate the model\n",
    "    print(classification_report(y_test, y_pred))\n",
    "\n",
    "    # not determining confidence for each class nor saving any diagnostics any more\n",
    "\n",
    "    return clf.predict(X)"
   ],
   "id": "f03bb583dfde6d48",
   "outputs": [],
   "execution_count": 36
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.940577Z",
     "start_time": "2024-11-15T19:43:40.935847Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def parse_curie_label(text, approved_prefixes=['ENVO']):\n",
    "    # Case-insensitive pattern for matching approved prefixes followed by an ID\n",
    "    pattern = r'\\b(?:' + '|'.join(approved_prefixes) + r')\\s*[:_]\\s*(\\d+)\\b'\n",
    "    curie_match = re.search(pattern, text, re.IGNORECASE)\n",
    "\n",
    "    if curie_match:\n",
    "        curie = f\"{approved_prefixes[0].upper()}:{curie_match.group(1)}\"  # standardize prefix to 'ENVO:ID'\n",
    "        label = re.sub(pattern, \"\", text).strip(\"[]() \")\n",
    "        # replace any colons in the label with a whitespace\n",
    "        return pd.Series([label, curie])\n",
    "    else:\n",
    "        label = re.sub(r':', ' ', text)\n",
    "        return pd.Series([label, None])  # No CURIE found, return original label and None for CURIE\n"
   ],
   "id": "9edf0d5477083503",
   "outputs": [],
   "execution_count": 37
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:40.987447Z",
     "start_time": "2024-11-15T19:43:40.983285Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def get_longest_annotation_curie(text, adapter):\n",
    "    annotations = adapter.annotate_text(text)\n",
    "    if not annotations:  # Check if annotations list is empty\n",
    "        return None\n",
    "    try:\n",
    "        longest_annotation = max(annotations, key=lambda x: x.subject_end - x.subject_start)\n",
    "\n",
    "        if longest_annotation.subject_end - longest_annotation.subject_start < MIN_ANNOTATION_LEN:\n",
    "            return None\n",
    "        return longest_annotation.object_id\n",
    "    except ValueError:\n",
    "        return None  # Return None if there's an unexpected issue with finding the max\n"
   ],
   "id": "8db01eb510b460c2",
   "outputs": [],
   "execution_count": 38
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:43:41.037181Z",
     "start_time": "2024-11-15T19:43:41.032993Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Determine the filenames and target directory\n",
    "ncbi_compressed_filename = urlparse(ncbi_duckdb_url).path.split('/')[-1]\n",
    "ncbi_filename = os.path.splitext(ncbi_compressed_filename)[0]\n",
    "target_dir = os.path.join(\"..\", \"..\")  # Two levels up"
   ],
   "id": "6b934c793818f5bc",
   "outputs": [],
   "execution_count": 39
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:45:16.346370Z",
     "start_time": "2024-11-15T19:43:41.080637Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Fetch the contents from the URL and save compressed file in target directory\n",
    "ncbi_response = requests.get(ncbi_duckdb_url)\n",
    "ncbi_compressed_file_path = os.path.join(target_dir, ncbi_compressed_filename)\n",
    "with open(ncbi_compressed_file_path, \"wb\") as f:\n",
    "    f.write(ncbi_response.content)\n",
    "\n",
    "# ~ 2 minutes @ 250 Mbps"
   ],
   "id": "8c6147ebd9dcd69d",
   "outputs": [],
   "execution_count": 40
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:28.907917Z",
     "start_time": "2024-11-15T19:45:16.392428Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Unzip the compressed file and save the extracted file in target directory\n",
    "ncbi_uncompressed_file_path = os.path.join(target_dir, ncbi_filename)\n",
    "with gzip.open(ncbi_compressed_file_path, \"rb\") as f_in:\n",
    "    with open(ncbi_uncompressed_file_path, \"wb\") as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)\n",
    "\n",
    "# ~ 2 minutes"
   ],
   "id": "a65971b89486ffe8",
   "outputs": [],
   "execution_count": 41
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:28.986441Z",
     "start_time": "2024-11-15T19:46:28.953064Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_conn = duckdb.connect(database=ncbi_uncompressed_file_path, read_only=True)",
   "id": "a3c0ec1db5a4f847",
   "outputs": [],
   "execution_count": 42
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:29.037920Z",
     "start_time": "2024-11-15T19:46:29.000577Z"
    }
   },
   "cell_type": "code",
   "source": "envo_adapter = get_adapter(envo_adapter_string)",
   "id": "2f5ec46c2396082c",
   "outputs": [],
   "execution_count": 43
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:29.165850Z",
     "start_time": "2024-11-15T19:46:29.052132Z"
    }
   },
   "cell_type": "code",
   "source": "biome_descendants = get_curie_descendants_label_dict(BIOME_CURIE, [IS_A], envo_adapter)",
   "id": "67808c8134cd1f9a",
   "outputs": [],
   "execution_count": 44
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:29.180235Z",
     "start_time": "2024-11-15T19:46:29.177110Z"
    }
   },
   "cell_type": "code",
   "source": "biome_descendants_lod = curie_descendants_label_dict_to_lod(biome_descendants)",
   "id": "97f53975986137a2",
   "outputs": [],
   "execution_count": 45
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:29.223474Z",
     "start_time": "2024-11-15T19:46:29.219040Z"
    }
   },
   "cell_type": "code",
   "source": "biome_descendants_frame = curie_descendants_label_lod_to_df(biome_descendants_lod)",
   "id": "6ff81c86ce7c4064",
   "outputs": [],
   "execution_count": 46
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:29.287423Z",
     "start_time": "2024-11-15T19:46:29.267121Z"
    }
   },
   "cell_type": "code",
   "source": "biome_descendants_frame",
   "id": "7d9f5e51cb21844",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "             curie                          label\n",
       "0    ENVO:01001505            alpine tundra biome\n",
       "1    ENVO:01000024           marine benthic biome\n",
       "2    ENVO:01000252          freshwater lake biome\n",
       "3    ENVO:01000180                   tundra biome\n",
       "4    ENVO:01000123       marine sponge reef biome\n",
       "..             ...                            ...\n",
       "123  ENVO:01000858         marine upwelling biome\n",
       "124  ENVO:01000188         tropical savanna biome\n",
       "125  ENVO:01000042  neritic epipelagic zone biome\n",
       "126  ENVO:01000045              epeiric sea biome\n",
       "127  ENVO:01000049        marine coral reef biome\n",
       "\n",
       "[128 rows x 2 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:01001505</td>\n",
       "      <td>alpine tundra biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>marine benthic biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:01000252</td>\n",
       "      <td>freshwater lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENVO:01000180</td>\n",
       "      <td>tundra biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:01000123</td>\n",
       "      <td>marine sponge reef biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>ENVO:01000858</td>\n",
       "      <td>marine upwelling biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>ENVO:01000188</td>\n",
       "      <td>tropical savanna biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>ENVO:01000042</td>\n",
       "      <td>neritic epipelagic zone biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>ENVO:01000045</td>\n",
       "      <td>epeiric sea biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>ENVO:01000049</td>\n",
       "      <td>marine coral reef biome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128 rows × 2 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 47
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "----",
   "id": "4c418e6f7c75da52"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:33.167030Z",
     "start_time": "2024-11-15T19:46:29.362892Z"
    }
   },
   "cell_type": "code",
   "source": "sv = get_schemaview_from_source(previous_submission_schema_url)",
   "id": "3f73cf88902b5ca6",
   "outputs": [],
   "execution_count": 48
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:33.209960Z",
     "start_time": "2024-11-15T19:46:33.205109Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo break out slow steps into its own cell\n",
    "\n",
    "try:\n",
    "    env_broad_scale_enum_def = sv.get_enum(ENV_BROAD_SCALE_ENUM)\n",
    "    env_broad_scale_pvs_keys = list(env_broad_scale_enum_def.permissible_values.keys())\n",
    "except AttributeError as e:\n",
    "    # Handle the AttributeError\n",
    "    print(f\"An AttributeError occurred: {e}\")\n",
    "    env_broad_scale_pvs_keys =[]\n",
    "    "
   ],
   "id": "c6355aee16d6a340",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An AttributeError occurred: 'NoneType' object has no attribute 'permissible_values'\n"
     ]
    }
   ],
   "execution_count": 49
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:38.527046Z",
     "start_time": "2024-11-15T19:46:38.524019Z"
    }
   },
   "cell_type": "code",
   "source": "initially_parsed_env_broad_scale_pvs = parse_hierarchically_underscored_strings(env_broad_scale_pvs_keys)",
   "id": "f0631ea631f6c807",
   "outputs": [],
   "execution_count": 50
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:43.698058Z",
     "start_time": "2024-11-15T19:46:43.694790Z"
    }
   },
   "cell_type": "code",
   "source": "deduped_env_broad_scale_pvs = dedupe_underscoreless_pvs(initially_parsed_env_broad_scale_pvs)",
   "id": "4f9efbd6e1e2c834",
   "outputs": [],
   "execution_count": 51
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:48.990756Z",
     "start_time": "2024-11-15T19:46:48.987893Z"
    }
   },
   "cell_type": "code",
   "source": "pv_validation_results = validate_curie_label_list_dict(deduped_env_broad_scale_pvs, envo_adapter, print_flag=True)",
   "id": "eb81442383577618",
   "outputs": [],
   "execution_count": 52
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:46:54.986128Z",
     "start_time": "2024-11-15T19:46:54.982330Z"
    }
   },
   "cell_type": "code",
   "source": "pv_validation_results",
   "id": "aa0d1bdc45bb3f3b",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'problems': [], 'valids': []}"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 53
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "----",
   "id": "cd05c55ab2a84b0d"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:47:47.656301Z",
     "start_time": "2024-11-15T19:47:00.450912Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo rename to all_nmdc_samples etc\n",
    "all_biosamples = get_docs_from_nmdc_collection(NMDC_RUNTIME_BASE_URL,\n",
    "                                               BIOSAMPLE_SET_COLLECTION)  # Example with stop_afte"
   ],
   "id": "53d8c3995e597527",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched page 1 with 1000 documents. Total fetched: 1000\n",
      "Fetched page 2 with 1000 documents. Total fetched: 2000\n",
      "Fetched page 3 with 1000 documents. Total fetched: 3000\n",
      "Fetched page 4 with 1000 documents. Total fetched: 4000\n",
      "Fetched page 5 with 1000 documents. Total fetched: 5000\n",
      "Fetched page 6 with 1000 documents. Total fetched: 6000\n",
      "Fetched page 7 with 1000 documents. Total fetched: 7000\n",
      "Fetched page 8 with 1000 documents. Total fetched: 8000\n",
      "Fetched page 9 with 320 documents. Total fetched: 8320\n",
      "All documents fetched.\n"
     ]
    }
   ],
   "execution_count": 54
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:47:53.113068Z",
     "start_time": "2024-11-15T19:47:52.549672Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo I don't think we're actually using this\n",
    "all_studies = get_docs_from_nmdc_collection(NMDC_RUNTIME_BASE_URL, STUDY_SET_COLLECTION)  # Example with stop_after"
   ],
   "id": "95a21f388c30f596",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched page 1 with 29 documents. Total fetched: 29\n",
      "All documents fetched.\n"
     ]
    }
   ],
   "execution_count": 55
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:48:02.789200Z",
     "start_time": "2024-11-15T19:48:02.776814Z"
    }
   },
   "cell_type": "code",
   "source": "env_pacakge_overrides = tsv_to_dict_of_dicts(env_package_override_file, 'id')",
   "id": "c8f6e276e7e29c15",
   "outputs": [],
   "execution_count": 56
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:48:02.804642Z",
     "start_time": "2024-11-15T19:48:02.793525Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# env_pacakge_overrides\n",
    "# todo or show as frame\n",
    "# todo include some other columns for context?"
   ],
   "id": "d3bd77c4549f2db3",
   "outputs": [],
   "execution_count": 57
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:48:15.078124Z",
     "start_time": "2024-11-15T19:48:07.694889Z"
    }
   },
   "cell_type": "code",
   "source": [
    "biosample_contexts_lod = biosamples_lod_context_extractor(all_biosamples, envo_adapter,\n",
    "                                                          env_pacakge_overrides=env_pacakge_overrides)"
   ],
   "id": "cdd0699ed7bfa22d",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overriding env_package for biosample nmdc:bsm-11-0k8nkx16 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-19v98823 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-1yvac190 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-28kgw077 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-2hswww54 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-34przm31 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-35m0rm03 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-3636w778 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3nffqc45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3nhng665 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3r4g4610 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-3tr7f795 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3xc5ft48 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-3xthwx28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-42c9kj17 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-43rwvm93 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-4838sk73 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5ak8ja54 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5g2kjf44 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5j742c82 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5qq80d09 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5rgay328 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-5w4ba396 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-61xfks69 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-62fxv990 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-674m9194 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-678ndj04 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-69g1rw32 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6k8cmy53 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6m5dma62 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6qgwc522 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6xa4vj59 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6y26bz06 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6ya0cs71 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-6z11g163 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-73zqmh40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-766dwa91 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7b566r47 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-7cfwd964 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7pyrek52 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7yg2ne22 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-7z1qvf63 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-82qw5v38 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-899bm454 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8aktkq28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8bewbp90 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8cgpqn03 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8e43d898 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-8g7xbe66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-91bdea61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-94c2xj88 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-9c0hk014 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-9f2ahd94 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-a6p0kd54 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-acehgb61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-acfvxq24 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-aesqd647 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-af8kqq09 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-ahfq0n74 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-b2w6cr26 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-b7pv2a25 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-b9cjnj29 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bcns6e03 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bgar5122 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bja22k98 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bm9haf23 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-bqp0m764 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-byzpm179 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-c12hyq86 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-c52npq37 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-c756vg97 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-can3rz84 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cjp6kd42 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ck98jg55 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cqbpqf62 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cv5vxb38 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cvqsze66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cxprmv88 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-cz68sy30 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-d8cgz835 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-djqze645 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dppkv536 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dqnfvd06 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dtyb1w45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dva76848 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-dz264p50 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-e0rz6p82 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-e747ch69 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-epzqzg45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-evfbfp31 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-f8tzvk82 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fbm8fa50 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-feynq214 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fgjpsy56 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fm4p4558 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fqchmc77 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-fy3j8w19 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fyey4107 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-fzvzb345 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g35dnf26 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g3r2dm95 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g4my3419 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-g4rxzh95 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-gch9pp73 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-gd31q856 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-gnbsqd30 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-gvp5zp51 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-gwx12029 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-h2k7nv93 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-h3e54f51 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-h96d6881 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ht2c1k71 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ht2v0s47 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-hv1hv533 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j1bay453 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j1hqcz15 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j2pfz414 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j5y27f16 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j6h8pw46 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-j6j1p639 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jjchgw36 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jp678560 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jq2e3v76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jx6ar740 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-jxazjs92 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-jy6ece20 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-k0vz1j87 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-k1twxb65 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-k9192006 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kdja3v28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kesw6757 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kmpwgz16 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-kqbbyb62 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-krcjqt76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-kv966g71 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-m0h9cb45 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-m4e83773 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-m4p1ck60 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-m5prkp97 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-m69k6p77 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-m9k8tx16 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ma7ry826 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-mb9s8c17 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-mbcv4b48 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n2a7p974 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n6vwc542 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n8t7zs15 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-n9sx1a80 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nher1m25 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-nm187g36 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nr7vje40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nrkha596 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nwh9ez71 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-nx7zwq61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-p4wss361 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-p9j7rw40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pebqj065 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pgmhxj86 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pgxc4f08 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ppp4fk41 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pq105b36 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pvr0ez88 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-pygdem51 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-q3qsx507 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-q3sa0t28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-q94v7385 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qam3hk28 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qe13q836 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qg6kxr61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qhq2kh40 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qms4zn61 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qqc8e352 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qvm0vt03 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qy2fef59 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-qzncy838 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r3b0ak56 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r3tm7493 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r4kwpw90 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-r7k97q10 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-rk62cv25 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-rpj4zm44 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-rzz0z710 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-s6r1q327 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-s70cv630 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sa4z4f71 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-sdjt4464 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sdjyhj66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-shjn2p67 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sjtd0y05 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-spav5m37 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-sv771k17 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-syb42s66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-t7xb8h63 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ta8dt754 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tcxb3w32 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-te5fxe60 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-th3sr643 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tkwn9650 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tn8q1269 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-tnbabs47 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-tpatza67 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-trbf6m66 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-trpbtt49 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-ttg2t449 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-tz96dg48 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-v0p16d89 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-v9k6q078 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-veffrc25 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-vme3w877 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-vp7n5531 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-vpe6j733 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-vrsw1c02 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-w1t8bv92 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wh35sj59 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wnrp0p76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wqy7mc31 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-wr7h1t97 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-x0sc5x81 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-xy5ern23 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-xy8t6808 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-y31mxv90 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-y5w1j489 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-y80cxc26 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-yd4dqe75 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-yfc2at70 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-yvzvbr38 from  to hydrocarbon resources-fluids_swabs\n",
      "Overriding env_package for biosample nmdc:bsm-11-z4hspp43 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-z66zfk92 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-z75j2v52 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-zf5qxz83 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-zsdrmw76 from  to plant-associated\n",
      "Overriding env_package for biosample nmdc:bsm-11-zvh2hc82 from  to plant-associated\n"
     ]
    }
   ],
   "execution_count": 58
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:48:20.773993Z",
     "start_time": "2024-11-15T19:48:20.751829Z"
    }
   },
   "cell_type": "code",
   "source": "nmdc_biosample_contexts_frame = pd.DataFrame(biosample_contexts_lod)",
   "id": "4b93dbbb4912a825",
   "outputs": [],
   "execution_count": 59
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:48:25.653113Z",
     "start_time": "2024-11-15T19:48:25.647974Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# print a value count for the normalized_env_package column\n",
    "print(\"Value counts for normalized_env_package column:\")\n",
    "print(nmdc_biosample_contexts_frame['normalized_env_package'].value_counts(dropna=False))"
   ],
   "id": "f7cc947dcc620637",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Value counts for normalized_env_package column:\n",
      "normalized_env_package\n",
      "                                                   5838\n",
      "soil                                               1665\n",
      "plant-associated                                    401\n",
      "water                                               192\n",
      "miscellaneous natural or artificial environment     140\n",
      "host-associated                                      61\n",
      "hydrocarbon resources-fluids_swabs                   23\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "execution_count": 60
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:48:37.796595Z",
     "start_time": "2024-11-15T19:48:30.708925Z"
    }
   },
   "cell_type": "code",
   "source": "package_predictions = predict_from_normalized_env_packages(nmdc_biosample_contexts_frame, envo_adapter)",
   "id": "c09274a258eaa131",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                 precision    recall  f1-score   support\n",
      "\n",
      "                                host-associated       1.00      1.00      1.00        21\n",
      "             hydrocarbon resources-fluids_swabs       1.00      0.83      0.91         6\n",
      "miscellaneous natural or artificial environment       1.00      1.00      1.00        44\n",
      "                               plant-associated       1.00      1.00      1.00       132\n",
      "                                           soil       1.00      1.00      1.00       489\n",
      "                                          water       0.98      1.00      0.99        53\n",
      "\n",
      "                                       accuracy                           1.00       745\n",
      "                                      macro avg       1.00      0.97      0.98       745\n",
      "                                   weighted avg       1.00      1.00      1.00       745\n",
      "\n"
     ]
    }
   ],
   "execution_count": 61
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:48:43.660283Z",
     "start_time": "2024-11-15T19:48:43.656635Z"
    }
   },
   "cell_type": "code",
   "source": "nmdc_biosample_contexts_frame['predicted_env_package'] = package_predictions",
   "id": "8db3619186f07681",
   "outputs": [],
   "execution_count": 62
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:48:48.840289Z",
     "start_time": "2024-11-15T19:48:48.836392Z"
    }
   },
   "cell_type": "code",
   "source": "nmdc_biosample_contexts_frame.shape",
   "id": "764a152a8c9fb9ff",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8320, 15)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 63
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:48:53.944625Z",
     "start_time": "2024-11-15T19:48:53.938947Z"
    }
   },
   "cell_type": "code",
   "source": [
    "nmdc_biosample_contexts_frame = nmdc_biosample_contexts_frame[\n",
    "    nmdc_biosample_contexts_frame['predicted_env_package'] == ENV_LC]"
   ],
   "id": "f8e447afbe4734f",
   "outputs": [],
   "execution_count": 64
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:48:59.031690Z",
     "start_time": "2024-11-15T19:48:59.027515Z"
    }
   },
   "cell_type": "code",
   "source": "nmdc_biosample_contexts_frame.shape",
   "id": "8c10b65e05f9df26",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 15)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 65
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "----",
   "id": "38d09a3bb6bfb3c3"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:49:06.619295Z",
     "start_time": "2024-11-15T19:49:04.954288Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame = ncbi_conn.execute(ncbi_query).fetchdf()",
   "id": "5789ecb82c53fd3a",
   "outputs": [],
   "execution_count": 66
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:49:11.618171Z",
     "start_time": "2024-11-15T19:49:11.614158Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame.insert(0, 'serial_number', range(1, len(ncbi_frame) + 1))",
   "id": "dd689a5d175c9fd9",
   "outputs": [],
   "execution_count": 67
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:49:16.837297Z",
     "start_time": "2024-11-15T19:49:16.834248Z"
    }
   },
   "cell_type": "code",
   "source": "# includes env broad scale values with counts of one... useful for discovering drag-down submissions?",
   "id": "c33fd385877863f3",
   "outputs": [],
   "execution_count": 68
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:49:23.356299Z",
     "start_time": "2024-11-15T19:49:23.351601Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['content_list'] = ncbi_frame['content'].str.split('|')",
   "id": "a2492565d3934896",
   "outputs": [],
   "execution_count": 69
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:49:29.244855Z",
     "start_time": "2024-11-15T19:49:29.241045Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['content_count'] = ncbi_frame['content_list'].apply(len)",
   "id": "20aa3fa342102bef",
   "outputs": [],
   "execution_count": 70
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:49:34.752415Z",
     "start_time": "2024-11-15T19:49:34.748665Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame.shape",
   "id": "34682934e6d4b38d",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1370, 5)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 71
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:49:40.220361Z",
     "start_time": "2024-11-15T19:49:40.214272Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame = ncbi_frame.explode('content_list').reset_index(drop=True)",
   "id": "d708e65b1ef28344",
   "outputs": [],
   "execution_count": 72
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:49:46.336068Z",
     "start_time": "2024-11-15T19:49:46.332329Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame.shape",
   "id": "3feabb64849b8402",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1398, 5)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 73
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:49:51.664523Z",
     "start_time": "2024-11-15T19:49:51.661778Z"
    }
   },
   "cell_type": "code",
   "source": "# how many content_list strings contain envo multiple times now?",
   "id": "a76934c8555f4923",
   "outputs": [],
   "execution_count": 74
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:49:57.000754Z",
     "start_time": "2024-11-15T19:49:56.995637Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['envo_count'] = ncbi_frame['content_list'].str.lower().str.count(\"envo\")",
   "id": "ec3f672a6982e836",
   "outputs": [],
   "execution_count": 75
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:50:02.461180Z",
     "start_time": "2024-11-15T19:50:02.455393Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['envo_count'].value_counts()",
   "id": "927b2041141d82ad",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "envo_count\n",
       "0    1189\n",
       "1     209\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 76
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "doesn't account for multiple label strings delimited with something other than '|'",
   "id": "36d4d1c696c1a5d5"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:50:08.736598Z",
     "start_time": "2024-11-15T19:50:08.556671Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame[['extracted_label', 'extracted_curie']] = ncbi_frame['content_list'].apply(parse_curie_label)",
   "id": "393d2f52a7b7155c",
   "outputs": [],
   "execution_count": 77
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:50:08.778914Z",
     "start_time": "2024-11-15T19:50:08.774792Z"
    }
   },
   "cell_type": "code",
   "source": [
    "parse_failures = ncbi_frame[\n",
    "    (ncbi_frame['envo_count'] > 0) & (ncbi_frame['extracted_curie'].isna() | (ncbi_frame['extracted_curie'] == ''))]\n"
   ],
   "id": "2b310df2655ee870",
   "outputs": [],
   "execution_count": 78
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:50:08.821367Z",
     "start_time": "2024-11-15T19:50:08.814222Z"
    }
   },
   "cell_type": "code",
   "source": "parse_failures",
   "id": "f0ec687bf5e3843c",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [serial_number, content, sample_count, content_list, content_count, envo_count, extracted_label, extracted_curie]\n",
       "Index: []"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>serial_number</th>\n",
       "      <th>content</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>content_list</th>\n",
       "      <th>content_count</th>\n",
       "      <th>envo_count</th>\n",
       "      <th>extracted_label</th>\n",
       "      <th>extracted_curie</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 79
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:50:14.794221Z",
     "start_time": "2024-11-15T19:50:14.381080Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['real_label'] = ncbi_frame['extracted_curie'].apply(envo_adapter.label)",
   "id": "6be66d60a992144b",
   "outputs": [],
   "execution_count": 80
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:50:35.584664Z",
     "start_time": "2024-11-15T19:50:14.798937Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Apply the function to each row in the 'label' column\n",
    "ncbi_frame['longest_annotation_curie'] = ncbi_frame['extracted_label'].apply(\n",
    "    lambda x: get_longest_annotation_curie(x, envo_adapter))\n"
   ],
   "id": "8bc7d8d00aeab56",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Skipping statements(subject=ENVO:00000112,predicate=oio:hasDbXref,object=<http://www.eionet.europa.eu/gemet/concept/8704>,value=None,datatype=None,language=None,); ValueError: <http://www.eionet.europa.eu/gemet/concept/8704> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:00001996,predicate=oio:hasDbXref,object=<https://en.wikipedia.org/wiki/Acid_mine_drainage>,value=None,datatype=None,language=None,); ValueError: <https://en.wikipedia.org/wiki/Acid_mine_drainage> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01000225,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01000227,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001495,predicate=oio:hasDbXref,object=<https://www.goosocean.org/components/com_oe/oe.php?task=download&id=35904&version=2.0&lang=1&format=1>,value=None,datatype=None,language=None,); ValueError: <https://www.goosocean.org/components/com_oe/oe.php?task=download&id=35904&version=2.0&lang=1&format=1> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001621,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001626,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/deserts-and-xeric-shrublands>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/deserts-and-xeric-shrublands> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001627,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/ecoregions/im1304>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/ecoregions/im1304> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001628,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/ecoregions/im1303>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/ecoregions/im1303> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001644,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C021/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C021/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001644,predicate=oio:hasDbXref,object=None,value=Carbonate which is formed as the result of some biological process.,datatype=None,language=None,); ValueError: Carbonate which is formed as the result of some biological process. is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001645,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C013/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C013/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001646,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C014/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C014/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001647,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C004/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C004/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001649,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C011/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C011/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001650,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C012/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C012/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001651,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C003/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C003/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001652,predicate=oio:hasDbXref,object=<http://vocab.nerc.ac.uk/collection/S23/current/S23C005/>,value=None,datatype=None,language=None,); ValueError: <http://vocab.nerc.ac.uk/collection/S23/current/S23C005/> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001685,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001692,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001743,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001799,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001801,predicate=oio:hasDbXref,object=<https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests>,value=None,datatype=None,language=None,); ValueError: <https://www.worldwildlife.org/biomes/tropical-and-subtropical-dry-broadleaf-forests> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001823,predicate=oio:hasDbXref,object=<https://github.com/EnvironmentOntology/envo/issues/1319>,value=None,datatype=None,language=None,); ValueError: <https://github.com/EnvironmentOntology/envo/issues/1319> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001878,predicate=oio:hasDbXref,object=<https://www.goosocean.org/index.php?option=com_oe&task=viewDocumentRecord&docID=17476>,value=None,datatype=None,language=None,); ValueError: <https://www.goosocean.org/index.php?option=com_oe&task=viewDocumentRecord&docID=17476> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001879,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001880,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:01001881,predicate=oio:hasDbXref,object=<https://github.com/ESIPFed/sweet/issues/219>,value=None,datatype=None,language=None,); ValueError: <https://github.com/ESIPFed/sweet/issues/219> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:02000093,predicate=oio:hasDbXref,object=<https://en.wikipedia.org/wiki/Uranium_mining>,value=None,datatype=None,language=None,); ValueError: <https://en.wikipedia.org/wiki/Uranium_mining> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:02000094,predicate=oio:hasDbXref,object=<https://en.wikipedia.org/wiki/Ore>,value=None,datatype=None,language=None,); ValueError: <https://en.wikipedia.org/wiki/Ore> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:03501363,predicate=oio:hasDbXref,object=<https://www.conveyorsystemsltd.co.uk/news/168/126/Roller-Conveyor-vs-Belt-Conveyor.htm>,value=None,datatype=None,language=None,); ValueError: <https://www.conveyorsystemsltd.co.uk/news/168/126/Roller-Conveyor-vs-Belt-Conveyor.htm> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:06105112,predicate=oio:hasDbXref,object=<http://eurovoc.europa.eu/1837>,value=None,datatype=None,language=None,); ValueError: <http://eurovoc.europa.eu/1837> is not a valid URI or CURIE\n",
      "ERROR:root:Skipping statements(subject=ENVO:06105112,predicate=oio:hasDbXref,object=<http://aims.fao.org/aos/agrovoc/c_5998>,value=None,datatype=None,language=None,); ValueError: <http://aims.fao.org/aos/agrovoc/c_5998> is not a valid URI or CURIE\n",
      "WARNING:root:Skipping <http://geneontology.org/foo/applies-pattern> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <http://schema.org/image> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <http://swrl.stanford.edu/ontologies/3.3/swrla.owl#isRuleEnabled> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q2306597> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q2> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q525> as it is not a valid CURIE\n",
      "WARNING:root:Skipping <https://www.wikidata.org/wiki/Q715269> as it is not a valid CURIE\n"
     ]
    }
   ],
   "execution_count": 81
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:50:41.739913Z",
     "start_time": "2024-11-15T19:50:41.312152Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['longest_annotation_label'] = ncbi_frame['longest_annotation_curie'].apply(envo_adapter.label)",
   "id": "38e6a45ffdafc058",
   "outputs": [],
   "execution_count": 82
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:50:47.595405Z",
     "start_time": "2024-11-15T19:50:47.582638Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame",
   "id": "d393c31729bdfb34",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      serial_number                          content  sample_count  \\\n",
       "0                 1                Small river biome          1819   \n",
       "1                 2                     marine biome          1644   \n",
       "2                 3                    ENVO:01000047           805   \n",
       "3                 4                         sediment           729   \n",
       "4                 5  estuarine biome [ENVO:01000020]           669   \n",
       "...             ...                              ...           ...   \n",
       "1393           1366   estuarine biome[Envo:01000031]             1   \n",
       "1394           1367   estuarine biome[Envo:01000035]             1   \n",
       "1395           1368                    ENVO:01000026             1   \n",
       "1396           1369                    ENVO:01000035             1   \n",
       "1397           1370                    ENVO:01000040             1   \n",
       "\n",
       "                         content_list  content_count  envo_count  \\\n",
       "0                   Small river biome              1           0   \n",
       "1                        marine biome              1           0   \n",
       "2                       ENVO:01000047              1           1   \n",
       "3                            sediment              1           0   \n",
       "4     estuarine biome [ENVO:01000020]              1           1   \n",
       "...                               ...            ...         ...   \n",
       "1393   estuarine biome[Envo:01000031]              1           1   \n",
       "1394   estuarine biome[Envo:01000035]              1           1   \n",
       "1395                    ENVO:01000026              1           1   \n",
       "1396                    ENVO:01000035              1           1   \n",
       "1397                    ENVO:01000040              1           1   \n",
       "\n",
       "                    extracted_label extracted_curie  \\\n",
       "0                 Small river biome            None   \n",
       "1                      marine biome            None   \n",
       "2                                     ENVO:01000047   \n",
       "3                          sediment            None   \n",
       "4                   estuarine biome   ENVO:01000020   \n",
       "...                             ...             ...   \n",
       "1393  estuarine biome[Envo:01000031   ENVO:01000031   \n",
       "1394  estuarine biome[Envo:01000035   ENVO:01000035   \n",
       "1395                                  ENVO:01000026   \n",
       "1396                                  ENVO:01000035   \n",
       "1397                                  ENVO:01000040   \n",
       "\n",
       "                            real_label longest_annotation_curie  \\\n",
       "0                                 None            ENVO:00000890   \n",
       "1                                 None            ENVO:00000447   \n",
       "2              mediterranean sea biome                     None   \n",
       "3                                 None            ENVO:00002007   \n",
       "4                      estuarine biome            ENVO:01000020   \n",
       "...                                ...                      ...   \n",
       "1393           obsolete marine feature            ENVO:01000020   \n",
       "1394     oceanic epipelagic zone biome            ENVO:01000020   \n",
       "1395         marine bathyal zone biome                     None   \n",
       "1396     oceanic epipelagic zone biome                     None   \n",
       "1397  oceanic benthopelagic zone biome                     None   \n",
       "\n",
       "     longest_annotation_label  \n",
       "0           small river biome  \n",
       "1                marine biome  \n",
       "2                        None  \n",
       "3                    sediment  \n",
       "4             estuarine biome  \n",
       "...                       ...  \n",
       "1393          estuarine biome  \n",
       "1394          estuarine biome  \n",
       "1395                     None  \n",
       "1396                     None  \n",
       "1397                     None  \n",
       "\n",
       "[1398 rows x 11 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>serial_number</th>\n",
       "      <th>content</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>content_list</th>\n",
       "      <th>content_count</th>\n",
       "      <th>envo_count</th>\n",
       "      <th>extracted_label</th>\n",
       "      <th>extracted_curie</th>\n",
       "      <th>real_label</th>\n",
       "      <th>longest_annotation_curie</th>\n",
       "      <th>longest_annotation_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Small river biome</td>\n",
       "      <td>1819</td>\n",
       "      <td>Small river biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Small river biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000890</td>\n",
       "      <td>small river biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>1644</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>ENVO:01000047</td>\n",
       "      <td>805</td>\n",
       "      <td>ENVO:01000047</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:01000047</td>\n",
       "      <td>mediterranean sea biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>sediment</td>\n",
       "      <td>729</td>\n",
       "      <td>sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sediment</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>estuarine biome [ENVO:01000020]</td>\n",
       "      <td>669</td>\n",
       "      <td>estuarine biome [ENVO:01000020]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>estuarine biome</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>estuarine biome</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>estuarine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>1366</td>\n",
       "      <td>estuarine biome[Envo:01000031]</td>\n",
       "      <td>1</td>\n",
       "      <td>estuarine biome[Envo:01000031]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>estuarine biome[Envo:01000031</td>\n",
       "      <td>ENVO:01000031</td>\n",
       "      <td>obsolete marine feature</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>estuarine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>1367</td>\n",
       "      <td>estuarine biome[Envo:01000035]</td>\n",
       "      <td>1</td>\n",
       "      <td>estuarine biome[Envo:01000035]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>estuarine biome[Envo:01000035</td>\n",
       "      <td>ENVO:01000035</td>\n",
       "      <td>oceanic epipelagic zone biome</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>estuarine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>1368</td>\n",
       "      <td>ENVO:01000026</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO:01000026</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:01000026</td>\n",
       "      <td>marine bathyal zone biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1396</th>\n",
       "      <td>1369</td>\n",
       "      <td>ENVO:01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO:01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:01000035</td>\n",
       "      <td>oceanic epipelagic zone biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1397</th>\n",
       "      <td>1370</td>\n",
       "      <td>ENVO:01000040</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO:01000040</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:01000040</td>\n",
       "      <td>oceanic benthopelagic zone biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1398 rows × 11 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 83
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "----",
   "id": "a34f812b53b3b945"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:52:51.464164Z",
     "start_time": "2024-11-15T19:50:52.816977Z"
    }
   },
   "cell_type": "code",
   "source": [
    "gold_biosamples_frame = pd.read_excel(gold_data_url, sheet_name=BIOSAMPLES_SHEET)\n",
    "# 2 minutes"
   ],
   "id": "a66b83e1abdbe5c",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mark/.cache/pypoetry/virtualenvs/nmdc-submission-schema-DC6HKp4p-py3.10/lib/python3.10/site-packages/openpyxl/styles/stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default\n",
      "  warn(\"Workbook contains no default style, apply openpyxl's default\")\n"
     ]
    }
   ],
   "execution_count": 84
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:52:56.988886Z",
     "start_time": "2024-11-15T19:52:56.983064Z"
    }
   },
   "cell_type": "code",
   "source": [
    "gold_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'] = gold_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'].fillna(\n",
    "    0).astype(int)\n"
   ],
   "id": "e56753c932a76fc9",
   "outputs": [],
   "execution_count": 85
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:02.456059Z",
     "start_time": "2024-11-15T19:53:02.438831Z"
    }
   },
   "cell_type": "code",
   "source": "gold_biosamples_frame",
   "id": "6a4bc761aa25612e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "       BIOSAMPLE GOLD ID                                     BIOSAMPLE NAME  \\\n",
       "0              Gb0011929  GEBA_MDM Biosample from Great Boiling Spring, ...   \n",
       "1              Gb0035601  Compost enrichment cellulose adapted microbial...   \n",
       "2              Gb0035602  Compost enrichment cellulose adapted microbial...   \n",
       "3              Gb0035635  Compost enrichment cellulose adapted microbial...   \n",
       "4              Gb0035638  Compost enrichment cellulose adapted microbial...   \n",
       "...                  ...                                                ...   \n",
       "210873         Gb0405291  Freshwater biofilm microbial communities from ...   \n",
       "210874         Gb0405292  Freshwater biofilm microbial communities from ...   \n",
       "210875         Gb0405293  Freshwater microbial communities from Lake Sug...   \n",
       "210876         Gb0405294  Freshwater microbial communities from Prairie ...   \n",
       "210877         Gb0405295  Freshwater biofilm microbial communities from ...   \n",
       "\n",
       "        BIOSAMPLE NCBI TAX ID BIOSAMPLE NCBI TAX NAME  \\\n",
       "0                    749907.0     sediment metagenome   \n",
       "1                    702656.0      compost metagenome   \n",
       "2                    702656.0      compost metagenome   \n",
       "3                    702656.0      compost metagenome   \n",
       "4                    702656.0      compost metagenome   \n",
       "...                       ...                     ...   \n",
       "210873               718308.0      biofilm metagenome   \n",
       "210874               718308.0      biofilm metagenome   \n",
       "210875               449393.0   freshwater metagenome   \n",
       "210876               449393.0   freshwater metagenome   \n",
       "210877               718308.0      biofilm metagenome   \n",
       "\n",
       "       BIOSAMPLE SAMPLE COLLECTION SITE BIOSAMPLE SAMPLE COLLECTION DATE  \\\n",
       "0                              Sediment                              NaN   \n",
       "1                           single cell                       2012-10-31   \n",
       "2                           single cell                       2012-10-31   \n",
       "3                           single cell                       2012-10-31   \n",
       "4                           single cell                       2012-10-31   \n",
       "...                                 ...                              ...   \n",
       "210873                    creek biofilm                       2023-07-26   \n",
       "210874                    creek biofilm                       2023-07-26   \n",
       "210875                       lake water                       2023-08-01   \n",
       "210876                       lake water                       2023-08-08   \n",
       "210877                    creek biofilm                       2023-07-24   \n",
       "\n",
       "                            BIOSAMPLE GEOGRAPHIC LOCATION  BIOSAMPLE LATITUDE  \\\n",
       "0                      Great Boiling Spring (GBS), Nevada           40.661433   \n",
       "1       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "2       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "3       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "4       USA: California: Emeryville: Joint Bio-Energy ...           37.840800   \n",
       "...                                                   ...                 ...   \n",
       "210873   USA: Lewis Run NEON Field Site, Briggs, Virginia           39.095630   \n",
       "210874   USA: Lewis Run NEON Field Site, Briggs, Virginia           39.095630   \n",
       "210875  USA: Lake Suggs NEON Field Site, Melrose, Florida           29.688140   \n",
       "210876  USA: Prairie Lake NEON Field Site, Vashti, Nor...           47.159710   \n",
       "210877  USA: Kings Creek NEON Field Site, Manhattan, K...           39.105060   \n",
       "\n",
       "        BIOSAMPLE LONGITUDE  BIOSAMPLE ECOSYSTEM PATH ID BIOSAMPLE ECOSYSTEM  \\\n",
       "0               -119.366250                         3992       Environmental   \n",
       "1               -122.289600                         6039          Engineered   \n",
       "2               -122.289600                         6039          Engineered   \n",
       "3               -122.289600                         6039          Engineered   \n",
       "4               -122.289600                         6039          Engineered   \n",
       "...                     ...                          ...                 ...   \n",
       "210873           -77.983216                         8389       Environmental   \n",
       "210874           -77.983216                         8389       Environmental   \n",
       "210875           -82.017318                         4179       Environmental   \n",
       "210876           -99.118723                         4179       Environmental   \n",
       "210877           -96.603829                         8389       Environmental   \n",
       "\n",
       "       BIOSAMPLE ECOSYSTEM CATEGORY BIOSAMPLE ECOSYSTEM TYPE  \\\n",
       "0                           Aquatic          Thermal springs   \n",
       "1                       Solid waste              Green waste   \n",
       "2                       Solid waste              Green waste   \n",
       "3                       Solid waste              Green waste   \n",
       "4                       Solid waste              Green waste   \n",
       "...                             ...                      ...   \n",
       "210873                      Aquatic               Freshwater   \n",
       "210874                      Aquatic               Freshwater   \n",
       "210875                      Aquatic               Freshwater   \n",
       "210876                      Aquatic               Freshwater   \n",
       "210877                      Aquatic               Freshwater   \n",
       "\n",
       "       BIOSAMPLE ECOSYSTEM SUBTYPE BIOSAMPLE SPECIFIC ECOSYSTEM  \n",
       "0                     Hot (42-90C)                 Unclassified  \n",
       "1                       Composting                 Unclassified  \n",
       "2                       Composting                 Unclassified  \n",
       "3                       Composting                 Unclassified  \n",
       "4                       Composting                 Unclassified  \n",
       "...                            ...                          ...  \n",
       "210873                       Creek                      Biofilm  \n",
       "210874                       Creek                      Biofilm  \n",
       "210875                        Lake                 Unclassified  \n",
       "210876                        Lake                 Unclassified  \n",
       "210877                       Creek                      Biofilm  \n",
       "\n",
       "[210878 rows x 15 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOSAMPLE GOLD ID</th>\n",
       "      <th>BIOSAMPLE NAME</th>\n",
       "      <th>BIOSAMPLE NCBI TAX ID</th>\n",
       "      <th>BIOSAMPLE NCBI TAX NAME</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION SITE</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION DATE</th>\n",
       "      <th>BIOSAMPLE GEOGRAPHIC LOCATION</th>\n",
       "      <th>BIOSAMPLE LATITUDE</th>\n",
       "      <th>BIOSAMPLE LONGITUDE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM PATH ID</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM CATEGORY</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM TYPE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM SUBTYPE</th>\n",
       "      <th>BIOSAMPLE SPECIFIC ECOSYSTEM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gb0011929</td>\n",
       "      <td>GEBA_MDM Biosample from Great Boiling Spring, ...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Great Boiling Spring (GBS), Nevada</td>\n",
       "      <td>40.661433</td>\n",
       "      <td>-119.366250</td>\n",
       "      <td>3992</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal springs</td>\n",
       "      <td>Hot (42-90C)</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gb0035601</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gb0035602</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gb0035635</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gb0035638</td>\n",
       "      <td>Compost enrichment cellulose adapted microbial...</td>\n",
       "      <td>702656.0</td>\n",
       "      <td>compost metagenome</td>\n",
       "      <td>single cell</td>\n",
       "      <td>2012-10-31</td>\n",
       "      <td>USA: California: Emeryville: Joint Bio-Energy ...</td>\n",
       "      <td>37.840800</td>\n",
       "      <td>-122.289600</td>\n",
       "      <td>6039</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>Green waste</td>\n",
       "      <td>Composting</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210873</th>\n",
       "      <td>Gb0405291</td>\n",
       "      <td>Freshwater biofilm microbial communities from ...</td>\n",
       "      <td>718308.0</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "      <td>creek biofilm</td>\n",
       "      <td>2023-07-26</td>\n",
       "      <td>USA: Lewis Run NEON Field Site, Briggs, Virginia</td>\n",
       "      <td>39.095630</td>\n",
       "      <td>-77.983216</td>\n",
       "      <td>8389</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Creek</td>\n",
       "      <td>Biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210874</th>\n",
       "      <td>Gb0405292</td>\n",
       "      <td>Freshwater biofilm microbial communities from ...</td>\n",
       "      <td>718308.0</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "      <td>creek biofilm</td>\n",
       "      <td>2023-07-26</td>\n",
       "      <td>USA: Lewis Run NEON Field Site, Briggs, Virginia</td>\n",
       "      <td>39.095630</td>\n",
       "      <td>-77.983216</td>\n",
       "      <td>8389</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Creek</td>\n",
       "      <td>Biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210875</th>\n",
       "      <td>Gb0405293</td>\n",
       "      <td>Freshwater microbial communities from Lake Sug...</td>\n",
       "      <td>449393.0</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>lake water</td>\n",
       "      <td>2023-08-01</td>\n",
       "      <td>USA: Lake Suggs NEON Field Site, Melrose, Florida</td>\n",
       "      <td>29.688140</td>\n",
       "      <td>-82.017318</td>\n",
       "      <td>4179</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Lake</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210876</th>\n",
       "      <td>Gb0405294</td>\n",
       "      <td>Freshwater microbial communities from Prairie ...</td>\n",
       "      <td>449393.0</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>lake water</td>\n",
       "      <td>2023-08-08</td>\n",
       "      <td>USA: Prairie Lake NEON Field Site, Vashti, Nor...</td>\n",
       "      <td>47.159710</td>\n",
       "      <td>-99.118723</td>\n",
       "      <td>4179</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Lake</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210877</th>\n",
       "      <td>Gb0405295</td>\n",
       "      <td>Freshwater biofilm microbial communities from ...</td>\n",
       "      <td>718308.0</td>\n",
       "      <td>biofilm metagenome</td>\n",
       "      <td>creek biofilm</td>\n",
       "      <td>2023-07-24</td>\n",
       "      <td>USA: Kings Creek NEON Field Site, Manhattan, K...</td>\n",
       "      <td>39.105060</td>\n",
       "      <td>-96.603829</td>\n",
       "      <td>8389</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Creek</td>\n",
       "      <td>Biofilm</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210878 rows × 15 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 86
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:08.670929Z",
     "start_time": "2024-11-15T19:53:08.666382Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Determine the filenames and target directory\n",
    "goldterms_compressed_filename = urlparse(goldterms_semsql_url).path.split('/')[-1]\n",
    "goldterms_filename = os.path.splitext(goldterms_compressed_filename)[0]\n",
    "target_dir = os.path.join(\"..\", \"..\")  # Two levels up\n",
    "\n",
    "# Print to confirm the filenames\n",
    "print(goldterms_filename)"
   ],
   "id": "900a1ccb2e55da85",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "goldterms.db\n"
     ]
    }
   ],
   "execution_count": 87
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:09.157698Z",
     "start_time": "2024-11-15T19:53:08.728381Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Fetch the contents from the URL and save compressed file in target directory\n",
    "goldterms_response = requests.get(goldterms_semsql_url)\n",
    "goldterms_compressed_file_path = os.path.join(target_dir, goldterms_compressed_filename)\n",
    "with open(goldterms_compressed_file_path, \"wb\") as f:\n",
    "    f.write(goldterms_response.content)"
   ],
   "id": "ba5fb12553bf4825",
   "outputs": [],
   "execution_count": 88
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:09.222009Z",
     "start_time": "2024-11-15T19:53:09.161775Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Unzip the compressed file and save the extracted file in target directory\n",
    "goldterms_uncompressed_file_path = os.path.join(target_dir, goldterms_filename)\n",
    "with gzip.open(goldterms_compressed_file_path, \"rb\") as f_in:\n",
    "    with open(goldterms_uncompressed_file_path, \"wb\") as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)"
   ],
   "id": "f60c064a587dd407",
   "outputs": [],
   "execution_count": 89
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:14.895869Z",
     "start_time": "2024-11-15T19:53:14.892015Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_conn = sqlite3.connect(goldterms_uncompressed_file_path)",
   "id": "1f663ead96b88e75",
   "outputs": [],
   "execution_count": 90
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:20.357547Z",
     "start_time": "2024-11-15T19:53:20.351989Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_subjects = pd.read_sql_query(goldterms_subclass_query, goldterms_conn)",
   "id": "bcedfe9ddb7254fd",
   "outputs": [],
   "execution_count": 91
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:25.713582Z",
     "start_time": "2024-11-15T19:53:25.709264Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_subjects['path_id'] = goldterms_subjects['subject'].str.extract(r'GOLDTERMS:(\\d+)')",
   "id": "82a240e419eb635d",
   "outputs": [],
   "execution_count": 92
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:31.632523Z",
     "start_time": "2024-11-15T19:53:31.626395Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_subjects",
   "id": "8e90c8dfb6b820b4",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "          subject path_id\n",
       "0  GOLDTERMS:3985    3985"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>path_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 93
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:36.946927Z",
     "start_time": "2024-11-15T19:53:36.934106Z"
    }
   },
   "cell_type": "code",
   "source": [
    "gold_path_ids = goldterms_subjects['path_id'].dropna().unique().tolist()\n",
    "gold_path_ids = [int(id) for id in gold_path_ids]\n"
   ],
   "id": "9e70094c5c28b5ae",
   "outputs": [],
   "execution_count": 94
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:36.993072Z",
     "start_time": "2024-11-15T19:53:36.988817Z"
    }
   },
   "cell_type": "code",
   "source": [
    "gold_env_filtered_biosamples_frame = gold_biosamples_frame[\n",
    "    gold_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'].isin(gold_path_ids)]\n"
   ],
   "id": "d294ca1680c31685",
   "outputs": [],
   "execution_count": 95
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:37.061770Z",
     "start_time": "2024-11-15T19:53:37.044916Z"
    }
   },
   "cell_type": "code",
   "source": "gold_env_filtered_biosamples_frame",
   "id": "ea40acff447df533",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "       BIOSAMPLE GOLD ID                                     BIOSAMPLE NAME  \\\n",
       "7590           Gb0112951  Trench sediment microbial communities from the...   \n",
       "7893           Gb0115403  Microbial communities from Baker Bay sediment,...   \n",
       "7894           Gb0115404  Microbial communities from Youngs Bay mouth se...   \n",
       "7895           Gb0115405  Microbial communities from Youngs Bay mouth se...   \n",
       "7896           Gb0115406  Microbial communities from Cathlamet Bay sedim...   \n",
       "...                  ...                                                ...   \n",
       "88480          Gb0241798  Cryoconite hole sediment microbial community f...   \n",
       "89714          Gb0243078  Mudflat Water microbial community from Shangha...   \n",
       "135134         Gb0291122  Sediment microbial communites from ultra-acidi...   \n",
       "135949         Gb0292281  Sediment microbial communities from Churince l...   \n",
       "203760         Gb0377973  Sediment microbial communities from Los Azufre...   \n",
       "\n",
       "        BIOSAMPLE NCBI TAX ID     BIOSAMPLE NCBI TAX NAME  \\\n",
       "7590                 412755.0  marine sediment metagenome   \n",
       "7893                 749907.0         sediment metagenome   \n",
       "7894                 749907.0         sediment metagenome   \n",
       "7895                 749907.0         sediment metagenome   \n",
       "7896                 749907.0         sediment metagenome   \n",
       "...                       ...                         ...   \n",
       "88480                     NaN                         NaN   \n",
       "89714                     NaN                         NaN   \n",
       "135134               749908.0     Japonacteon nipponensis   \n",
       "135949               749907.0         sediment metagenome   \n",
       "203760               749907.0         sediment metagenome   \n",
       "\n",
       "                         BIOSAMPLE SAMPLE COLLECTION SITE  \\\n",
       "7590                                      Deep-Sea Trench   \n",
       "7893                                Baker Bay boat launch   \n",
       "7894                                     Youngs Bay mouth   \n",
       "7895                                      Youngs Bay back   \n",
       "7896                                        Cathlamet Bay   \n",
       "...                                                   ...   \n",
       "88480                            Cryoconite hole sediment   \n",
       "89714                                       Mudflat Water   \n",
       "135134  sediment from ultra-acidic sulfur-rich Arctic ...   \n",
       "135949                                   Aquatic sediment   \n",
       "203760                         Sediment around a fumarole   \n",
       "\n",
       "       BIOSAMPLE SAMPLE COLLECTION DATE  \\\n",
       "7590                                NaN   \n",
       "7893                         2011-08-22   \n",
       "7894                         2011-08-22   \n",
       "7895                         2011-08-22   \n",
       "7896                         2013-08-19   \n",
       "...                                 ...   \n",
       "88480                        2010-06-01   \n",
       "89714                        2015-10-07   \n",
       "135134                       2018-12-07   \n",
       "135949                          2011-02   \n",
       "203760                       2019-03-01   \n",
       "\n",
       "                       BIOSAMPLE GEOGRAPHIC LOCATION  BIOSAMPLE LATITUDE  \\\n",
       "7590                                   Pacific Ocean           29.000000   \n",
       "7893                         Ilwaco, Washington, USA           46.285510   \n",
       "7894                            USA: Astoria, Oregon           46.174690   \n",
       "7895                            USA: Astoria, Oregon           46.159680   \n",
       "7896                                   Cathlamet Bay           46.187490   \n",
       "...                                              ...                 ...   \n",
       "88480                       Greenland: Kangerlussuaq           67.161020   \n",
       "89714                                China: Shanghai           31.550000   \n",
       "135134  Smoking Hills, Northwest Territories, Canada           69.665243   \n",
       "135949             Mexico: Cuatro Cienegas, Coahuila           26.840370   \n",
       "203760                Mexico: Los Azufres, Michoacan           19.781583   \n",
       "\n",
       "        BIOSAMPLE LONGITUDE  BIOSAMPLE ECOSYSTEM PATH ID BIOSAMPLE ECOSYSTEM  \\\n",
       "7590             142.000000                         3985       Environmental   \n",
       "7893            -124.051870                         3985       Environmental   \n",
       "7894            -123.849330                         3985       Environmental   \n",
       "7895            -123.806510                         3985       Environmental   \n",
       "7896            -123.588920                         3985       Environmental   \n",
       "...                     ...                          ...                 ...   \n",
       "88480            -50.014630                         3985       Environmental   \n",
       "89714            121.950000                         3985       Environmental   \n",
       "135134          -126.579435                         3985       Environmental   \n",
       "135949          -102.134120                         3985       Environmental   \n",
       "203760          -100.658130                         3985       Environmental   \n",
       "\n",
       "       BIOSAMPLE ECOSYSTEM CATEGORY BIOSAMPLE ECOSYSTEM TYPE  \\\n",
       "7590                        Aquatic                 Sediment   \n",
       "7893                        Aquatic                 Sediment   \n",
       "7894                        Aquatic                 Sediment   \n",
       "7895                        Aquatic                 Sediment   \n",
       "7896                        Aquatic                 Sediment   \n",
       "...                             ...                      ...   \n",
       "88480                       Aquatic                 Sediment   \n",
       "89714                       Aquatic                 Sediment   \n",
       "135134                      Aquatic                 Sediment   \n",
       "135949                      Aquatic                 Sediment   \n",
       "203760                      Aquatic                 Sediment   \n",
       "\n",
       "       BIOSAMPLE ECOSYSTEM SUBTYPE BIOSAMPLE SPECIFIC ECOSYSTEM  \n",
       "7590                  Unclassified                 Unclassified  \n",
       "7893                  Unclassified                 Unclassified  \n",
       "7894                  Unclassified                 Unclassified  \n",
       "7895                  Unclassified                 Unclassified  \n",
       "7896                  Unclassified                 Unclassified  \n",
       "...                            ...                          ...  \n",
       "88480                 Unclassified                 Unclassified  \n",
       "89714                 Unclassified                 Unclassified  \n",
       "135134                Unclassified                 Unclassified  \n",
       "135949                Unclassified                 Unclassified  \n",
       "203760                Unclassified                 Unclassified  \n",
       "\n",
       "[88 rows x 15 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOSAMPLE GOLD ID</th>\n",
       "      <th>BIOSAMPLE NAME</th>\n",
       "      <th>BIOSAMPLE NCBI TAX ID</th>\n",
       "      <th>BIOSAMPLE NCBI TAX NAME</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION SITE</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION DATE</th>\n",
       "      <th>BIOSAMPLE GEOGRAPHIC LOCATION</th>\n",
       "      <th>BIOSAMPLE LATITUDE</th>\n",
       "      <th>BIOSAMPLE LONGITUDE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM PATH ID</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM CATEGORY</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM TYPE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM SUBTYPE</th>\n",
       "      <th>BIOSAMPLE SPECIFIC ECOSYSTEM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7590</th>\n",
       "      <td>Gb0112951</td>\n",
       "      <td>Trench sediment microbial communities from the...</td>\n",
       "      <td>412755.0</td>\n",
       "      <td>marine sediment metagenome</td>\n",
       "      <td>Deep-Sea Trench</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pacific Ocean</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>142.000000</td>\n",
       "      <td>3985</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7893</th>\n",
       "      <td>Gb0115403</td>\n",
       "      <td>Microbial communities from Baker Bay sediment,...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Baker Bay boat launch</td>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>Ilwaco, Washington, USA</td>\n",
       "      <td>46.285510</td>\n",
       "      <td>-124.051870</td>\n",
       "      <td>3985</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7894</th>\n",
       "      <td>Gb0115404</td>\n",
       "      <td>Microbial communities from Youngs Bay mouth se...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Youngs Bay mouth</td>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>USA: Astoria, Oregon</td>\n",
       "      <td>46.174690</td>\n",
       "      <td>-123.849330</td>\n",
       "      <td>3985</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7895</th>\n",
       "      <td>Gb0115405</td>\n",
       "      <td>Microbial communities from Youngs Bay mouth se...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Youngs Bay back</td>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>USA: Astoria, Oregon</td>\n",
       "      <td>46.159680</td>\n",
       "      <td>-123.806510</td>\n",
       "      <td>3985</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7896</th>\n",
       "      <td>Gb0115406</td>\n",
       "      <td>Microbial communities from Cathlamet Bay sedim...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Cathlamet Bay</td>\n",
       "      <td>2013-08-19</td>\n",
       "      <td>Cathlamet Bay</td>\n",
       "      <td>46.187490</td>\n",
       "      <td>-123.588920</td>\n",
       "      <td>3985</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88480</th>\n",
       "      <td>Gb0241798</td>\n",
       "      <td>Cryoconite hole sediment microbial community f...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cryoconite hole sediment</td>\n",
       "      <td>2010-06-01</td>\n",
       "      <td>Greenland: Kangerlussuaq</td>\n",
       "      <td>67.161020</td>\n",
       "      <td>-50.014630</td>\n",
       "      <td>3985</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89714</th>\n",
       "      <td>Gb0243078</td>\n",
       "      <td>Mudflat Water microbial community from Shangha...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mudflat Water</td>\n",
       "      <td>2015-10-07</td>\n",
       "      <td>China: Shanghai</td>\n",
       "      <td>31.550000</td>\n",
       "      <td>121.950000</td>\n",
       "      <td>3985</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135134</th>\n",
       "      <td>Gb0291122</td>\n",
       "      <td>Sediment microbial communites from ultra-acidi...</td>\n",
       "      <td>749908.0</td>\n",
       "      <td>Japonacteon nipponensis</td>\n",
       "      <td>sediment from ultra-acidic sulfur-rich Arctic ...</td>\n",
       "      <td>2018-12-07</td>\n",
       "      <td>Smoking Hills, Northwest Territories, Canada</td>\n",
       "      <td>69.665243</td>\n",
       "      <td>-126.579435</td>\n",
       "      <td>3985</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135949</th>\n",
       "      <td>Gb0292281</td>\n",
       "      <td>Sediment microbial communities from Churince l...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Aquatic sediment</td>\n",
       "      <td>2011-02</td>\n",
       "      <td>Mexico: Cuatro Cienegas, Coahuila</td>\n",
       "      <td>26.840370</td>\n",
       "      <td>-102.134120</td>\n",
       "      <td>3985</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203760</th>\n",
       "      <td>Gb0377973</td>\n",
       "      <td>Sediment microbial communities from Los Azufre...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Sediment around a fumarole</td>\n",
       "      <td>2019-03-01</td>\n",
       "      <td>Mexico: Los Azufres, Michoacan</td>\n",
       "      <td>19.781583</td>\n",
       "      <td>-100.658130</td>\n",
       "      <td>3985</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 15 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 96
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:42.547340Z",
     "start_time": "2024-11-15T19:53:42.540590Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_mixs_broad_frame = pd.read_sql_query(goldterms_envo_query, goldterms_conn)",
   "id": "4b4ed63d71bf1e65",
   "outputs": [],
   "execution_count": 97
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:42.838302Z",
     "start_time": "2024-11-15T19:53:42.602396Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_mixs_broad_frame['mixs_broad_label'] = goldterms_mixs_broad_frame['object'].apply(envo_adapter.label)",
   "id": "e736fe9475ea3d",
   "outputs": [],
   "execution_count": 98
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:42.847909Z",
     "start_time": "2024-11-15T19:53:42.842882Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_mixs_broad_frame['path_id'] = goldterms_mixs_broad_frame['subject'].str.extract(r'GOLDTERMS:(\\d+)')",
   "id": "ce93f9d692231ed8",
   "outputs": [],
   "execution_count": 99
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:47.977087Z",
     "start_time": "2024-11-15T19:53:47.965530Z"
    }
   },
   "cell_type": "code",
   "source": "goldterms_mixs_broad_frame",
   "id": "f0e775fd25205bcd",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                                stanza  \\\n",
       "0            GOLDTERMS:Engineered-Artificial-ecosystem   \n",
       "1    GOLDTERMS:Engineered-Artificial-ecosystem-Plan...   \n",
       "2    GOLDTERMS:Engineered-Artificial-ecosystem-Plan...   \n",
       "3    GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...   \n",
       "4          GOLDTERMS:Engineered-Bioreactor-DHS-reactor   \n",
       "..                                                 ...   \n",
       "643                                     GOLDTERMS:5841   \n",
       "644                                     GOLDTERMS:5843   \n",
       "645                                     GOLDTERMS:5846   \n",
       "646                                     GOLDTERMS:5849   \n",
       "647                                     GOLDTERMS:5850   \n",
       "\n",
       "                                               subject       predicate  \\\n",
       "0            GOLDTERMS:Engineered-Artificial-ecosystem  mixs:env_broad   \n",
       "1    GOLDTERMS:Engineered-Artificial-ecosystem-Plan...  mixs:env_broad   \n",
       "2    GOLDTERMS:Engineered-Artificial-ecosystem-Plan...  mixs:env_broad   \n",
       "3    GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...  mixs:env_broad   \n",
       "4          GOLDTERMS:Engineered-Bioreactor-DHS-reactor  mixs:env_broad   \n",
       "..                                                 ...             ...   \n",
       "643                                     GOLDTERMS:5841  mixs:env_broad   \n",
       "644                                     GOLDTERMS:5843  mixs:env_broad   \n",
       "645                                     GOLDTERMS:5846  mixs:env_broad   \n",
       "646                                     GOLDTERMS:5849  mixs:env_broad   \n",
       "647                                     GOLDTERMS:5850  mixs:env_broad   \n",
       "\n",
       "            object value datatype language graph           mixs_broad_label  \\\n",
       "0    ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "1    ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "2    ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "3    ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "4    ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "..             ...   ...      ...      ...   ...                        ...   \n",
       "643  ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "644  ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "645  ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "646  ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "647  ENVO:01000313  None     None     None  None  anthropogenic environment   \n",
       "\n",
       "    path_id  \n",
       "0       NaN  \n",
       "1       NaN  \n",
       "2       NaN  \n",
       "3       NaN  \n",
       "4       NaN  \n",
       "..      ...  \n",
       "643    5841  \n",
       "644    5843  \n",
       "645    5846  \n",
       "646    5849  \n",
       "647    5850  \n",
       "\n",
       "[648 rows x 10 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stanza</th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "      <th>value</th>\n",
       "      <th>datatype</th>\n",
       "      <th>language</th>\n",
       "      <th>graph</th>\n",
       "      <th>mixs_broad_label</th>\n",
       "      <th>path_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem</td>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem-Plan...</td>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem-Plan...</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem-Plan...</td>\n",
       "      <td>GOLDTERMS:Engineered-Artificial-ecosystem-Plan...</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...</td>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-Anaerobic-Soft...</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-DHS-reactor</td>\n",
       "      <td>GOLDTERMS:Engineered-Bioreactor-DHS-reactor</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>643</th>\n",
       "      <td>GOLDTERMS:5841</td>\n",
       "      <td>GOLDTERMS:5841</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>5841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>GOLDTERMS:5843</td>\n",
       "      <td>GOLDTERMS:5843</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>5843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>GOLDTERMS:5846</td>\n",
       "      <td>GOLDTERMS:5846</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>5846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>GOLDTERMS:5849</td>\n",
       "      <td>GOLDTERMS:5849</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>5849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>GOLDTERMS:5850</td>\n",
       "      <td>GOLDTERMS:5850</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000313</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>anthropogenic environment</td>\n",
       "      <td>5850</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>648 rows × 10 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 100
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:54.089747Z",
     "start_time": "2024-11-15T19:53:54.081691Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Fill NaN values in 'BIOSAMPLE ECOSYSTEM PATH ID' with 0 and convert to int\n",
    "gold_env_filtered_biosamples_frame['BIOSAMPLE ECOSYSTEM PATH ID'] = gold_env_filtered_biosamples_frame[\n",
    "    'BIOSAMPLE ECOSYSTEM PATH ID'].fillna(0).astype(int)\n",
    "\n",
    "# Drop rows with NaN in 'path_id' in goldterms_mixs_broad_frame\n",
    "goldterms_mixs_broad_frame = goldterms_mixs_broad_frame.dropna(subset=['path_id'])\n",
    "\n",
    "# Convert 'path_id' to int\n",
    "goldterms_mixs_broad_frame['path_id'] = goldterms_mixs_broad_frame['path_id'].astype(int)\n",
    "\n",
    "# Perform the left merge\n",
    "gold_env_filtered_biosamples_inferred_broad = gold_env_filtered_biosamples_frame.merge(\n",
    "    goldterms_mixs_broad_frame,\n",
    "    left_on='BIOSAMPLE ECOSYSTEM PATH ID',\n",
    "    right_on='path_id',\n",
    "    how='left'\n",
    ")\n"
   ],
   "id": "7a3f4269e0736579",
   "outputs": [],
   "execution_count": 101
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:54.176642Z",
     "start_time": "2024-11-15T19:53:54.156275Z"
    }
   },
   "cell_type": "code",
   "source": "gold_env_filtered_biosamples_inferred_broad",
   "id": "27e616ea4cbd27ab",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   BIOSAMPLE GOLD ID                                     BIOSAMPLE NAME  \\\n",
       "0          Gb0112951  Trench sediment microbial communities from the...   \n",
       "1          Gb0115403  Microbial communities from Baker Bay sediment,...   \n",
       "2          Gb0115404  Microbial communities from Youngs Bay mouth se...   \n",
       "3          Gb0115405  Microbial communities from Youngs Bay mouth se...   \n",
       "4          Gb0115406  Microbial communities from Cathlamet Bay sedim...   \n",
       "..               ...                                                ...   \n",
       "83         Gb0241798  Cryoconite hole sediment microbial community f...   \n",
       "84         Gb0243078  Mudflat Water microbial community from Shangha...   \n",
       "85         Gb0291122  Sediment microbial communites from ultra-acidi...   \n",
       "86         Gb0292281  Sediment microbial communities from Churince l...   \n",
       "87         Gb0377973  Sediment microbial communities from Los Azufre...   \n",
       "\n",
       "    BIOSAMPLE NCBI TAX ID     BIOSAMPLE NCBI TAX NAME  \\\n",
       "0                412755.0  marine sediment metagenome   \n",
       "1                749907.0         sediment metagenome   \n",
       "2                749907.0         sediment metagenome   \n",
       "3                749907.0         sediment metagenome   \n",
       "4                749907.0         sediment metagenome   \n",
       "..                    ...                         ...   \n",
       "83                    NaN                         NaN   \n",
       "84                    NaN                         NaN   \n",
       "85               749908.0     Japonacteon nipponensis   \n",
       "86               749907.0         sediment metagenome   \n",
       "87               749907.0         sediment metagenome   \n",
       "\n",
       "                     BIOSAMPLE SAMPLE COLLECTION SITE  \\\n",
       "0                                     Deep-Sea Trench   \n",
       "1                               Baker Bay boat launch   \n",
       "2                                    Youngs Bay mouth   \n",
       "3                                     Youngs Bay back   \n",
       "4                                       Cathlamet Bay   \n",
       "..                                                ...   \n",
       "83                           Cryoconite hole sediment   \n",
       "84                                      Mudflat Water   \n",
       "85  sediment from ultra-acidic sulfur-rich Arctic ...   \n",
       "86                                   Aquatic sediment   \n",
       "87                         Sediment around a fumarole   \n",
       "\n",
       "   BIOSAMPLE SAMPLE COLLECTION DATE  \\\n",
       "0                               NaN   \n",
       "1                        2011-08-22   \n",
       "2                        2011-08-22   \n",
       "3                        2011-08-22   \n",
       "4                        2013-08-19   \n",
       "..                              ...   \n",
       "83                       2010-06-01   \n",
       "84                       2015-10-07   \n",
       "85                       2018-12-07   \n",
       "86                          2011-02   \n",
       "87                       2019-03-01   \n",
       "\n",
       "                   BIOSAMPLE GEOGRAPHIC LOCATION  BIOSAMPLE LATITUDE  \\\n",
       "0                                  Pacific Ocean           29.000000   \n",
       "1                        Ilwaco, Washington, USA           46.285510   \n",
       "2                           USA: Astoria, Oregon           46.174690   \n",
       "3                           USA: Astoria, Oregon           46.159680   \n",
       "4                                  Cathlamet Bay           46.187490   \n",
       "..                                           ...                 ...   \n",
       "83                      Greenland: Kangerlussuaq           67.161020   \n",
       "84                               China: Shanghai           31.550000   \n",
       "85  Smoking Hills, Northwest Territories, Canada           69.665243   \n",
       "86             Mexico: Cuatro Cienegas, Coahuila           26.840370   \n",
       "87                Mexico: Los Azufres, Michoacan           19.781583   \n",
       "\n",
       "    BIOSAMPLE LONGITUDE  BIOSAMPLE ECOSYSTEM PATH ID  ...          stanza  \\\n",
       "0            142.000000                         3985  ...  GOLDTERMS:3985   \n",
       "1           -124.051870                         3985  ...  GOLDTERMS:3985   \n",
       "2           -123.849330                         3985  ...  GOLDTERMS:3985   \n",
       "3           -123.806510                         3985  ...  GOLDTERMS:3985   \n",
       "4           -123.588920                         3985  ...  GOLDTERMS:3985   \n",
       "..                  ...                          ...  ...             ...   \n",
       "83           -50.014630                         3985  ...  GOLDTERMS:3985   \n",
       "84           121.950000                         3985  ...  GOLDTERMS:3985   \n",
       "85          -126.579435                         3985  ...  GOLDTERMS:3985   \n",
       "86          -102.134120                         3985  ...  GOLDTERMS:3985   \n",
       "87          -100.658130                         3985  ...  GOLDTERMS:3985   \n",
       "\n",
       "           subject       predicate         object value datatype language  \\\n",
       "0   GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "1   GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "2   GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "3   GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "4   GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "..             ...             ...            ...   ...      ...      ...   \n",
       "83  GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "84  GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "85  GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "86  GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "87  GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "\n",
       "   graph      mixs_broad_label path_id  \n",
       "0   None  environmental system    3985  \n",
       "1   None  environmental system    3985  \n",
       "2   None  environmental system    3985  \n",
       "3   None  environmental system    3985  \n",
       "4   None  environmental system    3985  \n",
       "..   ...                   ...     ...  \n",
       "83  None  environmental system    3985  \n",
       "84  None  environmental system    3985  \n",
       "85  None  environmental system    3985  \n",
       "86  None  environmental system    3985  \n",
       "87  None  environmental system    3985  \n",
       "\n",
       "[88 rows x 25 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOSAMPLE GOLD ID</th>\n",
       "      <th>BIOSAMPLE NAME</th>\n",
       "      <th>BIOSAMPLE NCBI TAX ID</th>\n",
       "      <th>BIOSAMPLE NCBI TAX NAME</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION SITE</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION DATE</th>\n",
       "      <th>BIOSAMPLE GEOGRAPHIC LOCATION</th>\n",
       "      <th>BIOSAMPLE LATITUDE</th>\n",
       "      <th>BIOSAMPLE LONGITUDE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM PATH ID</th>\n",
       "      <th>...</th>\n",
       "      <th>stanza</th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "      <th>value</th>\n",
       "      <th>datatype</th>\n",
       "      <th>language</th>\n",
       "      <th>graph</th>\n",
       "      <th>mixs_broad_label</th>\n",
       "      <th>path_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gb0112951</td>\n",
       "      <td>Trench sediment microbial communities from the...</td>\n",
       "      <td>412755.0</td>\n",
       "      <td>marine sediment metagenome</td>\n",
       "      <td>Deep-Sea Trench</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pacific Ocean</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>142.000000</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gb0115403</td>\n",
       "      <td>Microbial communities from Baker Bay sediment,...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Baker Bay boat launch</td>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>Ilwaco, Washington, USA</td>\n",
       "      <td>46.285510</td>\n",
       "      <td>-124.051870</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gb0115404</td>\n",
       "      <td>Microbial communities from Youngs Bay mouth se...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Youngs Bay mouth</td>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>USA: Astoria, Oregon</td>\n",
       "      <td>46.174690</td>\n",
       "      <td>-123.849330</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gb0115405</td>\n",
       "      <td>Microbial communities from Youngs Bay mouth se...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Youngs Bay back</td>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>USA: Astoria, Oregon</td>\n",
       "      <td>46.159680</td>\n",
       "      <td>-123.806510</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gb0115406</td>\n",
       "      <td>Microbial communities from Cathlamet Bay sedim...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Cathlamet Bay</td>\n",
       "      <td>2013-08-19</td>\n",
       "      <td>Cathlamet Bay</td>\n",
       "      <td>46.187490</td>\n",
       "      <td>-123.588920</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Gb0241798</td>\n",
       "      <td>Cryoconite hole sediment microbial community f...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cryoconite hole sediment</td>\n",
       "      <td>2010-06-01</td>\n",
       "      <td>Greenland: Kangerlussuaq</td>\n",
       "      <td>67.161020</td>\n",
       "      <td>-50.014630</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>Gb0243078</td>\n",
       "      <td>Mudflat Water microbial community from Shangha...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mudflat Water</td>\n",
       "      <td>2015-10-07</td>\n",
       "      <td>China: Shanghai</td>\n",
       "      <td>31.550000</td>\n",
       "      <td>121.950000</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Gb0291122</td>\n",
       "      <td>Sediment microbial communites from ultra-acidi...</td>\n",
       "      <td>749908.0</td>\n",
       "      <td>Japonacteon nipponensis</td>\n",
       "      <td>sediment from ultra-acidic sulfur-rich Arctic ...</td>\n",
       "      <td>2018-12-07</td>\n",
       "      <td>Smoking Hills, Northwest Territories, Canada</td>\n",
       "      <td>69.665243</td>\n",
       "      <td>-126.579435</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Gb0292281</td>\n",
       "      <td>Sediment microbial communities from Churince l...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Aquatic sediment</td>\n",
       "      <td>2011-02</td>\n",
       "      <td>Mexico: Cuatro Cienegas, Coahuila</td>\n",
       "      <td>26.840370</td>\n",
       "      <td>-102.134120</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Gb0377973</td>\n",
       "      <td>Sediment microbial communities from Los Azufre...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Sediment around a fumarole</td>\n",
       "      <td>2019-03-01</td>\n",
       "      <td>Mexico: Los Azufres, Michoacan</td>\n",
       "      <td>19.781583</td>\n",
       "      <td>-100.658130</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 25 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 102
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "----",
   "id": "7e6055a76a17daf0"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:54.214452Z",
     "start_time": "2024-11-15T19:53:54.211521Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows = set()",
   "id": "74d28ee43bc2df13",
   "outputs": [],
   "execution_count": 103
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:54.279149Z",
     "start_time": "2024-11-15T19:53:54.275924Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update(biome_descendants_frame['curie'])",
   "id": "4a5858d51d55c1f3",
   "outputs": [],
   "execution_count": 104
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:54.368929Z",
     "start_time": "2024-11-15T19:53:54.365112Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update([i['curie'] for i in pv_validation_results['valids']])",
   "id": "3a206d349955474b",
   "outputs": [],
   "execution_count": 105
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:54.444394Z",
     "start_time": "2024-11-15T19:53:54.439996Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update(nmdc_biosample_contexts_frame['env_broad_scale_id'])",
   "id": "6612c4f580111c69",
   "outputs": [],
   "execution_count": 106
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:53:59.927272Z",
     "start_time": "2024-11-15T19:53:59.923689Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update(ncbi_frame['extracted_curie'])",
   "id": "4c9a01436eb4376",
   "outputs": [],
   "execution_count": 107
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:05.476141Z",
     "start_time": "2024-11-15T19:54:05.472712Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update(ncbi_frame['longest_annotation_curie'])",
   "id": "79fc8410c9ad2cb1",
   "outputs": [],
   "execution_count": 108
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:11.680240Z",
     "start_time": "2024-11-15T19:54:11.677328Z"
    }
   },
   "cell_type": "code",
   "source": "include_in_rows.update(gold_env_filtered_biosamples_inferred_broad['object'])",
   "id": "3a349c2bc9dc144d",
   "outputs": [],
   "execution_count": 109
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:16.982603Z",
     "start_time": "2024-11-15T19:54:16.979859Z"
    }
   },
   "cell_type": "code",
   "source": "rows_lod = []",
   "id": "8cc32502bceef8ea",
   "outputs": [],
   "execution_count": 110
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:22.687803Z",
     "start_time": "2024-11-15T19:54:22.318825Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# TODO MOVE THESE UP, because the expressions are already being used above\n",
    "\n",
    "biome_curies = list(biome_descendants_frame['curie'])\n",
    "legacy_pv_curies = [i['curie'] for i in pv_validation_results['valids']]\n",
    "terrestrial_biome_curies = list(envo_adapter.descendants('ENVO:00000446', predicates=[IS_A]))\n",
    "aquatic_biome_curies = list(envo_adapter.descendants('ENVO:00002030', predicates=[IS_A]))\n",
    "abp_curies = list(envo_adapter.descendants('ENVO:01000813', predicates=[IS_A]))\n",
    "env_sys_curies = list(envo_adapter.descendants('ENVO:01000254', predicates=[IS_A]))\n",
    "env_mat_curies = list(envo_adapter.descendants('ENVO:00010483', predicates=[IS_A]))\n",
    "obsoletes_curies = list(envo_adapter.obsoletes())\n",
    "\n",
    "for curie in include_in_rows:\n",
    "    if curie is None:\n",
    "        continue\n",
    "    row = {\n",
    "        'curie': curie,\n",
    "        'label': envo_adapter.label(curie),\n",
    "        'envo_native': False,\n",
    "        'obsolete': False,\n",
    "        'legacy_pv': False,\n",
    "        'abp': False,\n",
    "        'env_sys': False,\n",
    "        'biome': False,\n",
    "        'terrestrial_biome': False,\n",
    "        'aquatic_biome': False,\n",
    "        'env_mat': False,\n",
    "    }\n",
    "    prefix, local_id = curie.split(':')\n",
    "    if prefix and prefix == 'ENVO' and row['label'] is not None:\n",
    "        row['envo_native'] = True\n",
    "    if curie in biome_curies:\n",
    "        row['biome'] = True\n",
    "    if curie in terrestrial_biome_curies:\n",
    "        row['terrestrial_biome'] = True\n",
    "    if curie in aquatic_biome_curies:\n",
    "        row['aquatic_biome'] = True\n",
    "    if curie in abp_curies:\n",
    "        row['abp'] = True\n",
    "    if curie in env_sys_curies:\n",
    "        row['env_sys'] = True\n",
    "    if curie in env_mat_curies:\n",
    "        row['env_mat'] = True\n",
    "    if curie in legacy_pv_curies:\n",
    "        row['legacy_pv'] = True\n",
    "    if curie in obsoletes_curies:\n",
    "        row['obsolete'] = True\n",
    "    rows_lod.append(row)\n",
    "\n",
    "# todo terrestrial biome, aquatic biome, ABP, environmental material\n"
   ],
   "id": "2a88bc28334675b9",
   "outputs": [],
   "execution_count": 111
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:28.223441Z",
     "start_time": "2024-11-15T19:54:28.206913Z"
    }
   },
   "cell_type": "code",
   "source": "rows_frame = pd.DataFrame(rows_lod)",
   "id": "e3c1c8acd0d86bd8",
   "outputs": [],
   "execution_count": 112
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:33.892317Z",
     "start_time": "2024-11-15T19:54:33.880277Z"
    }
   },
   "cell_type": "code",
   "source": "rows_frame",
   "id": "1b64138f2a698553",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "              curie                      label  envo_native  obsolete  \\\n",
       "0     ENVO:01000052  marine white smoker biome         True     False   \n",
       "1     ENVO:00000111              forested area         True     False   \n",
       "2     ENVO:00000020                       lake         True     False   \n",
       "3    UBERON:0001988                      feces        False     False   \n",
       "4     ENVO:01001830             tropical biome         True     False   \n",
       "..              ...                        ...          ...       ...   \n",
       "321   ENVO:00000053           freshwater marsh         True     False   \n",
       "322  NCBITaxon:2157                    Archaea        False     False   \n",
       "323   ENVO:03000033            marine sediment         True     False   \n",
       "324   ENVO:01001835               alpine biome         True     False   \n",
       "325   ENVO:03500005       anthropogenic litter         True     False   \n",
       "\n",
       "     legacy_pv    abp  env_sys  biome  terrestrial_biome  aquatic_biome  \\\n",
       "0        False   True     True   True              False           True   \n",
       "1        False   True    False  False              False          False   \n",
       "2        False   True    False  False              False          False   \n",
       "3        False  False    False  False              False          False   \n",
       "4        False   True     True   True              False          False   \n",
       "..         ...    ...      ...    ...                ...            ...   \n",
       "321      False  False     True  False              False          False   \n",
       "322      False  False    False  False              False          False   \n",
       "323      False  False    False  False              False          False   \n",
       "324      False   True     True   True              False          False   \n",
       "325      False  False    False  False              False          False   \n",
       "\n",
       "     env_mat  \n",
       "0      False  \n",
       "1      False  \n",
       "2      False  \n",
       "3      False  \n",
       "4      False  \n",
       "..       ...  \n",
       "321    False  \n",
       "322    False  \n",
       "323     True  \n",
       "324    False  \n",
       "325    False  \n",
       "\n",
       "[326 rows x 11 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>label</th>\n",
       "      <th>envo_native</th>\n",
       "      <th>obsolete</th>\n",
       "      <th>legacy_pv</th>\n",
       "      <th>abp</th>\n",
       "      <th>env_sys</th>\n",
       "      <th>biome</th>\n",
       "      <th>terrestrial_biome</th>\n",
       "      <th>aquatic_biome</th>\n",
       "      <th>env_mat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:01000052</td>\n",
       "      <td>marine white smoker biome</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:00000111</td>\n",
       "      <td>forested area</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:00000020</td>\n",
       "      <td>lake</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UBERON:0001988</td>\n",
       "      <td>feces</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:01001830</td>\n",
       "      <td>tropical biome</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>ENVO:00000053</td>\n",
       "      <td>freshwater marsh</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>NCBITaxon:2157</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>ENVO:01001835</td>\n",
       "      <td>alpine biome</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>ENVO:03500005</td>\n",
       "      <td>anthropogenic litter</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>326 rows × 11 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 113
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:40.088158Z",
     "start_time": "2024-11-15T19:54:40.084423Z"
    }
   },
   "cell_type": "code",
   "source": [
    "nmdc_biosample_ebs_counts = nmdc_biosample_contexts_frame['env_broad_scale_id'].value_counts().reset_index()\n",
    "nmdc_biosample_ebs_counts.columns = ['curie', 'nmdc_ebs_count']\n"
   ],
   "id": "8392ef84a3254d56",
   "outputs": [],
   "execution_count": 114
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:40.149561Z",
     "start_time": "2024-11-15T19:54:40.144315Z"
    }
   },
   "cell_type": "code",
   "source": "nmdc_biosample_ebs_counts",
   "id": "fd4a155b6cb06235",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [curie, nmdc_ebs_count]\n",
       "Index: []"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>nmdc_ebs_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 115
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:40.204522Z",
     "start_time": "2024-11-15T19:54:40.199634Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Perform the left merge\n",
    "rows_frame = rows_frame.merge(\n",
    "    nmdc_biosample_ebs_counts,\n",
    "    left_on='curie',\n",
    "    right_on='curie',\n",
    "    how='left'\n",
    ")"
   ],
   "id": "3bd7ce9b1a42f70a",
   "outputs": [],
   "execution_count": 116
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:40.311865Z",
     "start_time": "2024-11-15T19:54:40.291247Z"
    }
   },
   "cell_type": "code",
   "source": "gold_env_filtered_biosamples_inferred_broad",
   "id": "335b3ab91d029ba",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   BIOSAMPLE GOLD ID                                     BIOSAMPLE NAME  \\\n",
       "0          Gb0112951  Trench sediment microbial communities from the...   \n",
       "1          Gb0115403  Microbial communities from Baker Bay sediment,...   \n",
       "2          Gb0115404  Microbial communities from Youngs Bay mouth se...   \n",
       "3          Gb0115405  Microbial communities from Youngs Bay mouth se...   \n",
       "4          Gb0115406  Microbial communities from Cathlamet Bay sedim...   \n",
       "..               ...                                                ...   \n",
       "83         Gb0241798  Cryoconite hole sediment microbial community f...   \n",
       "84         Gb0243078  Mudflat Water microbial community from Shangha...   \n",
       "85         Gb0291122  Sediment microbial communites from ultra-acidi...   \n",
       "86         Gb0292281  Sediment microbial communities from Churince l...   \n",
       "87         Gb0377973  Sediment microbial communities from Los Azufre...   \n",
       "\n",
       "    BIOSAMPLE NCBI TAX ID     BIOSAMPLE NCBI TAX NAME  \\\n",
       "0                412755.0  marine sediment metagenome   \n",
       "1                749907.0         sediment metagenome   \n",
       "2                749907.0         sediment metagenome   \n",
       "3                749907.0         sediment metagenome   \n",
       "4                749907.0         sediment metagenome   \n",
       "..                    ...                         ...   \n",
       "83                    NaN                         NaN   \n",
       "84                    NaN                         NaN   \n",
       "85               749908.0     Japonacteon nipponensis   \n",
       "86               749907.0         sediment metagenome   \n",
       "87               749907.0         sediment metagenome   \n",
       "\n",
       "                     BIOSAMPLE SAMPLE COLLECTION SITE  \\\n",
       "0                                     Deep-Sea Trench   \n",
       "1                               Baker Bay boat launch   \n",
       "2                                    Youngs Bay mouth   \n",
       "3                                     Youngs Bay back   \n",
       "4                                       Cathlamet Bay   \n",
       "..                                                ...   \n",
       "83                           Cryoconite hole sediment   \n",
       "84                                      Mudflat Water   \n",
       "85  sediment from ultra-acidic sulfur-rich Arctic ...   \n",
       "86                                   Aquatic sediment   \n",
       "87                         Sediment around a fumarole   \n",
       "\n",
       "   BIOSAMPLE SAMPLE COLLECTION DATE  \\\n",
       "0                               NaN   \n",
       "1                        2011-08-22   \n",
       "2                        2011-08-22   \n",
       "3                        2011-08-22   \n",
       "4                        2013-08-19   \n",
       "..                              ...   \n",
       "83                       2010-06-01   \n",
       "84                       2015-10-07   \n",
       "85                       2018-12-07   \n",
       "86                          2011-02   \n",
       "87                       2019-03-01   \n",
       "\n",
       "                   BIOSAMPLE GEOGRAPHIC LOCATION  BIOSAMPLE LATITUDE  \\\n",
       "0                                  Pacific Ocean           29.000000   \n",
       "1                        Ilwaco, Washington, USA           46.285510   \n",
       "2                           USA: Astoria, Oregon           46.174690   \n",
       "3                           USA: Astoria, Oregon           46.159680   \n",
       "4                                  Cathlamet Bay           46.187490   \n",
       "..                                           ...                 ...   \n",
       "83                      Greenland: Kangerlussuaq           67.161020   \n",
       "84                               China: Shanghai           31.550000   \n",
       "85  Smoking Hills, Northwest Territories, Canada           69.665243   \n",
       "86             Mexico: Cuatro Cienegas, Coahuila           26.840370   \n",
       "87                Mexico: Los Azufres, Michoacan           19.781583   \n",
       "\n",
       "    BIOSAMPLE LONGITUDE  BIOSAMPLE ECOSYSTEM PATH ID  ...          stanza  \\\n",
       "0            142.000000                         3985  ...  GOLDTERMS:3985   \n",
       "1           -124.051870                         3985  ...  GOLDTERMS:3985   \n",
       "2           -123.849330                         3985  ...  GOLDTERMS:3985   \n",
       "3           -123.806510                         3985  ...  GOLDTERMS:3985   \n",
       "4           -123.588920                         3985  ...  GOLDTERMS:3985   \n",
       "..                  ...                          ...  ...             ...   \n",
       "83           -50.014630                         3985  ...  GOLDTERMS:3985   \n",
       "84           121.950000                         3985  ...  GOLDTERMS:3985   \n",
       "85          -126.579435                         3985  ...  GOLDTERMS:3985   \n",
       "86          -102.134120                         3985  ...  GOLDTERMS:3985   \n",
       "87          -100.658130                         3985  ...  GOLDTERMS:3985   \n",
       "\n",
       "           subject       predicate         object value datatype language  \\\n",
       "0   GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "1   GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "2   GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "3   GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "4   GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "..             ...             ...            ...   ...      ...      ...   \n",
       "83  GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "84  GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "85  GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "86  GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "87  GOLDTERMS:3985  mixs:env_broad  ENVO:01000254  None     None     None   \n",
       "\n",
       "   graph      mixs_broad_label path_id  \n",
       "0   None  environmental system    3985  \n",
       "1   None  environmental system    3985  \n",
       "2   None  environmental system    3985  \n",
       "3   None  environmental system    3985  \n",
       "4   None  environmental system    3985  \n",
       "..   ...                   ...     ...  \n",
       "83  None  environmental system    3985  \n",
       "84  None  environmental system    3985  \n",
       "85  None  environmental system    3985  \n",
       "86  None  environmental system    3985  \n",
       "87  None  environmental system    3985  \n",
       "\n",
       "[88 rows x 25 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOSAMPLE GOLD ID</th>\n",
       "      <th>BIOSAMPLE NAME</th>\n",
       "      <th>BIOSAMPLE NCBI TAX ID</th>\n",
       "      <th>BIOSAMPLE NCBI TAX NAME</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION SITE</th>\n",
       "      <th>BIOSAMPLE SAMPLE COLLECTION DATE</th>\n",
       "      <th>BIOSAMPLE GEOGRAPHIC LOCATION</th>\n",
       "      <th>BIOSAMPLE LATITUDE</th>\n",
       "      <th>BIOSAMPLE LONGITUDE</th>\n",
       "      <th>BIOSAMPLE ECOSYSTEM PATH ID</th>\n",
       "      <th>...</th>\n",
       "      <th>stanza</th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "      <th>value</th>\n",
       "      <th>datatype</th>\n",
       "      <th>language</th>\n",
       "      <th>graph</th>\n",
       "      <th>mixs_broad_label</th>\n",
       "      <th>path_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gb0112951</td>\n",
       "      <td>Trench sediment microbial communities from the...</td>\n",
       "      <td>412755.0</td>\n",
       "      <td>marine sediment metagenome</td>\n",
       "      <td>Deep-Sea Trench</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pacific Ocean</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>142.000000</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gb0115403</td>\n",
       "      <td>Microbial communities from Baker Bay sediment,...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Baker Bay boat launch</td>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>Ilwaco, Washington, USA</td>\n",
       "      <td>46.285510</td>\n",
       "      <td>-124.051870</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gb0115404</td>\n",
       "      <td>Microbial communities from Youngs Bay mouth se...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Youngs Bay mouth</td>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>USA: Astoria, Oregon</td>\n",
       "      <td>46.174690</td>\n",
       "      <td>-123.849330</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gb0115405</td>\n",
       "      <td>Microbial communities from Youngs Bay mouth se...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Youngs Bay back</td>\n",
       "      <td>2011-08-22</td>\n",
       "      <td>USA: Astoria, Oregon</td>\n",
       "      <td>46.159680</td>\n",
       "      <td>-123.806510</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gb0115406</td>\n",
       "      <td>Microbial communities from Cathlamet Bay sedim...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Cathlamet Bay</td>\n",
       "      <td>2013-08-19</td>\n",
       "      <td>Cathlamet Bay</td>\n",
       "      <td>46.187490</td>\n",
       "      <td>-123.588920</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Gb0241798</td>\n",
       "      <td>Cryoconite hole sediment microbial community f...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cryoconite hole sediment</td>\n",
       "      <td>2010-06-01</td>\n",
       "      <td>Greenland: Kangerlussuaq</td>\n",
       "      <td>67.161020</td>\n",
       "      <td>-50.014630</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>Gb0243078</td>\n",
       "      <td>Mudflat Water microbial community from Shangha...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mudflat Water</td>\n",
       "      <td>2015-10-07</td>\n",
       "      <td>China: Shanghai</td>\n",
       "      <td>31.550000</td>\n",
       "      <td>121.950000</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Gb0291122</td>\n",
       "      <td>Sediment microbial communites from ultra-acidi...</td>\n",
       "      <td>749908.0</td>\n",
       "      <td>Japonacteon nipponensis</td>\n",
       "      <td>sediment from ultra-acidic sulfur-rich Arctic ...</td>\n",
       "      <td>2018-12-07</td>\n",
       "      <td>Smoking Hills, Northwest Territories, Canada</td>\n",
       "      <td>69.665243</td>\n",
       "      <td>-126.579435</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Gb0292281</td>\n",
       "      <td>Sediment microbial communities from Churince l...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Aquatic sediment</td>\n",
       "      <td>2011-02</td>\n",
       "      <td>Mexico: Cuatro Cienegas, Coahuila</td>\n",
       "      <td>26.840370</td>\n",
       "      <td>-102.134120</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Gb0377973</td>\n",
       "      <td>Sediment microbial communities from Los Azufre...</td>\n",
       "      <td>749907.0</td>\n",
       "      <td>sediment metagenome</td>\n",
       "      <td>Sediment around a fumarole</td>\n",
       "      <td>2019-03-01</td>\n",
       "      <td>Mexico: Los Azufres, Michoacan</td>\n",
       "      <td>19.781583</td>\n",
       "      <td>-100.658130</td>\n",
       "      <td>3985</td>\n",
       "      <td>...</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>GOLDTERMS:3985</td>\n",
       "      <td>mixs:env_broad</td>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>environmental system</td>\n",
       "      <td>3985</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 25 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 117
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:40.421645Z",
     "start_time": "2024-11-15T19:54:40.417538Z"
    }
   },
   "cell_type": "code",
   "source": [
    "gold_biosample_ebs_counts = gold_env_filtered_biosamples_inferred_broad['object'].value_counts().reset_index()\n",
    "gold_biosample_ebs_counts.columns = ['curie', 'gold_ebs_count']"
   ],
   "id": "c7a4d5fd716efd3c",
   "outputs": [],
   "execution_count": 118
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:40.482917Z",
     "start_time": "2024-11-15T19:54:40.475791Z"
    }
   },
   "cell_type": "code",
   "source": "gold_biosample_ebs_counts",
   "id": "da2433c6d19907a1",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "           curie  gold_ebs_count\n",
       "0  ENVO:01000254              88"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>gold_ebs_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:01000254</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 119
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:40.854243Z",
     "start_time": "2024-11-15T19:54:40.839023Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Perform the left merge\n",
    "rows_frame = rows_frame.merge(\n",
    "    gold_biosample_ebs_counts,\n",
    "    left_on='curie',\n",
    "    right_on='curie',\n",
    "    how='left'\n",
    ")"
   ],
   "id": "f5c673de7e5fd32a",
   "outputs": [],
   "execution_count": 120
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:40.914363Z",
     "start_time": "2024-11-15T19:54:40.899526Z"
    }
   },
   "cell_type": "code",
   "source": "rows_frame",
   "id": "cdde35e2977eaf9a",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "              curie                      label  envo_native  obsolete  \\\n",
       "0     ENVO:01000052  marine white smoker biome         True     False   \n",
       "1     ENVO:00000111              forested area         True     False   \n",
       "2     ENVO:00000020                       lake         True     False   \n",
       "3    UBERON:0001988                      feces        False     False   \n",
       "4     ENVO:01001830             tropical biome         True     False   \n",
       "..              ...                        ...          ...       ...   \n",
       "321   ENVO:00000053           freshwater marsh         True     False   \n",
       "322  NCBITaxon:2157                    Archaea        False     False   \n",
       "323   ENVO:03000033            marine sediment         True     False   \n",
       "324   ENVO:01001835               alpine biome         True     False   \n",
       "325   ENVO:03500005       anthropogenic litter         True     False   \n",
       "\n",
       "     legacy_pv    abp  env_sys  biome  terrestrial_biome  aquatic_biome  \\\n",
       "0        False   True     True   True              False           True   \n",
       "1        False   True    False  False              False          False   \n",
       "2        False   True    False  False              False          False   \n",
       "3        False  False    False  False              False          False   \n",
       "4        False   True     True   True              False          False   \n",
       "..         ...    ...      ...    ...                ...            ...   \n",
       "321      False  False     True  False              False          False   \n",
       "322      False  False    False  False              False          False   \n",
       "323      False  False    False  False              False          False   \n",
       "324      False   True     True   True              False          False   \n",
       "325      False  False    False  False              False          False   \n",
       "\n",
       "     env_mat  nmdc_ebs_count  gold_ebs_count  \n",
       "0      False             NaN             NaN  \n",
       "1      False             NaN             NaN  \n",
       "2      False             NaN             NaN  \n",
       "3      False             NaN             NaN  \n",
       "4      False             NaN             NaN  \n",
       "..       ...             ...             ...  \n",
       "321    False             NaN             NaN  \n",
       "322    False             NaN             NaN  \n",
       "323     True             NaN             NaN  \n",
       "324    False             NaN             NaN  \n",
       "325    False             NaN             NaN  \n",
       "\n",
       "[326 rows x 13 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>label</th>\n",
       "      <th>envo_native</th>\n",
       "      <th>obsolete</th>\n",
       "      <th>legacy_pv</th>\n",
       "      <th>abp</th>\n",
       "      <th>env_sys</th>\n",
       "      <th>biome</th>\n",
       "      <th>terrestrial_biome</th>\n",
       "      <th>aquatic_biome</th>\n",
       "      <th>env_mat</th>\n",
       "      <th>nmdc_ebs_count</th>\n",
       "      <th>gold_ebs_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:01000052</td>\n",
       "      <td>marine white smoker biome</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:00000111</td>\n",
       "      <td>forested area</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:00000020</td>\n",
       "      <td>lake</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UBERON:0001988</td>\n",
       "      <td>feces</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:01001830</td>\n",
       "      <td>tropical biome</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>ENVO:00000053</td>\n",
       "      <td>freshwater marsh</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>NCBITaxon:2157</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>ENVO:01001835</td>\n",
       "      <td>alpine biome</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>ENVO:03500005</td>\n",
       "      <td>anthropogenic litter</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>326 rows × 13 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 121
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:40.932781Z",
     "start_time": "2024-11-15T19:54:40.929610Z"
    }
   },
   "cell_type": "code",
   "source": "# 990 rows in https://docs.google.com/spreadsheets/d/12WH3eduBq2qSTy9zVF3n7fyajn6ssLZL/edit?gid=546570706#gid=546570706",
   "id": "9baa9d73af8159ca",
   "outputs": [],
   "execution_count": 122
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:41.114640Z",
     "start_time": "2024-11-15T19:54:41.111484Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# gold and ncbi counts are slightly trickier\n",
    "# for gold may want to include presence or mapping in goldterms in addition to biosamples counts\n",
    "# ncbi: we have extracted curies and annotated curies"
   ],
   "id": "460de8efcd09385d",
   "outputs": [],
   "execution_count": 123
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:41.265274Z",
     "start_time": "2024-11-15T19:54:41.242772Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# todo move this stuff up to immediately after the creation of ncbi_frame ?\n",
    "\n",
    "# todo don't accept extracted curie if no real label?\n",
    "# any kind of string similarity checking for label of annotated curie vs extracted label ?\n",
    "# look for long runs of curies?\n",
    "# can we measure the beneficial impact of any of this? current crux: how to distribute counts\n",
    "\n",
    "ncbi_frame['curie_list'] = ncbi_frame.apply(\n",
    "    lambda row: list({row['extracted_curie'], row['longest_annotation_curie']} - {None}),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "ncbi_frame['unique_curie_count'] = ncbi_frame['curie_list'].apply(len)"
   ],
   "id": "f416a6fe17c57df8",
   "outputs": [],
   "execution_count": 124
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:41.322846Z",
     "start_time": "2024-11-15T19:54:41.305835Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame",
   "id": "8e61714bd08518e1",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      serial_number                          content  sample_count  \\\n",
       "0                 1                Small river biome          1819   \n",
       "1                 2                     marine biome          1644   \n",
       "2                 3                    ENVO:01000047           805   \n",
       "3                 4                         sediment           729   \n",
       "4                 5  estuarine biome [ENVO:01000020]           669   \n",
       "...             ...                              ...           ...   \n",
       "1393           1366   estuarine biome[Envo:01000031]             1   \n",
       "1394           1367   estuarine biome[Envo:01000035]             1   \n",
       "1395           1368                    ENVO:01000026             1   \n",
       "1396           1369                    ENVO:01000035             1   \n",
       "1397           1370                    ENVO:01000040             1   \n",
       "\n",
       "                         content_list  content_count  envo_count  \\\n",
       "0                   Small river biome              1           0   \n",
       "1                        marine biome              1           0   \n",
       "2                       ENVO:01000047              1           1   \n",
       "3                            sediment              1           0   \n",
       "4     estuarine biome [ENVO:01000020]              1           1   \n",
       "...                               ...            ...         ...   \n",
       "1393   estuarine biome[Envo:01000031]              1           1   \n",
       "1394   estuarine biome[Envo:01000035]              1           1   \n",
       "1395                    ENVO:01000026              1           1   \n",
       "1396                    ENVO:01000035              1           1   \n",
       "1397                    ENVO:01000040              1           1   \n",
       "\n",
       "                    extracted_label extracted_curie  \\\n",
       "0                 Small river biome            None   \n",
       "1                      marine biome            None   \n",
       "2                                     ENVO:01000047   \n",
       "3                          sediment            None   \n",
       "4                   estuarine biome   ENVO:01000020   \n",
       "...                             ...             ...   \n",
       "1393  estuarine biome[Envo:01000031   ENVO:01000031   \n",
       "1394  estuarine biome[Envo:01000035   ENVO:01000035   \n",
       "1395                                  ENVO:01000026   \n",
       "1396                                  ENVO:01000035   \n",
       "1397                                  ENVO:01000040   \n",
       "\n",
       "                            real_label longest_annotation_curie  \\\n",
       "0                                 None            ENVO:00000890   \n",
       "1                                 None            ENVO:00000447   \n",
       "2              mediterranean sea biome                     None   \n",
       "3                                 None            ENVO:00002007   \n",
       "4                      estuarine biome            ENVO:01000020   \n",
       "...                                ...                      ...   \n",
       "1393           obsolete marine feature            ENVO:01000020   \n",
       "1394     oceanic epipelagic zone biome            ENVO:01000020   \n",
       "1395         marine bathyal zone biome                     None   \n",
       "1396     oceanic epipelagic zone biome                     None   \n",
       "1397  oceanic benthopelagic zone biome                     None   \n",
       "\n",
       "     longest_annotation_label                      curie_list  \\\n",
       "0           small river biome                 [ENVO:00000890]   \n",
       "1                marine biome                 [ENVO:00000447]   \n",
       "2                        None                 [ENVO:01000047]   \n",
       "3                    sediment                 [ENVO:00002007]   \n",
       "4             estuarine biome                 [ENVO:01000020]   \n",
       "...                       ...                             ...   \n",
       "1393          estuarine biome  [ENVO:01000031, ENVO:01000020]   \n",
       "1394          estuarine biome  [ENVO:01000035, ENVO:01000020]   \n",
       "1395                     None                 [ENVO:01000026]   \n",
       "1396                     None                 [ENVO:01000035]   \n",
       "1397                     None                 [ENVO:01000040]   \n",
       "\n",
       "      unique_curie_count  \n",
       "0                      1  \n",
       "1                      1  \n",
       "2                      1  \n",
       "3                      1  \n",
       "4                      1  \n",
       "...                  ...  \n",
       "1393                   2  \n",
       "1394                   2  \n",
       "1395                   1  \n",
       "1396                   1  \n",
       "1397                   1  \n",
       "\n",
       "[1398 rows x 13 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>serial_number</th>\n",
       "      <th>content</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>content_list</th>\n",
       "      <th>content_count</th>\n",
       "      <th>envo_count</th>\n",
       "      <th>extracted_label</th>\n",
       "      <th>extracted_curie</th>\n",
       "      <th>real_label</th>\n",
       "      <th>longest_annotation_curie</th>\n",
       "      <th>longest_annotation_label</th>\n",
       "      <th>curie_list</th>\n",
       "      <th>unique_curie_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Small river biome</td>\n",
       "      <td>1819</td>\n",
       "      <td>Small river biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Small river biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000890</td>\n",
       "      <td>small river biome</td>\n",
       "      <td>[ENVO:00000890]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>1644</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>[ENVO:00000447]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>ENVO:01000047</td>\n",
       "      <td>805</td>\n",
       "      <td>ENVO:01000047</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:01000047</td>\n",
       "      <td>mediterranean sea biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[ENVO:01000047]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>sediment</td>\n",
       "      <td>729</td>\n",
       "      <td>sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sediment</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "      <td>[ENVO:00002007]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>estuarine biome [ENVO:01000020]</td>\n",
       "      <td>669</td>\n",
       "      <td>estuarine biome [ENVO:01000020]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>estuarine biome</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>estuarine biome</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>estuarine biome</td>\n",
       "      <td>[ENVO:01000020]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>1366</td>\n",
       "      <td>estuarine biome[Envo:01000031]</td>\n",
       "      <td>1</td>\n",
       "      <td>estuarine biome[Envo:01000031]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>estuarine biome[Envo:01000031</td>\n",
       "      <td>ENVO:01000031</td>\n",
       "      <td>obsolete marine feature</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>estuarine biome</td>\n",
       "      <td>[ENVO:01000031, ENVO:01000020]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>1367</td>\n",
       "      <td>estuarine biome[Envo:01000035]</td>\n",
       "      <td>1</td>\n",
       "      <td>estuarine biome[Envo:01000035]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>estuarine biome[Envo:01000035</td>\n",
       "      <td>ENVO:01000035</td>\n",
       "      <td>oceanic epipelagic zone biome</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>estuarine biome</td>\n",
       "      <td>[ENVO:01000035, ENVO:01000020]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>1368</td>\n",
       "      <td>ENVO:01000026</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO:01000026</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:01000026</td>\n",
       "      <td>marine bathyal zone biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[ENVO:01000026]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1396</th>\n",
       "      <td>1369</td>\n",
       "      <td>ENVO:01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO:01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:01000035</td>\n",
       "      <td>oceanic epipelagic zone biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[ENVO:01000035]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1397</th>\n",
       "      <td>1370</td>\n",
       "      <td>ENVO:01000040</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO:01000040</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:01000040</td>\n",
       "      <td>oceanic benthopelagic zone biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[ENVO:01000040]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1398 rows × 13 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 125
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:41.412205Z",
     "start_time": "2024-11-15T19:54:41.406688Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['unique_curie_count'].value_counts()",
   "id": "7a69fc0bd5e16703",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "unique_curie_count\n",
       "1    769\n",
       "0    582\n",
       "2     47\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 126
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:41.607566Z",
     "start_time": "2024-11-15T19:54:41.603335Z"
    }
   },
   "cell_type": "code",
   "source": "double_curie_frame = ncbi_frame[ncbi_frame['unique_curie_count'] > 1]",
   "id": "2b43b2226a208cf2",
   "outputs": [],
   "execution_count": 127
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:41.729938Z",
     "start_time": "2024-11-15T19:54:41.726186Z"
    }
   },
   "cell_type": "code",
   "source": "double_curie_frame = double_curie_frame[['extracted_curie', 'longest_annotation_curie']]",
   "id": "9e74fd42072e1242",
   "outputs": [],
   "execution_count": 128
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:41.813857Z",
     "start_time": "2024-11-15T19:54:41.809496Z"
    }
   },
   "cell_type": "code",
   "source": "double_curie_frame = double_curie_frame.drop_duplicates()",
   "id": "3406bb4668dc4ecc",
   "outputs": [],
   "execution_count": 129
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:41.937530Z",
     "start_time": "2024-11-15T19:54:41.931175Z"
    }
   },
   "cell_type": "code",
   "source": "double_curie_frame[['extracted_prefix', 'extracted_local_id']] = double_curie_frame['extracted_curie'].str.split(':', expand=True)",
   "id": "79a954a6c86b49c7",
   "outputs": [],
   "execution_count": 130
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:47.549872Z",
     "start_time": "2024-11-15T19:54:47.545936Z"
    }
   },
   "cell_type": "code",
   "source": "double_curie_frame['extracted_local_id_int'] = pd.to_numeric(double_curie_frame['extracted_local_id'], errors='coerce').astype('Int64')",
   "id": "9822b8635cb794c",
   "outputs": [],
   "execution_count": 131
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:53.009657Z",
     "start_time": "2024-11-15T19:54:53.003958Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def find_consecutive_stretches_dict(series):\n",
    "    \"\"\"\n",
    "    Detect consecutive stretches of integer values in a pandas Series.\n",
    "    Returns a dictionary where the keys are serial numbers (starting at 0),\n",
    "    and the values are lists of consecutive integers.\n",
    "    \"\"\"\n",
    "    # Ensure the series is clean: drop NaN, duplicates, and non-integer values\n",
    "    series = series.dropna().drop_duplicates()\n",
    "    series = series[series.apply(lambda x: isinstance(x, (int, float)) and (x == int(x)))].astype(int)\n",
    "    series = series.sort_values()\n",
    "\n",
    "    stretches_dict = {}\n",
    "    current_stretch = []\n",
    "    stretch_index = 1\n",
    "\n",
    "    for i in range(len(series)):\n",
    "        if i == 0 or series.iloc[i] - series.iloc[i - 1] == 1:\n",
    "            current_stretch.append(series.iloc[i])\n",
    "        else:\n",
    "            if len(current_stretch) >= 3:\n",
    "                stretches_dict[stretch_index] = current_stretch\n",
    "                stretch_index += 1\n",
    "            current_stretch = [series.iloc[i]]\n",
    "\n",
    "    if len(current_stretch) >= 3:\n",
    "        stretches_dict[stretch_index] = current_stretch\n",
    "\n",
    "    return stretches_dict\n"
   ],
   "id": "14b9721ab1bdae08",
   "outputs": [],
   "execution_count": 132
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:54:58.969914Z",
     "start_time": "2024-11-15T19:54:58.966132Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Function to convert stretches dict to long-format DataFrame\n",
    "def stretches_dict_to_long_dataframe(stretches_dict):\n",
    "    \"\"\"\n",
    "    Convert a dictionary of consecutive stretches into a long-format DataFrame.\n",
    "    Each row corresponds to an individual integer value in a stretch, with:\n",
    "    - 'stretch_id': The key from the dictionary.\n",
    "    - 'value': The integer value within the stretch.\n",
    "    \"\"\"\n",
    "    rows = []\n",
    "    for stretch_id, stretch_values in stretches_dict.items():\n",
    "        for value in stretch_values:\n",
    "            rows.append({'stretch_id': stretch_id, 'value': int(value)})  # Ensure integers only\n",
    "    return pd.DataFrame(rows)"
   ],
   "id": "93aeb0ebbaf0984c",
   "outputs": [],
   "execution_count": 133
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:04.202909Z",
     "start_time": "2024-11-15T19:55:04.198766Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Ensure extracted_local_id_int is unique and sorted\n",
    "unique_sorted_series = double_curie_frame['extracted_local_id_int'].dropna().drop_duplicates().sort_values()\n"
   ],
   "id": "7cd8a092fdd31038",
   "outputs": [],
   "execution_count": 134
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:09.492906Z",
     "start_time": "2024-11-15T19:55:09.488308Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Find stretches\n",
    "stretches_dict = find_consecutive_stretches_dict(unique_sorted_series)\n",
    "\n",
    "# pprint.pprint(stretches_dict)"
   ],
   "id": "97c87784915ccb88",
   "outputs": [],
   "execution_count": 135
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:14.810522Z",
     "start_time": "2024-11-15T19:55:14.807245Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Convert the stretches dictionary into a DataFrame\n",
    "stretches_df = stretches_dict_to_long_dataframe(stretches_dict)"
   ],
   "id": "4dc6b1dc0c012b66",
   "outputs": [],
   "execution_count": 136
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:20.144355Z",
     "start_time": "2024-11-15T19:55:20.137366Z"
    }
   },
   "cell_type": "code",
   "source": "stretches_df",
   "id": "64f939c9fdeadd27",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    stretch_id    value\n",
       "0            1  1000020\n",
       "1            1  1000021\n",
       "2            1  1000022\n",
       "3            1  1000023\n",
       "4            1  1000024\n",
       "5            1  1000025\n",
       "6            1  1000026\n",
       "7            1  1000027\n",
       "8            1  1000028\n",
       "9            1  1000029\n",
       "10           1  1000030\n",
       "11           1  1000031\n",
       "12           1  1000032\n",
       "13           1  1000033\n",
       "14           1  1000034\n",
       "15           1  1000035\n",
       "16           1  1000036\n",
       "17           1  1000037\n",
       "18           1  1000038\n",
       "19           1  1000039\n",
       "20           1  1000040\n",
       "21           2  1000048\n",
       "22           2  1000049\n",
       "23           2  1000050\n",
       "24           3  1000052\n",
       "25           3  1000053\n",
       "26           3  1000054"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stretch_id</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1000020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1000022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1000023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1000024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>1000025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>1000026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>1000027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>1000028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>1000029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>1000030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>1000031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>1000032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>1000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>1000034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>1000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>1000036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>1000037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>1000038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>1000039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>1000040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>2</td>\n",
       "      <td>1000048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2</td>\n",
       "      <td>1000049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2</td>\n",
       "      <td>1000050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>3</td>\n",
       "      <td>1000052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>3</td>\n",
       "      <td>1000053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>3</td>\n",
       "      <td>1000054</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 137
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:26.281916Z",
     "start_time": "2024-11-15T19:55:26.268466Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Perform the left merge\n",
    "double_curie_frame = double_curie_frame.merge(\n",
    "    stretches_df,\n",
    "    left_on='extracted_local_id_int',\n",
    "    right_on='value',\n",
    "    how='left'\n",
    ")"
   ],
   "id": "8c04d9d078886956",
   "outputs": [],
   "execution_count": 138
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:26.376913Z",
     "start_time": "2024-11-15T19:55:26.372364Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Create a new DataFrame summarizing each stretch_id with the most common longest_annotation_curie\n",
    "def summarize_stretch_groups(df):\n",
    "    summary_rows = []\n",
    "    \n",
    "    # Iterate through each group of rows by stretch_id\n",
    "    for stretch_id, group in df.dropna(subset=['stretch_id']).groupby('stretch_id'):\n",
    "        # Calculate the most common longest_annotation_curie and its fraction\n",
    "        most_common_curie = group['longest_annotation_curie'].value_counts().idxmax()\n",
    "        fraction = group['longest_annotation_curie'].value_counts(normalize=True).max()\n",
    "        \n",
    "        # Append the summary row\n",
    "        summary_rows.append({\n",
    "            'stretch_id': stretch_id,\n",
    "            'most_common_longest_annotation_curie': most_common_curie,\n",
    "            'fraction': fraction\n",
    "        })\n",
    "    \n",
    "    # Convert the summary rows into a new DataFrame\n",
    "    return pd.DataFrame(summary_rows)\n"
   ],
   "id": "1ffd89fc81da1568",
   "outputs": [],
   "execution_count": 139
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:26.440714Z",
     "start_time": "2024-11-15T19:55:26.432755Z"
    }
   },
   "cell_type": "code",
   "source": "stretch_summary_df = summarize_stretch_groups(double_curie_frame)\n",
   "id": "1e339dc7cb20377d",
   "outputs": [],
   "execution_count": 140
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:31.854573Z",
     "start_time": "2024-11-15T19:55:31.847510Z"
    }
   },
   "cell_type": "code",
   "source": "stretch_summary_df",
   "id": "dfd6a50c15372421",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   stretch_id most_common_longest_annotation_curie  fraction\n",
       "0         1.0                        ENVO:01000020  0.678571\n",
       "1         2.0                        ENVO:01000024  1.000000\n",
       "2         3.0                        ENVO:01000024  1.000000"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stretch_id</th>\n",
       "      <th>most_common_longest_annotation_curie</th>\n",
       "      <th>fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>0.678571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 141
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:38.308873Z",
     "start_time": "2024-11-15T19:55:38.301687Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Perform the left merge\n",
    "double_curie_frame = double_curie_frame.merge(\n",
    "    stretch_summary_df,\n",
    "    left_on='stretch_id',\n",
    "    right_on='stretch_id',\n",
    "    how='left'\n",
    ")"
   ],
   "id": "33d066472e719724",
   "outputs": [],
   "execution_count": 142
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:38.375583Z",
     "start_time": "2024-11-15T19:55:38.371505Z"
    }
   },
   "cell_type": "code",
   "source": "# ncbi_frame.to_csv('ncbi_frame.tsv', sep='\\t', index=False)",
   "id": "a0fcaee2ee390915",
   "outputs": [],
   "execution_count": 143
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:38.443001Z",
     "start_time": "2024-11-15T19:55:38.434111Z"
    }
   },
   "cell_type": "code",
   "source": [
    "drag_evidence_frame = double_curie_frame[double_curie_frame['stretch_id'] >= 1]\n",
    "drag_evidence_frame = drag_evidence_frame[['extracted_curie', 'longest_annotation_curie']]\n",
    "drag_evidence_frame['drag_evidence'] = True"
   ],
   "id": "4ee4a3b5e3d57324",
   "outputs": [],
   "execution_count": 144
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:45.383514Z",
     "start_time": "2024-11-15T19:55:45.372081Z"
    }
   },
   "cell_type": "code",
   "source": "drag_evidence_frame",
   "id": "b538401f9eba6f23",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   extracted_curie longest_annotation_curie  drag_evidence\n",
       "4    ENVO:01000022            ENVO:00000054           True\n",
       "5    ENVO:01000020            ENVO:00000045           True\n",
       "10   ENVO:01000031            ENVO:01000024           True\n",
       "11   ENVO:01000040            ENVO:01000024           True\n",
       "12   ENVO:01000053            ENVO:01000024           True\n",
       "13   ENVO:01000049            ENVO:01000024           True\n",
       "14   ENVO:01000052            ENVO:01000024           True\n",
       "15   ENVO:01000034            ENVO:01000024           True\n",
       "16   ENVO:01000036            ENVO:01000024           True\n",
       "19   ENVO:01000022            ENVO:01000020           True\n",
       "20   ENVO:01000033            ENVO:01000020           True\n",
       "21   ENVO:01000037            ENVO:01000020           True\n",
       "22   ENVO:01000038            ENVO:01000020           True\n",
       "23   ENVO:01000023            ENVO:01000020           True\n",
       "24   ENVO:01000030            ENVO:01000020           True\n",
       "25   ENVO:01000033            ENVO:01000024           True\n",
       "26   ENVO:01000050            ENVO:01000024           True\n",
       "27   ENVO:01000054            ENVO:01000024           True\n",
       "28   ENVO:01000038            ENVO:01000024           True\n",
       "29   ENVO:01000032            ENVO:01000024           True\n",
       "30   ENVO:01000048            ENVO:01000024           True\n",
       "32   ENVO:01000027            ENVO:01000020           True\n",
       "33   ENVO:01000029            ENVO:01000020           True\n",
       "34   ENVO:01000032            ENVO:01000020           True\n",
       "35   ENVO:01000025            ENVO:01000020           True\n",
       "36   ENVO:01000034            ENVO:01000020           True\n",
       "37   ENVO:01000036            ENVO:01000020           True\n",
       "39   ENVO:01000039            ENVO:01000020           True\n",
       "40   ENVO:01000024            ENVO:01000020           True\n",
       "41   ENVO:01000026            ENVO:01000020           True\n",
       "42   ENVO:01000021            ENVO:01000020           True\n",
       "43   ENVO:01000028            ENVO:01000020           True\n",
       "44   ENVO:01000031            ENVO:01000020           True\n",
       "45   ENVO:01000035            ENVO:01000020           True"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>extracted_curie</th>\n",
       "      <th>longest_annotation_curie</th>\n",
       "      <th>drag_evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:01000022</td>\n",
       "      <td>ENVO:00000054</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>ENVO:00000045</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENVO:01000031</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENVO:01000040</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENVO:01000053</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENVO:01000049</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENVO:01000052</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENVO:01000034</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENVO:01000036</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENVO:01000022</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENVO:01000033</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENVO:01000037</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENVO:01000038</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENVO:01000023</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENVO:01000030</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENVO:01000033</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENVO:01000050</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENVO:01000054</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENVO:01000038</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENVO:01000032</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENVO:01000048</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ENVO:01000027</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENVO:01000029</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENVO:01000032</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENVO:01000025</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENVO:01000034</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ENVO:01000036</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>ENVO:01000039</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>ENVO:01000026</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>ENVO:01000021</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>ENVO:01000028</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ENVO:01000031</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>ENVO:01000035</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 145
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:52.538792Z",
     "start_time": "2024-11-15T19:55:52.531167Z"
    }
   },
   "cell_type": "code",
   "source": [
    "ncbi_frame = ncbi_frame.merge(\n",
    "    drag_evidence_frame,\n",
    "    left_on=['extracted_curie', 'longest_annotation_curie'],\n",
    "    right_on=['extracted_curie', 'longest_annotation_curie'],\n",
    "    how='left'\n",
    ")"
   ],
   "id": "6c204f542b99cc8e",
   "outputs": [],
   "execution_count": 146
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:52.814094Z",
     "start_time": "2024-11-15T19:55:52.668965Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Initialize dragless_curie_list with curie_list values\n",
    "ncbi_frame[\"dragless_curie_list\"] = ncbi_frame[\"curie_list\"]\n",
    "\n",
    "# Update dragless_curie_list based on the condition\n",
    "for index, row in ncbi_frame.iterrows():\n",
    "    if row[\"drag_evidence\"] is True:\n",
    "        if row[\"longest_annotation_curie\"] is not None:\n",
    "            ncbi_frame.at[index, \"dragless_curie_list\"] = [row[\"longest_annotation_curie\"]]\n",
    "        else:\n",
    "            ncbi_frame.at[index, \"dragless_curie_list\"] = []\n",
    "\n",
    "ncbi_frame['dragless_curie_count'] = ncbi_frame['dragless_curie_list'].apply(len)"
   ],
   "id": "60cb2591f94be794",
   "outputs": [],
   "execution_count": 147
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:55:52.884682Z",
     "start_time": "2024-11-15T19:55:52.877455Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['unique_curie_count'].value_counts()",
   "id": "ad797f2c47ad4d2e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "unique_curie_count\n",
       "1    769\n",
       "0    582\n",
       "2     47\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 148
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:01.469942Z",
     "start_time": "2024-11-15T19:56:01.462043Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame['dragless_curie_count'].value_counts()",
   "id": "db64af727a390bde",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dragless_curie_count\n",
       "1    803\n",
       "0    582\n",
       "2     13\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 149
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:01.564724Z",
     "start_time": "2024-11-15T19:56:01.560602Z"
    }
   },
   "cell_type": "code",
   "source": "# ncbi_frame.to_csv('ncbi_frame.tsv', sep='\\t', index=False)",
   "id": "7c8010b9d0d4c140",
   "outputs": [],
   "execution_count": 150
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:01.612723Z",
     "start_time": "2024-11-15T19:56:01.605372Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame.shape",
   "id": "e7369d5d29ddd57e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1398, 16)"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 151
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:01.789218Z",
     "start_time": "2024-11-15T19:56:01.781951Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame_undisputed = ncbi_frame[ncbi_frame['dragless_curie_count'] <= 1]",
   "id": "f43289b34183795d",
   "outputs": [],
   "execution_count": 152
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:01.853909Z",
     "start_time": "2024-11-15T19:56:01.846539Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame_undisputed.shape",
   "id": "abc66f266ab36d4",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1385, 16)"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 153
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:01.993245Z",
     "start_time": "2024-11-15T19:56:01.986025Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame_disputed = ncbi_frame[ncbi_frame['dragless_curie_count'] > 1]",
   "id": "df155f2a8e6f043e",
   "outputs": [],
   "execution_count": 154
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:02.034586Z",
     "start_time": "2024-11-15T19:56:02.029294Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame_disputed.shape",
   "id": "1e255b1b388190ae",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13, 16)"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 155
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:10.450136Z",
     "start_time": "2024-11-15T19:56:10.442658Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame_disputed = ncbi_frame_disputed.explode(\"dragless_curie_list\", ignore_index=True)\n",
   "id": "f572b9083b2c0183",
   "outputs": [],
   "execution_count": 156
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:17.915431Z",
     "start_time": "2024-11-15T19:56:17.909978Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame_disputed.shape",
   "id": "b5166d98cf8b8c35",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(26, 16)"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 157
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:25.055577Z",
     "start_time": "2024-11-15T19:56:25.051002Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_frame_disputed[\"dragless_curie_list\"] = ncbi_frame_disputed[\"dragless_curie_list\"].apply(lambda x: [x])",
   "id": "dbbf498f0dc1e3e7",
   "outputs": [],
   "execution_count": 158
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:33.379189Z",
     "start_time": "2024-11-15T19:56:33.374081Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Combine the rows of ncbi_frame_undisputed and ncbi_frame_disputed into a new DataFrame\n",
    "ncbi_disputes_exploded_frame = pd.concat([ncbi_frame_undisputed, ncbi_frame_disputed], ignore_index=True)\n"
   ],
   "id": "cffbde835766f6fd",
   "outputs": [],
   "execution_count": 159
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:41.563232Z",
     "start_time": "2024-11-15T19:56:41.558651Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_disputes_exploded_frame.shape",
   "id": "be0bc9e01b0223b2",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1411, 16)"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 160
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:47.717647Z",
     "start_time": "2024-11-15T19:56:47.695261Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_disputes_exploded_frame",
   "id": "30054792f2183507",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      serial_number                                            content  \\\n",
       "0                 1                                  Small river biome   \n",
       "1                 2                                       marine biome   \n",
       "2                 3                                      ENVO:01000047   \n",
       "3                 4                                           sediment   \n",
       "4                 5                    estuarine biome [ENVO:01000020]   \n",
       "...             ...                                                ...   \n",
       "1406            732               marine benthic biome [ENVO:01000046]   \n",
       "1407           1121  Hydothermal field [ENVO:01001854] | Sea [ENVO:...   \n",
       "1408           1121  Hydothermal field [ENVO:01001854] | Sea [ENVO:...   \n",
       "1409           1251                Groundwater aquifer [ENVO:00012408]   \n",
       "1410           1251                Groundwater aquifer [ENVO:00012408]   \n",
       "\n",
       "      sample_count                          content_list  content_count  \\\n",
       "0             1819                     Small river biome              1   \n",
       "1             1644                          marine biome              1   \n",
       "2              805                         ENVO:01000047              1   \n",
       "3              729                              sediment              1   \n",
       "4              669       estuarine biome [ENVO:01000020]              1   \n",
       "...            ...                                   ...            ...   \n",
       "1406             1  marine benthic biome [ENVO:01000046]              1   \n",
       "1407             1    Hydothermal field [ENVO:01001854]               2   \n",
       "1408             1    Hydothermal field [ENVO:01001854]               2   \n",
       "1409             1   Groundwater aquifer [ENVO:00012408]              1   \n",
       "1410             1   Groundwater aquifer [ENVO:00012408]              1   \n",
       "\n",
       "      envo_count       extracted_label extracted_curie  \\\n",
       "0              0     Small river biome            None   \n",
       "1              0          marine biome            None   \n",
       "2              1                         ENVO:01000047   \n",
       "3              0              sediment            None   \n",
       "4              1       estuarine biome   ENVO:01000020   \n",
       "...          ...                   ...             ...   \n",
       "1406           1  marine benthic biome   ENVO:01000046   \n",
       "1407           1     Hydothermal field   ENVO:01001854   \n",
       "1408           1     Hydothermal field   ENVO:01001854   \n",
       "1409           1   Groundwater aquifer   ENVO:00012408   \n",
       "1410           1   Groundwater aquifer   ENVO:00012408   \n",
       "\n",
       "                   real_label longest_annotation_curie  \\\n",
       "0                        None            ENVO:00000890   \n",
       "1                        None            ENVO:00000447   \n",
       "2     mediterranean sea biome                     None   \n",
       "3                        None            ENVO:00002007   \n",
       "4             estuarine biome            ENVO:01000020   \n",
       "...                       ...                      ...   \n",
       "1406       marginal sea biome            ENVO:01000024   \n",
       "1407       hydrothermal field            ENVO:01000352   \n",
       "1408       hydrothermal field            ENVO:01000352   \n",
       "1409                  aquifer            ENVO:01001004   \n",
       "1410                  aquifer            ENVO:01001004   \n",
       "\n",
       "     longest_annotation_label                      curie_list  \\\n",
       "0           small river biome                 [ENVO:00000890]   \n",
       "1                marine biome                 [ENVO:00000447]   \n",
       "2                        None                 [ENVO:01000047]   \n",
       "3                    sediment                 [ENVO:00002007]   \n",
       "4             estuarine biome                 [ENVO:01000020]   \n",
       "...                       ...                             ...   \n",
       "1406     marine benthic biome  [ENVO:01000046, ENVO:01000024]   \n",
       "1407                    field  [ENVO:01000352, ENVO:01001854]   \n",
       "1408                    field  [ENVO:01000352, ENVO:01001854]   \n",
       "1409              groundwater  [ENVO:00012408, ENVO:01001004]   \n",
       "1410              groundwater  [ENVO:00012408, ENVO:01001004]   \n",
       "\n",
       "      unique_curie_count drag_evidence dragless_curie_list  \\\n",
       "0                      1           NaN     [ENVO:00000890]   \n",
       "1                      1           NaN     [ENVO:00000447]   \n",
       "2                      1           NaN     [ENVO:01000047]   \n",
       "3                      1           NaN     [ENVO:00002007]   \n",
       "4                      1           NaN     [ENVO:01000020]   \n",
       "...                  ...           ...                 ...   \n",
       "1406                   2           NaN     [ENVO:01000024]   \n",
       "1407                   2           NaN     [ENVO:01000352]   \n",
       "1408                   2           NaN     [ENVO:01001854]   \n",
       "1409                   2           NaN     [ENVO:00012408]   \n",
       "1410                   2           NaN     [ENVO:01001004]   \n",
       "\n",
       "      dragless_curie_count  \n",
       "0                        1  \n",
       "1                        1  \n",
       "2                        1  \n",
       "3                        1  \n",
       "4                        1  \n",
       "...                    ...  \n",
       "1406                     2  \n",
       "1407                     2  \n",
       "1408                     2  \n",
       "1409                     2  \n",
       "1410                     2  \n",
       "\n",
       "[1411 rows x 16 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>serial_number</th>\n",
       "      <th>content</th>\n",
       "      <th>sample_count</th>\n",
       "      <th>content_list</th>\n",
       "      <th>content_count</th>\n",
       "      <th>envo_count</th>\n",
       "      <th>extracted_label</th>\n",
       "      <th>extracted_curie</th>\n",
       "      <th>real_label</th>\n",
       "      <th>longest_annotation_curie</th>\n",
       "      <th>longest_annotation_label</th>\n",
       "      <th>curie_list</th>\n",
       "      <th>unique_curie_count</th>\n",
       "      <th>drag_evidence</th>\n",
       "      <th>dragless_curie_list</th>\n",
       "      <th>dragless_curie_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Small river biome</td>\n",
       "      <td>1819</td>\n",
       "      <td>Small river biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Small river biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000890</td>\n",
       "      <td>small river biome</td>\n",
       "      <td>[ENVO:00000890]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:00000890]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>1644</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>[ENVO:00000447]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:00000447]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>ENVO:01000047</td>\n",
       "      <td>805</td>\n",
       "      <td>ENVO:01000047</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:01000047</td>\n",
       "      <td>mediterranean sea biome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[ENVO:01000047]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:01000047]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>sediment</td>\n",
       "      <td>729</td>\n",
       "      <td>sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>sediment</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "      <td>[ENVO:00002007]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:00002007]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>estuarine biome [ENVO:01000020]</td>\n",
       "      <td>669</td>\n",
       "      <td>estuarine biome [ENVO:01000020]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>estuarine biome</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>estuarine biome</td>\n",
       "      <td>ENVO:01000020</td>\n",
       "      <td>estuarine biome</td>\n",
       "      <td>[ENVO:01000020]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:01000020]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1406</th>\n",
       "      <td>732</td>\n",
       "      <td>marine benthic biome [ENVO:01000046]</td>\n",
       "      <td>1</td>\n",
       "      <td>marine benthic biome [ENVO:01000046]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>marine benthic biome</td>\n",
       "      <td>ENVO:01000046</td>\n",
       "      <td>marginal sea biome</td>\n",
       "      <td>ENVO:01000024</td>\n",
       "      <td>marine benthic biome</td>\n",
       "      <td>[ENVO:01000046, ENVO:01000024]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:01000024]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1407</th>\n",
       "      <td>1121</td>\n",
       "      <td>Hydothermal field [ENVO:01001854] | Sea [ENVO:...</td>\n",
       "      <td>1</td>\n",
       "      <td>Hydothermal field [ENVO:01001854]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>Hydothermal field</td>\n",
       "      <td>ENVO:01001854</td>\n",
       "      <td>hydrothermal field</td>\n",
       "      <td>ENVO:01000352</td>\n",
       "      <td>field</td>\n",
       "      <td>[ENVO:01000352, ENVO:01001854]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:01000352]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1408</th>\n",
       "      <td>1121</td>\n",
       "      <td>Hydothermal field [ENVO:01001854] | Sea [ENVO:...</td>\n",
       "      <td>1</td>\n",
       "      <td>Hydothermal field [ENVO:01001854]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>Hydothermal field</td>\n",
       "      <td>ENVO:01001854</td>\n",
       "      <td>hydrothermal field</td>\n",
       "      <td>ENVO:01000352</td>\n",
       "      <td>field</td>\n",
       "      <td>[ENVO:01000352, ENVO:01001854]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:01001854]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1409</th>\n",
       "      <td>1251</td>\n",
       "      <td>Groundwater aquifer [ENVO:00012408]</td>\n",
       "      <td>1</td>\n",
       "      <td>Groundwater aquifer [ENVO:00012408]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Groundwater aquifer</td>\n",
       "      <td>ENVO:00012408</td>\n",
       "      <td>aquifer</td>\n",
       "      <td>ENVO:01001004</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>[ENVO:00012408, ENVO:01001004]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:00012408]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1410</th>\n",
       "      <td>1251</td>\n",
       "      <td>Groundwater aquifer [ENVO:00012408]</td>\n",
       "      <td>1</td>\n",
       "      <td>Groundwater aquifer [ENVO:00012408]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Groundwater aquifer</td>\n",
       "      <td>ENVO:00012408</td>\n",
       "      <td>aquifer</td>\n",
       "      <td>ENVO:01001004</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>[ENVO:00012408, ENVO:01001004]</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENVO:01001004]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1411 rows × 16 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 161
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:54.050783Z",
     "start_time": "2024-11-15T19:56:54.045412Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_disputes_exploded_frame['post_explode_curie_count'] = ncbi_disputes_exploded_frame['dragless_curie_list'].apply(len)",
   "id": "6c4bcc094860c1ca",
   "outputs": [],
   "execution_count": 162
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:54.140159Z",
     "start_time": "2024-11-15T19:56:54.133427Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_disputes_exploded_frame['post_explode_curie_count'].value_counts()",
   "id": "2c017b4a72fea50d",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "post_explode_curie_count\n",
       "1    829\n",
       "0    582\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 163
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:54.279830Z",
     "start_time": "2024-11-15T19:56:54.272997Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Set 'post_explode_curie' to the 0th item in 'dragless_curie_list'\n",
    "ncbi_disputes_exploded_frame[\"post_explode_curie\"] = ncbi_disputes_exploded_frame[\"dragless_curie_list\"].apply(\n",
    "    lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None\n",
    ")"
   ],
   "id": "53565dbf2f899386",
   "outputs": [],
   "execution_count": 164
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:54.473704Z",
     "start_time": "2024-11-15T19:56:54.464702Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# ncbi_biosample_ebs_counts = ncbi_disputes_exploded_frame['post_explode_curie'].value_counts().reset_index()\n",
    "\n",
    "ncbi_biosample_ebs_counts = (\n",
    "    ncbi_disputes_exploded_frame.groupby(\"post_explode_curie\")[\"sample_count\"].sum().reset_index()\n",
    ")\n",
    "\n",
    "ncbi_biosample_ebs_counts.columns = ['curie', 'ncbi_ebs_count']"
   ],
   "id": "b7aad78ffe14089a",
   "outputs": [],
   "execution_count": 165
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:54.550087Z",
     "start_time": "2024-11-15T19:56:54.538853Z"
    }
   },
   "cell_type": "code",
   "source": "ncbi_biosample_ebs_counts",
   "id": "3fdf963c16419e30",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "              curie  ncbi_ebs_count\n",
       "0       CHEBI:15377             113\n",
       "1       CHEBI:15379               7\n",
       "2       CHEBI:16134               2\n",
       "3       CHEBI:22695              14\n",
       "4     ENVO:00000003              42\n",
       "..              ...             ...\n",
       "241    PATO:0001456             236\n",
       "242    PATO:0001657               1\n",
       "243    PATO:0045093               8\n",
       "244     PCO:1000004              43\n",
       "245  UBERON:0001988               3\n",
       "\n",
       "[246 rows x 2 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>ncbi_ebs_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEBI:15377</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEBI:15379</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEBI:16134</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEBI:22695</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:00000003</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>PATO:0001456</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>PATO:0001657</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>PATO:0045093</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>PCO:1000004</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>UBERON:0001988</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246 rows × 2 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 166
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:54.643800Z",
     "start_time": "2024-11-15T19:56:54.636619Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Perform the left merge\n",
    "rows_frame = rows_frame.merge(\n",
    "    ncbi_biosample_ebs_counts,\n",
    "    left_on='curie',\n",
    "    right_on='curie',\n",
    "    how='left'\n",
    ")"
   ],
   "id": "a27bbaecf2e606f6",
   "outputs": [],
   "execution_count": 167
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:54.908175Z",
     "start_time": "2024-11-15T19:56:54.886314Z"
    }
   },
   "cell_type": "code",
   "source": "rows_frame",
   "id": "f684c7a677faf979",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "              curie                      label  envo_native  obsolete  \\\n",
       "0     ENVO:01000052  marine white smoker biome         True     False   \n",
       "1     ENVO:00000111              forested area         True     False   \n",
       "2     ENVO:00000020                       lake         True     False   \n",
       "3    UBERON:0001988                      feces        False     False   \n",
       "4     ENVO:01001830             tropical biome         True     False   \n",
       "..              ...                        ...          ...       ...   \n",
       "321   ENVO:00000053           freshwater marsh         True     False   \n",
       "322  NCBITaxon:2157                    Archaea        False     False   \n",
       "323   ENVO:03000033            marine sediment         True     False   \n",
       "324   ENVO:01001835               alpine biome         True     False   \n",
       "325   ENVO:03500005       anthropogenic litter         True     False   \n",
       "\n",
       "     legacy_pv    abp  env_sys  biome  terrestrial_biome  aquatic_biome  \\\n",
       "0        False   True     True   True              False           True   \n",
       "1        False   True    False  False              False          False   \n",
       "2        False   True    False  False              False          False   \n",
       "3        False  False    False  False              False          False   \n",
       "4        False   True     True   True              False          False   \n",
       "..         ...    ...      ...    ...                ...            ...   \n",
       "321      False  False     True  False              False          False   \n",
       "322      False  False    False  False              False          False   \n",
       "323      False  False    False  False              False          False   \n",
       "324      False   True     True   True              False          False   \n",
       "325      False  False    False  False              False          False   \n",
       "\n",
       "     env_mat  nmdc_ebs_count  gold_ebs_count  ncbi_ebs_count  \n",
       "0      False             NaN             NaN             NaN  \n",
       "1      False             NaN             NaN            23.0  \n",
       "2      False             NaN             NaN           244.0  \n",
       "3      False             NaN             NaN             3.0  \n",
       "4      False             NaN             NaN             NaN  \n",
       "..       ...             ...             ...             ...  \n",
       "321    False             NaN             NaN             1.0  \n",
       "322    False             NaN             NaN            40.0  \n",
       "323     True             NaN             NaN           701.0  \n",
       "324    False             NaN             NaN             NaN  \n",
       "325    False             NaN             NaN            48.0  \n",
       "\n",
       "[326 rows x 14 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>label</th>\n",
       "      <th>envo_native</th>\n",
       "      <th>obsolete</th>\n",
       "      <th>legacy_pv</th>\n",
       "      <th>abp</th>\n",
       "      <th>env_sys</th>\n",
       "      <th>biome</th>\n",
       "      <th>terrestrial_biome</th>\n",
       "      <th>aquatic_biome</th>\n",
       "      <th>env_mat</th>\n",
       "      <th>nmdc_ebs_count</th>\n",
       "      <th>gold_ebs_count</th>\n",
       "      <th>ncbi_ebs_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:01000052</td>\n",
       "      <td>marine white smoker biome</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENVO:00000111</td>\n",
       "      <td>forested area</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:00000020</td>\n",
       "      <td>lake</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>244.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UBERON:0001988</td>\n",
       "      <td>feces</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:01001830</td>\n",
       "      <td>tropical biome</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>ENVO:00000053</td>\n",
       "      <td>freshwater marsh</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>NCBITaxon:2157</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>701.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>ENVO:01001835</td>\n",
       "      <td>alpine biome</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>ENVO:03500005</td>\n",
       "      <td>anthropogenic litter</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>326 rows × 14 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 168
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:57:18.095736Z",
     "start_time": "2024-11-15T19:57:18.088298Z"
    }
   },
   "cell_type": "code",
   "source": "rows_frame.to_csv('sediment_ebs_rows_frame.tsv', sep='\\t', index=False)",
   "id": "6a822a334470c1fb",
   "outputs": [],
   "execution_count": 170
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-15T19:56:55.223778Z",
     "start_time": "2024-11-15T19:56:55.220736Z"
    }
   },
   "cell_type": "code",
   "source": "",
   "id": "733bb74b08c29519",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
